<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: Downloading and organizing files</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/rcac/favicon-16x16.png"><link rel="manifest" href="favicons/rcac/site.webmanifest"><link rel="mask-icon" href="favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/02-downloading-and-organizing-files.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 17%" class="percentage">
    17%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 17%" aria-valuenow="17" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/02-downloading-and-organizing-files.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Downloading and organizing files
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#the-workshop-dataset-p53-mediated-response-to-ionizing-radiation">The workshop dataset: p53-mediated response to ionizing
radiation</a></li>
<li><a href="#project-organization">Project organization</a></li>
<li><a href="#what-files-do-i-need-for-rna-seq-analysis">What files do I need for RNA-seq analysis?</a></li>
<li><a href="#downloading-the-data">Downloading the data</a></li>
<li><a href="#raw-reads">Raw reads</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Kallisto)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05b-deseq2-expression-analyses-kallisto.html">7. B. Differential expression using DESeq2 (Kallisto pathway)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-gene-set-enrichment-analyses.html">8. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-intro-to-rnaseq.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-qc-and-trimming.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-intro-to-rnaseq.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="03-qc-and-trimming.html" rel="next">
          Next: Quality control of...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Downloading and organizing files</h1>
        <p>Last updated on 2026-01-20 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/02-downloading-and-organizing-files.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What files are required to process raw RNA seq reads?</li>
<li>Where do we obtain raw reads, reference genomes, annotations, and
transcript sequences?</li>
<li>How should project directories be organized to support a smooth
workflow?</li>
<li>What practical considerations matter when downloading and preparing
RNA seq data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Identify the essential inputs for RNA seq analysis: FASTQ files,
reference genome, annotation, and transcript sequences.</li>
<li>Learn where and how to download RNA seq datasets and reference
materials.</li>
<li>Create a clean project directory structure suitable for quality
control, alignment, and quantification.</li>
<li>Understand practical issues such as compressed FASTQ files and
adapter trimming.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>Before we can perform quality control, mapping, or quantification, we
must first gather the files required for RNA seq analysis and organize
them in a reproducible way. This episode introduces the essential inputs
of an RNA seq workflow and demonstrates how to obtain and structure them
on a computing system.</p>
<p>We begin by discussing raw FASTQ reads, reference genomes, annotation
files, and transcript sequences. We then download the dataset used
throughout the workshop and set up the directory structure that supports
downstream steps.</p>
<p>Most FASTQ files arrive compressed as <code>.fastq.gz</code> files.
Modern RNA seq tools accept compressed files directly, so uncompressing
them is usually unnecessary.</p>
<div id="should-i-uncompress-fastq-files" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-uncompress-fastq-files" class="callout-inner">
<h3 class="callout-title">Should I uncompress FASTQ files?</h3>
<div class="callout-content">
<p>In almost all RNA-seq workflows, you should keep FASTQ files
compressed. Tools such as FastQC, HISAT2, STAR, Salmon, and fastp can
read <code>.gz</code> files directly. Keeping files compressed saves
space and reduces input and output overhead.</p>
</div>
</div>
</div>
<p>Many library preparation protocols introduce adapter sequences at the
ends of reads. For most alignment based analyses, explicit trimming is
not required because aligners soft clip adapter sequences
automatically.</p>
<div id="should-i-trim-adapters-for-rna-seq-analysis" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-trim-adapters-for-rna-seq-analysis" class="callout-inner">
<h3 class="callout-title">Should I trim adapters for RNA seq
analysis?</h3>
<div class="callout-content">
<p>In most cases, no. Aligners such as HISAT2 and STAR soft clip adapter
sequences. Overly aggressive trimming can reduce mapping rates or
distort read length distributions. Trimming is needed only for specific
applications such as transcript assembly where uniform read lengths are
important.<br>
More information: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/" class="external-link uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/</a></p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-workshop-dataset-p53-mediated-response-to-ionizing-radiation">The workshop dataset: p53-mediated response to ionizing
radiation<a class="anchor" aria-label="anchor" href="#the-workshop-dataset-p53-mediated-response-to-ionizing-radiation"></a></h2>
<hr class="half-width"><p>For this workshop, we use a subset of a publicly available dataset
(GEO accession <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71176" class="external-link">GSE71176</a>)
that investigates the transcriptional response to ionizing radiation
(IR) in mouse B cells. The original study by <a href="https://doi.org/10.18632/oncotarget.5232" class="external-link">Tonelli et
al. (2015)</a> examined how the tumor suppressor p53 (encoded by
<em>Trp53</em>) regulates gene expression following DNA damage.</p>
<div class="section level3">
<h3 id="biological-context">Biological context<a class="anchor" aria-label="anchor" href="#biological-context"></a></h3>
<p>The <strong>TP53</strong> gene (called <em>Trp53</em> in mice)
encodes the p53 protein, often called the “guardian of the genome.” When
cells experience DNA damage, such as double-strand breaks caused by
ionizing radiation, p53 activates transcriptional programs that lead
to:</p>
<ul><li>
<strong>Cell cycle arrest</strong> (allowing time for DNA
repair)</li>
<li>
<strong>Apoptosis</strong> (programmed cell death if damage is
irreparable)</li>
<li>
<strong>Senescence</strong> (permanent growth arrest)</li>
<li><strong>Metabolic reprogramming</strong></li>
</ul><p>Mutations in <em>TP53</em> are among the most common alterations in
human cancers, occurring in approximately 50% of all tumors.
Understanding p53-regulated genes is therefore central to cancer
biology.</p>
</div>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a></h3>
<p>The full dataset includes wild-type (WT) and p53-knockout (KO) mouse
B cells, with and without IR treatment (7 Gy, 4 hours post-exposure).
For simplicity, we will focus on <strong>8 samples from wild-type B
cells only</strong>:</p>
<table class="table"><colgroup><col width="24%"><col width="28%"><col width="20%"><col width="26%"></colgroup><thead><tr class="header"><th>Condition</th>
<th>Description</th>
<th>Samples</th>
<th>Replicates</th>
</tr></thead><tbody><tr class="odd"><td>
<strong>WT_mock</strong> (control)</td>
<td>Wild-type B cells, no radiation</td>
<td>4</td>
<td>SRR2121778-81</td>
</tr><tr class="even"><td>
<strong>WT_IR</strong> (treatment)</td>
<td>Wild-type B cells, 4h post 7Gy IR</td>
<td>4</td>
<td>SRR2121786-89</td>
</tr></tbody></table><p>This balanced design (n=4 per group) allows us to identify genes that
respond to ionizing radiation in cells with functional p53.</p>
</div>
<div class="section level3">
<h3 id="expected-biological-results">Expected biological results<a class="anchor" aria-label="anchor" href="#expected-biological-results"></a></h3>
<p>Because p53 is a transcription factor activated by DNA damage, we
expect to see upregulation of canonical p53 target genes in IR-treated
samples:</p>
<p><strong>Genes expected to be upregulated (IR vs mock):</strong></p>
<ul><li>
<em>Cdkn1a</em> (p21) - cell cycle arrest</li>
<li>
<em>Bax</em>, <em>Bbc3</em> (Puma), <em>Pmaip1</em> (Noxa) -
pro-apoptotic</li>
<li>
<em>Mdm2</em> - negative feedback regulator of p53</li>
<li>
<em>Gadd45a</em> - DNA damage response</li>
<li>
<em>Fas</em>, <em>Tnfrsf10b</em> (DR5) - death receptor
signaling</li>
</ul><p><strong>Pathways expected to be enriched:</strong></p>
<ul><li>p53 signaling pathway</li>
<li>Apoptosis</li>
<li>Cell cycle</li>
<li>DNA damage response</li>
</ul><p>The presence of these expected results serves as a positive control
that our analysis pipeline is working correctly.</p>
</div>
<div class="section level3">
<h3 id="why-this-dataset">Why this dataset?<a class="anchor" aria-label="anchor" href="#why-this-dataset"></a></h3>
<p>This dataset is ideal for teaching because:</p>
<ol style="list-style-type: decimal"><li>
<strong>Simple two-group comparison</strong>: Control vs. treatment
with balanced replicates</li>
<li>
<strong>Clear biological interpretation</strong>: DNA damage
response is well-characterized</li>
<li>
<strong>Expected positive controls</strong>: Known p53 targets
should appear as top hits</li>
<li>
<strong>Clinical relevance</strong>: p53 biology is fundamental to
cancer research</li>
<li>
<strong>Model organism</strong>: Mouse with excellent genome
annotation</li>
</ol></div>
</section><section><h2 class="section-heading" id="project-organization">Project organization<a class="anchor" aria-label="anchor" href="#project-organization"></a></h2>
<hr class="half-width"><p>Before downloading files, we create a reproducible directory layout
for raw data, scripts, mapping outputs, and count files.</p>
<div id="create-the-directories-needed-for-this-episode" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="create-the-directories-needed-for-this-episode" class="callout-inner">
<h3 class="callout-title">Create the directories needed for this
episode</h3>
<div class="callout-content">
<p>Create a working directory for the workshop (e.g.,
<code>rnaseq-workshop</code>). Inside it, create four
subdirectories:</p>
<ul><li>
<code>data</code> : raw FASTQ files, reference genome,
annotation<br></li>
<li>
<code>scripts</code> : custom scripts, SLURM job files<br></li>
<li>
<code>results</code> : alignment, counts and various other
outputs</li>
</ul></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RCAC_SCRATCH</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> rnaseq-workshop/<span class="dt">{data</span><span class="op">,</span><span class="dt">scripts</span><span class="op">,</span><span class="dt">results}</span></span></code></pre>
</div>
<p>The resulting directory structure should look like this:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">aseethar@scholar-fe03:[aseethar]</span> $ pwd</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">/scratch/scholar/aseethar</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">aseethar@scholar-fe03:[aseethar]</span> $ tree rnaseq-workshop/</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">rnaseq-workshop/</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">├──</span> results</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">├──</span> data</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="ex">└──</span> scripts</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="what-files-do-i-need-for-rna-seq-analysis">What files do I need for RNA-seq analysis?<a class="anchor" aria-label="anchor" href="#what-files-do-i-need-for-rna-seq-analysis"></a></h2>
<hr class="half-width"><div id="prereq3" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div class="callout-inner">
<div class="callout-content">
<p>You will need:</p>
<ol style="list-style-type: decimal"><li>Raw reads (FASTQ)
<ul><li>Usually <code>.fastq.gz</code> files containing nucleotide sequences
and per-base quality scores.</li>
</ul></li>
</ol><p><strong><em>For alignment based workflows</em></strong></p>
<ol start="2" style="list-style-type: decimal"><li>Reference genome (FASTA)
<ul><li>Contains chromosome or contig sequences.</li>
</ul></li>
<li>Annotation file (GTF/GFF)
<ul><li>Describes gene models: exons, transcripts, coordinates.</li>
</ul></li>
</ol><p><strong><em>For transcript-level quantification
workflows</em></strong></p>
<ol start="4" style="list-style-type: decimal"><li>Transcript sequences (FASTA)
<ul><li>Required for transcript-based quantification with tools such as
Salmon, Kallisto, and RSEM.</li>
</ul></li>
</ol><p>Reference files must be consistent with each other (same genome
version). To ensure a smooth analysis workflow, keep these files
logically organized. Good directory structure minimizes mistakes,
simplifies scripting, and supports reproducibility.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="downloading-the-data">Downloading the data<a class="anchor" aria-label="anchor" href="#downloading-the-data"></a></h2>
<hr class="half-width"><p>We now download:</p>
<ul><li>the GRCm39 reference genome<br></li>
<li>the corresponding GENCODE annotation</li>
<li>transcript sequences (if using transcript based quantification)</li>
<li>raw FASTQ files from SRA</li>
</ul><div class="section level3">
<h3 id="reference-genome-and-annotation">Reference genome and annotation<a class="anchor" aria-label="anchor" href="#reference-genome-and-annotation"></a></h3>
<p>We will use <strong>GENCODE GRCm39 (M38)</strong> as the
reference.</p>
<div id="annotation" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="annotation" class="callout-inner">
<h3 class="callout-title">Annotation</h3>
<div class="callout-content">
<p>Visit the GENCODE mouse page: <a href="https://www.gencodegenes.org/mouse/" class="external-link uri">https://www.gencodegenes.org/mouse/</a>.<br>
Which annotation file (GTF/GFF3 files section) should you download, and
why?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Use the <strong>basic gene annotation in GTF format</strong> (e.g.,
<code>gencode.vM38.primary_assembly.basic.annotation.gtf.gz</code>).<br>
It contains essential gene and transcript models without pseudogenes or
other biotypes not typically used for RNA-seq quantification.</p>
<figure style="text-align: center"><img src="fig/01_fq2counts/gencode-gtf.png" alt="Basic gene annotation in GTF format" class="figure mx-auto d-block"><figcaption>
Basic gene annotation in GTF format
</figcaption></figure></div>
</div>
</div>
</div>
<div id="reference-genome" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="reference-genome" class="callout-inner">
<h3 class="callout-title">Reference genome</h3>
<div class="callout-content">
<p>Select the corresponding <strong>Genome sequence (GRCm39)
FASTA</strong> (e.g.,
<code>GRCm39.primary_assembly.genome.fa.gz</code>).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<figure style="text-align: center"><img src="fig/01_fq2counts/gencode-genome.png" alt="Reference genome file in FASTA format" class="figure mx-auto d-block"><figcaption>
Reference genome file in FASTA format
</figcaption></figure></div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="transcript-sequences">Transcript sequences<a class="anchor" aria-label="anchor" href="#transcript-sequences"></a></h3>
<p>Transcript level quantification requires a transcript FASTA file
consistent with the annotation. GENCODE provides transcript sequences
for all transcripts in GRCm39
(<code>gencode.vM38.transcripts.fa.gz</code>).</p>
<div id="transcript-sequence-file" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="transcript-sequence-file" class="callout-inner">
<h3 class="callout-title">Transcript sequence file</h3>
<div class="callout-content">
<p>Where on the GENCODE page can you find the transcript FASTA file?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>The transcript FASTA file is available in the same release directory
as the genome and annotation files (under FASTA files). It is named
<code>gencode.vM38.transcripts.fa.gz</code>.</p>
<figure style="text-align: center"><img src="fig/01_fq2counts/gencode-transcripts.png" alt="Transcript sequences file in FASTA format" class="figure mx-auto d-block"><figcaption>
Transcript sequences file in FASTA format
</figcaption></figure></div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="downloading-all-reference-files">Downloading all reference files<a class="anchor" aria-label="anchor" href="#downloading-all-reference-files"></a></h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="va">GTFlink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/gencode.vM38.primary_assembly.basic.annotation.gtf.gz"</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="va">GENOMElink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/GRCm39.primary_assembly.genome.fa.gz"</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="va">CDSLink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/gencode.vM38.transcripts.fa.gz"</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Download to 'data'</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GENOMElink}</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GTFlink}</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${CDSLink}</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co"># Uncompress</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="bu">cd</span> data</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="fu">gunzip</span> GRCm39.primary_assembly.genome.fa.gz</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="fu">gunzip</span> gencode.vM38.primary_assembly.basic.annotation.gtf.gz</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="fu">gunzip</span> gencode.vM38.transcripts.fa.gz</span></code></pre>
</div>
<p>The transcript file header has multiple pieces of information which
often interfere with downstream analysis. We create a simplified version
containing only the transcript IDs.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> gencode.vM38.transcripts.fa <span class="kw">|</span><span class="fu">head</span> <span class="at">-1</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 <span class="at">-d</span> <span class="st">"|"</span> gencode.vM38.transcripts.fa <span class="op">&gt;</span> gencode.vM38.transcripts-clean.fa</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> gencode.vM38.transcripts-clean.fa <span class="kw">|</span><span class="fu">head</span> <span class="at">-1</span></span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="raw-reads">Raw reads<a class="anchor" aria-label="anchor" href="#raw-reads"></a></h2>
<hr class="half-width"><ul><li>
<strong>Study:</strong> Genome-wide analysis of p53 transcriptional
programs in B cells upon exposure to genotoxic stress <em>in
vivo</em><br></li>
<li>
<strong>Organism:</strong> Mus musculus (C57BL/6)</li>
<li>
<strong>Cell type:</strong> Primary B cells from spleen</li>
<li>
<strong>Design:</strong> Wild-type mock vs Wild-type IR-treated (n=4
per group)</li>
<li>
<strong>Treatment:</strong> 7 Gy ionizing radiation, harvested 4
hours post-exposure</li>
<li>
<strong>Sequencing:</strong> Illumina HiSeq 2000, paired-end
51bp</li>
<li>
<strong>GEO:</strong> <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71176" class="external-link">GSE71176</a>
</li>
<li>
<strong>Reference:</strong> Tonelli et al. <em>Oncotarget</em> 2015
Sep 22;6(28):24611-26. <a href="https://doi.org/10.18632/oncotarget.5232" class="external-link">PMID: 26372730</a>
</li>
</ul><figure style="text-align: center"><img src="fig/01_fq2counts/geo-db.png" alt="FASTQ files from GEO" class="figure mx-auto d-block"><figcaption>
FASTQ files from GEO
</figcaption></figure><div id="where-do-you-find-fastq-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="where-do-you-find-fastq-files" class="callout-inner">
<h3 class="callout-title">Where do you find FASTQ files?</h3>
<div class="callout-content">
<p>GEO provides metadata and experimental details, but FASTQ files are
stored in the <strong>SRA</strong>. How do you obtain the complete list
of sequencing runs?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>Navigate to the SRA page for the BioProject and use <strong>Run
Selector</strong> to list runs.<br>
On the Run Selector page, select the accessions (SRR2121778-81,
SRR2121786-89), toggle the “selected” and click <strong>Accession
List</strong> to download the selected SRR accession IDs.</p>
<figure style="text-align: center"><img src="fig/01_fq2counts/srr-runs.png" alt="FASTQ files from SRA Run Selector" class="figure mx-auto d-block"><figcaption>
FASTQ files from SRA Run Selector
</figcaption></figure></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="download-fastq-files-with-sra-toolkit">Download FASTQ files with SRA Toolkit<a class="anchor" aria-label="anchor" href="#download-fastq-files-with-sra-toolkit"></a></h3>
<p>[DO NOT RUN THIS COMMAND DURING THE WORKSHOP - IT TAKES TOO LONG TO
COMPLETE]. You already copied the pre-downloaded data to your scratch
space.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop/data</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">module</span> load sra-tools</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">SRR</span><span class="kw">;</span> <span class="cf">do</span> </span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>   <span class="ex">fasterq-dump</span> <span class="at">--threads</span> 4 <span class="at">--progress</span> <span class="va">$SRR</span><span class="kw">;</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>   <span class="fu">gzip</span> <span class="va">${SRR}</span>_1.fastq</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>   <span class="fu">gzip</span> <span class="va">${SRR}</span>_2.fastq</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="cf">done</span><span class="op">&lt;</span>SRR_Acc_List.txt</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co"># rename files to meaningful names</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># WT B cells - mock</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="fu">mv</span> SRR2121778_1.fastq.gz WT_Bcell_mock_rep1_R1.fastq.gz</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="fu">mv</span> SRR2121778_2.fastq.gz WT_Bcell_mock_rep1_R2.fastq.gz</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="fu">mv</span> SRR2121779_1.fastq.gz WT_Bcell_mock_rep2_R1.fastq.gz</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="fu">mv</span> SRR2121779_2.fastq.gz WT_Bcell_mock_rep2_R2.fastq.gz</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="fu">mv</span> SRR2121780_1.fastq.gz WT_Bcell_mock_rep3_R1.fastq.gz</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="fu">mv</span> SRR2121780_2.fastq.gz WT_Bcell_mock_rep3_R2.fastq.gz</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="fu">mv</span> SRR2121781_1.fastq.gz WT_Bcell_mock_rep4_R1.fastq.gz</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="fu">mv</span> SRR2121781_2.fastq.gz WT_Bcell_mock_rep4_R2.fastq.gz</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co"># WT B cells - IR</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="fu">mv</span> SRR2121786_1.fastq.gz WT_Bcell_IR_rep1_R1.fastq.gz</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="fu">mv</span> SRR2121786_2.fastq.gz WT_Bcell_IR_rep1_R2.fastq.gz</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="fu">mv</span> SRR2121787_1.fastq.gz WT_Bcell_IR_rep2_R1.fastq.gz</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="fu">mv</span> SRR2121787_2.fastq.gz WT_Bcell_IR_rep2_R2.fastq.gz</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a><span class="fu">mv</span> SRR2121788_1.fastq.gz WT_Bcell_IR_rep3_R1.fastq.gz</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a><span class="fu">mv</span> SRR2121788_2.fastq.gz WT_Bcell_IR_rep3_R2.fastq.gz</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a><span class="fu">mv</span> SRR2121789_1.fastq.gz WT_Bcell_IR_rep4_R1.fastq.gz</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="fu">mv</span> SRR2121789_2.fastq.gz WT_Bcell_IR_rep4_R2.fastq.gz</span></code></pre>
</div>
<div id="workshop-data-is-subsampled" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="workshop-data-is-subsampled" class="callout-inner">
<h3 class="callout-title">Workshop data is subsampled</h3>
<div class="callout-content">
<p>The FASTQ files provided for this workshop have been
<strong>subsampled to 20 million reads per sample</strong>. The original
samples contain 100+ million reads each, and running the full pipeline
would take many hours—far exceeding our workshop session time.</p>
<p><strong>Never subsample your real experimental data.</strong> This is
done exclusively for workshop time constraints. Subsampling reduces
statistical power and may affect differential expression results.</p>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Workshop-only: How reads were subsampled </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p>For reference, here is how the workshop data was prepared.
<strong>You do not need to run this</strong>—the pre-downloaded data is
already subsampled.</p>
<p>The original samples from SRA contain 80-100+ million paired-end
reads each. To enable completion of the full analysis pipeline within a
single workshop day, we subsample to 20 million read pairs per sample
using <code>seqtk</code>.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Subsample reads to 20 million per sample (WORKSHOP ONLY - do not do this with real data!)</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">module</span> load seqtk</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> WT_Bcell_mock_rep1 WT_Bcell_mock_rep2 WT_Bcell_mock_rep3 WT_Bcell_mock_rep4 <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>              WT_Bcell_IR_rep1 WT_Bcell_IR_rep2 WT_Bcell_IR_rep3 WT_Bcell_IR_rep4<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="co"># Use same seed for R1 and R2 to maintain pairing</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    <span class="ex">seqtk</span> sample <span class="at">-s</span> 42 <span class="va">${sample}</span>_R1.fastq.gz 20000000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> <span class="va">${sample}</span>_sub_R1.fastq.gz</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>    <span class="ex">seqtk</span> sample <span class="at">-s</span> 42 <span class="va">${sample}</span>_R2.fastq.gz 20000000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> <span class="va">${sample}</span>_sub_R2.fastq.gz</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co"># Archive original files</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> original_reads</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>_R1.fastq.gz <span class="pp">*</span>_R2.fastq.gz original_reads/</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co"># Rename subsampled files to original names for pipeline compatibility</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> WT_Bcell_mock_rep1 WT_Bcell_mock_rep2 WT_Bcell_mock_rep3 WT_Bcell_mock_rep4 <span class="dt">\</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>              WT_Bcell_IR_rep1 WT_Bcell_IR_rep2 WT_Bcell_IR_rep3 WT_Bcell_IR_rep4<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>    <span class="fu">mv</span> <span class="va">${sample}</span>_sub_R1.fastq.gz <span class="va">${sample}</span>_R1.fastq.gz</span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>    <span class="fu">mv</span> <span class="va">${sample}</span>_sub_R2.fastq.gz <span class="va">${sample}</span>_R2.fastq.gz</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p><strong>Why 20 million reads?</strong></p>
<ul><li>Sufficient depth for differential expression analysis of moderately
expressed genes</li>
<li>Reduces alignment time from ~30 minutes to ~5-10 minutes per
sample</li>
<li>Allows completion of the full workshop pipeline in a single day</li>
<li>Still produces biologically meaningful results for well-expressed
p53 target genes</li>
</ul><p><strong>What you lose by subsampling:</strong></p>
<ul><li>Statistical power to detect lowly expressed genes</li>
<li>Precision in abundance estimates</li>
<li>Ability to detect subtle fold changes</li>
<li>Some lowly expressed transcripts may appear as zeros</li>
</ul><p>For your own research, always use the full sequencing depth. Modern
RNA-seq experiments typically aim for 20-50 million reads for
differential expression, but the original unsubsampled data is always
preferred.</p>
</div>
</div>
</div>
</div>
<p>The directory, after downloading, should look like this:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">data</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">├──</span> gencode.vM38.primary_assembly.basic.annotation.gtf</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="ex">├──</span> gencode.vM38.transcripts.fa</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="ex">├──</span> gencode.vM38.transcripts-clean.fa</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="ex">├──</span> GRCm39.primary_assembly.genome.fa</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep1_R1.fastq.gz</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep1_R2.fastq.gz</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep2_R1.fastq.gz</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep2_R2.fastq.gz</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep3_R1.fastq.gz</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep3_R2.fastq.gz</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep4_R1.fastq.gz</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_mock_rep4_R2.fastq.gz</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep1_R1.fastq.gz</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep1_R2.fastq.gz</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep2_R1.fastq.gz</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep2_R2.fastq.gz</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep3_R1.fastq.gz</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep3_R2.fastq.gz</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep4_R1.fastq.gz</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="ex">├──</span> WT_Bcell_IR_rep4_R2.fastq.gz</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="ex">└──</span> SRR_Acc_List.txt</span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>RNA-seq requires three main inputs: FASTQ reads, a reference genome
(FASTA), and gene annotation (GTF/GFF).</li>
<li>
<em>In lieu</em> of a reference genome and annotation, transcript
sequences (FASTA) can be used for transcript-level quantification.</li>
<li>Keeping files compressed and well-organized supports reproducible
analysis.</li>
<li>Reference files must match in genome build and version.</li>
<li>Public data repositories such as GEO and SRA provide raw reads and
metadata.</li>
<li>Tools like <code>wget</code> and <code>fasterq-dump</code> enable
programmatic, reproducible data retrieval.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-intro-to-rnaseq.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-qc-and-trimming.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-intro-to-rnaseq.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="03-qc-and-trimming.html" rel="next">
          Next: Quality control of...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/02-downloading-and-organizing-files.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.18.3" class="external-link">sandpaper (0.18.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/02-downloading-and-organizing-files.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "Downloading and organizing files",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/02-downloading-and-organizing-files.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/02-downloading-and-organizing-files.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2026-01-20",
  "datePublished": "2026-01-20"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

