<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: B. Differential expression using DESeq2 (Kallisto pathway)</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/rcac/favicon-16x16.png"><link rel="manifest" href="favicons/rcac/site.webmanifest"><link rel="mask-icon" href="favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/05b-deseq2-expression-analyses-kallisto.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 69%" class="percentage">
    69%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 69%" aria-valuenow="69" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/05b-deseq2-expression-analyses-kallisto.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Kallisto)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. B. Differential expression using DESeq2 (Kallisto pathway)
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#attribution">Attribution</a></li>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#step-1-load-packages-and-data">Step 1: Load packages and data</a></li>
<li><a href="#step-2-create-deseq2-object-from-tximport">Step 2: Create DESeq2 object from tximport</a></li>
<li><a href="#step-3-exploratory-data-analysis">Step 3: Exploratory data analysis</a></li>
<li><a href="#step-4-differential-expression-analysis">Step 4: Differential expression analysis</a></li>
<li><a href="#step-5-summarize-and-visualize-results">Step 5: Summarize and visualize results</a></li>
<li><a href="#step-6-save-results">Step 6: Save results</a></li>
<li><a href="#summary">Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-gene-set-enrichment-analyses.html">8. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-deseq2-expression-analyses.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="06-gene-set-enrichment-analyses.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-deseq2-expression-analyses.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Gene-level QC and
        </a>
        <a class="chapter-link float-end" href="06-gene-set-enrichment-analyses.html" rel="next">
          Next: Gene set enrichment...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>B. Differential expression using DESeq2 (Kallisto pathway)</h1>
        <p>Last updated on 2026-01-17 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05b-deseq2-expression-analyses-kallisto.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do we import transcript-level quantification from Kallisto into
DESeq2?</li>
<li>What exploratory analyses should we perform before differential
expression testing?</li>
<li>How do we perform differential expression analysis with DESeq2 using
tximport data?</li>
<li>How do we visualize and interpret DE results from transcript-based
quantification?</li>
<li>What are the key differences between genome-based and
transcript-based DE workflows?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Load tximport data from Kallisto quantification into DESeq2.</li>
<li>Perform quality control visualizations (boxplots, density plots,
PCA).</li>
<li>Apply variance stabilizing transformation for exploratory
analysis.</li>
<li>Run differential expression analysis with appropriate
contrasts.</li>
<li>Create volcano plots to visualize DE results.</li>
<li>Export annotated DE results for downstream analysis.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="attribution">Attribution<a class="anchor" aria-label="anchor" href="#attribution"></a></h2>
<hr class="half-width"><p>This section is adapted from materials developed by <strong>Michael
Gribskov</strong>, Professor of Computational Genomics &amp; Systems
Biology at Purdue University.</p>
<p>Original and related materials are available via the <strong>CGSB
Wiki</strong>: <a href="https://cgsb.miraheze.org/wiki/Main_Page" class="external-link uri">https://cgsb.miraheze.org/wiki/Main_Page</a></p>
</section><section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>In Episode 04b, we quantified transcript expression using
<strong>Kallisto</strong> and summarized the results to gene-level
counts using <strong>tximport</strong>. In this episode, we use those
gene-level estimates for differential expression analysis with
<strong>DESeq2</strong>.</p>
<p>The workflow follows the same general pattern as Episode 05
(genome-based), but with important differences:</p>
<ol style="list-style-type: decimal"><li>
<strong>Input data</strong>: We use the <code>txi</code> object from
tximport rather than raw counts from featureCounts.</li>
<li>
<strong>Count handling</strong>: Kallisto estimates are model-based
(not integer counts), and DESeq2 handles this appropriately via
<code>DESeqDataSetFromTximport()</code>.</li>
<li>
<strong>Bootstrap support</strong>: Kallisto’s uncertainty estimates
(from bootstraps) can be used with sleuth for transcript-level DE.</li>
</ol><div id="when-to-use-this-pathway" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="when-to-use-this-pathway" class="callout-inner">
<h3 class="callout-title">When to use this pathway</h3>
<div class="callout-content">
<p>Use this transcript-based workflow when:</p>
<ul><li>You quantified with Salmon, Kallisto, or similar tools.</li>
<li>You want to leverage transcript-level bias corrections.</li>
<li>You did not generate BAM files and cannot use featureCounts.</li>
</ul><p>The genome-based workflow (Episode 05) is preferred when you have BAM
files and need splice junction information or plan to visualize
alignments.</p>
</div>
</div>
</div>
<div id="what-you-need-for-this-episode" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-you-need-for-this-episode" class="callout-inner">
<h3 class="callout-title">What you need for this episode</h3>
<div class="callout-content">
<ul><li>
<code>txi.rds</code> file generated from tximport in Episode
04b</li>
<li>A <code>samples.csv</code> file describing the experimental
groups</li>
<li>RStudio session via Open OnDemand</li>
</ul><p>If you haven’t created the sample metadata file, create
<code>scripts/samples.csv</code> with:</p>
<pre><code>sample,condition
WT_Bcell_mock_rep1,WT_mock
WT_Bcell_mock_rep2,WT_mock
WT_Bcell_mock_rep3,WT_mock
WT_Bcell_mock_rep4,WT_mock
WT_Bcell_IR_rep1,WT_IR
WT_Bcell_IR_rep2,WT_IR
WT_Bcell_IR_rep3,WT_IR
WT_Bcell_IR_rep4,WT_IR</code></pre>
<p>Create a results directory:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/deseq2_kallisto</span></code></pre>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-1-load-packages-and-data">Step 1: Load packages and data<a class="anchor" aria-label="anchor" href="#step-1-load-packages-and-data"></a></h2>
<hr class="half-width"><p>Start your RStudio session via Open OnDemand as described in Episode
05, then load the required packages:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tximportData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">reshape2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggrepel</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="co"># Construct the path dynamically</span></span>
<span><span class="va">work_dir</span> <span class="op">&lt;-</span> <span class="fu">file.path</span><span class="op">(</span><span class="st">"/scratch/negishi"</span>, <span class="fu">Sys.getenv</span><span class="op">(</span><span class="st">"USER"</span><span class="op">)</span>, <span class="st">"rnaseq-workshop"</span><span class="op">)</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="va">work_dir</span><span class="op">)</span></span></code></pre>
</div>
<p>Load the tximport object created in Episode 04b:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txi</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"results/kallisto_quant/txi.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Examine the structure of the tximport object:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">txi</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] "abundance"           "counts"              "infReps"             "length"              "countsFromAbundance"</code></pre>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">txi</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code>                      WT_Bcell_IR_rep1 WT_Bcell_IR_rep2 WT_Bcell_IR_rep3 WT_Bcell_IR_rep4 WT_Bcell_mock_rep1 WT_Bcell_mock_rep2 WT_Bcell_mock_rep3
ENSMUSG00000000001.5         387.39815         329.4356     737.97311964        654.17550           687.1189          620.35065          661.35411
ENSMUSG00000000003.16          0.00000           1.0000       0.00000000          0.00000             0.0000            0.00000            0.00000
ENSMUSG00000000028.16         32.81579          20.0000      36.63144821         29.71743            53.5750           36.92104           39.00000
ENSMUSG00000000031.20          0.00000           0.0000       0.09733145          0.00000             0.0000            0.00000            0.00000
ENSMUSG00000000037.18          2.00000           3.0000       1.28018005          8.00000             1.0000            0.00000            1.58711
ENSMUSG00000000049.12          0.00000           0.0000       0.00000000          0.00000             0.0000            0.00000            0.00000
                      WT_Bcell_mock_rep4
ENSMUSG00000000001.5            704.8851
ENSMUSG00000000003.16             0.0000
ENSMUSG00000000028.16            36.0000
ENSMUSG00000000031.20             0.0000
ENSMUSG00000000037.18             0.0000
ENSMUSG00000000049.12             0.0000</code></pre>
<div id="understanding-the-tximport-object" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="understanding-the-tximport-object" class="callout-inner">
<h3 class="callout-title">Understanding the tximport object</h3>
<div class="callout-content">
<p>The <code>txi</code> object contains several components:</p>
<ul><li>
<strong>counts</strong>: Gene-level estimated counts (used by
DESeq2).</li>
<li>
<strong>abundance</strong>: Gene-level TPM values (within-sample
normalized).</li>
<li>
<strong>length</strong>: Average transcript length per gene (used
for length bias correction).</li>
<li>
<strong>countsFromAbundance</strong>: Method used to generate counts
(default: <code>"no"</code>).</li>
</ul><p>When using <code>DESeqDataSetFromTximport()</code>, the
<code>length</code> matrix is automatically used to correct for
gene-length bias during normalization. This is why we use the default
<code>countsFromAbundance = "no"</code> in Episode 04b—DESeq2 handles
length correction internally, so pre-scaling counts would apply the
correction twice.</p>
</div>
</div>
</div>
<p>Load sample metadata:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"scripts/samples.csv"</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">txi</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="va">coldata</span></span></code></pre>
</div>
<pre><code>                   condition
WT_Bcell_IR_rep1       WT_IR
WT_Bcell_IR_rep2       WT_IR
WT_Bcell_IR_rep3       WT_IR
WT_Bcell_IR_rep4       WT_IR
WT_Bcell_mock_rep1   WT_mock
WT_Bcell_mock_rep2   WT_mock
WT_Bcell_mock_rep3   WT_mock
WT_Bcell_mock_rep4   WT_mock</code></pre>
<p>Verify that sample names match between tximport and metadata:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">all</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">txi</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] TRUE</code></pre>
</section><section><h2 class="section-heading" id="step-2-create-deseq2-object-from-tximport">Step 2: Create DESeq2 object from tximport<a class="anchor" aria-label="anchor" href="#step-2-create-deseq2-object-from-tximport"></a></h2>
<hr class="half-width"><p>The key difference from the genome-based workflow is using
<code>DESeqDataSetFromTximport()</code> instead of
<code>DESeqDataSetFromMatrix()</code>:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span><span class="op">(</span></span>
<span>    <span class="va">txi</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="why-use-deseqdatasetfromtximport" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-deseqdatasetfromtximport" class="callout-inner">
<h3 class="callout-title">Why use DESeqDataSetFromTximport?</h3>
<div class="callout-content">
<p>This function:</p>
<ul><li>Automatically handles non-integer counts from Salmon/Kallisto.</li>
<li>Preserves transcript length information for accurate
normalization.</li>
<li>Incorporates the <code>txi$length</code> matrix to account for
gene-length bias.</li>
</ul><p>Using <code>DESeqDataSetFromMatrix()</code> with tximport counts
would lose this information.</p>
</div>
</div>
</div>
<p>Filter lowly expressed genes using group-aware filtering:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Group-aware filtering: keep genes with &gt;= 10 counts in at least 4 samples</span></span>
<span><span class="co"># (the size of the smallest experimental group)</span></span>
<span><span class="va">min_samples</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">min_counts</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_counts</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_samples</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="va">dds</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 18924     8</code></pre>
<div id="why-use-group-aware-filtering" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-group-aware-filtering" class="callout-inner">
<h3 class="callout-title">Why use group-aware filtering?</h3>
<div class="callout-content">
<p>A simple sum filter (<code>rowSums(counts(dds)) &gt;= 10</code>) can
be problematic:</p>
<ul><li>A gene with 10 total counts across 8 samples averages ~1.25
counts/sample—too low to be informative.</li>
<li>It may remove genes expressed in only one condition (biologically
interesting!).</li>
</ul><p>Group-aware filtering
(<code>rowSums(counts &gt;= threshold) &gt;= min_group_size</code>)
ensures:</p>
<ul><li>Each kept gene has meaningful expression in at least one
experimental group.</li>
<li>Genes with condition-specific expression are retained.</li>
<li>The threshold is interpretable (e.g., “at least 10 counts in at
least 4 samples”).</li>
</ul></div>
</div>
</div>
<p>Estimate size factors:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">normalizationFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>                      WT_Bcell_IR_rep1 WT_Bcell_IR_rep2 WT_Bcell_IR_rep3 WT_Bcell_IR_rep4 WT_Bcell_mock_rep1 WT_Bcell_mock_rep2 WT_Bcell_mock_rep3
ENSMUSG00000000001.5         0.8578693        0.8520842         1.099577        1.0524525           1.059461          0.9946634          0.9552297
ENSMUSG00000000028.16        0.9455555        0.7479666         1.149805        0.9882356           1.071439          1.0191038          0.9154327
ENSMUSG00000000056.8         0.8580737        0.8528464         1.101744        1.0540656           1.058546          0.9930345          0.9540278
ENSMUSG00000000078.8         0.8580503        0.8527592         1.101496        1.0538811           1.058650          0.9932207          0.9541651
ENSMUSG00000000085.17        0.7941016        0.7697798         1.199288        1.0181546           1.163680          0.9580967          1.0675296
ENSMUSG00000000088.8         0.8537054        0.8364591         1.055812        1.0199231           1.078562          1.0286198          0.9803416
                      WT_Bcell_mock_rep4
ENSMUSG00000000001.5            1.174355
ENSMUSG00000000028.16           1.244899
ENSMUSG00000000056.8            1.173333
ENSMUSG00000000078.8            1.173450
ENSMUSG00000000085.17           1.125634
ENSMUSG00000000088.8            1.195682</code></pre>
</section><section><h2 class="section-heading" id="step-3-exploratory-data-analysis">Step 3: Exploratory data analysis<a class="anchor" aria-label="anchor" href="#step-3-exploratory-data-analysis"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="raw-count-distributions">Raw count distributions<a class="anchor" aria-label="anchor" href="#raw-count-distributions"></a></h3>
<p>Visualize the distribution of raw counts across samples:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_melted</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span></span>
<span>    <span class="fu">log10</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    varnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"sample"</span><span class="op">)</span>,</span>
<span>    value.name <span class="op">=</span> <span class="st">"log10_count"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">counts_melted</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">log10_count</span>, fill <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_boxplot</span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>        y <span class="op">=</span> <span class="st">"log10(count + 1)"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Raw count distributions (Kallisto)"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/raw-count-k.png" alt="Raw count distributions" class="figure mx-auto d-block"><figcaption>
Raw count distributions
</figcaption></figure><p>Density plot of raw counts:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">counts_melted</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log10_count</span>, fill <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="st">"log10(count + 1)"</span>,</span>
<span>        y <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Count density distributions"</span></span>
<span>    <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/count-density-k.png" alt="Count density distributions" class="figure mx-auto d-block"><figcaption>
Count density distributions
</figcaption></figure></div>
<div class="section level3">
<h3 id="variance-stabilizing-transformation">Variance stabilizing transformation<a class="anchor" aria-label="anchor" href="#variance-stabilizing-transformation"></a></h3>
<p>Apply VST for exploratory analysis:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div id="why-use-vst" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-vst" class="callout-inner">
<h3 class="callout-title">Why use VST?</h3>
<div class="callout-content">
<p>Raw counts have a strong mean-variance relationship: highly expressed
genes have higher variance. VST removes this dependency, making
distance-based methods (PCA, clustering) more reliable.</p>
<p>Setting <code>blind = TRUE</code> ensures the transformation is not
influenced by the experimental design, which is appropriate for quality
control.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sample-to-sample-distances">Sample-to-sample distances<a class="anchor" aria-label="anchor" href="#sample-to-sample-distances"></a></h3>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleDists</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sampleDistMatrix</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">sampleDists</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">rev</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Blues"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span></span>
<span>    <span class="va">sampleDistMatrix</span>,</span>
<span>    clustering_distance_rows <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>    clustering_distance_cols <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Sample distance heatmap (Kallisto)"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/pheatmap-k.png" alt="Sample distance heatmap" class="figure mx-auto d-block"><figcaption>
Sample distance heatmap
</figcaption></figure><p>The distance heatmap shows how similar samples are to each other.
Samples from the same condition should cluster together.</p>
</div>
<div class="section level3">
<h3 id="pca-plot">PCA plot<a class="anchor" aria-label="anchor" href="#pca-plot"></a></h3>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaData</span> <span class="op">&lt;-</span> <span class="fu">plotPCA</span><span class="op">(</span><span class="va">vsd</span>, intgroup <span class="op">=</span> <span class="st">"condition"</span>, returnData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaData</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaData</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text_repel</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"PCA of Kallisto-quantified samples"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/pcaplot-k.png" alt="PCA plot of Kallisto-quantified samples" class="figure mx-auto d-block"><figcaption>
PCA plot of Kallisto-quantified samples
</figcaption></figure><div id="exercise-interpret-exploratory-plots" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-interpret-exploratory-plots" class="callout-inner">
<h3 class="callout-title">Exercise: Interpret exploratory plots</h3>
<div class="callout-content">
<p>Using the distance heatmap and PCA plot:</p>
<ol style="list-style-type: decimal"><li>Do samples cluster by experimental condition?</li>
<li>Are there any outlier samples that don’t group with their
replicates?</li>
<li>How much variance is explained by PC1? What might this represent
biologically?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Example interpretation:</p>
<ol style="list-style-type: decimal"><li>Samples should cluster by condition (WT_mock vs WT_IR) in both the
heatmap and PCA.</li>
<li>If all replicates cluster together, there are no obvious
outliers.</li>
<li>PC1 typically captures the largest source of variation. If it
separates conditions, the experimental treatment is the dominant
signal.</li>
</ol></div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-differential-expression-analysis">Step 4: Differential expression analysis<a class="anchor" aria-label="anchor" href="#step-4-differential-expression-analysis"></a></h2>
<hr class="half-width"><p>Run the full DESeq2 pipeline:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>estimating size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing</code></pre>
<p>Inspect dispersion estimates:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/plotdispests-k.png" alt="Volcano plot of differential expression results" class="figure mx-auto d-block"><figcaption>
Volcano plot of differential expression results
</figcaption></figure><div id="interpreting-the-dispersion-plot" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-the-dispersion-plot" class="callout-inner">
<h3 class="callout-title">Interpreting the dispersion plot</h3>
<div class="callout-content">
<ul><li>
<strong>Black dots</strong>: Gene-wise dispersion estimates.</li>
<li>
<strong>Red line</strong>: Fitted trend (shrinkage target).</li>
<li>
<strong>Blue dots</strong>: Final shrunken estimates.</li>
</ul><p>A good fit shows the red line passing through the center of the black
cloud, with blue dots closer to the line than the original black
dots.</p>
</div>
</div>
</div>
<p>Extract results for the contrast of interest:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span></span>
<span>    <span class="va">dds</span>,</span>
<span>    contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"WT_mock"</span>, <span class="st">"WT_IR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>out of 19693 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 3482, 18%
LFC &lt; 0 (down)     : 3675, 19%
outliers [1]       : 10, 0.051%
low counts [2]     : 0, 0%
(mean count &lt; 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
<p>Order results by adjusted p-value:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_ordered</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res_ordered</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre class="text"><code>log2 fold change (MLE): condition WT_IR vs WT_mock
Wald test p-value: condition WT_IR vs WT_mock
DataFrame with 6 rows and 6 columns
                       baseMean log2FoldChange     lfcSE      stat       pvalue         padj
                      &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;
ENSMUSG00000004085.15   795.499        4.10779  0.157888   26.0172 3.16645e-149 6.23252e-145
ENSMUSG00000021701.9    840.251        6.34293  0.253829   24.9889 8.06388e-138 7.93607e-134
ENSMUSG00000072825.13   705.733        4.70979  0.190927   24.6680 2.35896e-134 1.54771e-130
ENSMUSG00000048458.9    820.666        5.79808  0.238203   24.3409 7.24251e-131 3.56386e-127
ENSMUSG00000028893.9    680.824        4.28171  0.179531   23.8495 1.02517e-125 4.03570e-122
ENSMUSG00000075122.6    836.194        3.74860  0.158501   23.6502 1.17375e-123 3.85048e-120</code></pre>
<div class="section level3">
<h3 id="apply-log-fold-change-shrinkage">Apply log fold change shrinkage<a class="anchor" aria-label="anchor" href="#apply-log-fold-change-shrinkage"></a></h3>
<p>LFC shrinkage improves estimates for genes with low counts or high
dispersion.</p>
<p>First, check the available coefficients:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">resultsNames</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] "Intercept"        "condition_WT_mock_vs_WT_IR"</code></pre>
<p>Apply shrinkage using <code>apeglm</code> (recommended for standard
contrasts):</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_shrunk</span> <span class="op">&lt;-</span> <span class="fu">lfcShrink</span><span class="op">(</span></span>
<span>    <span class="va">dds</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"condition_WT_mock_vs_WT_IR"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"apeglm"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="why-shrink-log-fold-changes" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-shrink-log-fold-changes" class="callout-inner">
<h3 class="callout-title">Why shrink log fold changes?</h3>
<div class="callout-content">
<p>Genes with low counts can have unreliably large fold changes.
Shrinkage:</p>
<ul><li>Reduces noise in LFC estimates.</li>
<li>Improves ranking for downstream analyses (e.g., GSEA).</li>
<li>Does not affect p-values or significance calls.</li>
</ul></div>
</div>
</div>
<div id="choosing-a-shrinkage-method" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="choosing-a-shrinkage-method" class="callout-inner">
<h3 class="callout-title">Choosing a shrinkage method</h3>
<div class="callout-content">
<ul><li>
<strong><code>apeglm</code></strong> (recommended): Fast,
well-calibrated, uses coefficient name (<code>coef</code>).</li>
<li>
<strong><code>ashr</code></strong>: Required for complex contrasts
that can’t be specified with <code>coef</code> (e.g., interaction
terms).</li>
<li>
<strong><code>normal</code></strong>: Legacy method, generally not
recommended.</li>
</ul><p>For simple two-group comparisons like ours, <code>apeglm</code> is
preferred.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-summarize-and-visualize-results">Step 5: Summarize and visualize results<a class="anchor" aria-label="anchor" href="#step-5-summarize-and-visualize-results"></a></h2>
<hr class="half-width"><p>Create a summary table:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res_shrunk</span><span class="op">)</span></span>
<span><span class="va">res_df</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">summary_table</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span></span>
<span>    total_genes <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    up <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    down <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">summary_table</span><span class="op">)</span></span></code></pre>
</div>
<p>output:</p>
<pre class="text"><code># A tibble: 1 × 4
  total_genes   sig    up  down
        &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1       19693  6078  2296  2448</code></pre>
<p>We will attach annotations so that results are interpretable. Load
gene annotation data (from Episode 02):</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/mart.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/annot.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">res_df</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span></code></pre>
</div>
<p>Join DE results, normalized counts, and annotation:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">mart</span>,</span>
<span>                   by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span></code></pre>
</div>
<p>Joining normalized counts and annotation makes the output
biologically interpretable. The final table becomes a comprehensive
results object containing:</p>
<ul><li><code>gene identifiers</code></li>
<li><code>gene symbols</code></li>
<li><code>functional descriptions</code></li>
<li><code>differential expression statistics</code></li>
</ul><p>This is the format most researchers expect when examining results or
importing them into downstream tools.</p>
<p>Define significance labels for plotting:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span></span>
<span>    label <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">coalesce</span><span class="op">(</span><span class="va">external_gene_name</span>, <span class="va">ensembl_gene_id_version</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&gt;=</span>  <span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"up"</span>,</span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&lt;=</span> <span class="op">-</span><span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"down"</span>,</span>
<span>      <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"ns"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="volcano-plot">Volcano plot<a class="anchor" aria-label="anchor" href="#volcano-plot"></a></h3>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span></span>
<span>  <span class="va">res_annot</span>,</span>
<span>  <span class="fu">aes</span><span class="op">(</span></span>
<span>    x     <span class="op">=</span> <span class="va">log2FoldChange</span>,</span>
<span>    y     <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>    col   <span class="op">=</span> <span class="va">sig</span>,</span>
<span>    label <span class="op">=</span> <span class="va">label</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"up"</span>   <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>    <span class="st">"down"</span> <span class="op">=</span> <span class="st">"dodgerblue3"</span>,</span>
<span>    <span class="st">"ns"</span>   <span class="op">=</span> <span class="st">"grey70"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text_repel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span><span class="va">res_annot</span>, <span class="va">sig</span> <span class="op">!=</span> <span class="st">"ns"</span><span class="op">)</span>,</span>
<span>    max.overlaps       <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    min.segment.length <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>    box.padding        <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    seed               <span class="op">=</span> <span class="fl">42</span>,</span>
<span>    show.legend        <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"grey40"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="va">log2fc_cut</span>, <span class="va">log2fc_cut</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"grey40"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10 adjusted p value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Volcano plot: WT_IR vs WT_mock"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/05_deseq/volcano-k.png" alt="Volcano plot of differential expression results" class="figure mx-auto d-block"><figcaption>
Volcano plot of differential expression results
</figcaption></figure><div id="interpreting-the-volcano-plot" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-the-volcano-plot" class="callout-inner">
<h3 class="callout-title">Interpreting the volcano plot</h3>
<div class="callout-content">
<ul><li>
<strong>X-axis</strong>: Direction and magnitude of change (positive
= higher in WT_IR, i.e., upregulated by radiation).</li>
<li>
<strong>Y-axis</strong>: Statistical significance (-log10 scale,
higher = more significant).</li>
<li>
<strong>Colored points</strong>: Genes passing both fold change and
significance thresholds.</li>
</ul></div>
</div>
</div>
<div id="exercise-compare-with-genome-based-results" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-compare-with-genome-based-results" class="callout-inner">
<h3 class="callout-title">Exercise: Compare with genome-based
results</h3>
<div class="callout-content">
<p>If you also ran Episode 05 (genome-based workflow):</p>
<ol style="list-style-type: decimal"><li>Are the number of DE genes similar between the two approaches?</li>
<li>Do the top DE genes overlap?</li>
<li>Are there systematic differences in fold change estimates?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Expected observations:</p>
<ol style="list-style-type: decimal"><li>The number of DE genes should be broadly similar, though not
identical.</li>
<li>Most top DE genes should overlap between methods.</li>
<li>Kallisto may produce slightly different LFC estimates due to its
different quantification algorithm.</li>
</ol><p>Both approaches are valid; consistency between them increases
confidence in the results.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-6-save-results">Step 6: Save results<a class="anchor" aria-label="anchor" href="#step-6-save-results"></a></h2>
<hr class="half-width"><p>Save the full results table:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>    <span class="va">res_df</span>,</span>
<span>    <span class="st">"results/deseq2_kallisto/DESeq2_kallisto_results.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Save significant genes only:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_res</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span></span>
<span>        <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>        <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">log2fc_cut</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>    <span class="va">sig_res</span>,</span>
<span>    <span class="st">"results/deseq2_kallisto/DESeq2_kallisto_sig.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Save the DESeq2 object for downstream analysis:</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"results/deseq2_kallisto/dds_kallisto.rds"</span><span class="op">)</span></span></code></pre>
</div>
<div id="genome-based-vs.-transcript-based-which-to-choose" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="genome-based-vs.-transcript-based-which-to-choose" class="callout-inner">
<h3 class="callout-title">Genome-based vs. transcript-based: Which to
choose?</h3>
<div class="callout-content">
<table class="table"><thead><tr class="header"><th>Aspect</th>
<th>Genome-based (Ep 05)</th>
<th>Transcript-based (Ep 05b)</th>
</tr></thead><tbody><tr class="odd"><td>Input</td>
<td>BAM files</td>
<td>FASTQ files</td>
</tr><tr class="even"><td>Speed</td>
<td>Slower (alignment + counting)</td>
<td>Faster (pseudo-alignment)</td>
</tr><tr class="odd"><td>Storage</td>
<td>Large (BAM files)</td>
<td>Small (abundance.tsv files)</td>
</tr><tr class="even"><td>Bias correction</td>
<td>Limited</td>
<td>Sequence + GC bias</td>
</tr><tr class="odd"><td>Novel transcripts</td>
<td>Can detect</td>
<td>Cannot detect</td>
</tr><tr class="even"><td>Visualization</td>
<td>IGV compatible</td>
<td>No BAM files</td>
</tr></tbody></table><p>For standard differential expression, both methods produce comparable
results. Choose based on your specific needs and available
resources.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Kallisto output is imported via <code>tximport</code> and loaded
with <code>DESeqDataSetFromTximport()</code>.</li>
<li>The tximport object preserves transcript length information for
accurate normalization.</li>
<li>Exploratory analysis (PCA, distance heatmaps) should precede
differential expression testing.</li>
<li>DESeq2 handles the statistical analysis identically to the
genome-based workflow.</li>
<li>LFC shrinkage improves fold change estimates for low-count
genes.</li>
<li>Results from transcript-based and genome-based workflows should be
broadly concordant.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-deseq2-expression-analyses.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="06-gene-set-enrichment-analyses.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-deseq2-expression-analyses.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Gene-level QC and
        </a>
        <a class="chapter-link float-end" href="06-gene-set-enrichment-analyses.html" rel="next">
          Next: Gene set enrichment...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05b-deseq2-expression-analyses-kallisto.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.18.0" class="external-link">sandpaper (0.18.0)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/05b-deseq2-expression-analyses-kallisto.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "B. Differential expression using DESeq2 (Kallisto pathway)",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/05b-deseq2-expression-analyses-kallisto.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/05b-deseq2-expression-analyses-kallisto.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2026-01-17",
  "datePublished": "2026-01-17"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

