<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: Reference</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/rcac/favicon-16x16.png"><link rel="manifest" href="favicons/rcac/site.webmanifest"><link rel="mask-icon" href="favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/reference.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: NA%" class="percentage">
    NA%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: NA%" aria-valuenow="NA" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/reference.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Salmon)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-gene-set-enrichment-analyses.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Reference</h1>
        <p>Last updated on 2025-12-02 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/learners/reference.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<p>Here are selected method-specific references covering the major
components of RNA-seq analysis, including alignment, quantification,
quality control, differential expression, functional interpretation, and
batch correction. These papers represent widely used tools and
foundational methods in the field.</p>
<div class="section level3">
<h3 id="general-rna-seq-references">General RNA-seq references<a class="anchor" aria-label="anchor" href="#general-rna-seq-references"></a></h3>
<ol style="list-style-type: decimal"><li>
<a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6096346/" class="external-link">A
beginner’s guide to analysis of RNA sequencing data</a> : K. M. Koch et
al. A practical overview of the key steps in a typical RNA-seq analysis:
library preparation, QC, alignment, counting, differential expression,
and pitfalls.</li>
<li>
<a href="https://www.illumina.com/techniques/sequencing/rna-sequencing.html" class="external-link">RNA
sequencing (RNA-seq) methods &amp; workflows</a> : Illumina A
vendor-agnostic guide describing different RNA-seq workflows, including
bulk mRNA-seq, total RNA-seq, strand specificity, and exploratory
considerations for experimental design.</li>
<li>
<a href="https://www.annualreviews.org/content/journals/10.1146/annurev-biodatasci-072018-021255" class="external-link">RNA
sequencing data: Hitchhiker’s guide to expression, interpretation and
beyond</a> : K. Van den Berge et al. A detailed review covering design,
quantification, differential expression, and emerging types of RNA-seq
(e.g., long-read, single-cell).</li>
</ol><hr></div>
<div class="section level3">
<h3 id="core-differential-expression-methods">Core differential expression methods<a class="anchor" aria-label="anchor" href="#core-differential-expression-methods"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>DESeq2</strong> Love MI, Huber W, Anders S. <em>Moderated
estimation of fold change and dispersion for RNA-seq data with
DESeq2.</em> Genome Biology, 2014. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8</a></p></li>
<li><p><strong>edgeR</strong> Robinson MD, McCarthy DJ, Smyth GK.
<em>edgeR: a Bioconductor package for differential expression analysis
of digital gene expression data.</em> Bioinformatics, 2010. <a href="https://academic.oup.com/bioinformatics/article/26/1/139/182458" class="external-link">https://academic.oup.com/bioinformatics/article/26/1/139/182458</a></p></li>
<li><p><strong>limma-voom</strong> Law CW, Chen Y, Shi W, Smyth GK.
<em>voom: precision weights unlock linear model analysis tools for
RNA-seq read counts.</em> Genome Biology, 2014. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29</a></p></li>
<li><p><strong>Sleuth (for Salmon/Kallisto quantification)</strong>
Pimentel H, Bray NL, Puente S, Melsted P, Pachter L. <em>Differential
analysis of RNA-seq incorporating quantification uncertainty.</em>
Nature Methods, 2017. <a href="https://www.nature.com/articles/nmeth.4324" class="external-link">https://www.nature.com/articles/nmeth.4324</a></p></li>
</ol><hr></div>
<div class="section level3">
<h3 id="alignment-methods">Alignment methods<a class="anchor" aria-label="anchor" href="#alignment-methods"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>STAR aligner</strong> Dobin A et al. <em>STAR: ultrafast
universal RNA-seq aligner.</em> Bioinformatics, 2013. <a href="https://academic.oup.com/bioinformatics/article/29/1/15/272537" class="external-link">https://academic.oup.com/bioinformatics/article/29/1/15/272537</a></p></li>
<li><p><strong>HISAT2</strong> Kim D, Langmead B, Salzberg SL.
<em>HISAT: a fast spliced aligner with low memory requirements.</em>
Nature Methods, 2015. <a href="https://www.nature.com/articles/nmeth.3317" class="external-link">https://www.nature.com/articles/nmeth.3317</a></p></li>
<li><p><strong>TopHat2 (historical, not recommended now but still
cited)</strong> Kim D et al. <em>TopHat2: accurate alignment of
transcriptomes in the presence of insertions, deletions and gene
fusions.</em> Genome Biology, 2013. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-4-r36" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-4-r36</a></p></li>
</ol><hr></div>
<div class="section level3">
<h3 id="quantification-methods">Quantification methods<a class="anchor" aria-label="anchor" href="#quantification-methods"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>Salmon (quasi-mapping and lightweight alignment)</strong>
Patro R, Duggal G, Love MI, Irizarry RA, Kingsford C. <em>Salmon
provides fast and bias-aware quantification of transcript
expression.</em> Nature Methods, 2017. <a href="https://www.nature.com/articles/nmeth.4197" class="external-link">https://www.nature.com/articles/nmeth.4197</a></p></li>
<li><p><strong>Kallisto (pseudo-alignment)</strong> Bray NL, Pimentel H,
Melsted P, Pachter L. <em>Near-optimal probabilistic RNA-seq
quantification.</em> Nature Biotechnology, 2016. <a href="https://www.nature.com/articles/nbt.3519" class="external-link">https://www.nature.com/articles/nbt.3519</a></p></li>
<li><p><strong>RSEM (alignment-based quantification)</strong> Li B,
Dewey CN. <em>RSEM: accurate transcript quantification from RNA-seq data
with or without a reference genome.</em> BMC Bioinformatics, 2011. <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323" class="external-link">https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323</a></p></li>
</ol><hr></div>
<div class="section level3">
<h3 id="quality-control-and-preprocessing">Quality control and preprocessing<a class="anchor" aria-label="anchor" href="#quality-control-and-preprocessing"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>FastQC</strong> Andrews S. <em>FastQC: A quality control
tool for high throughput sequence data.</em> Babraham Institute, 2010.
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p></li>
<li><p><strong>Trim Galore</strong> (cutadapt + FastQC wrapper) Krueger
F. <em>Trim Galore.</em> Babraham Institute. <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/" class="external-link">https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/</a></p></li>
<li><p><strong>cutadapt</strong> Martin M. <em>Cutadapt removes adapter
sequences from high-throughput sequencing reads.</em> EMBnet Journal,
2011. <a href="https://journal.embnet.org/index.php/embnetjournal/article/view/200" class="external-link">https://journal.embnet.org/index.php/embnetjournal/article/view/200</a></p></li>
<li><p><strong>MultiQC</strong> Ewels P, Magnusson M, Lundin S, Käller
M. <em>MultiQC: summarize analysis results for multiple tools and
samples in a single report.</em> Bioinformatics, 2016. <a href="https://academic.oup.com/bioinformatics/article/32/19/3047/2196507" class="external-link">https://academic.oup.com/bioinformatics/article/32/19/3047/2196507</a>
—</p></li>
</ol></div>
<div class="section level3">
<h3 id="functional-analysis-and-interpretation">Functional analysis and interpretation<a class="anchor" aria-label="anchor" href="#functional-analysis-and-interpretation"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>GOseq (for GO analysis accounting for transcript length
bias)</strong> Young MD, Wakefield MJ, Smyth GK, Oshlack A. <em>Gene
ontology analysis for RNA-seq: accounting for selection bias.</em>
Genome Biology, 2010. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14</a></p></li>
<li><p><strong>clusterProfiler (widely used for GO/KEGG)</strong> Yu G,
Wang LG, Han Y, He QY. <em>clusterProfiler: an R package for comparing
biological themes among gene clusters.</em> OMICS, 2012. <a href="https://www.liebertpub.com/doi/10.1089/omi.2011.0118" class="external-link">https://www.liebertpub.com/doi/10.1089/omi.2011.0118</a></p></li>
<li><p><strong>fgsea (fast GSEA)</strong> Korotkevich G, Sukhov V,
Sergushichev A. <em>Fast gene set enrichment analysis.</em> bioRxiv,
2016. <a href="https://www.biorxiv.org/content/10.1101/060012v3" class="external-link">https://www.biorxiv.org/content/10.1101/060012v3</a></p></li>
<li><p><strong>KEGG</strong> Kanehisa M, Goto S. <em>KEGG: Kyoto
Encyclopedia of Genes and Genomes.</em> Nucleic Acids Research, 2000. <a href="https://academic.oup.com/nar/article/28/1/27/2384398" class="external-link">https://academic.oup.com/nar/article/28/1/27/2384398</a></p></li>
</ol><hr></div>
<div class="section level3">
<h3 id="splicing-and-isoform-analysis-if-you-want-to-include-optional-advanced-topics">Splicing and isoform analysis (if you want to include optional
advanced topics)<a class="anchor" aria-label="anchor" href="#splicing-and-isoform-analysis-if-you-want-to-include-optional-advanced-topics"></a></h3>
<ol style="list-style-type: decimal"><li><p><strong>DEXSeq (differential exon usage)</strong> Anders S, Reyes
A, Huber W. <em>Detecting differential usage of exons from RNA-seq
data.</em> Genome Research, 2012. <a href="https://genome.cshlp.org/content/22/10/2008" class="external-link">https://genome.cshlp.org/content/22/10/2008</a></p></li>
<li><p><strong>SUPPA2 (isoform-level splicing changes)</strong> Trincado
JL et al. <em>SUPPA2 provides fast, accurate, and uncertainty-aware
differential splicing analysis.</em> Genome Biology, 2018. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1417-1" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1417-1</a></p></li>
<li><p><strong>StringTie2 (transcript assembly and
quantification)</strong> Kovaka S et al. <em>Transcriptome assembly from
long-read RNA-seq alignments with StringTie2.</em> Genome Biology, 2019.
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1910-1" class="external-link">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1910-1</a></p></li>
</ol><hr></div>
<div class="section level3">
<h3 id="batch-correction-and-confounder-modeling">Batch correction and confounder modeling<a class="anchor" aria-label="anchor" href="#batch-correction-and-confounder-modeling"></a></h3>
<ol style="list-style-type: decimal"><li>
<p><strong>ComBat</strong> Johnson WE, Li C, Rabinovic A.
<em>Adjusting batch effects in microarray expression data using
empirical Bayes methods.</em> Biostatistics, 2007. <a href="https://academic.oup.com/biostatistics/article/8/1/118/252073" class="external-link">https://academic.oup.com/biostatistics/article/8/1/118/252073</a></p>
<p>(Still widely used for RNA-seq after variance stabilizing
transformation.)</p>
</li>
<li><p><strong>RUVSeq</strong> Risso D, Ngai J, Speed TP, Dudoit S.
<em>Normalization of RNA-seq data using factor analysis of control genes
or samples.</em> Nature Biotechnology, 2014. <a href="https://www.nature.com/articles/nbt.2931" class="external-link">https://www.nature.com/articles/nbt.2931</a>
—</p></li>
</ol></div>
<section><h2 class="section-heading" id="glossary">Glossary<a class="anchor" aria-label="anchor" href="#glossary"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="experimental-concepts">Experimental concepts<a class="anchor" aria-label="anchor" href="#experimental-concepts"></a></h3>
<p><strong>adapter trimming</strong> Removal of sequencing adapters and
low-quality bases from raw reads before alignment, typically using tools
such as <code>fastp</code> or <code>Trim Galore</code>.</p>
<p><strong>alignment (read mapping)</strong> Assigning sequencing reads
to a reference genome or transcriptome using aligners like HISAT2 or
STAR.</p>
<p><strong>biological replicate</strong> Independent samples from the
same biological condition, used to estimate natural biological
variation.</p>
<p><strong>batch effects</strong> Confounding technical variation
introduced by differences in library prep, sequencing runs, or
operators; must be accounted for in experimental design and statistical
models.</p>
<p><strong>bulk RNA-seq</strong> Conventional RNA-seq performed on
pooled cells or tissues, as opposed to single-cell or spatial
RNA-seq.</p>
<p><strong>cDNA</strong> Complementary DNA synthesized from RNA
templates during library preparation.</p>
<p><strong>library preparation</strong> Experimental conversion of RNA
into a sequencing-ready library, including fragmentation, adapter
ligation, and amplification.</p>
<p><strong>paired-end sequencing</strong> Sequencing both ends of
DNA/cDNA fragments to improve mapping accuracy and detection of splice
junctions.</p>
<p><strong>read depth / sequencing depth</strong> The total number of
reads generated per sample; higher depth increases power to detect lowly
expressed or differentially expressed genes.</p>
<p><strong>stranded library / strand specificity</strong> Preservation
of the transcriptional strand of origin during library prep, allowing
direction-specific quantification.</p>
<hr></div>
<div class="section level3">
<h3 id="file-formats-and-data-organization">File formats and data organization<a class="anchor" aria-label="anchor" href="#file-formats-and-data-organization"></a></h3>
<p><strong>FASTQ</strong> A text format containing raw sequencing reads
with per-base quality scores.</p>
<p><strong>GTF / GFF</strong> Annotation file formats describing gene
and transcript coordinates, essential for read counting and feature
quantification.</p>
<p><strong>counts table</strong> A matrix of raw read counts (rows =
genes or transcripts, columns = samples) used for downstream
differential expression analysis.</p>
<hr></div>
<div class="section level3">
<h3 id="quantification-and-normalization">Quantification and normalization<a class="anchor" aria-label="anchor" href="#quantification-and-normalization"></a></h3>
<p><strong>featureCounts</strong> A program that summarizes aligned
reads to annotated genomic features, producing a counts table for DE
analysis.</p>
<p><strong>gene-level vs. transcript-level quantification</strong>
Gene-level quantification collapses isoforms into a single total count
per gene; transcript-level quantification measures individual isoform
expression, critical for splicing or isoform-specific analyses.</p>
<p><strong>normalization</strong> Adjusting for library size and
composition so samples are comparable. Common methods include TMM
(edgeR), median-ratio (DESeq2), and TPM for within-sample comparisons.
Use <strong>raw counts</strong> with DESeq2/edgeR for statistical
modeling—FPKM and TPM are not appropriate for DE testing.</p>
<p><strong>FPKM / TPM</strong> FPKM = Fragments Per Kilobase per
Million; TPM = Transcripts Per Million. TPM normalizes read counts
across transcript lengths and library size and is preferred for
<em>within-sample</em> comparisons, but <strong>not</strong> for
differential expression across samples.</p>
<p><strong>pseudo-alignment</strong> Lightweight transcript
quantification without full alignment, used by tools such as Salmon or
Kallisto.</p>
<p><strong>reference-based vs. de novo RNA-seq</strong> Reference-based
workflows align reads to a known genome or transcriptome. De novo
workflows assemble transcripts directly from reads when no reference
exists.</p>
<p><strong>transcriptome assembly</strong> Reconstruction of transcripts
from RNA-seq reads, either de novo or guided by a reference.</p>
<hr></div>
<div class="section level3">
<h3 id="differential-expression-and-statistics">Differential expression and statistics<a class="anchor" aria-label="anchor" href="#differential-expression-and-statistics"></a></h3>
<p><strong>differential expression (DE)</strong> Statistical
identification of genes or transcripts with significant changes in
expression between experimental conditions.</p>
<p><strong>dispersion</strong> A parameter in negative-binomial count
models representing variability beyond Poisson noise; estimated in
DESeq2 and edgeR.</p>
<p><strong>log₂ fold change (LFC)</strong> A measure of change in
expression between two conditions on a log₂ scale; values &gt; 1 or &lt;
–1 typically indicate meaningful differences.</p>
<p><strong>FDR / adjusted p-value</strong> The false discovery rate
controls for multiple testing (Benjamini–Hochberg correction). Genes are
typically considered significant at FDR &lt; 0.05.</p>
<p><strong>normalization factor</strong> Scaling coefficient applied
during model fitting to account for library size or composition bias
(internally computed by DESeq2/edgeR).</p>
<hr></div>
<div class="section level3">
<h3 id="interpretation-and-downstream-analysis">Interpretation and downstream analysis<a class="anchor" aria-label="anchor" href="#interpretation-and-downstream-analysis"></a></h3>
<p><strong>functional enrichment analysis</strong> Identification of
Gene Ontology (GO) terms, pathways, or functional categories
overrepresented among differentially expressed genes.</p>
<p><strong>gene ontology (GO)</strong> A controlled vocabulary
describing gene functions in three domains: biological process,
molecular function, and cellular component.</p>
<p><strong>KEGG</strong> The Kyoto Encyclopedia of Genes and Genomes;
maps genes to metabolic and signaling pathways.</p>
<p><strong>visualization</strong> Graphical summaries such as PCA plots,
volcano plots, or heatmaps used to assess overall patterns and highlight
DE genes.</p>
<p><strong>workflow reproducibility</strong> Ensuring analyses can be
replicated by others using automated, version-controlled workflows
(e.g., Snakemake, Nextflow, Apptainer).</p>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/learners/reference.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/reference.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "Reference",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/reference.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/reference.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2025-12-02",
  "datePublished": "2025-12-02"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

