<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: Quality control of RNA-seq reads</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png"><link rel="manifest" href="../favicons/rcac/site.webmanifest"><link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../03-qc-and-trimming.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 28%" class="percentage">
    28%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 28%" aria-valuenow="28" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../03-qc-and-trimming.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Quality control of RNA-seq reads
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#running-fastqc">Running FastQC</a></li>
<li><a href="#understanding-the-fastqc-report">Understanding the FastQC report</a></li>
<li><a href="#key-modules-explained">Key modules explained</a></li>
<li><a href="#aggregating-reports-with-multiqc">Aggregating reports with MultiQC</a></li>
<li><a href="#trimming-adapter-sequences">Trimming adapter sequences</a></li>
<li><a href="#exercise-evaluate-the-qc-results">Exercise: evaluate the QC results</a></li>
<li><a href="#summary">Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Salmon)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-gene-set-enrichment-analyses.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/02-downloading-and-organizing-files.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04a-genome-based-quantification.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/02-downloading-and-organizing-files.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Downloading and
        </a>
        <a class="chapter-link float-end" href="../instructor/04a-genome-based-quantification.html" rel="next">
          Next: A. Genome-based...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Quality control of RNA-seq reads</h1>
        <p>Last updated on 2025-11-20 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/03-qc-and-trimming.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do we check the quality of raw RNA-seq reads?</li>
<li>What information do FastQC and MultiQC provide?</li>
<li>How do we decide if trimming is required?</li>
<li>What QC issues are common in Illumina RNA-seq data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Run FastQC and MultiQC on RNA-seq FASTQ files.</li>
<li>Understand how to interpret the important FastQC modules.</li>
<li>Learn which FastQC warnings are expected for RNA-seq.</li>
<li>Understand why trimming is usually unnecessary for standard
RNA-seq.</li>
<li>Identify samples that may require closer inspection.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>Before mapping reads to a genome, it is important to evaluate the
quality of the raw RNA-seq data.<br>
Quality control helps reveal problems such as adapter contamination, low
quality cycles, and unexpected technical issues.</p>
<p>In this episode, we will run FastQC and MultiQC on the mouse heart
dataset and interpret the results.<br>
We will also explain when trimming is useful and when it can be harmful
for alignment based RNA-seq workflows.</p>
<p>FastQC evaluates individual FASTQ files. MultiQC summarizes all
FastQC reports into one view, which is helpful for comparing
replicates.</p>
<div id="why-do-we-perform-qc-on-fastq-files" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-do-we-perform-qc-on-fastq-files" class="callout-inner">
<h3 class="callout-title">Why do we perform QC on FASTQ files?</h3>
<div class="callout-content">
<p>Quality control answers several questions:</p>
<ul><li>Are sequencing qualities stable across cycles?<br></li>
<li>Are adapters present at high levels?<br></li>
<li>Do different replicates show similar patterns?<br></li>
<li>Are there outliers that indicate failed libraries?</li>
</ul><p>QC provides confidence that downstream steps will be reliable.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a></h2>
<hr class="half-width"><p>FastQC generates an HTML report for each FASTQ file. For paired-end
datasets, you can run it on both R1 and R2 together or separately.</p>
<p>We will create a directory for QC output and then run FastQC on all
FASTQ files.</p>
<div class="section level3">
<h3 id="run-fastqc-on-your-fastq-files">Run FastQC on your FASTQ files<a class="anchor" aria-label="anchor" href="#run-fastqc-on-your-fastq-files"></a></h3>
<p>Use the commands below to run FastQC on all reads in the
<code>data</code> directory.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/qc_fastq/</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">module</span> load fastqc</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="ex">fastqc</span> data/<span class="pp">*</span>.fastq.gz <span class="at">--outdir</span> results/qc_fastq/ <span class="at">--threads</span> 4</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">qc</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_1_fastqc.html</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_1_fastqc.zip</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_2_fastqc.html</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_2_fastqc.zip</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_1_fastqc.html</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_1_fastqc.zip</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_2_fastqc.html</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_2_fastqc.zip</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_1_fastqc.html</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_1_fastqc.zip</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_2_fastqc.html</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_2_fastqc.zip</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_1_fastqc.html</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_1_fastqc.zip</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_2_fastqc.html</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_2_fastqc.zip</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_1_fastqc.html</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_1_fastqc.zip</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_2_fastqc.html</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_2_fastqc.zip</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_1_fastqc.html</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_1_fastqc.zip</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_2_fastqc.html</span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="ex">└──</span> SRR33253290_2_fastqc.zip</span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="understanding-the-fastqc-report">Understanding the FastQC report<a class="anchor" aria-label="anchor" href="#understanding-the-fastqc-report"></a></h2>
<hr class="half-width"><p>FastQC produces several diagnostic modules. RNA-seq users often
misinterpret certain modules, so we focus on the ones that matter
most.</p>
<div id="which-fastqc-modules-are-important-for-rna-seq" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="which-fastqc-modules-are-important-for-rna-seq" class="callout-inner">
<h3 class="callout-title">Which FastQC modules are important for
RNA-seq?</h3>
<div class="callout-content">
<p>Most informative modules:</p>
<ul><li>Per base sequence quality</li>
<li>Per sequence quality scores</li>
<li>Per base sequence content</li>
<li>Per sequence GC content</li>
<li>Adapter content</li>
<li>Overrepresented sequences</li>
</ul><p>Modules that commonly show warnings for RNA-seq, even when libraries
are fine:</p>
<ul><li>Per sequence GC content (RNA-seq reflects transcript GC bias, not
whole genome)</li>
<li>K-mer content (poly A tails and priming artifacts often trigger
warnings)</li>
</ul><p>Warnings do not always indicate problems. Context and comparison
across samples matter more than individual icons.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="key-modules-explained">Key modules explained<a class="anchor" aria-label="anchor" href="#key-modules-explained"></a></h2>
<hr class="half-width"><p>Below are guidelines for interpreting the most useful modules.</p>
<div class="section level3">
<h3 id="per-base-sequence-quality">Per base sequence quality<a class="anchor" aria-label="anchor" href="#per-base-sequence-quality"></a></h3>
<p>Shows the distribution of quality scores at each position across all
reads. You want high and stable quality across the read. A small quality
drop at the end of R2 is common and usually not a problem.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_sequence_quality_histogram.png" alt="Per base sequence quality for a typical RNA-seq dataset" class="figure mx-auto d-block"><figcaption>
Per base sequence quality for a typical RNA-seq dataset
</figcaption></figure></div>
<div class="section level3">
<h3 id="per-tile-sequence-quality">Per tile sequence quality<a class="anchor" aria-label="anchor" href="#per-tile-sequence-quality"></a></h3>
<p>Reports whether specific regions of the flowcell produced lower
quality reads. Uneven tiles can indicate instrument issues or bubble
formation, but this module rarely shows problems in modern NovaSeq
data.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_sequence_content.png" alt="Per tile sequence quality plot" class="figure mx-auto d-block"><figcaption>
Per tile sequence quality plot
</figcaption></figure></div>
<div class="section level3">
<h3 id="per-sequence-quality-scores">Per sequence quality scores<a class="anchor" aria-label="anchor" href="#per-sequence-quality-scores"></a></h3>
<p>Shows how many reads have high or low overall quality. A good dataset
has most reads with high scores and very few reads with low quality.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_sequence_quality_score.png" alt="Distribution of per sequence quality scores" class="figure mx-auto d-block"><figcaption>
Distribution of per sequence quality scores
</figcaption></figure></div>
<div class="section level3">
<h3 id="per-base-sequence-content">Per base sequence content<a class="anchor" aria-label="anchor" href="#per-base-sequence-content"></a></h3>
<p>Shows the proportion of A, C, G, and T at each cycle. For genomic
resequencing, all four lines should be close to 25 percent. RNA-seq
usually shows imbalanced composition in early cycles due to biased
priming and transcript composition. Mild imbalance at the start of the
read is normal and does not require trimming.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_atgc_content.png" alt="Per base sequence content for RNA-seq reads" class="figure mx-auto d-block"><figcaption>
Per base sequence content for RNA-seq reads
</figcaption></figure></div>
<div class="section level3">
<h3 id="per-sequence-gc-content">Per sequence GC content<a class="anchor" aria-label="anchor" href="#per-sequence-gc-content"></a></h3>
<p>Shows the GC distribution across all reads. RNA-seq often fails this
test because expressed transcripts are not GC neutral. Only pronounced
multi-modal or extremely shifted distributions are concerning, for
example when contamination from another organism is suspected.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_sequence_GC_content.png" alt="Per sequence GC content distribution" class="figure mx-auto d-block"><figcaption>
Per sequence GC content distribution
</figcaption></figure></div>
<div class="section level3">
<h3 id="per-base-n-content">Per base N content<a class="anchor" aria-label="anchor" href="#per-base-n-content"></a></h3>
<p>Reports the percentage of unidentified bases (N) at each position.
This should be near zero for modern Illumina runs. High N content
indicates failed cycles or poor sequencing chemistry.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_n_content.png" alt="Per base N content plot" class="figure mx-auto d-block"><figcaption>
Per base N content plot
</figcaption></figure></div>
<div class="section level3">
<h3 id="sequence-length-distribution">Sequence length distribution<a class="anchor" aria-label="anchor" href="#sequence-length-distribution"></a></h3>
<p>Shows the distribution of read lengths. Standard RNA-seq libraries
should have a narrow peak at the expected read length. Variable lengths
usually appear after trimming or for specialized protocols such as small
RNA.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_length_distribution.png" alt="Sequence length distribution for RNA-seq data" class="figure mx-auto d-block"><figcaption>
Sequence length distribution for RNA-seq data
</figcaption></figure></div>
<div class="section level3">
<h3 id="sequence-duplication-levels">Sequence duplication levels<a class="anchor" aria-label="anchor" href="#sequence-duplication-levels"></a></h3>
<p>Reports how many reads are exact duplicates. Some duplication is
normal in RNA-seq because highly expressed genes produce many identical
fragments. Very high duplication, especially combined with low library
size, can indicate low library complexity or overamplification.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_sequence_duplication_level.png" alt="Sequence duplication levels" class="figure mx-auto d-block"><figcaption>
Sequence duplication levels
</figcaption></figure></div>
<div class="section level3">
<h3 id="overrepresented-sequences">Overrepresented sequences<a class="anchor" aria-label="anchor" href="#overrepresented-sequences"></a></h3>
<p>Lists specific sequences that occur more often than expected. Common
sources include ribosomal RNA fragments, poly A tails, adapter sequence,
or reads from highly expressed transcripts. Some overrepresented
sequences are expected, but very high levels may indicate contamination
or poor rRNA depletion.</p>
</div>
<div class="section level3">
<h3 id="adapter-content">Adapter content<a class="anchor" aria-label="anchor" href="#adapter-content"></a></h3>
<p>Shows whether any read cycles contain adapter sequences. Small
amounts at the extreme ends are common and usually harmless.
Consistently high adapter content across many cycles suggests short
inserts or poorly prepared libraries.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_adapter_content.png" alt="Adapter content across the read" class="figure mx-auto d-block"><figcaption>
Adapter content across the read
</figcaption></figure></div>
</section><section><h2 class="section-heading" id="aggregating-reports-with-multiqc">Aggregating reports with MultiQC<a class="anchor" aria-label="anchor" href="#aggregating-reports-with-multiqc"></a></h2>
<hr class="half-width"><p>FastQC generates a separate HTML file per sample. MultiQC
consolidates all reports and allows easy comparison of replicates.</p>
<div class="section level3">
<h3 id="run-multiqc-on-the-fastqc-outputs">Run MultiQC on the FastQC outputs<a class="anchor" aria-label="anchor" href="#run-multiqc-on-the-fastqc-outputs"></a></h3>
<p>Generate a single combined report using MultiQC.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> rnaseq-workshop</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="ex">multiqc</span> results/qc_fastq/ <span class="at">-o</span> results/qc_fastq/</span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/02_qc/multiqc-summary.png" alt="Example MultiQC summary across all samples" class="figure mx-auto d-block"><figcaption>
Example MultiQC summary across all samples
</figcaption></figure></div>
</section><section><h2 class="section-heading" id="trimming-adapter-sequences">Trimming adapter sequences<a class="anchor" aria-label="anchor" href="#trimming-adapter-sequences"></a></h2>
<hr class="half-width"><p>Before deciding whether to trim reads, it is essential to understand
what adapters are and why they appear in sequencing data.</p>
<p>Adapters are short synthetic DNA sequences ligated to both ends of
each fragment during library preparation. They provide primer binding
sites needed for cluster amplification and sequencing. Ideally, the
sequenced fragment is long enough that the machine reads only biological
DNA. However, when the insert is shorter than the read length, the
sequencer runs into the adapter sequence, which produces adapter
contamination at the ends of reads.</p>
<p>Many users assume that adapters must always be trimmed, but this is
not true for typical RNA-seq workflows.</p>
<div id="do-we-need-to-trim-reads-for-standard-rna-seq-workflows" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="do-we-need-to-trim-reads-for-standard-rna-seq-workflows" class="callout-inner">
<h3 class="callout-title">Do we need to trim reads for standard RNA-seq
workflows?</h3>
<div class="callout-content">
<p>For alignment based RNA-seq, adapter trimming is usually not
required. Modern aligners such as HISAT2 and STAR detect and soft clip
adapter sequence automatically. Soft-clipped bases do not contribute to
mismatches or mapping penalties, so adapters generally do not harm
alignment quality.</p>
<p>In contrast, unnecessary trimming can introduce new problems:</p>
<ul><li>shorter reads and lower mapping power</li>
<li>uneven trimming across samples and artificial differences</li>
<li>reduction in read complexity</li>
<li>broken read pairing for paired-end sequencing</li>
<li>more parameters to track for reproducibility</li>
</ul><p>Trimming is recommended only in specific situations:</p>
<ul><li>very short inserts, for example degraded RNA or ancient DNA</li>
<li>adapter sequence dominating read tails, for example more than 10-15
percent adapter content</li>
<li>transcript assembly workflows such as StringTie or Scallop, which
benefit from uniform read lengths</li>
<li>small RNA or miRNA libraries, where inserts are intentionally very
short and adapter presence is guaranteed</li>
</ul></div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-evaluate-the-qc-results">Exercise: evaluate the QC results<a class="anchor" aria-label="anchor" href="#exercise-evaluate-the-qc-results"></a></h2>
<hr class="half-width"><div id="inspect-your-multiqc-report" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="inspect-your-multiqc-report" class="callout-inner">
<h3 class="callout-title">Inspect your MultiQC report</h3>
<div class="callout-content">
<p>Using the MultiQC HTML report, answer:</p>
<ol style="list-style-type: decimal"><li>Do all replicates show similar per base sequence quality
profiles?</li>
<li>Is adapter content high enough to impact analysis?</li>
<li>Is any sample behaving differently from the rest, for example in
number of reads, quality, duplication, or GC percentage?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Typical interpretation for this dataset (students should compare with
their own):</p>
<ul><li>All replicates show high, stable sequence quality.</li>
<li>Adapter signal appears only at the extreme ends of reads for
selected samples and is less than 4 percent.</li>
<li>No samples show concerning patterns or biases.</li>
<li>Trimming is therefore unnecessary for this dataset.</li>
</ul></div>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What if trimming is required? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p><em>Optional reference: how to trim adapters with <code>fastp</code>
(not part of the hands-on)</em></p>
<p>Sometimes you may encounter datasets where trimming is appropriate,
for example short inserts, small RNA, or visibly high adapter content in
QC. Although we will not trim reads in this workshop, here is a minimal
example you can use in your own projects if trimming becomes
necessary.</p>
<p><code>fastp</code> is a fast, widely used tool that can automatically
detect adapters, trim them, filter low quality reads, and generate QC
reports.</p>
<div class="section level3">
<h3 id="example-trimming-paired-end-reads-with-fastp">Example: trimming paired-end reads with <code>fastp</code>
<a class="anchor" aria-label="anchor" href="#example-trimming-paired-end-reads-with-fastp"></a></h3>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">module</span> load fastp</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">-i</span> SRRXXXXXXX_1.fastq.gz <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">-I</span> SRRXXXXXXX_2.fastq.gz <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">-o</span> SRRXXXXXXX_1.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">-O</span> SRRXXXXXXX_2.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>  <span class="at">--thread</span> 8 <span class="dt">\</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>  <span class="at">--html</span> fastp_report.html <span class="dt">\</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  <span class="at">--json</span> fastp_report.json</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="what-this-command-does">What this command does<a class="anchor" aria-label="anchor" href="#what-this-command-does"></a></h3>
<ul><li>automatically detects adapters
(<code>--detect_adapter_for_pe</code>)</li>
<li>trims low quality bases (default behaviour)</li>
<li>keeps paired-end reads synchronized</li>
<li>writes trimmed output (<code>*.trimmed.fastq.gz</code>)</li>
<li>generates QC reports (HTML and JSON)</li>
</ul></div>
<div class="section level3">
<h3 id="when-to-use-this">When to use this<a class="anchor" aria-label="anchor" href="#when-to-use-this"></a></h3>
<p>Use trimming only if:</p>
<ul><li>QC shows strong adapter presence</li>
<li>your library prep is known to produce short inserts</li>
<li>you are working with small RNA or degraded RNA</li>
<li>you are doing transcript assembly where uniform read length
matters</li>
</ul><p>For standard RNA-seq, trimming is usually unnecessary.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>FastQC and MultiQC provide essential diagnostics for RNA-seq
data.</li>
<li>Some FastQC warnings are normal for RNA-seq and do not indicate
problems.</li>
<li>Aligners soft clip adapters, so trimming is usually unnecessary for
alignment based RNA-seq.</li>
<li>QC helps detect outliers before alignment and quantification.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/02-downloading-and-organizing-files.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04a-genome-based-quantification.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/02-downloading-and-organizing-files.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Downloading and
        </a>
        <a class="chapter-link float-end" href="../instructor/04a-genome-based-quantification.html" rel="next">
          Next: A. Genome-based...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/03-qc-and-trimming.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/03-qc-and-trimming.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "Quality control of RNA-seq reads",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/03-qc-and-trimming.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/03-qc-and-trimming.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2025-11-20",
  "datePublished": "2025-11-20"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

