<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: A. Genome-based quantification (STAR + featureCounts)</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png"><link rel="manifest" href="../favicons/rcac/site.webmanifest"><link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../04a-genome-based-quantification.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 30%" class="percentage">
    30%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 30%" aria-valuenow="30" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../04a-genome-based-quantification.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. A. Genome-based quantification (STAR + featureCounts)
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#step-1-checking-rna-seq-library-strandness">Step 1: Checking RNA-seq library strandness</a></li>
<li><a href="#step-2-preparing-the-reference-genome-for-alignment">Step 2: Preparing the reference genome for alignment</a></li>
<li><a href="#step-3-mapping-reads-with-star">Step 3: Mapping reads with STAR</a></li>
<li><a href="#step-4-assessing-alignment-quality">Step 4: Assessing alignment quality</a></li>
<li><a href="#step-5-quantifying-gene-counts-with-featurecounts">Step 5: Quantifying gene counts with featureCounts</a></li>
<li><a href="#step-6-qc-on-counts">Step 6: QC on counts</a></li>
<li><a href="#summary">Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Kallisto)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05b-deseq2-expression-analyses-kallisto.html">7. B. Differential expression using DESeq2 (Kallisto pathway)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-gene-set-enrichment-analyses.html">8. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/03-qc-and-trimming.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04b-transcript-based-quantification.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/03-qc-and-trimming.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Quality control of
        </a>
        <a class="chapter-link float-end" href="../instructor/04b-transcript-based-quantification.html" rel="next">
          Next: B. Transcript-based...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>A. Genome-based quantification (STAR + featureCounts)</h1>
        <p>Last updated on 2026-01-17 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/04a-genome-based-quantification.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 70 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do we map RNA-seq reads to a reference genome?</li>
<li>How do we determine library strandness before alignment?</li>
<li>How do we quantify reads per gene using featureCounts?</li>
<li>How do we submit mapping jobs on RCAC clusters with SLURM?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Check RNA-seq library strandness using Salmon.</li>
<li>Index a reference genome for STAR.</li>
<li>Map reads to the genome using STAR.</li>
<li>Quantify gene level counts with featureCounts.</li>
<li>Interpret basic alignment and counting statistics.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>In this episode we move from raw FASTQ files to genome based gene
counts.<br>
This workflow has three major steps:</p>
<ol style="list-style-type: decimal"><li>Determine library strandness: STAR does not infer strandness
automatically. <code>featureCounts</code> requires the correct
setting.</li>
<li>Map reads to the genome using STAR: STAR is a splice aware aligner
designed for RNA-seq data.</li>
<li>Count mapped reads per gene using <code>featureCounts</code>
</li>
</ol><p>This represents one of the two standard quantification strategies in
RNA-seq.<br>
The alternative method (B. transcript based quantification with Salmon
or Kallisto) will be covered in Episode 4B.</p>
</section><section><h2 class="section-heading" id="step-1-checking-rna-seq-library-strandness">Step 1: Checking RNA-seq library strandness<a class="anchor" aria-label="anchor" href="#step-1-checking-rna-seq-library-strandness"></a></h2>
<hr class="half-width"><div id="why-strandness-matters" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-strandness-matters" class="callout-inner">
<h3 class="callout-title">Why strandness matters</h3>
<div class="callout-content">
<p>RNA-seq libraries can be:</p>
<ul><li>unstranded<br></li>
<li>stranded forward<br></li>
<li>stranded reverse</li>
</ul><p>Stranded protocols are common in modern kits, but old GEO and SRA
datasets are often unstranded.<br>
featureCounts must be given the correct strand setting. If the wrong
setting is used, most reads will be counted against incorrect genes.</p>
<p>STAR does not infer strandness automatically, so we must determine it
<strong>before</strong> alignment.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="detecting-strandness-with-salmon">Detecting strandness with Salmon<a class="anchor" aria-label="anchor" href="#detecting-strandness-with-salmon"></a></h3>
<p>Salmon can infer strandness directly from raw FASTQ files without any
alignment.<br>
We run Salmon with <code>-l A</code>, which tells it to explore all
library types.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-q</span> standby <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/strand_check</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">module</span> load salmon</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="ex">salmon</span> index <span class="at">--transcripts</span> data/gencode.vM38.transcripts-clean.fa <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>             <span class="at">--index</span> data/salmon_index <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>             <span class="at">--threads</span> 4</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="ex">salmon</span> quant <span class="at">--index</span> data/salmon_index <span class="dt">\</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>             <span class="at">--libType</span> A <span class="dt">\</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>             <span class="at">--mates1</span> data/WT_Bcell_mock_rep1_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>             <span class="at">--mates2</span> data/WT_Bcell_mock_rep1_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>             <span class="at">--output</span> results/strand_check <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>             <span class="at">--threads</span> 4</span></code></pre>
</div>
<p>The important output is in
<code>results/strand_check/lib_format_counts.json</code>:</p>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What does this file look like? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">JSON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode json" tabindex="0"><code class="sourceCode json"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>    <span class="dt">"read_files"</span><span class="fu">:</span> <span class="st">"[ data/WT_Bcell_mock_rep1_R1.fastq.gz, data/WT_Bcell_mock_rep1_R2.fastq.gz]"</span><span class="fu">,</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>    <span class="dt">"expected_format"</span><span class="fu">:</span> <span class="st">"IU"</span><span class="fu">,</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="dt">"compatible_fragment_ratio"</span><span class="fu">:</span> <span class="fl">1.0</span><span class="fu">,</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>    <span class="dt">"num_compatible_fragments"</span><span class="fu">:</span> <span class="dv">11994681</span><span class="fu">,</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>    <span class="dt">"num_assigned_fragments"</span><span class="fu">:</span> <span class="dv">11994681</span><span class="fu">,</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>    <span class="dt">"num_frags_with_concordant_consistent_mappings"</span><span class="fu">:</span> <span class="dv">10472924</span><span class="fu">,</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>    <span class="dt">"num_frags_with_inconsistent_or_orphan_mappings"</span><span class="fu">:</span> <span class="dv">2280227</span><span class="fu">,</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>    <span class="dt">"strand_mapping_bias"</span><span class="fu">:</span> <span class="dv">0</span><span class="er">.</span><span class="dv">5082214861866657</span><span class="fu">,</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>    <span class="dt">"MSF"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>    <span class="dt">"OSF"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>    <span class="dt">"ISF"</span><span class="fu">:</span> <span class="dv">5322565</span><span class="fu">,</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>    <span class="dt">"MSR"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>    <span class="dt">"OSR"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>    <span class="dt">"ISR"</span><span class="fu">:</span> <span class="dv">5150359</span><span class="fu">,</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>    <span class="dt">"SF"</span><span class="fu">:</span> <span class="dv">1149290</span><span class="fu">,</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>    <span class="dt">"SR"</span><span class="fu">:</span> <span class="dv">1130937</span><span class="fu">,</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>    <span class="dt">"MU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>    <span class="dt">"OU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>    <span class="dt">"IU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>    <span class="dt">"U"</span><span class="fu">:</span> <span class="dv">0</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="fu">}</span></span></code></pre>
</div>
<blockquote>
<h3 id="interpreting-lib_format_counts.json">Interpreting
<code>lib_format_counts.json</code>
</h3>
<p>This JSON file reports Salmon’s automatic library type detection.</p>
<ul><li>
<strong><code>"expected_format": "IU"</code></strong>: This is
Salmon’s conclusion. “I” means <strong>Inward</strong> (correct for
paired-end reads) and “U” means <strong>Unstranded</strong>.</li>
<li>
<strong><code>"strand_mapping_bias": 0.508</code></strong>: This is
the key evidence. A value near 0.5 (50%) indicates that reads mapped
equally to both the sense and antisense strands, the definitive sign of
an <strong>unstranded</strong> library.</li>
<li>
<strong><code>"ISF"</code></strong> and
<strong><code>"ISR"</code></strong>: These are the counts for
Inward-Sense-Forward (36.4M) and Inward-Sense-Reverse (35.2M) fragments.
Because these values are almost equal, they confirm the ~50/50 split
seen in the strand bias.</li>
</ul></blockquote>
<p>Common results:</p>
<table class="table"><thead><tr class="header"><th>Code</th>
<th>Meaning</th>
</tr></thead><tbody><tr class="odd"><td>IU</td>
<td>unstranded</td>
</tr><tr class="even"><td>ISR</td>
<td>stranded reverse (paired end)</td>
</tr><tr class="odd"><td>ISF</td>
<td>stranded forward (paired end)</td>
</tr><tr class="even"><td>SR</td>
<td>stranded reverse (single end)</td>
</tr><tr class="odd"><td>SF</td>
<td>stranded forward (single end)</td>
</tr></tbody></table><p><strong>Conclusion:</strong> The data is <strong>unstranded</strong>.
We will record this and supply it to STAR/featureCounts.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-2-preparing-the-reference-genome-for-alignment">Step 2: Preparing the reference genome for alignment<a class="anchor" aria-label="anchor" href="#step-2-preparing-the-reference-genome-for-alignment"></a></h2>
<hr class="half-width"><p>STAR requires a genome index. Indexing builds a searchable data
structure that speeds up alignment.</p>
<p>Required input:</p>
<ul><li>genome FASTA file</li>
<li>annotation file (GTF or GFF3)</li>
<li>output directory for the index</li>
</ul><p>Optionally, annotation improves splice junction detection and yields
better alignments.</p>
<div id="why-star" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-star" class="callout-inner">
<h3 class="callout-title">Why STAR?</h3>
<div class="callout-content">
<ul><li>Splice-aware<br></li>
<li>Extremely fast<br></li>
<li>Produces high-quality alignments<br></li>
<li>Well supported and widely used<br></li>
<li>Compatible with downstream tools such as featureCounts, StringTie,
and RSEM</li>
</ul><p>Optional alternatives include HISAT2 and Bowtie2, but STAR remains
the standard for high-depth Illumina RNA-seq.</p>
</div>
</div>
</div>
<div id="exercise-read-the-star-manual" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-read-the-star-manual" class="callout-inner">
<h3 class="callout-title">Exercise: Read the STAR manual</h3>
<div class="callout-content">
<p>Check the <strong>“Generating genome indexes”</strong> section of
the<br><a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" class="external-link">STAR
manual</a></p>
<p>Which options are required for indexing?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Key options:</p>
<ul><li>
<code>--runMode genomeGenerate</code><br></li>
<li>
<code>--genomeDir</code> directory to store the index<br></li>
<li>
<code>--genomeFastaFiles</code> the FASTA file<br></li>
<li>
<code>--sjdbGTFfile</code> annotation (recommended)<br></li>
<li>
<code>--runThreadN</code> number of threads<br></li>
<li>
<code>--sjdbOverhang</code> read length minus one (100 is a safe
default)</li>
</ul></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="indexing-the-genome-with-star">Indexing the Genome with STAR<a class="anchor" aria-label="anchor" href="#indexing-the-genome-with-star"></a></h3>
<p>Below is a minimal SLURM job script for building the STAR genome
index. Create a file named <code>index_genome.sh</code> in
<code>$SCRATCH/rcac_rnaseq/scripts</code> with this content:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=20</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=star_index</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="ex">module</span> load star</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="va">data_dir</span><span class="op">=</span><span class="st">"</span><span class="va">${SCRATCH}</span><span class="st">/rnaseq-workshop/data"</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/GRCm39.primary_assembly.genome.fa"</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/gencode.vM38.primary_assembly.basic.annotation.gtf"</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="va">indexdir</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/star_index"</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>    <span class="at">--runMode</span> genomeGenerate <span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>    <span class="at">--runThreadN</span> <span class="va">${SLURM_CPUS_ON_NODE}</span> <span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>    <span class="at">--genomeDir</span> <span class="va">${indexdir}</span> <span class="dt">\</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>    <span class="at">--genomeFastaFiles</span> <span class="va">${genome}</span> <span class="dt">\</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>    <span class="at">--sjdbGTFfile</span> <span class="va">${gtf}</span></span></code></pre>
</div>
<p>Submit the job:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">sbatch</span> index_genome.sh</span></code></pre>
</div>
<p>Indexing requires about 30 to 45 minutes.</p>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What will the output look like? </h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" aria-labelledby="headingSpoiler2" data-bs-parent="#accordionSpoiler2">
<div class="accordion-body">
<p>The index directory is now ready for use in mapping and should have
contents like this: Location: <code>$SCRATCH/rcac_rnaseq/data</code>
Contents of <code>star_index/</code>:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">star_index/</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">├──</span> chrLength.txt</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">├──</span> chrNameLength.txt</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">├──</span> chrName.txt</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="ex">├──</span> chrStart.txt</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="ex">├──</span> exonGeTrInfo.tab</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="ex">├──</span> exonInfo.tab</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="ex">├──</span> geneInfo.tab</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="ex">├──</span> Genome</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="ex">├──</span> genomeParameters.txt</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="ex">├──</span> Log.out</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="ex">├──</span> SA</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="ex">├──</span> SAindex</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="ex">├──</span> sjdbInfo.txt</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="ex">├──</span> sjdbList.fromGTF.out.tab</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="ex">├──</span> sjdbList.out.tab</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="ex">└──</span> transcriptInfo.tab</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="do-i-need-the-gtf-for-indexing" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="do-i-need-the-gtf-for-indexing" class="callout-inner">
<h3 class="callout-title">Do I need the GTF for indexing?</h3>
<div class="callout-content">
<p>No, but it is <em>strongly recommended</em>.<br>
Including it improves splice junction detection and increases mapping
accuracy.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-3-mapping-reads-with-star">Step 3: Mapping reads with STAR<a class="anchor" aria-label="anchor" href="#step-3-mapping-reads-with-star"></a></h2>
<hr class="half-width"><p>Once the genome index is ready, we can align reads.</p>
<p>For aligning reads, important STAR options are:</p>
<ul><li><code>--runThreadN</code></li>
<li><code>--genomeDir</code></li>
<li><code>--readFilesIn</code></li>
<li>
<code>--readFilesCommand zcat</code> for gzipped FASTQ files</li>
<li><code>--outSAMtype BAM SortedByCoordinate</code></li>
<li><code>--outSAMunmapped Within</code></li>
</ul><p>STAR’s defaults are tuned for mammalian genomes. If working with
plants, fungi, or repetitive genomes, additional tuning may be
required.</p>
<div id="exercise-write-a-star-alignment-command" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-write-a-star-alignment-command" class="callout-inner">
<h3 class="callout-title">Exercise: Write a STAR alignment command</h3>
<div class="callout-content">
<p>Write a basic STAR alignment command using your genome index and
paired FASTQ files.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>    <span class="at">--runThreadN</span> 12 <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>    <span class="at">--genomeDir</span> data/star_index <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>    <span class="at">--readFilesIn</span> data/SRR1234567_1.fastq.gz data/SRR1234567_2.fastq.gz <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> results/mapping/SRR1234567 <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    <span class="at">--outSAMunmapped</span> Within</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="mapping-many-fastq-files-with-a-slurm-array">Mapping many FASTQ files with a SLURM array<a class="anchor" aria-label="anchor" href="#mapping-many-fastq-files-with-a-slurm-array"></a></h3>
<p>When having multiple samples, array jobs simplify submission.</p>
<p>First, we will create <code>samples.txt</code> file with just the
sample names</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/data</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>_R1.fastq.gz <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_R1.fastq.gz//'</span> <span class="op">&gt;</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop/scripts/samples.txt</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=20</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#SBATCH --time=8:00:00</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=read_mapping</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#SBATCH --array=1-8</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="ex">module</span> load star</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="va">FASTQ_DIR</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data"</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="va">GENOME_INDEX</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data/star_index"</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="va">OUTDIR</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/mapping"</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUTDIR</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span><span class="va">$(</span><span class="fu">sed</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${SLURM_ARRAY_TASK_ID}</span><span class="st">p"</span> samples.txt<span class="va">)</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="va">R1</span><span class="op">=</span><span class="va">${FASTQ_DIR}</span>/<span class="va">${SAMPLE}</span>_R1.fastq.gz</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="va">R2</span><span class="op">=</span><span class="va">${FASTQ_DIR}</span>/<span class="va">${SAMPLE}</span>_R2.fastq.gz</span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a>    <span class="at">--runThreadN</span> <span class="va">${SLURM_CPUS_ON_NODE}</span> <span class="dt">\</span></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a>    <span class="at">--genomeDir</span> <span class="va">$GENOME_INDEX</span> <span class="dt">\</span></span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a>    <span class="at">--readFilesIn</span> <span class="va">$R1</span> <span class="va">$R2</span> <span class="dt">\</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> <span class="va">$OUTDIR</span>/<span class="va">$SAMPLE</span> <span class="dt">\</span></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a>    <span class="at">--outSAMunmapped</span> Within <span class="dt">\</span></span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span></code></pre>
</div>
<p>Submit:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">sbatch</span> map_reads.sh</span></code></pre>
</div>
<div id="why-use-an-array-job" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="why-use-an-array-job" class="callout-inner">
<h3 class="callout-title">Why use an array job?</h3>
<div class="callout-content">
<p>Think about this:</p>
<ul><li>We have <em>eight</em> samples, each with two FASTQ files</li>
<li>Having eight separate slurm jobs means:
<ul><li>Copy-pasting the same command eight times</li>
<li>Higher chance of typos or mistakes</li>
</ul></li>
<li>if mapping with single slurm file, sequential execution =
slower</li>
<li>Samples can run independently</li>
</ul><p>Why is a job array a good choice?</p>
<p>Arrays allow us to run identical commands across many inputs:</p>
<ul><li>All samples are processed the same way<br></li>
<li>Runs in parallel<br></li>
<li>Simplifies submission<br></li>
<li>Reduces copy-paste errors<br></li>
<li>Minimizes runtime cost on the cluster</li>
</ul></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-assessing-alignment-quality">Step 4: Assessing alignment quality<a class="anchor" aria-label="anchor" href="#step-4-assessing-alignment-quality"></a></h2>
<hr class="half-width"><p>STAR produces several useful files, particularly:</p>
<ul><li><p><code>Log.final.out</code><br>
Contains mapping statistics, including % uniquely mapped.</p></li>
<li><p><code>Aligned.sortedByCoord.out.bam</code><br>
Ready for downstream quantification.</p></li>
</ul><p>To summarize alignment statistics across samples, we use MultiQC:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="ex">multiqc</span> results/mapping <span class="at">-o</span> results/qc_alignment</span></code></pre>
</div>
<div class="section level3">
<h3 id="key-metrics-to-review">Key metrics to review<a class="anchor" aria-label="anchor" href="#key-metrics-to-review"></a></h3>
<ul><li>
<strong>Uniquely mapped reads</strong> — typically 60–90% is
good<br></li>
<li>
<strong>Multi-mapped reads</strong> — depends on genome, but &gt;20%
is concerning<br></li>
<li>
<strong>Unmapped reads</strong> — check reasons (mismatches? too
short?)<br></li>
<li>
<strong>Mismatch rate</strong> — high values indicate quality or
contamination issues</li>
</ul><div id="exercise-examine-your-alignment-metrics" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-examine-your-alignment-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Examine your alignment metrics</h3>
<div class="callout-content">
<p>Using your MultiQC alignment report:</p>
<ol style="list-style-type: decimal"><li>Which sample has the highest uniquely mapped percentage?<br></li>
<li>Are any samples clear outliers?<br></li>
<li>Does mapped/unmapped reads appear consistent across samples?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Example interpretation for this dataset:</p>
<ul><li>All samples have ~80% uniquely mapped reads.</li>
<li>No sample is an outlier.</li>
<li>mapped/unmapped reads consistent across samples</li>
</ul></div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-quantifying-gene-counts-with-featurecounts">Step 5: Quantifying gene counts with featureCounts<a class="anchor" aria-label="anchor" href="#step-5-quantifying-gene-counts-with-featurecounts"></a></h2>
<hr class="half-width"><p>Once the reads have been aligned and sorted, we can quantify how many
read pairs map to each gene. This gives us gene-level counts that we
will later use for differential expression. For alignment-based RNA-seq
workflows, <code>featureCounts</code> is the recommended tool because it
is fast, robust, and compatible with most gene annotation formats.</p>
<p><code>featureCounts</code> uses two inputs:</p>
<ol style="list-style-type: decimal"><li>A set of aligned BAM files (sorted by coordinate)</li>
<li>A GTF annotation file describing gene models</li>
</ol><p>It assigns each read (or read pair) to a gene based on the exon
boundaries.</p>
<div id="important-stranded-or-unstranded" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="important-stranded-or-unstranded" class="callout-inner">
<h3 class="callout-title">Important: stranded or unstranded?</h3>
<div class="callout-content">
<p><code>featureCounts</code> needs to know whether your dataset is
stranded.</p>
<ul><li>
<code>-s 0</code> unstranded</li>
<li>
<code>-s 1</code> stranded forward</li>
<li>
<code>-s 2</code> stranded reverse</li>
</ul><p>From our Salmon strandness check, this dataset behaves as
<strong>unstranded</strong>, so we will use <code>-s 0</code>.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="running-featurecounts">Running featureCounts<a class="anchor" aria-label="anchor" href="#running-featurecounts"></a></h3>
<p>We will create a new directory for count output:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/scripts</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/results/counts</span></code></pre>
</div>
<p>Below is an example SLURM script to count all BAM files at once.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=16</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=featurecounts</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="ex">module</span> load subread</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="va">GTF</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data/gencode.vM38.primary_assembly.basic.annotation.gtf"</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="va">BAMS</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/mapping/*.bam"</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="va">output_dir</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/counts"</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="dt">\</span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>  <span class="at">-T</span> <span class="va">${SLURM_CPUS_PER_TASK}</span> <span class="dt">\</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>  <span class="at">-p</span> <span class="dt">\</span></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a>  <span class="at">-s</span> 0 <span class="dt">\</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>  <span class="at">-t</span> exon <span class="dt">\</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a>  <span class="at">-g</span> gene_id <span class="dt">\</span></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>  <span class="at">-a</span> <span class="va">${GTF}</span> <span class="dt">\</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>  <span class="at">-o</span> <span class="va">${output_dir}</span>/gene_counts.txt <span class="dt">\</span></span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>  <span class="va">${BAMS}</span></span></code></pre>
</div>
<p>Submit the job:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">sbatch</span> count_features.sh</span></code></pre>
</div>
<p><code>gene_counts.txt</code> will contain:</p>
<ul><li>one row per gene</li>
<li>one column per sample</li>
</ul><div id="generate-a-clean-count-matrix" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="generate-a-clean-count-matrix" class="callout-inner">
<h3 class="callout-title">Generate a clean count matrix</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/results/counts</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1,7- gene_counts.txt <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  <span class="fu">sed</span> <span class="st">'s+/scratch/negishi/'</span><span class="va">$USER</span><span class="st">'/rnaseq-workshop/results/mapping/++g'</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="fu">grep</span> <span class="at">-v</span> <span class="st">"^#"</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>  <span class="fu">sed</span> <span class="st">'s/Aligned.sortedByCoord.out.bam//g'</span> <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>  <span class="op">&gt;</span> gene_counts_clean.txt</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-6-qc-on-counts">Step 6: QC on counts<a class="anchor" aria-label="anchor" href="#step-6-qc-on-counts"></a></h2>
<hr class="half-width"><p>Before proceeding to differential expression, it is good practice to
perform some QC on the count data. We can again use MultiQC to summarize
featureCounts statistics:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/qc_counts</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="ex">multiqc</span> results/counts <span class="at">-o</span> results/qc_counts</span></code></pre>
</div>
<div id="exercise-examine-your-counting-metrics" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-examine-your-counting-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Examine your counting metrics</h3>
<div class="callout-content">
<p>Using your MultiQC featureCounts report:</p>
<ol style="list-style-type: decimal"><li>Which sample has the lowest percent assigned, and what are common
reasons for lower assignment?</li>
<li>Which unassigned category is the largest, and what does it
indicate?</li>
<li>Does the sample with the most assigned reads also have the highest
percent assigned?</li>
<li>What does “Unassigned: No Features” mean, and why might it
occur?</li>
<li>Are multi-mapped reads a concern for RNA-seq differential
expression?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>Example interpretation for this dataset:</p>
<ol style="list-style-type: decimal"><li>The sample with the lowest assignment rate may vary. Causes include
low complexity, incomplete annotation, or more intronic reads.</li>
<li>Unmapped or No Features dominate. These reflect reads that do not
align or do not overlap annotated exons.</li>
<li>No. Total assigned reads depend on depth, while percent assigned
reflects library quality.</li>
<li>Reads mapped outside annotated exons, often from intronic regions,
unannotated transcripts, or incomplete annotation.</li>
<li>Multi-mapping is expected for paralogs and repeats. It is usually
fine as long as the rate is consistent across samples.</li>
</ol></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>STAR is a fast splice-aware aligner widely used for RNA-seq.</li>
<li>Genome indexing requires FASTA and optionally GTF for improved
splice detection.</li>
<li>Mapping is efficiently performed using SLURM array jobs.</li>
<li>Alignment statistics (from <code>Log.final.out</code> + MultiQC)
must be reviewed.</li>
<li>Strandness should be inferred using aligned BAM files.</li>
<li>Final BAM files are ready for downstream counting.</li>
<li>
<code>featureCounts</code> is used to obtain gene-level counts for
differential expression.</li>
<li>Exons are counted and summed per gene.</li>
<li>The output is a simple matrix for downstream statistical
analysis.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/03-qc-and-trimming.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04b-transcript-based-quantification.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/03-qc-and-trimming.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Quality control of
        </a>
        <a class="chapter-link float-end" href="../instructor/04b-transcript-based-quantification.html" rel="next">
          Next: B. Transcript-based...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/04a-genome-based-quantification.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.18.0" class="external-link">sandpaper (0.18.0)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/04a-genome-based-quantification.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "A. Genome-based quantification (STAR + featureCounts)",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/04a-genome-based-quantification.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/04a-genome-based-quantification.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2026-01-17",
  "datePublished": "2026-01-17"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

