<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: Gene-level QC and differential expression (DESeq2)</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png"><link rel="manifest" href="../favicons/rcac/site.webmanifest"><link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../05-deseq2-expression-analyses.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 55%" class="percentage">
    55%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 55%" aria-valuenow="55" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../05-deseq2-expression-analyses.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Salmon)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Gene-level QC and differential expression (DESeq2)
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#step-1-start-open-ondemand-r-session-and-prepare-data">Step 1: Start Open OnDemand R session and prepare data</a></li>
<li><a href="#exploratory-analysis">Exploratory analysis</a></li>
<li><a href="#differential-gene-expression-analysis">Differential gene expression analysis</a></li>
<li><a href="#joining-results-with-expression-and-annotation">Joining results with expression and annotation</a></li>
<li><a href="#visualization-volcano-plot">Visualization: volcano plot</a></li>
<li><a href="#saving-results">Saving results</a></li>
<li><a href="#summary">Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05b-deseq2-expression-analyses-kallisto.html">7. Differential expression using DESeq2</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-gene-set-enrichment-analyses.html">8. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/04b-transcript-based-quantification.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/05b-deseq2-expression-analyses-kallisto.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/04b-transcript-based-quantification.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: B. Transcript-based
        </a>
        <a class="chapter-link float-end" href="../instructor/05b-deseq2-expression-analyses-kallisto.html" rel="next">
          Next: Differential...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Gene-level QC and differential expression (DESeq2)</h1>
        <p>Last updated on 2025-12-30 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05-deseq2-expression-analyses.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 85 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do we explore RNA seq count data before running DESeq2.</li>
<li>How do we restrict analysis to protein coding genes.</li>
<li>How do we perform differential expression with DESeq2.</li>
<li>How do we visualize sample relationships and DE genes.</li>
<li>How do we save a useful DE results table with annotation and
expression values.</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Import count and sample metadata into R.</li>
<li>Attach biotype and gene symbol annotation to Ensembl gene IDs.</li>
<li>Explore library sizes, variance stabilizing transforms, distances,
and PCA.</li>
<li>Run DESeq2 for a two group comparison.</li>
<li>Create a volcano plot with gene symbols.</li>
<li>Export joined DE results including normalized counts and
annotation.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>In this episode we move from <strong>count matrices</strong> to
<strong>differential expression analysis</strong> and
<strong>visualization</strong>.</p>
<p>The workflow has four parts:</p>
<ol style="list-style-type: decimal"><li>Import counts and sample metadata.</li>
<li>Exploratory analysis on protein coding genes (quality checks).</li>
<li>Differential expression analysis with DESeq2.</li>
<li>Visualization and export of annotated DE results.</li>
</ol><div id="what-you-need-for-this-episode" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-you-need-for-this-episode" class="callout-inner">
<h3 class="callout-title">What you need for this episode</h3>
<div class="callout-content">
<ul><li>
<code>gene_counts_clean.txt</code> generated from
featureCounts<br></li>
<li>a <code>samples.csv</code> file describing the experimental
groups<br></li>
<li>RStudio session on Scholar using Open OnDemand</li>
</ul><p>While we created the count matrix in the previous episode, we still
need to create a sample metadata file. This file should contain at least
two columns: sample names matching the count matrix column names, and
the experimental condition (e.g., control vs treatment). Simply
copy/paste the following into a text file and save it in your
<code>scripts</code> directory as <code>samples.csv</code>:</p>
<pre><code>sample,condition
SRR33253285,Nat10-CKO
SRR33253286,Nat10-CKO
SRR33253287,Nat10-CKO
SRR33253288,Nat10floxflox
SRR33253289,Nat10flox/flox
SRR33253290,Nat10flox/flox</code></pre>
<p>Also, create a direcotry for DESeq2 results:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/deseq2</span></code></pre>
</div>
</div>
</div>
</div>
<p>All analyses are performed in R through Open OnDemand</p>
</section><section><h2 class="section-heading" id="step-1-start-open-ondemand-r-session-and-prepare-data">Step 1: Start Open OnDemand R session and prepare data<a class="anchor" aria-label="anchor" href="#step-1-start-open-ondemand-r-session-and-prepare-data"></a></h2>
<hr class="half-width"><p>We will be using the OOD to start an interactive session on Scholar:
<a href="https://gateway.negishi.rcac.purdue.edu/" class="external-link">https://gateway.negishi.rcac.purdue.edu</a></p>
<ol style="list-style-type: decimal"><li>Login using your Purdue credentials after clicking the above
link</li>
<li>Click on “Interactive Apps” in the top menu, and select “RStudio
(Bioconductor)”</li>
<li>Fill the job submission form as follows:
<ul><li>queue: <code>workshop</code>
</li>
<li>Walltime: <code>4</code>
</li>
<li>Number of cores: <code>4</code>
</li>
</ul></li>
<li>Click “Launch” and wait for the RStudio session to start</li>
<li>Once the session starts, you’ll will be able to click on “Connect to
RStudio server” which will open the RStudio interface.</li>
</ol><figure style="text-align: center"><img src="../fig/05_deseq/open-on-demand.png" alt="Open OnDemand interface" class="figure mx-auto d-block"><figcaption>
Open OnDemand interface
</figcaption></figure><p>Once in RStudio, load the necessary libraries:</p>
<div class="section level3">
<h3 id="setup-load-packages-and-data">Setup: load packages and data<a class="anchor" aria-label="anchor" href="#setup-load-packages-and-data"></a></h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Mmusculus.v79</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ensembldb</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggrepel</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">vsn</span><span class="op">)</span></span>
<span><span class="co"># Modify the path as needed</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="st">"/scratch/negishi/aseethar/rnaseq-workshop"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">countsFile</span> <span class="op">&lt;-</span> <span class="st">"results/counts/gene_counts_clean.txt"</span></span>
<span><span class="co"># prepare this file first!</span></span>
<span><span class="va">groupFile</span>  <span class="op">&lt;-</span> <span class="st">"scripts/samples.csv"</span></span>
<span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="va">groupFile</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cts</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">read.delim</span><span class="op">(</span><span class="va">countsFile</span>, row.names <span class="op">=</span> <span class="fl">1</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>We also load gene level annotation tables prepared earlier.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/mart.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/annot.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> How were these data prepared? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<p>Since the gene IDs in the count matrix are Ensembl IDs, it will be
hard to interpret the results without annotation. We will need to use
the <code>org.Mm.eg.db</code> package for attaching gene symbols, so we
can interpret the results better.0</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># select Mart</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>ensembl <span class="ot">=</span> <span class="fu">useMart</span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># we need to use the correct dataset for mouse</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co"># we will query the available datasets and filter for mouse (GRCm39)</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="fu">listDatasets</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(description, <span class="st">"GRCm39"</span>))</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># we can use the dataset name to set the dataset</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>ensembl <span class="ot">=</span> <span class="fu">useDataset</span>(<span class="st">"mmusculus_gene_ensembl"</span>, <span class="at">mart =</span> ensembl)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># our counts have Ensembl gene IDs with version numbers</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co"># lets find how it is referred in biomaRt</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="fu">listFilters</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(name, <span class="st">"ensembl"</span>))</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co"># so we will now set the filter type accordingly</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>filterType <span class="ot">&lt;-</span> <span class="st">"ensembl_gene_id_version"</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co"># from our counts data, get the list of gene IDs</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>filterValues <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co"># take a look at the available attributes (first 20)</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="fu">listAttributes</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>     <span class="fu">head</span>(<span class="dv">20</span>)</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co"># we will retrieve ensembl gene id, ensembl gene id with version and external gene name</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>attributeNames <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'ensembl_gene_id'</span>,</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>                    <span class="st">'ensembl_gene_id_version'</span>,</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>                    <span class="st">'external_gene_name'</span>,</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>                    <span class="st">'gene_biotype'</span>,</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>                    <span class="st">'description'</span>)</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="co"># get the annotation</span></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>annot <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>  <span class="at">attributes =</span> attributeNames,</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>  <span class="at">filters =</span> filterType,</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>  <span class="at">values =</span> filterValues,</span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>  <span class="at">mart =</span> ensembl</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>)</span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">saveRDS</span>(annot, <span class="at">file =</span> <span class="st">"results/counts/gene_annotation.rds"</span>)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co"># also save as tsv for future use</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a><span class="fu">write.table</span>(</span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a>  annot,</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"data/mart.tsv"</span>,</span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>  <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="cn">FALSE</span></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a>)</span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="co"># save a simplified annotation file with only essential columns</span></span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a>mart <span class="sc">%&gt;%</span> </span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a><span class="fu">select</span>(ensembl_gene_id, ensembl_gene_id_version, external_gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a>  <span class="fu">write.table</span>(</span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a>    <span class="at">file =</span> <span class="st">"data/annot.tsv"</span>,</span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a>    <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">FALSE</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The objects loaded here form the foundation for the entire workflow.
<code>cts</code> contains raw gene counts from featureCounts.
<code>coldata</code> contains the metadata that describes each sample
and defines the comparison groups. <code>mart</code> and
<code>annot</code> contain functional information for each gene so we
can attach gene symbols and biotypes later in the analysis.</p>
<div id="sample-metadata-and-design" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="sample-metadata-and-design" class="callout-inner">
<h3 class="callout-title">Sample metadata and design</h3>
<div class="callout-content">
<p><code>coldata</code> must have row names that match the count matrix
column names. The <code>condition</code> column defines the groups for
DESeq2 (<code>Nat10-CKO</code> vs <code>Nat10flox/flox</code>).</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exploratory-analysis">Exploratory analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="attach-annotation-and-filter-to-protein-coding-genes">Attach annotation and filter to protein coding genes<a class="anchor" aria-label="anchor" href="#attach-annotation-and-filter-to-protein-coding-genes"></a></h3>
<p>We first join counts with biotype information from <code>mart</code>
and filter to protein coding genes.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cts_tbl</span> <span class="op">&lt;-</span> <span class="va">cts</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cts_annot</span> <span class="op">&lt;-</span> <span class="va">cts_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">mart</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="fu">is.na</span><span class="op">(</span><span class="va">cts_annot</span><span class="op">$</span><span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Joining annotation allows us to move beyond raw Ensembl IDs. This
step attaches gene symbols, biotypes, and descriptions so that later
plots and tables are interpretable. At this point the dataset is still
large and includes many noncoding genes, pseudogenes, and biotypes we
won’t analyze further.</p>
<p>Summarize gene biotypes:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biotype_df</span> <span class="op">&lt;-</span> <span class="va">cts_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">count</span><span class="op">(</span><span class="va">gene_biotype</span>, name <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">biotype_df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu">reorder</span><span class="op">(</span><span class="va">gene_biotype</span>, <span class="va">n</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_col</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Gene biotype"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Number of genes"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Distribution of gene biotypes in annotation"</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/gene-biotypes.png" alt="Distribution of gene biotypes in annotation" class="figure mx-auto d-block"><figcaption>
Distribution of gene biotypes in annotation
</figcaption></figure><p>Visualizing biotype composition gives a sense of how many genes
belong to each functional class. RNA-seq quantifies all transcribed
loci, but many biotypes (snRNA, TEC, pseudogenes) are not suitable for
differential expression with DESeq2 because they tend to be lowly
expressed or unstable. This motivates filtering to protein-coding genes
for QC and visualization.</p>
<p>Filter to protein coding genes and lowly expressed genes:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cts_coding</span> <span class="op">&lt;-</span> <span class="va">cts_annot</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span>, <span class="fu">all_of</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">cts_coding</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 15839     6</code></pre>
<div id="why-filter-protein-coding-genes-and-low-expression" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-filter-protein-coding-genes-and-low-expression" class="callout-inner">
<h3 class="callout-title">Why filter protein coding genes and low
expression</h3>
<div class="callout-content">
<p>Noncoding biotypes (lncRNA, pseudogene, etc.) can be biologically
important. Filtering lowly expressed genes reduces noise in exploratory
analyses. Genes with near-zero counts across all samples inflate
variance estimates and obscure real biological structure. The threshold
of 5 total counts is a conservative filter that keeps genes with minimal
but genuine expression while removing background noise.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="library-size-differences">Library size differences<a class="anchor" aria-label="anchor" href="#library-size-differences"></a></h3>
<p>We summarize total counts per sample for protein coding genes.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">libSize</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="va">cts_coding</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rename</span><span class="op">(</span>total_counts <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>total_millions <span class="op">=</span> <span class="va">total_counts</span> <span class="op">/</span> <span class="fl">1e6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">libSize</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">total_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_col</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Total assigned reads"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Total reads per sample"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/libsize-difference.png" alt="Total assigned reads per sample" class="figure mx-auto d-block"><figcaption>
Total assigned reads per sample
</figcaption></figure><p>Library size differences affect normalization. Although DESeq2
corrects for this through size factors, inspecting raw library sizes
ensures no sample has abnormally low sequencing depth, which could
compromise downstream analyses.</p>
</div>
<div class="section level3">
<h3 id="create-deseq2-object-for-exploratory-plots">Create DESeq2 object for exploratory plots<a class="anchor" aria-label="anchor" href="#create-deseq2-object-for-exploratory-plots"></a></h3>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>  countData <span class="op">=</span> <span class="va">cts_coding</span>,</span>
<span>  colData   <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>  design    <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<p>Inspect size factors vs library size:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span></span>
<span>  <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    libSize    <span class="op">=</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    sizeFactor <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,</span>
<span>    Group      <span class="op">=</span> <span class="va">dds</span><span class="op">$</span><span class="va">condition</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">libSize</span>, y <span class="op">=</span> <span class="va">sizeFactor</span>, col <span class="op">=</span> <span class="va">Group</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size"</span>, y <span class="op">=</span> <span class="st">"Size factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/size-factors-vs-lib-size.png" alt="Size factors vs library size" class="figure mx-auto d-block"><figcaption>
Size factors vs library size
</figcaption></figure></div>
<div class="section level3">
<h3 id="variance-stabilization-and-mean-sd-plots">Variance stabilization and mean SD plots<a class="anchor" aria-label="anchor" href="#variance-stabilization-and-mean-sd-plots"></a></h3>
<p>Raw count data have a strong mean–variance dependency: low count
genes have high variance, and high count genes show lower relative
variability. This makes Euclidean distance and PCA misleading on raw
counts. VST removes this dependency so that downstream clustering is
biologically meaningful rather than driven by count scale.</p>
<p>Inspect mean vs standard deviation plot before transformation:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/before-vsd.png" alt="average counts vs variance" class="figure mx-auto d-block"><figcaption>
average counts vs variance
</figcaption></figure><p>Apply variance stabilizing transform (VST):</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/after-vsd.png" alt="average counts vs variance after transformation" class="figure mx-auto d-block"><figcaption>
average counts vs variance after transformation
</figcaption></figure><div id="why-use-vst" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-vst" class="callout-inner">
<h3 class="callout-title">Why use VST</h3>
<div class="callout-content">
<p>Counts have mean dependent variance. Variance stabilizing transform
(VST) makes the variance roughly constant across the range, which
improves distance based methods and PCA.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sample-to-sample-distances">Sample to sample distances<a class="anchor" aria-label="anchor" href="#sample-to-sample-distances"></a></h3>
<p>The distance heatmap helps answer a key question: Do samples group by
biological condition or by an unwanted variable (batch, sequencing run,
contamination)? This is one of the fastest ways to detect outliers.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleDists</span>       <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sampleDistMatrix</span>  <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">sampleDists</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">rev</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Blues"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span></span>
<span>  <span class="va">sampleDistMatrix</span>,</span>
<span>  clustering_distance_rows <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>  clustering_distance_cols <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>  col                      <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/clustering.png" alt="Eucledean distance heatmap" class="figure mx-auto d-block"><figcaption>
Eucledean distance heatmap
</figcaption></figure></div>
<div class="section level3">
<h3 id="pca-on-top-variable-genes">PCA on top variable genes<a class="anchor" aria-label="anchor" href="#pca-on-top-variable-genes"></a></h3>
<p>PCA projects samples into low-dimensional space and reveals dominant
sources of variation. Ideally, samples separate by biological condition
along one of the top PCs. If not, the signal from the condition may be
weak or confounded by other factors.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">select</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">rv</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fu">seq_len</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="fl">500</span>, <span class="fu">length</span><span class="op">(</span><span class="va">rv</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">select</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="va">pca</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">intgroup</span>    <span class="op">&lt;-</span> <span class="st">"condition"</span></span>
<span><span class="va">intgroup.df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span>, <span class="va">intgroup</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">intgroup</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">factor</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">intgroup.df</span>, <span class="fl">1</span>, <span class="va">paste</span>, collapse <span class="op">=</span> <span class="st">" : "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>  PC1 <span class="op">=</span> <span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  PC2 <span class="op">=</span> <span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  <span class="va">intgroup.df</span>,</span>
<span>  name <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">d</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_shape_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"Nat10-CKO"</span>       <span class="op">=</span> <span class="st">"#00b462"</span>,</span>
<span>    <span class="st">"Nat10flox/flox"</span>  <span class="op">=</span> <span class="st">"#980c80"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, stroke <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text_repel</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"PC1"</span>, <span class="fu">round</span><span class="op">(</span><span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"PC2"</span>, <span class="fu">round</span><span class="op">(</span><span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/pca-plot.png" alt="PCA plot showing sample clustering" class="figure mx-auto d-block"><figcaption>
PCA plot showing sample clustering
</figcaption></figure><div id="exercise-interpret-exploratory-plots" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-interpret-exploratory-plots" class="callout-inner">
<h3 class="callout-title">Exercise: interpret exploratory plots</h3>
<div class="callout-content">
<p>Using the library size, distance heatmap, and PCA:</p>
<ol style="list-style-type: decimal"><li>Are there any obvious outlier samples.</li>
<li>Do replicates from the same condition cluster together.</li>
<li>Does any sample look inconsistent with its metadata label.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Example interpretation:</p>
<ol style="list-style-type: decimal"><li>Library sizes are comparable across samples and fall within a narrow
range.</li>
<li>Samples cluster by condition in both distance heatmaps and PCA.</li>
<li>No sample appears as a clear outlier relative to its group.</li>
</ol></div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-gene-expression-analysis">Differential gene expression analysis<a class="anchor" aria-label="anchor" href="#differential-gene-expression-analysis"></a></h2>
<hr class="half-width"><p>For DE testing we use the <strong>full count matrix</strong>
(<code>cts</code>) without filtering by biotype. This preserves all
features counted by featureCounts.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>  countData <span class="op">=</span> <span class="va">cts_coding</span>,</span>
<span>  colData   <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>  design    <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>You can either run the individual DESeq2 steps:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateDispersions</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">nbinomWaldTest</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Single command to run the DESeq2 pipeline </h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" aria-labelledby="headingSpoiler2" data-bs-parent="#accordionSpoiler2">
<div class="accordion-body">
<p>Above three steps can be run together using the <code>DESeq</code>
function:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code>estimating size factors
  Note: levels of factors in the design contain characters other than
  letters, numbers, '_' and '.'. It is recommended (but not required) to use
  only letters, numbers, and delimiters '_' or '.', as these are safe characters
  for column names in R. [This is a message, not a warning or an error]
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
  Note: levels of factors in the design contain characters other than
  letters, numbers, '_' and '.'. It is recommended (but not required) to use
  only letters, numbers, and delimiters '_' or '.', as these are safe characters
  for column names in R. [This is a message, not a warning or an error]
final dispersion estimates
fitting model and testing</code></pre>
</div>
</div>
</div>
</div>
<p>Inspect dispersion estimates:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/dispersion-plot.png" alt="Dispersion estimates from DESeq2" class="figure mx-auto d-block"><figcaption>
Dispersion estimates from DESeq2
</figcaption></figure><p>Obtain results for the contrast of interest
(<code>Nat10flox/flox</code> vs <code>Nat10-CKO</code>):</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">results</span><span class="op">(</span></span>
<span>    <span class="va">dds</span>,</span>
<span>    contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>      <span class="st">"condition"</span>,</span>
<span>      <span class="st">"Nat10flox/flox"</span>,</span>
<span>      <span class="st">"Nat10-CKO"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">TXT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode txt" tabindex="0"><code class="sourceCode default"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>out of 15839 with nonzero total read count</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>adjusted p-value &lt; 0.1</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>LFC &gt; 0 (up)       : 3195, 20%</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>LFC &lt; 0 (down)     : 3398, 21%</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>outliers [1]       : 362, 2.3%</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>low counts [2]     : 308, 1.9%</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>(mean count &lt; 1)</span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>[1] see 'cooksCutoff' argument of ?results</span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>[2] see 'independentFiltering' argument of ?results</span></code></pre>
</div>
<p>Order the results by p value to see the top DE genes:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code>log2 fold change (MLE): condition Nat10flox/flox vs Nat10-CKO
Wald test p-value: condition Nat10flox.flox vs Nat10.CKO
DataFrame with 6 rows and 6 columns
       baseMean log2FoldChange     lfcSE      stat       pvalue         padj
      &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;
18512   4645.33       -5.17453 0.1335387  -38.7493  0.00000e+00  0.00000e+00
940     5041.68       -2.88015 0.1039939  -27.6953 7.94364e-169 6.02485e-165
13789  14362.27       -2.40229 0.0951377  -25.2506 1.11514e-140 5.63852e-137
16183  42766.77       -4.68585 0.1923780  -24.3575 4.82857e-131 1.83111e-127
18278   7055.06       -3.48409 0.1535305  -22.6932 5.23170e-114 1.58719e-110
270     1469.83       -2.87951 0.1298316  -22.1788 5.50606e-109 1.39202e-105</code></pre>
<p>Each row represents one gene. Key columns: <code>baseMean</code>:
average normalized count across all samples <code>log2FoldChange</code>:
direction and magnitude of change between groups <code>pvalue</code> /
<code>padj</code>: statistical significance after multiple testing
correction Genes with <code>NA</code> values often have too few reads to
estimate dispersion and should not be interpreted.</p>
<p>Create a summary table:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="va">summary_table</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span></span>
<span>    total_genes <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    up  <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span>  <span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    down<span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">summary_table</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code># A tibble: 1 × 4
  total_genes   sig    up  down
        &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1       15839  5658  2180  2363</code></pre>
<p>Keep gene IDs:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">res_df</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="joining-results-with-expression-and-annotation">Joining results with expression and annotation<a class="anchor" aria-label="anchor" href="#joining-results-with-expression-and-annotation"></a></h2>
<hr class="half-width"><p>Get normalized counts:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_counts</span>    <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">norm_counts_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">norm_counts</span><span class="op">)</span></span>
<span><span class="va">norm_counts_df</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">norm_counts_df</span><span class="op">)</span></span></code></pre>
</div>
<p>Join DE results, normalized counts, and annotation:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">norm_counts_df</span>,</span>
<span>                   by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">mart</span>,</span>
<span>                   by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span></code></pre>
</div>
<p>Joining normalized counts and annotation makes the output
biologically interpretable. The final table becomes a comprehensive
results object containing: - <code>gene identifiers</code> -
<code>gene symbols</code> - <code>functional descriptions</code> -
<code>differential expression statistics</code> -
<code>normalized expression values</code> This is the format most
researchers expect when examining results or importing them into
downstream tools.</p>
<p>Define significance labels for plotting:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span></span>
<span>    label <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">coalesce</span><span class="op">(</span><span class="va">external_gene_name</span>, <span class="va">ensembl_gene_id_version</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&gt;=</span>  <span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"up"</span>,</span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&lt;=</span> <span class="op">-</span><span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"down"</span>,</span>
<span>      <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"ns"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="visualization-volcano-plot">Visualization: volcano plot<a class="anchor" aria-label="anchor" href="#visualization-volcano-plot"></a></h2>
<hr class="half-width"><p>The volcano plot summarizes statistical significance
(<code>padj</code>) and effect size (<code>log2FoldChange</code>)
simultaneously. Genes far left or right show strong expression
differences. Genes high on the plot pass stringent significance
thresholds. Labels help identify the most biologically meaningful
genes.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span></span>
<span>  <span class="va">res_annot</span>,</span>
<span>  <span class="fu">aes</span><span class="op">(</span></span>
<span>    x     <span class="op">=</span> <span class="va">log2FoldChange</span>,</span>
<span>    y     <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>    col   <span class="op">=</span> <span class="va">sig</span>,</span>
<span>    label <span class="op">=</span> <span class="va">label</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"up"</span>   <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>    <span class="st">"down"</span> <span class="op">=</span> <span class="st">"dodgerblue3"</span>,</span>
<span>    <span class="st">"ns"</span>   <span class="op">=</span> <span class="st">"grey70"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text_repel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span><span class="va">res_annot</span>, <span class="va">sig</span> <span class="op">!=</span> <span class="st">"ns"</span><span class="op">)</span>,</span>
<span>    max.overlaps       <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    min.segment.length <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>    box.padding        <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    seed               <span class="op">=</span> <span class="fl">42</span>,</span>
<span>    show.legend        <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10 adjusted p value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Volcano plot: Nat10-CKO vs Nat10flox/flox"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/volcano-plot.png" alt="Volcano plot of differential expression results" class="figure mx-auto d-block"><figcaption>
Volcano plot of differential expression results
</figcaption></figure><div id="interpreting-the-volcano" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-the-volcano" class="callout-inner">
<h3 class="callout-title">Interpreting the volcano</h3>
<div class="callout-content">
<p>Points to the right are higher in <code>Nat10flox/flox</code>, to the
left are higher in <code>Nat10-CKO</code>. Points near the top have
stronger statistical support (small adjusted p values). Colored points
mark genes that pass both fold change and padj thresholds.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="saving-results">Saving results<a class="anchor" aria-label="anchor" href="#saving-results"></a></h2>
<hr class="half-width"><p>Saving both the full table and the significant subset allows
flexibility: - The full table is needed for pathway analyses and
reproducibility. - The subset table focuses on interpretable and
biologically meaningful DE genes.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">readr</span><span class="fu">::</span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>  <span class="va">res_annot</span>,</span>
<span>  <span class="st">"results/deseq2/DESeq2_results_joined.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Filter to significant genes (padj ≤ 0.05 and fold change ≥ 1.5):</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_res</span> <span class="op">&lt;-</span> <span class="va">res_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span></span>
<span>    <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>    <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">log2fc_cut</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>  <span class="va">sig_res</span>,</span>
<span>  <span class="st">"results/deseq2/DESeq2_results_sig.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="exercise-explore-top-de-genes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-explore-top-de-genes" class="callout-inner">
<h3 class="callout-title">Exercise: explore top DE genes</h3>
<div class="callout-content">
<p>Using <code>sig_res</code>:</p>
<ol style="list-style-type: decimal"><li>Which genes have the largest positive log2 fold change.</li>
<li>Which genes have the largest negative log2 fold change.</li>
<li>Are any of the top genes known to be related to Nat10 function or
the studied phenotype.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Example interpretation:</p>
<ul><li>Several strongly up regulated genes are involved in cell cycle and
RNA processing.</li>
<li>Strongly down regulated genes may include transcription factors and
signaling components.</li>
<li>Cross referencing top hits with prior literature or databases can
reveal mechanistic hypotheses.</li>
</ul></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Counts and sample metadata must be aligned before building DESeq2
objects.</li>
<li>Biotype annotation allows restriction to protein coding genes for
exploratory QC.</li>
<li>Variance stabilizing transform, distance heatmaps, and PCA help
detect outliers and batch effects.</li>
<li>DESeq2 provides a complete pipeline from raw counts to differential
expression.</li>
<li>Joining DE results with normalized counts and gene annotation yields
a useful final table.</li>
<li>Volcano plots summarize differential expression using fold change
and adjusted p values.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/04b-transcript-based-quantification.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/05b-deseq2-expression-analyses-kallisto.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/04b-transcript-based-quantification.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: B. Transcript-based
        </a>
        <a class="chapter-link float-end" href="../instructor/05b-deseq2-expression-analyses-kallisto.html" rel="next">
          Next: Differential...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05-deseq2-expression-analyses.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.3" class="external-link">sandpaper (0.17.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/05-deseq2-expression-analyses.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "Gene-level QC and differential expression (DESeq2)",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/05-deseq2-expression-analyses.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/05-deseq2-expression-analyses.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2025-12-30",
  "datePublished": "2025-12-30"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

