<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq in practice: Hands-on workshop on RCAC systems: Gene set enrichment analysis</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png"><link rel="manifest" href="../favicons/rcac/site.webmanifest"><link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../06-gene-set-enrichment-analyses.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 83%" class="percentage">
    83%
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: 83%" aria-valuenow="83" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../06-gene-set-enrichment-analyses.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Kallisto)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05b-deseq2-expression-analyses-kallisto.html">7. B. Differential expression using DESeq2 (Kallisto pathway)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        8. Gene set enrichment analysis
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#background-understanding-over-representation-analysis">Background: Understanding over-representation analysis</a></li>
<li><a href="#step-1-load-de-results-and-prepare-gene-lists">Step 1: Load DE results and prepare gene lists</a></li>
<li><a href="#step-2-go-enrichment-analysis">Step 2: GO enrichment analysis</a></li>
<li><a href="#step-3-kegg-pathway-enrichment">Step 3: KEGG pathway enrichment</a></li>
<li><a href="#step-4-msigdb-hallmark-enrichment">Step 4: MSigDB Hallmark enrichment</a></li>
<li><a href="#step-5-compare-results-across-databases">Step 5: Compare results across databases</a></li>
<li><a href="#step-5b-direction-aware-enrichment-analysis">Step 5b: Direction-aware enrichment analysis</a></li>
<li><a href="#step-5c-gene-set-enrichment-analysis-gsea">Step 5c: Gene Set Enrichment Analysis (GSEA)</a></li>
<li><a href="#step-6-save-results">Step 6: Save results</a></li>
<li><a href="#interpretation-guidelines">Interpretation guidelines</a></li>
<li><a href="#summary">Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/05b-deseq2-expression-analyses-kallisto.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/05b-deseq2-expression-analyses-kallisto.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: B. Differential
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Gene set enrichment analysis</h1>
        <p>Last updated on 2026-01-22 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/06-gene-set-enrichment-analyses.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 95 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is over-representation analysis and how does it work
statistically?</li>
<li>How do we identify functional pathways and biological themes from DE
genes?</li>
<li>How do we run enrichment analysis for GO, KEGG, and MSigDB gene
sets?</li>
<li>How do we interpret and compare enrichment results across different
databases?</li>
<li>What are the limitations and best practices for enrichment
analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand the statistical basis of over-representation analysis
(ORA).</li>
<li>Prepare gene lists and background universes from DESeq2
results.</li>
<li>Run GO enrichment using clusterProfiler and interpret ontology
categories.</li>
<li>Run KEGG pathway enrichment with organism-specific data.</li>
<li>Perform MSigDB Hallmark enrichment for curated biological
signatures.</li>
<li>Produce and interpret barplots, dotplots, and enrichment maps.</li>
<li>Compare results across multiple gene set databases.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>Differential expression analysis identifies individual genes that
change between conditions, but interpreting long gene lists is
challenging. <strong>Gene set enrichment analysis</strong> provides
biological context by asking: <em>Are genes with specific functions
over-represented among the differentially expressed genes?</em></p>
<p>This episode covers <strong>over-representation analysis
(ORA)</strong>, the most common approach for interpreting DE
results.</p>
<div id="what-you-need-for-this-episode" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-you-need-for-this-episode" class="callout-inner">
<h3 class="callout-title">What you need for this episode</h3>
<div class="callout-content">
<ul><li>DE results from Episode 05 or 05b
(<code>DESeq2_results_joined.tsv</code> or
<code>DESeq2_kallisto_results.tsv</code>)</li>
<li>RStudio session via Open OnDemand</li>
<li>Internet connection (for KEGG queries)</li>
</ul><div class="section level2">
<h2 id="required-packages">Required packages<a class="anchor" aria-label="anchor" href="#required-packages"></a></h2>
<ul><li>clusterProfiler (Bioconductor)</li>
<li>org.Mm.eg.db (mouse annotation)</li>
<li>msigdbr (MSigDB gene sets)</li>
<li>enrichplot (visualization)</li>
<li>dplyr, ggplot2, readr</li>
</ul></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="background-understanding-over-representation-analysis">Background: Understanding over-representation analysis<a class="anchor" aria-label="anchor" href="#background-understanding-over-representation-analysis"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="what-is-ora">What is ORA?<a class="anchor" aria-label="anchor" href="#what-is-ora"></a></h3>
<p>ORA tests whether a predefined gene set contains more DE genes than
expected by chance. It answers the question: <em>If I randomly selected
the same number of genes from my background, how often would I see this
many genes from pathway X?</em></p>
</div>
<div class="section level3">
<h3 id="the-statistical-test">The statistical test<a class="anchor" aria-label="anchor" href="#the-statistical-test"></a></h3>
<p>ORA uses the <strong>hypergeometric test</strong> (equivalent to
Fisher’s exact test), which calculates the probability of observing
<em>k</em> or more genes from a pathway given:</p>
<ul><li>
<strong>N</strong>: Total genes in the background (universe)</li>
<li>
<strong>K</strong>: Genes in the pathway of interest</li>
<li>
<strong>n</strong>: Total DE genes in your list</li>
<li>
<strong>k</strong>: DE genes that are in the pathway</li>
</ul><pre><code>                    Pathway genes    Non-pathway genes
DE genes                 k              n - k
Non-DE genes           K - k          N - K - n + k</code></pre>
<div id="the-hypergeometric-distribution" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="the-hypergeometric-distribution" class="callout-inner">
<h3 class="callout-title">The hypergeometric distribution</h3>
<div class="callout-content">
<p>The p-value represents the probability of drawing <em>k</em> or more
pathway genes when randomly selecting <em>n</em> genes from a population
of <em>N</em> genes containing <em>K</em> pathway members:</p>
<p><span class="math display">\[P(X \geq k) = \sum_{i=k}^{\min(K,n)}
\frac{\binom{K}{i}\binom{N-K}{n-i}}{\binom{N}{n}}\]</span></p>
<p>This is a one-tailed test asking if the pathway is over-represented
(enriched) in your gene list.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="multiple-testing-correction">Multiple testing correction<a class="anchor" aria-label="anchor" href="#multiple-testing-correction"></a></h3>
<p>With thousands of gene sets tested, many false positives would occur
by chance. We apply <strong>false discovery rate (FDR)</strong>
correction using the Benjamini-Hochberg method:</p>
<ol style="list-style-type: decimal"><li>Rank all p-values from smallest to largest.</li>
<li>Calculate adjusted p-value: <span class="math inline">\(p_{adj} = p
\times \frac{n}{rank}\)</span>
</li>
<li>Filter results by adjusted p-value threshold (typically 0.05).</li>
</ol><div id="why-the-universe-matters" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-the-universe-matters" class="callout-inner">
<h3 class="callout-title">Why the universe matters</h3>
<div class="callout-content">
<p>The background universe should include <strong>all genes that could
have been detected as DE</strong> in your experiment, typically all
genes tested by DESeq2.</p>
<p>Common mistakes: - Using the whole genome (inflates significance). -
Using only expressed genes (may miss relevant pathways). - Using a
different species’ gene list.</p>
<p>The universe defines what “expected by chance” means. An
inappropriate universe leads to misleading results.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="gene-set-databases">Gene set databases<a class="anchor" aria-label="anchor" href="#gene-set-databases"></a></h3>
<p>Different databases capture different aspects of biology:</p>
<table class="table"><colgroup><col width="30%"><col width="39%"><col width="30%"></colgroup><thead><tr class="header"><th>Database</th>
<th>Description</th>
<th>Best for</th>
</tr></thead><tbody><tr class="odd"><td>GO BP</td>
<td>Biological Process ontology</td>
<td>Functional processes</td>
</tr><tr class="even"><td>GO MF</td>
<td>Molecular Function ontology</td>
<td>Protein activities</td>
</tr><tr class="odd"><td>GO CC</td>
<td>Cellular Component ontology</td>
<td>Subcellular localization</td>
</tr><tr class="even"><td>KEGG</td>
<td>Curated pathways</td>
<td>Metabolic/signaling pathways</td>
</tr><tr class="odd"><td>Reactome</td>
<td>Detailed pathway reactions</td>
<td>Mechanistic understanding</td>
</tr><tr class="even"><td>MSigDB Hallmark</td>
<td>50 curated signatures</td>
<td>High-level biological states</td>
</tr><tr class="odd"><td>MSigDB C2</td>
<td>Curated gene sets</td>
<td>Canonical pathways</td>
</tr></tbody></table><div id="which-database-should-i-use" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="which-database-should-i-use" class="callout-inner">
<h3 class="callout-title">Which database should I use?</h3>
<div class="callout-content">
<p>There’s no single “best” database. Consider:</p>
<ul><li>
<strong>GO</strong>: Comprehensive but redundant; many overlapping
terms.</li>
<li>
<strong>KEGG</strong>: Well-curated pathways; limited coverage;
requires Entrez IDs.</li>
<li>
<strong>Hallmark</strong>: Reduced redundancy; easier to interpret;
may miss specific pathways.</li>
</ul><p>Best practice: Run multiple databases and look for <strong>convergent
signals</strong> across them.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-1-load-de-results-and-prepare-gene-lists">Step 1: Load DE results and prepare gene lists<a class="anchor" aria-label="anchor" href="#step-1-load-de-results-and-prepare-gene-lists"></a></h2>
<hr class="half-width"><p>Start your RStudio session via Open OnDemand as described in Episode
05, then load the required packages:</p>
<p>Load packages:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span>
<span><span class="co"># Construct the path dynamically</span></span>
<span><span class="va">work_dir</span> <span class="op">&lt;-</span> <span class="fu">file.path</span><span class="op">(</span><span class="st">"/scratch/negishi"</span>, <span class="fu">Sys.getenv</span><span class="op">(</span><span class="st">"USER"</span><span class="op">)</span>, <span class="st">"rnaseq-workshop"</span><span class="op">)</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="va">work_dir</span><span class="op">)</span></span></code></pre>
</div>
<p>Load DESeq2 results from Episode 05:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"results/deseq2/DESeq2_results_joined.tsv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code># A tibble: 6 × 20
  baseMean log2FoldChange  lfcSE        pvalue         padj ensembl_gene_id_version WT_Bcell_mock_rep1 WT_Bcell_mock_rep2 WT_Bcell_mock_rep3 WT_Bcell_mock_rep4
     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;                                &lt;dbl&gt;              &lt;dbl&gt;              &lt;dbl&gt;              &lt;dbl&gt;
1     259.         -0.703 0.123  0.00000000431 0.0000000268 ENSMUSG00000033845.14                 195.               212.               202.               174.
2     142.          0.244 0.192  0.177         0.256        ENSMUSG00000025903.15                 157.               141.               167.               160.
3     305.         -0.218 0.121  0.0652        0.111        ENSMUSG00000033813.16                 293.               269.               321.               242.
4     241.          0.224 0.129  0.0741        0.123        ENSMUSG00000033793.13                 279.               250.               291.               224.
5     773.          0.231 0.0952 0.0140        0.0286       ENSMUSG00000025907.15                 831.               760.               887.               865.
6     556.          0.126 0.0892 0.152         0.226        ENSMUSG00000051285.18                 608.               616.               560.               540.
# ℹ 10 more variables: WT_Bcell_IR_rep1 &lt;dbl&gt;, WT_Bcell_IR_rep2 &lt;dbl&gt;, WT_Bcell_IR_rep3 &lt;dbl&gt;, WT_Bcell_IR_rep4 &lt;dbl&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_gene_name &lt;chr&gt;, gene_biotype &lt;chr&gt;, description &lt;chr&gt;, label &lt;chr&gt;, sig &lt;chr&gt;</code></pre>
<p>Define the gene lists:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Universe: all genes tested</span></span>
<span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span></span>
<span></span>
<span><span class="co"># Significant genes: padj &lt; 0.05 and |log2FC| &gt; log2(1.5)</span></span>
<span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="va">sig_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">log2fc_cut</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">sig_genes</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 11330
[1] 3597</code></pre>
<div id="choosing-significance-thresholds" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="choosing-significance-thresholds" class="callout-inner">
<h3 class="callout-title">Choosing significance thresholds</h3>
<div class="callout-content">
<p>The threshold for “significant” genes affects enrichment results:</p>
<ul><li>
<strong>Stringent (padj &lt; 0.01, |LFC| &gt; 1)</strong>: Fewer
genes, clearer signals, may miss subtle effects.</li>
<li>
<strong>Lenient (padj &lt; 0.1, no LFC cutoff)</strong>: More genes,
noisier results, higher sensitivity.</li>
</ul><p>There’s no universally correct threshold. Consider your experimental
question and validate key findings.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="convert-gene-ids">Convert gene IDs<a class="anchor" aria-label="anchor" href="#convert-gene-ids"></a></h3>
<p>Most enrichment tools require <strong>Entrez IDs</strong>. Our data
uses Ensembl IDs with version numbers, so we need to convert:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove version suffix for mapping</span></span>
<span><span class="va">universe_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">universe</span><span class="op">)</span></span>
<span><span class="va">sig_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">sig_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Map Ensembl to Entrez</span></span>
<span><span class="va">id_map</span> <span class="op">&lt;-</span> <span class="fu">bitr</span><span class="op">(</span></span>
<span>    <span class="va">universe_clean</span>,</span>
<span>    fromType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>    toType <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">id_map</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>             ENSEMBL ENTREZID
1 ENSMUSG00000033845    27395
2 ENSMUSG00000025903    18777
3 ENSMUSG00000033813    21399
4 ENSMUSG00000033793   108664
5 ENSMUSG00000025907    12421
6 ENSMUSG00000051285   319263</code></pre>
<div id="gene-id-conversion-challenges" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="gene-id-conversion-challenges" class="callout-inner">
<h3 class="callout-title">Gene ID conversion challenges</h3>
<div class="callout-content">
<p>Not all genes map successfully:</p>
<ul><li>Some Ensembl IDs lack Entrez equivalents.</li>
<li>ID mapping databases may be outdated.</li>
<li>Multi-mapping (one Ensembl → multiple Entrez) can occur.</li>
</ul><p><code>bitr()</code> drops unmapped genes silently. Check how many
genes were lost:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">message</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Mapped %d of %d genes (%.1f%%)"</span>,</span>
<span>    <span class="fu">nrow</span><span class="op">(</span><span class="va">id_map</span><span class="op">)</span>, <span class="fu">length</span><span class="op">(</span><span class="va">universe_clean</span><span class="op">)</span>,</span>
<span>    <span class="fl">100</span> <span class="op">*</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">id_map</span><span class="op">)</span> <span class="op">/</span> <span class="fu">length</span><span class="op">(</span><span class="va">universe_clean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code>Mapped 11358 of 11330 genes (100.2%)</code></pre>
</div>
</div>
</div>
<p>Create final gene lists with Entrez IDs:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Universe with Entrez IDs</span></span>
<span><span class="va">universe_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span><span class="op">$</span><span class="va">ENTREZID</span></span>
<span></span>
<span><span class="co"># Significant genes with Entrez IDs</span></span>
<span><span class="va">sig_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">ENSEMBL</span> <span class="op">%in%</span> <span class="va">sig_clean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">sig_entrez</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code>[1] 3608</code></pre>
<div id="exercise-prepare-up-regulated-and-down-regulated-gene-lists" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-prepare-up-regulated-and-down-regulated-gene-lists" class="callout-inner">
<h3 class="callout-title">Exercise: Prepare up-regulated and
down-regulated gene lists</h3>
<div class="callout-content">
<p>Create separate gene lists for: 1. Up-regulated genes (log2FC &gt; 0
and padj &lt; 0.05) 2. Down-regulated genes (log2FC &lt; 0 and padj &lt;
0.05)</p>
<p>Why might analyzing these separately be informative?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Up-regulated genes</span></span>
<span><span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="va">log2fc_cut</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span><span class="op">)</span></span>
<span><span class="va">up_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">up_genes</span><span class="op">)</span></span>
<span><span class="va">up_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">ENSEMBL</span> <span class="op">%in%</span> <span class="va">up_clean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Down-regulated genes</span></span>
<span><span class="va">down_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">log2fc_cut</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span><span class="op">)</span></span>
<span><span class="va">down_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">down_genes</span><span class="op">)</span></span>
<span><span class="va">down_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">ENSEMBL</span> <span class="op">%in%</span> <span class="va">down_clean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span></code></pre>
</div>
<p>Check number of genes:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">up_entrez</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 1810</code></pre>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">down_entrez</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 1798</code></pre>
<p>Analyzing up and down separately can reveal: - Different biological
processes activated vs. suppressed. - Clearer signals when combining
masks opposing effects. - Direction-specific pathway involvement.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-2-go-enrichment-analysis">Step 2: GO enrichment analysis<a class="anchor" aria-label="anchor" href="#step-2-go-enrichment-analysis"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="run-go-biological-process-enrichment">Run GO Biological Process enrichment<a class="anchor" aria-label="anchor" href="#run-go-biological-process-enrichment"></a></h3>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_bp</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>    ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>    pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    qvalueCutoff <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">ego_bp</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<p>Output (columns truncated for brevity):</p>
<pre><code>                   ID                                    Description GeneRatio   BgRatio RichFactor FoldEnrichment   zScore       pvalue    p.adjust      qvalue
GO:0002683 GO:0002683   negative regulation of immune system process  174/3494 405/10989  0.4296296       1.351231 4.917337 1.003906e-06 0.005247415 0.004941330
GO:0042100 GO:0042100                           B cell proliferation   44/3494  77/10989  0.5714286       1.797203 4.792888 3.631972e-06 0.009166217 0.008631545
GO:0032943 GO:0032943                 mononuclear cell proliferation  119/3494 266/10989  0.4473684       1.407021 4.588117 5.260886e-06 0.009166217 0.008631545
GO:0051250 GO:0051250   negative regulation of lymphocyte activation   65/3494 130/10989  0.5000000       1.572553 4.483605 1.087013e-05 0.012782384 0.012036778
GO:0046651 GO:0046651                       lymphocyte proliferation  115/3494 260/10989  0.4423077       1.391105 4.357475 1.428670e-05 0.012782384 0.012036778
GO:0002698 GO:0002698 negative regulation of immune effector process   51/3494  97/10989  0.5257732       1.653612 4.414558 1.636446e-05 0.012782384 0.012036778
GO:0042254 GO:0042254                            ribosome biogenesis  131/3494 304/10989  0.4309211       1.355292 4.289141 1.808829e-05 0.012782384 0.012036778
GO:0042113 GO:0042113                              B cell activation  107/3494 241/10989  0.4439834       1.396375 4.248012 2.305909e-05 0.012782384 0.012036778
GO:0006364 GO:0006364                                rRNA processing   95/3494 210/10989  0.4523810       1.422786 4.223534 2.665078e-05 0.012782384 0.012036778
GO:0002694 GO:0002694             regulation of leukocyte activation  189/3494 466/10989  0.4055794       1.275590 4.150705 2.844730e-05 0.012782384 0.012036778</code></pre>
<div id="understanding-enrichgo-output" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="understanding-enrichgo-output" class="callout-inner">
<h3 class="callout-title">Understanding enrichGO output</h3>
<div class="callout-content">
<p>Key columns in the results:</p>
<ul><li>
<strong>ID</strong>: GO term identifier (e.g., <a href="GO:0006915" class="uri">GO:0006915</a>)</li>
<li>
<strong>Description</strong>: Human-readable term name</li>
<li>
<strong>GeneRatio</strong>: DE genes in term / total DE genes with
annotation</li>
<li>
<strong>BgRatio</strong>: Term genes in universe / total annotated
genes in universe</li>
<li>
<strong>pvalue</strong>: Raw hypergeometric p-value</li>
<li>
<strong>p.adjust</strong>: BH-corrected p-value</li>
<li>
<strong>qvalue</strong>: Alternative FDR estimate</li>
<li>
<strong>geneID</strong>: List of DE genes in this term</li>
<li>
<strong>Count</strong>: Number of DE genes in this term</li>
</ul></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="visualize-go-results">Visualize GO results<a class="anchor" aria-label="anchor" href="#visualize-go-results"></a></h3>
<p>Dotplot showing top enriched terms:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ego_bp</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO Biological Process enrichment"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/go-bp-up.png" alt="GO Biological Process enrichment dotplot for upregulated genes" class="figure mx-auto d-block"><figcaption>
GO Biological Process enrichment dotplot for upregulated genes
</figcaption></figure><p>The dotplot shows: - <strong>X-axis</strong>: Gene ratio (proportion
of DE genes in the term) - <strong>Y-axis</strong>: GO term description
- <strong>Color</strong>: Adjusted p-value - <strong>Size</strong>:
Number of genes</p>
<p>Barplot alternative:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">ego_bp</span>, showCategory <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO BP enrichment - Top 15 terms"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/go-bp-up-bar.png" alt="GO Biological Process enrichment barplot for upregulated genes" class="figure mx-auto d-block"><figcaption>
GO Biological Process enrichment barplot for upregulated genes
</figcaption></figure></div>
<div class="section level3">
<h3 id="reduce-go-redundancy">Reduce GO redundancy<a class="anchor" aria-label="anchor" href="#reduce-go-redundancy"></a></h3>
<p>GO terms are hierarchical and often redundant. Use
<code>simplify()</code> to remove similar terms:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_bp_simple</span> <span class="op">&lt;-</span> <span class="fu">simplify</span><span class="op">(</span></span>
<span>    <span class="va">ego_bp</span>,</span>
<span>    cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"p.adjust"</span>,</span>
<span>    select_fun <span class="op">=</span> <span class="va">min</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ego_bp_simple</span>, showCategory <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO BP enrichment (simplified)"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/go-bp-up-simple.png" alt="Simplified GO Biological Process enrichment dotplot" class="figure mx-auto d-block"><figcaption>
Simplified GO Biological Process enrichment dotplot
</figcaption></figure><figure style="text-align: center"><img src="../fig/06-enrich/go-bp-up-bar-simple.png" alt="Simplified GO Biological Process enrichment barplot" class="figure mx-auto d-block"><figcaption>
Simplified GO Biological Process enrichment barplot
</figcaption></figure><div id="why-simplify-go-results" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-simplify-go-results" class="callout-inner">
<h3 class="callout-title">Why simplify GO results?</h3>
<div class="callout-content">
<p>The GO hierarchy means related terms often appear together: -
“regulation of cell death” and “positive regulation of cell death” -
“response to stimulus” and “response to external stimulus”</p>
<p><code>simplify()</code> removes redundant terms based on semantic
similarity, making results easier to interpret. The <code>cutoff</code>
parameter (0-1) controls how aggressively terms are merged.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="enrichment-map-visualization">Enrichment map visualization<a class="anchor" aria-label="anchor" href="#enrichment-map-visualization"></a></h3>
<p>For many significant terms, an enrichment map shows
relationships:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_bp_em</span> <span class="op">&lt;-</span> <span class="fu">pairwise_termsim</span><span class="op">(</span><span class="va">ego_bp_simple</span><span class="op">)</span></span>
<span><span class="fu">emapplot</span><span class="op">(</span><span class="va">ego_bp_em</span>, showCategory <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO BP enrichment map"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/go-bp-up-enrichment-map.png" alt="GO Biological Process enrichment map showing term relationships" class="figure mx-auto d-block"><figcaption>
GO Biological Process enrichment map showing term relationships
</figcaption></figure><p>The enrichment map clusters similar terms together, revealing
functional themes.</p>
<div id="exercise-compare-go-ontologies" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-compare-go-ontologies" class="callout-inner">
<h3 class="callout-title">Exercise: Compare GO ontologies</h3>
<div class="callout-content">
<p>Run enrichment for GO Molecular Function (MF) and Cellular Component
(CC):</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_mf</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">sig_entrez</span>, universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>                   OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>, ont <span class="op">=</span> <span class="st">"MF"</span>, readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ego_cc</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">sig_entrez</span>, universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>                   OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>, ont <span class="op">=</span> <span class="st">"CC"</span>, readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<ol style="list-style-type: decimal"><li>How do the top terms differ between BP, MF, and CC?</li>
<li>Are there consistent themes across ontologies?</li>
<li>Which ontology provides the most interpretable results for your
experiment?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Example interpretation:</p>
<ul><li>
<strong>BP</strong> terms describe processes (e.g., “cell cycle”,
“RNA processing”).</li>
<li>
<strong>MF</strong> terms describe molecular activities (e.g., “RNA
binding”, “kinase activity”).</li>
<li>
<strong>CC</strong> terms describe locations (e.g., “nucleus”,
“mitochondrion”).</li>
</ul><p>Consistent themes across ontologies strengthen biological
interpretation. For example, if BP shows “translation” enrichment, MF
might show “ribosome binding”, and CC might show “ribosome” or
“cytoplasm”.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-3-kegg-pathway-enrichment">Step 3: KEGG pathway enrichment<a class="anchor" aria-label="anchor" href="#step-3-kegg-pathway-enrichment"></a></h2>
<hr class="half-width"><p>KEGG provides curated pathway maps connecting genes to biological
systems.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ekegg</span> <span class="op">&lt;-</span> <span class="fu">enrichKEGG</span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>    organism <span class="op">=</span> <span class="st">"mmu"</span>,  <span class="co"># Mouse</span></span>
<span>    pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>    pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">ekegg</span><span class="op">)</span></span></code></pre>
</div>
<p>Output (geneID column not shown for brevity):</p>
<pre><code>                                    category                         subcategory       ID                              Description GeneRatio BgRatio RichFactor
mmu04115                   Cellular Processes               Cell growth and death mmu04115                    p53 signaling pathway   41/1686 64/5208  0.6406250
mmu05322                       Human Diseases                      Immune disease mmu05322             Systemic lupus erythematosus   46/1686 87/5208  0.5287356
mmu04977                   Organismal Systems                    Digestive system mmu04977         Vitamin digestion and absorption   12/1686 15/5208  0.8000000
mmu04514 Environmental Information Processing Signaling molecules and interaction mmu04514 Cell adhesion molecule (CAM) interaction   43/1686 85/5208  0.5058824
mmu05150                       Human Diseases       Infectious disease: bacterial mmu05150          Staphylococcus aureus infection   19/1686 30/5208  0.6333333
mmu04981                                 &lt;NA&gt;                                &lt;NA&gt; mmu04981          Folate transport and metabolism   13/1686 18/5208  0.7222222
         FoldEnrichment   zScore       pvalue     p.adjust       qvalue
mmu04115       1.978870 5.451204 1.688415e-07 5.605537e-05 5.491791e-05
mmu05322       1.633247 4.120819 5.329885e-05 8.847610e-03 8.668077e-03
mmu04977       2.471174 3.947558 2.042589e-04 2.260466e-02 2.214597e-02
mmu04514       1.562654 3.618403 3.393132e-04 2.816300e-02 2.759152e-02
mmu05150       1.956346 3.634316 4.705801e-04 3.124652e-02 3.061248e-02
mmu04981       2.230921 3.619185 6.082405e-04 3.365597e-02 3.297304e-02
         Count
mmu04115    41
mmu05322    46
mmu04977    12
mmu04514    43
mmu05150    19
mmu04981    13</code></pre>
<div id="kegg-organism-codes" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="kegg-organism-codes" class="callout-inner">
<h3 class="callout-title">KEGG organism codes</h3>
<div class="callout-content">
<p>Common organism codes: - <code>"hsa"</code>: Human -
<code>"mmu"</code>: Mouse - <code>"rno"</code>: Rat -
<code>"dme"</code>: Drosophila - <code>"sce"</code>: Yeast -
<code>"ath"</code>: Arabidopsis</p>
<p>Find your organism:
<code>search_kegg_organism("species name")</code></p>
</div>
</div>
</div>
<p>Visualize KEGG results:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">ekegg</span>, showCategory <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"KEGG pathway enrichment"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/kegg-bar.png" alt="KEGG pathway enrichment barplot" class="figure mx-auto d-block"><figcaption>
KEGG pathway enrichment barplot
</figcaption></figure><div id="kegg-limitations" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="kegg-limitations" class="callout-inner">
<h3 class="callout-title">KEGG limitations</h3>
<div class="callout-content">
<p>Be aware of KEGG-specific issues:</p>
<ol style="list-style-type: decimal"><li>
<strong>ID coverage</strong>: KEGG uses its own gene IDs; some
Entrez IDs may not map.</li>
<li>
<strong>Update frequency</strong>: KEGG pathways are updated
periodically; results may vary.</li>
<li>
<strong>License</strong>: KEGG has usage restrictions for commercial
applications.</li>
<li>
<strong>Incomplete mapping</strong>: Some genes lack pathway
assignments.</li>
</ol><p>Check dropped genes:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">message</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"KEGG analysis used %d of %d input genes"</span>,</span>
<span>    <span class="fu">length</span><span class="op">(</span><span class="va">ekegg</span><span class="op">@</span><span class="va">gene</span><span class="op">)</span>, <span class="fu">length</span><span class="op">(</span><span class="va">sig_entrez</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code>KEGG analysis used 3607 of 3608 input genes</code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-msigdb-hallmark-enrichment">Step 4: MSigDB Hallmark enrichment<a class="anchor" aria-label="anchor" href="#step-4-msigdb-hallmark-enrichment"></a></h2>
<hr class="half-width"><p>MSigDB Hallmark gene sets are 50 curated signatures representing
well-defined biological states and processes.</p>
<div class="section level3">
<h3 id="load-hallmark-gene-sets">Load Hallmark gene sets<a class="anchor" aria-label="anchor" href="#load-hallmark-gene-sets"></a></h3>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mus musculus"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">hallmark</span><span class="op">)</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code># A tibble: 6 × 26
  gene_symbol ncbi_gene ensembl_gene       db_gene_symbol db_ncbi_gene db_ensembl_gene source_gene gs_id gs_name  gs_collection gs_subcollection gs_collection_name
  &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;              &lt;chr&gt;          &lt;chr&gt;        &lt;chr&gt;           &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt;            &lt;chr&gt;
1 Abca1       11303     ENSMUSG00000015243 ABCA1          19           ENSG00000165029 ABCA1       M5905 HALLMAR… H             ""               Hallmark
2 Abcb8       74610     ENSMUSG00000028973 ABCB8          11194        ENSG00000197150 ABCB8       M5905 HALLMAR… H             ""               Hallmark
3 Acaa2       52538     ENSMUSG00000036880 ACAA2          10449        ENSG00000167315 ACAA2       M5905 HALLMAR… H             ""               Hallmark
4 Acadl       11363     ENSMUSG00000026003 ACADL          33           ENSG00000115361 ACADL       M5905 HALLMAR… H             ""               Hallmark
5 Acadm       11364     ENSMUSG00000062908 ACADM          34           ENSG00000117054 ACADM       M5905 HALLMAR… H             ""               Hallmark
6 Acads       11409     ENSMUSG00000029545 ACADS          35           ENSG00000122971 ACADS       M5905 HALLMAR… H             ""               Hallmark
# ℹ 14 more variables: gs_description &lt;chr&gt;, gs_source_species &lt;chr&gt;, gs_pmid &lt;chr&gt;, gs_geoid &lt;chr&gt;, gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, db_version &lt;chr&gt;,
#   db_target_species &lt;chr&gt;, ortholog_taxon_id &lt;int&gt;, ortholog_sources &lt;chr&gt;, num_ortholog_sources &lt;dbl&gt;, entrez_gene &lt;chr&gt;, gs_cat &lt;chr&gt;, gs_subcat &lt;chr&gt;</code></pre>
<p>Prepare gene set list:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark_list</span> <span class="op">&lt;-</span> <span class="va">hallmark</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">entrez_gene</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span>term <span class="op">=</span> <span class="va">gs_name</span>, gene <span class="op">=</span> <span class="va">entrez_gene</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="run-hallmark-enrichment">Run Hallmark enrichment<a class="anchor" aria-label="anchor" href="#run-hallmark-enrichment"></a></h3>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ehall</span> <span class="op">&lt;-</span> <span class="fu">enricher</span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>    TERM2GENE <span class="op">=</span> <span class="va">hallmark_list</span>,</span>
<span>    pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>    pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">ehall</span><span class="op">)</span></span></code></pre>
</div>
<p>Output (columns truncated for brevity):</p>
<pre><code>                                             ID             Description GeneRatio  BgRatio RichFactor FoldEnrichment   zScore       pvalue     p.adjust
HALLMARK_MYC_TARGETS_V2 HALLMARK_MYC_TARGETS_V2 HALLMARK_MYC_TARGETS_V2   40/1134  57/3216  0.7017544       1.990161 5.565765 6.573980e-08 3.286990e-06
HALLMARK_P53_PATHWAY       HALLMARK_P53_PATHWAY    HALLMARK_P53_PATHWAY   86/1134 164/3216  0.5243902       1.487160 4.725607 2.814573e-06 7.036433e-05</code></pre>
<p>Visualize:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ehall</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"MSigDB Hallmark enrichment"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/hallmark1.png" alt="MSigDB Hallmark gene set enrichment dotplot" class="figure mx-auto d-block"><figcaption>
MSigDB Hallmark gene set enrichment dotplot
</figcaption></figure><div id="why-use-hallmark-gene-sets" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-hallmark-gene-sets" class="callout-inner">
<h3 class="callout-title">Why use Hallmark gene sets?</h3>
<div class="callout-content">
<p>Advantages of Hallmark over GO/KEGG:</p>
<ol style="list-style-type: decimal"><li>
<strong>Reduced redundancy</strong>: Each signature is distinct and
non-overlapping.</li>
<li>
<strong>Expert curation</strong>: Refined from multiple sources to
represent coherent biology.</li>
<li>
<strong>Interpretability</strong>: 50 sets is manageable for
detailed examination.</li>
<li>
<strong>Reproducibility</strong>: Well-documented and stable
definitions.</li>
</ol><p>Common Hallmark sets include: - HALLMARK_INFLAMMATORY_RESPONSE -
HALLMARK_P53_PATHWAY - HALLMARK_OXIDATIVE_PHOSPHORYLATION -
HALLMARK_MYC_TARGETS_V1</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-compare-results-across-databases">Step 5: Compare results across databases<a class="anchor" aria-label="anchor" href="#step-5-compare-results-across-databases"></a></h2>
<hr class="half-width"><p>Create a summary of top enriched terms from each database:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract top 10 from each</span></span>
<span><span class="va">top_go</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="va">ego_bp_simple</span><span class="op">@</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>Database <span class="op">=</span> <span class="st">"GO BP"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">Database</span>, <span class="va">Description</span>, <span class="va">p.adjust</span>, <span class="va">Count</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_kegg</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="va">ekegg</span><span class="op">@</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>Database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">Database</span>, <span class="va">Description</span>, <span class="va">p.adjust</span>, <span class="va">Count</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_hall</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="va">ehall</span><span class="op">@</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>Database <span class="op">=</span> <span class="st">"Hallmark"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">Database</span>, <span class="va">ID</span>, <span class="va">p.adjust</span>, <span class="va">Count</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span>Description <span class="op">=</span> <span class="va">ID</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine</span></span>
<span><span class="va">comparison</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">top_go</span>, <span class="va">top_kegg</span>, <span class="va">top_hall</span><span class="op">)</span></span>
<span><span class="va">comparison</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code>                                 Database                                  Description     p.adjust Count
GO:0002683                          GO BP negative regulation of immune system process 5.247415e-03   174
GO:0042100                          GO BP                         B cell proliferation 9.166217e-03    44
GO:0032943                          GO BP               mononuclear cell proliferation 9.166217e-03   119
GO:0051250                          GO BP negative regulation of lymphocyte activation 1.278238e-02    65
GO:0042254                          GO BP                          ribosome biogenesis 1.278238e-02   131
GO:0042113                          GO BP                            B cell activation 1.278238e-02   107
GO:0006364                          GO BP                              rRNA processing 1.278238e-02    95
GO:0002252                          GO BP                      immune effector process 1.278238e-02   199
GO:0002437                          GO BP  inflammatory response to antigenic stimulus 1.390022e-02    29
GO:0016072                          GO BP                       rRNA metabolic process 1.390022e-02   106
mmu04115                             KEGG                        p53 signaling pathway 5.605537e-05    41
mmu05322                             KEGG                 Systemic lupus erythematosus 8.847610e-03    46
mmu04977                             KEGG             Vitamin digestion and absorption 2.260466e-02    12
mmu04514                             KEGG     Cell adhesion molecule (CAM) interaction 2.816300e-02    43
mmu05150                             KEGG              Staphylococcus aureus infection 3.124652e-02    19
mmu04981                             KEGG              Folate transport and metabolism 3.365597e-02    13
mmu02010                             KEGG                             ABC transporters 4.462944e-02    18
mmu03008                             KEGG            Ribosome biogenesis in eukaryotes 8.758267e-02    35
mmu05202                             KEGG      Transcriptional misregulation in cancer 9.020508e-02    60
mmu04068                             KEGG                       FoxO signaling pathway 1.135249e-01    48
HALLMARK_MYC_TARGETS_V2          Hallmark                      HALLMARK_MYC_TARGETS_V2 3.286990e-06    40
HALLMARK_P53_PATHWAY             Hallmark                         HALLMARK_P53_PATHWAY 7.036433e-05    86
HALLMARK_XENOBIOTIC_METABOLISM   Hallmark               HALLMARK_XENOBIOTIC_METABOLISM 1.034831e-01    57
HALLMARK_HEME_METABOLISM         Hallmark                     HALLMARK_HEME_METABOLISM 1.034831e-01    72
HALLMARK_IL2_STAT5_SIGNALING     Hallmark                 HALLMARK_IL2_STAT5_SIGNALING 1.611589e-01    69
HALLMARK_ALLOGRAFT_REJECTION     Hallmark                 HALLMARK_ALLOGRAFT_REJECTION 1.611589e-01    70
HALLMARK_MYC_TARGETS_V1          Hallmark                      HALLMARK_MYC_TARGETS_V1 1.757524e-01    84
HALLMARK_TNFA_SIGNALING_VIA_NFKB Hallmark             HALLMARK_TNFA_SIGNALING_VIA_NFKB 2.021817e-01    67
HALLMARK_ESTROGEN_RESPONSE_EARLY Hallmark             HALLMARK_ESTROGEN_RESPONSE_EARLY 2.021817e-01    56
HALLMARK_APOPTOSIS               Hallmark                           HALLMARK_APOPTOSIS 2.527274e-01    54</code></pre>
<div id="exercise-identify-convergent-themes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-identify-convergent-themes" class="callout-inner">
<h3 class="callout-title">Exercise: Identify convergent themes</h3>
<div class="callout-content">
<p>Looking at your GO, KEGG, and Hallmark results:</p>
<ol style="list-style-type: decimal"><li>Are there biological themes that appear across multiple
databases?</li>
<li>Are there database-specific findings that don’t replicate
elsewhere?</li>
<li>How would you prioritize which findings to investigate further?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>Convergent signals to look for:</p>
<ul><li>If GO shows “cell cycle” terms, KEGG might show “Cell cycle”
pathway, and Hallmark might show “HALLMARK_G2M_CHECKPOINT”.</li>
<li>If GO shows “immune response”, KEGG might show “Cytokine-cytokine
receptor interaction”, and Hallmark might show
“HALLMARK_INFLAMMATORY_RESPONSE”.</li>
</ul><p>Database-specific findings may indicate: - Pathway coverage
differences. - Annotation biases. - Potentially spurious signals
(require validation).</p>
<p>Prioritization strategy: 1. Focus on themes appearing in multiple
databases. 2. Consider effect size (how many genes, fold enrichment). 3.
Assess biological plausibility given experimental design. 4. Validate
key findings with independent methods.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5b-direction-aware-enrichment-analysis">Step 5b: Direction-aware enrichment analysis<a class="anchor" aria-label="anchor" href="#step-5b-direction-aware-enrichment-analysis"></a></h2>
<hr class="half-width"><p>ORA on all significant genes (up + down combined) can mask biological
signals when up-regulated and down-regulated genes have different
functions. Analyzing directions separately often reveals clearer
patterns.</p>
<div class="section level3">
<h3 id="prepare-direction-specific-gene-lists">Prepare direction-specific gene lists<a class="anchor" aria-label="anchor" href="#prepare-direction-specific-gene-lists"></a></h3>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Up-regulated genes</span></span>
<span><span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="va">log2fc_cut</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span><span class="op">)</span></span>
<span><span class="va">up_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">up_genes</span><span class="op">)</span></span>
<span><span class="va">up_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">ENSEMBL</span> <span class="op">%in%</span> <span class="va">up_clean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Down-regulated genes</span></span>
<span><span class="va">down_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">log2fc_cut</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span><span class="op">)</span></span>
<span><span class="va">down_clean</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">down_genes</span><span class="op">)</span></span>
<span><span class="va">down_entrez</span> <span class="op">&lt;-</span> <span class="va">id_map</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">ENSEMBL</span> <span class="op">%in%</span> <span class="va">down_clean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">message</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Up-regulated: %d genes, Down-regulated: %d genes"</span>,</span>
<span>    <span class="fu">length</span><span class="op">(</span><span class="va">up_entrez</span><span class="op">)</span>, <span class="fu">length</span><span class="op">(</span><span class="va">down_entrez</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>Up-regulated: 1810 genes, Down-regulated: 1798 genes</code></pre>
</div>
<div class="section level3">
<h3 id="run-enrichment-for-each-direction">Run enrichment for each direction<a class="anchor" aria-label="anchor" href="#run-enrichment-for-each-direction"></a></h3>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># GO enrichment for up-regulated genes</span></span>
<span><span class="va">ego_up</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="va">up_entrez</span>,</span>
<span>    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>    ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GO enrichment for down-regulated genes</span></span>
<span><span class="va">ego_down</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="va">down_entrez</span>,</span>
<span>    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>    ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="compare-directions">Compare directions<a class="anchor" aria-label="anchor" href="#compare-directions"></a></h3>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine top results for comparison</span></span>
<span><span class="va">up_top</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="va">ego_up</span><span class="op">@</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>Direction <span class="op">=</span> <span class="st">"Up-regulated"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">Direction</span>, <span class="va">Description</span>, <span class="va">p.adjust</span>, <span class="va">Count</span><span class="op">)</span></span>
<span></span>
<span><span class="va">down_top</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="va">ego_down</span><span class="op">@</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>Direction <span class="op">=</span> <span class="st">"Down-regulated"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">Direction</span>, <span class="va">Description</span>, <span class="va">p.adjust</span>, <span class="va">Count</span><span class="op">)</span></span>
<span></span>
<span><span class="va">direction_comparison</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">up_top</span>, <span class="va">down_top</span><span class="op">)</span></span>
<span><span class="va">direction_comparison</span></span></code></pre>
</div>
<p>Output:</p>
<pre><code>                Direction                                                                              Description     p.adjust Count
GO:0002694   Up-regulated                                                       regulation of leukocyte activation 7.705961e-07   127
GO:0002683   Up-regulated                                             negative regulation of immune system process 9.130747e-07   113
GO:0050865   Up-regulated                                                            regulation of cell activation 1.389200e-06   131
GO:0002252   Up-regulated                                                                  immune effector process 1.389200e-06   130
GO:0007264   Up-regulated                                                small GTPase-mediated signal transduction 2.781020e-06    95
GO:0051249   Up-regulated                                                      regulation of lymphocyte activation 4.543747e-06   111
GO:0009617   Up-regulated                                                                    response to bacterium 4.720694e-06   120
GO:0002697   Up-regulated                                                    regulation of immune effector process 1.175944e-05    88
GO:0007186   Up-regulated                                             G protein-coupled receptor signaling pathway 1.175944e-05    97
GO:0030036   Up-regulated                                                          actin cytoskeleton organization 3.108071e-05   123
GO:0022613 Down-regulated                                                     ribonucleoprotein complex biogenesis 2.154137e-27   164
GO:0042254 Down-regulated                                                                      ribosome biogenesis 3.115494e-25   128
GO:0006364 Down-regulated                                                                          rRNA processing 4.368810e-19    92
GO:0016072 Down-regulated                                                                   rRNA metabolic process 6.776964e-19   100
GO:0042273 Down-regulated                                                       ribosomal large subunit biogenesis 1.090959e-08    33
GO:0042274 Down-regulated                                                       ribosomal small subunit biogenesis 4.296891e-08    45
GO:0006399 Down-regulated                                                                   tRNA metabolic process 1.002785e-06    62
GO:0000470 Down-regulated                                                                   maturation of LSU-rRNA 2.496138e-05    17
GO:0009451 Down-regulated                                                                         RNA modification 4.970122e-05    48
GO:0000463 Down-regulated maturation of LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA) 7.457449e-05    13</code></pre>
<div id="when-to-analyze-directions-separately" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="when-to-analyze-directions-separately" class="callout-inner">
<h3 class="callout-title">When to analyze directions separately</h3>
<div class="callout-content">
<p>Separate analysis is particularly useful when:</p>
<ul><li>You expect different biological processes to be activated
vs. suppressed.</li>
<li>Combined analysis shows few significant results (opposing signals
may cancel out).</li>
<li>You want to understand mechanism (what’s gained vs. lost).</li>
</ul><p>Separate analysis is less useful when:</p>
<ul><li>You have very few DE genes in one direction.</li>
<li>The biological question is about overall pathway disruption.</li>
</ul></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5c-gene-set-enrichment-analysis-gsea">Step 5c: Gene Set Enrichment Analysis (GSEA)<a class="anchor" aria-label="anchor" href="#step-5c-gene-set-enrichment-analysis-gsea"></a></h2>
<hr class="half-width"><p>ORA has a fundamental limitation: it requires an arbitrary cutoff to
define “significant” genes. <strong>GSEA</strong> uses the full ranked
gene list without cutoffs.</p>
<div class="section level3">
<h3 id="how-gsea-differs-from-ora">How GSEA differs from ORA<a class="anchor" aria-label="anchor" href="#how-gsea-differs-from-ora"></a></h3>
<table style="width:100%;" class="table"><colgroup><col width="42%"><col width="26%"><col width="31%"></colgroup><thead><tr class="header"><th>Aspect</th>
<th>ORA</th>
<th>GSEA</th>
</tr></thead><tbody><tr class="odd"><td>Input</td>
<td>List of significant genes</td>
<td>All genes, ranked by fold change</td>
</tr><tr class="even"><td>Cutoff</td>
<td>Required (padj, LFC)</td>
<td>None</td>
</tr><tr class="odd"><td>Sensitivity</td>
<td>May miss coordinated small changes</td>
<td>Detects subtle but consistent shifts</td>
</tr><tr class="even"><td>Gene weights</td>
<td>All DE genes treated equally</td>
<td>Magnitude of change matters</td>
</tr></tbody></table></div>
<div class="section level3">
<h3 id="prepare-ranked-gene-list-for-gsea">Prepare ranked gene list for GSEA<a class="anchor" aria-label="anchor" href="#prepare-ranked-gene-list-for-gsea"></a></h3>
<p>GSEA requires a named vector of gene scores (typically log2 fold
change or a signed significance score):</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create ranked list using signed -log10(pvalue) * sign(LFC)</span></span>
<span><span class="co"># This weights by both significance and direction</span></span>
<span><span class="va">res_gsea</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span></span>
<span>        ensembl_clean <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"\\.\\d+$"</span>, <span class="st">""</span>, <span class="va">ensembl_gene_id_version</span><span class="op">)</span>,</span>
<span>        rank_score <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">pvalue</span><span class="op">)</span> <span class="op">*</span> <span class="fu">sign</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Map to Entrez IDs</span></span>
<span><span class="va">res_gsea</span> <span class="op">&lt;-</span> <span class="va">res_gsea</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">inner_join</span><span class="op">(</span><span class="va">id_map</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"ensembl_clean"</span> <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create named vector (required format for GSEA)</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="va">res_gsea</span><span class="op">$</span><span class="va">rank_score</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">gene_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res_gsea</span><span class="op">$</span><span class="va">ENTREZID</span></span>
<span></span>
<span><span class="co"># Sort by rank score (required for GSEA)</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="fu">sort</span><span class="op">(</span><span class="va">gene_list</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">head</span>(gene_list)</span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>   <span class="dv">20343</span>   <span class="dv">108105</span>    <span class="dv">12481</span>    <span class="dv">66222</span>    <span class="dv">23943</span>    <span class="dv">15002</span> </span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a><span class="fl">261.1024</span> <span class="fl">126.5745</span> <span class="fl">124.5315</span> <span class="fl">108.3381</span> <span class="fl">103.0995</span> <span class="fl">101.4362</span> </span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="fu">tail</span>(gene_list)</span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>   <span class="dv">230784</span>    <span class="dv">217882</span>     <span class="dv">17246</span>     <span class="dv">65964</span>     <span class="dv">12519</span>     <span class="dv">27015</span> </span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a><span class="sc">-</span><span class="fl">211.2999</span> <span class="sc">-</span><span class="fl">212.3574</span> <span class="sc">-</span><span class="fl">213.2033</span> <span class="sc">-</span><span class="fl">231.8965</span> <span class="sc">-</span><span class="fl">244.4231</span> <span class="sc">-</span><span class="fl">250.3965</span> </span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="run-gsea-with-clusterprofiler">Run GSEA with clusterProfiler<a class="anchor" aria-label="anchor" href="#run-gsea-with-clusterprofiler"></a></h3>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Sort the list in decreasing order (required by clusterProfiler)</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="fu">sort</span><span class="op">(</span><span class="va">gene_list</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># 2. Remove duplicate Entrez IDs </span></span>
<span><span class="co"># (Since we sorted by value, this keeps the instance with the highest absolute value)</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="va">gene_list</span><span class="op">[</span><span class="op">!</span><span class="fu">duplicated</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">gene_list</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># 3. Verify it is a numeric vector</span></span>
<span><span class="co"># (It should look like: 261.1, 126.5... with names "20343", "108105"...)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">gene_list</span><span class="op">)</span></span>
<span><span class="co"># 4. Run GSEA</span></span>
<span><span class="va">gsea_go</span> <span class="op">&lt;-</span> <span class="fu">gseGO</span><span class="op">(</span></span>
<span>    geneList     <span class="op">=</span> <span class="va">gene_list</span>,      <span class="co"># Pass the numeric vector, NOT names(gene_list)</span></span>
<span>    OrgDb        <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>    ont          <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    minGSSize    <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    maxGSSize    <span class="op">=</span> <span class="fl">500</span>,</span>
<span>    pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    verbose      <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">gsea_go</span><span class="op">)</span></span></code></pre>
</div>
<p>Output (columns truncated for brevity):</p>
<pre><code>                   ID                                   Description setSize enrichmentScore       NES       pvalue     p.adjust       qvalue rank
GO:0042254 GO:0042254                           ribosome biogenesis     304      -0.7713857 -2.022231 1.000000e-10 0.0000001312 1.169211e-07 1867
GO:0006364 GO:0006364                               rRNA processing     210      -0.7694754 -1.950534 1.000000e-10 0.0000001312 1.169211e-07 1867
GO:0016072 GO:0016072                        rRNA metabolic process     241      -0.7525383 -1.933838 1.000000e-10 0.0000001312 1.169211e-07 1867
GO:0022613 GO:0022613          ribonucleoprotein complex biogenesis     425      -0.7186644 -1.919595 1.000000e-10 0.0000001312 1.169211e-07 1867
GO:0042770 GO:0042770 signal transduction in response to DNA damage     170      -0.7420666 -1.849975 1.315390e-07 0.0001380633 1.230374e-04  892
GO:0042274 GO:0042274            ribosomal small subunit biogenesis     106      -0.7953618 -1.887333 2.644409e-07 0.0002312976 2.061247e-04  948</code></pre>
</div>
<div class="section level3">
<h3 id="visualize-gsea-results">Visualize GSEA results<a class="anchor" aria-label="anchor" href="#visualize-gsea-results"></a></h3>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Dot plot of enriched gene sets</span></span>
<span><span class="fu">dotplot</span><span class="op">(</span><span class="va">gsea_go</span>, showCategory <span class="op">=</span> <span class="fl">20</span>, split <span class="op">=</span> <span class="st">".sign"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">facet_grid</span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">.sign</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GSEA: GO Biological Process"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/gsea-1.png" alt="GSEA dotplot for GO Biological Process showing activated and suppressed gene sets" class="figure mx-auto d-block"><figcaption>
GSEA dotplot for GO Biological Process showing activated and suppressed
gene sets
</figcaption></figure><p>The <code>.sign</code> column indicates whether the gene set is
enriched in up-regulated (<code>activated</code>) or down-regulated
(<code>suppressed</code>) genes.</p>
</div>
<div class="section level3">
<h3 id="gsea-enrichment-plot-for-a-specific-term">GSEA enrichment plot for a specific term<a class="anchor" aria-label="anchor" href="#gsea-enrichment-plot-for-a-specific-term"></a></h3>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Enrichment plot for top gene set</span></span>
<span><span class="fu">gseaplot2</span><span class="op">(</span><span class="va">gsea_go</span>, geneSetID <span class="op">=</span> <span class="fl">1</span>, title <span class="op">=</span> <span class="va">gsea_go</span><span class="op">$</span><span class="va">Description</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-enrich/gsea-2b.png" alt="GSEA enrichment plot showing running enrichment score for top gene set" class="figure mx-auto d-block"><figcaption>
GSEA enrichment plot showing running enrichment score for top gene set
</figcaption></figure><div id="interpreting-gsea-results" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-gsea-results" class="callout-inner">
<h3 class="callout-title">Interpreting GSEA results</h3>
<div class="callout-content">
<p>Key metrics:</p>
<ul><li>
<strong>NES (Normalized Enrichment Score)</strong>: Direction and
strength of enrichment. Positive = enriched in up-regulated genes;
Negative = enriched in down-regulated genes.</li>
<li>
<strong>pvalue/p.adjust</strong>: Statistical significance.</li>
<li>
<strong>core_enrichment</strong>: Genes contributing most to the
enrichment signal (“leading edge”).</li>
</ul><p>GSEA is particularly valuable when:</p>
<ul><li>Changes are subtle but coordinated across many genes.</li>
<li>You want to avoid arbitrary significance cutoffs.</li>
<li>You’re comparing with published gene signatures.</li>
</ul></div>
</div>
</div>
<div id="exercise-compare-ora-and-gsea-results" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-compare-ora-and-gsea-results" class="callout-inner">
<h3 class="callout-title">Exercise: Compare ORA and GSEA results</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Do the same pathways appear in both ORA and GSEA results?</li>
<li>Does GSEA identify any pathways that ORA missed?</li>
<li>For pathways appearing in both, is the direction (NES sign)
consistent with the ORA direction analysis?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>Expected observations:</p>
<ul><li>Many top pathways should overlap between ORA and GSEA.</li>
<li>GSEA may identify additional pathways where changes are subtle but
coordinated.</li>
<li>The NES sign (positive/negative) should match whether the pathway
was enriched in up- or down-regulated genes in ORA.</li>
<li>Discrepancies may indicate pathways where the signal comes from many
small changes rather than a few large ones.</li>
</ul></div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-6-save-results">Step 6: Save results<a class="anchor" aria-label="anchor" href="#step-6-save-results"></a></h2>
<hr class="half-width"><p>Save all enrichment results:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create output directory</span></span>
<span><span class="fu">dir.create</span><span class="op">(</span><span class="st">"results/enrichment"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GO results</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ego_bp</span><span class="op">)</span>, <span class="st">"results/enrichment/GO_BP_enrichment.csv"</span><span class="op">)</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ego_bp_simple</span><span class="op">)</span>, <span class="st">"results/enrichment/GO_BP_simplified.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># KEGG results</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ekegg</span><span class="op">)</span>, <span class="st">"results/enrichment/KEGG_enrichment.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Hallmark results</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ehall</span><span class="op">)</span>, <span class="st">"results/enrichment/Hallmark_enrichment.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Direction-aware results</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span>, <span class="st">"results/enrichment/GO_BP_up_regulated.csv"</span><span class="op">)</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ego_down</span><span class="op">)</span>, <span class="st">"results/enrichment/GO_BP_down_regulated.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GSEA results</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">gsea_go</span><span class="op">)</span>, <span class="st">"results/enrichment/GSEA_GO_BP.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>Save plots:</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># GO dotplot</span></span>
<span><span class="fu">pdf</span><span class="op">(</span><span class="st">"results/enrichment/GO_BP_dotplot.pdf"</span>, width <span class="op">=</span> <span class="fl">10</span>, height <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ego_bp_simple</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO Biological Process enrichment"</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># KEGG barplot</span></span>
<span><span class="fu">pdf</span><span class="op">(</span><span class="st">"results/enrichment/KEGG_barplot.pdf"</span>, width <span class="op">=</span> <span class="fl">10</span>, height <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">barplot</span><span class="op">(</span><span class="va">ekegg</span>, showCategory <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"KEGG pathway enrichment"</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Hallmark dotplot</span></span>
<span><span class="fu">pdf</span><span class="op">(</span><span class="st">"results/enrichment/Hallmark_dotplot.pdf"</span>, width <span class="op">=</span> <span class="fl">10</span>, height <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ehall</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"MSigDB Hallmark enrichment"</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="interpretation-guidelines">Interpretation guidelines<a class="anchor" aria-label="anchor" href="#interpretation-guidelines"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="signs-of-meaningful-enrichment">Signs of meaningful enrichment<a class="anchor" aria-label="anchor" href="#signs-of-meaningful-enrichment"></a></h3>
<p>Enrichment results are more reliable when:</p>
<ul><li>
<strong>Multiple related terms</strong> appear (not just one
isolated hit).</li>
<li>
<strong>Consistent themes</strong> emerge across databases.</li>
<li>
<strong>Biological plausibility</strong>: Results align with known
biology of your system.</li>
<li>
<strong>Reasonable gene counts</strong>: Terms with many genes
(&gt;5-10) are more robust.</li>
<li>
<strong>Moderate p-values</strong>: Extremely low p-values (e.g.,
10^-50) may indicate annotation bias.</li>
</ul></div>
<div class="section level3">
<h3 id="common-pitfalls">Common pitfalls<a class="anchor" aria-label="anchor" href="#common-pitfalls"></a></h3>
<ol style="list-style-type: decimal"><li>
<strong>Over-interpretation</strong>: A significant p-value doesn’t
prove causation.</li>
<li>
<strong>Ignoring effect size</strong>: A term with 3/1000 genes is
less meaningful than 30/100.</li>
<li>
<strong>Cherry-picking</strong>: Report all significant results, not
just expected ones.</li>
<li>
<strong>Universe mismatch</strong>: Using wrong background inflates
or deflates significance.</li>
<li>
<strong>Outdated annotations</strong>: Gene function annotations
improve over time.</li>
</ol></div>
<div class="section level3">
<h3 id="reporting-enrichment-results">Reporting enrichment results<a class="anchor" aria-label="anchor" href="#reporting-enrichment-results"></a></h3>
<p>In publications, include:</p>
<ul><li>Method and software versions used.</li>
<li>Background universe definition.</li>
<li>Multiple testing correction method.</li>
<li>Full results tables (supplementary).</li>
<li>Number of input genes and those successfully mapped.</li>
</ul></div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>ORA tests if gene sets contain more DE genes than expected by chance
using the hypergeometric test.</li>
<li>The background universe must include all genes that could have been
detected as DE.</li>
<li>Multiple testing correction (FDR) is essential when testing
thousands of gene sets.</li>
<li>GO provides comprehensive but redundant functional annotation; use
<code>simplify()</code> to reduce redundancy.</li>
<li>KEGG provides curated pathway maps but has limited gene coverage and
requires Entrez IDs.</li>
<li>MSigDB Hallmark sets offer 50 high-quality, non-redundant biological
signatures.</li>
<li>
<strong>Direction-aware analysis</strong> (up vs. down separately)
often reveals clearer biological signals.</li>
<li>
<strong>GSEA</strong> avoids arbitrary cutoffs by using the full
ranked gene list; it detects subtle but coordinated changes.</li>
<li>Convergent findings across multiple databases and methods strengthen
biological interpretation.</li>
<li>Always report methods, thresholds, and full results for
reproducibility.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/05b-deseq2-expression-analyses-kallisto.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/05b-deseq2-expression-analyses-kallisto.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: B. Differential
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/06-gene-set-enrichment-analyses.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.18.3" class="external-link">sandpaper (0.18.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/06-gene-set-enrichment-analyses.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "Gene set enrichment analysis",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/06-gene-set-enrichment-analyses.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/06-gene-set-enrichment-analyses.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2026-01-22",
  "datePublished": "2026-01-22"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

