<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>RNA-seq in practice: Hands-on workshop on RCAC systems: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png">
<link rel="manifest" href="../favicons/rcac/site.webmanifest">
<link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01a-from-reads-to-counts.html">2. Downloading and organizing files</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="01b-mapping-reads.html">3. Mapping reads to a reference genome</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="01c-count-features.html">4. Quantifying Gene Expression</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="02-setup.html">5. RStudio Project and Experimental Data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="03-import-annotate.html">6. Importing and annotating quantified data into R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="04-exploratory-qc.html">7. Exploratory analysis and quality control</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="05-differential-expression.html">8. Differential expression analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="06-extra-design.html">9. Extra exploration of design matrices</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="07-gene-set-analysis.html">10. Gene set enrichment analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="08-next-steps.html">11. Next steps</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-intro-to-rnaseq"><p>Content from <a href="01-intro-to-rnaseq.html">Introduction to RNA-seq</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/01-intro-to-rnaseq.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 100 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the different choices to consider when planning an RNA-seq
experiment?</li>
<li>How does one process the raw fastq files to generate a table with
read counts per gene and sample?</li>
<li>Where does one find information about annotated genes for a given
organism?</li>
<li>What are the typical steps in an RNA-seq analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what RNA-seq is.</li>
<li>Describe some of the most common design choices that have to be made
before running an RNA-seq experiment.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
<li>Show some common types of results and visualizations generated in
RNA-seq analyses.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-are-we-measuring-in-an-rna-seq-experiment">What are we measuring in an RNA-seq experiment?<a class="anchor" aria-label="anchor" href="#what-are-we-measuring-in-an-rna-seq-experiment"></a>
</h2>
<hr class="half-width">
<figure><img src="../fig/alternative_splicing_sub.png" alt="Illustration of part of the central dogma of molecular biology, where DNA is transcribed to RNA, and intronic sequences are spliced out" class="figure mx-auto d-block"></figure><p>In order to produce an RNA molecule, a stretch of DNA is first
transcribed into mRNA. Subsequently, intronic regions are spliced out,
and exonic regions are combined into different <em>isoforms</em> of a
gene.</p>
<figure><img src="../fig/rnaseq_martin_wang.png" alt="Illustration of the major experimental steps of an RNA-seq experiment" class="figure mx-auto d-block"></figure><p>(figure adapted from <a href="https://www.nature.com/articles/nrg3068" class="external-link">Martin &amp; Wang
(2011)</a>).</p>
<p>In a typical RNA-seq experiment, RNA molecules are first collected
from a sample of interest. After a potential enrichment for molecules
with polyA tails (predominantly mRNA), or depletion of otherwise highly
abundant ribosomal RNA, the remaining molecules are fragmented into
smaller pieces (there are also long-read protocols where entire
molecules are considered, but those are not the focus of this lesson).
It is important to keep in mind that because of the splicing excluding
intronic sequences, an RNA molecule (and hence a generated fragment) may
not correspond to an uninterrupted region of the genome. The RNA
fragments are then reverse transcribed into cDNA, whereafter sequencing
adapters are added to each end. These adapters allow the fragments to
attach to the flow cell. Once attached, each fragment will be heavily
amplified to generate a cluster of identical sequences on the flow cell.
The sequencer then determines the sequence of the first 50-200
nucleotides of the cDNA fragments in each such cluster, starting from
one end, which corresponds to a <em>read</em>. Many data sets are
generated with so called paired-end protocols, in which the fragments
are read from both ends. Millions of such reads (or pairs of reads) will
be generated in an experiment, and these will be represented in a (pair
of) FASTQ files. Each read is represented by four consecutive lines in
such a file: first a line with a unique read identifier, next the
inferred sequence of the read, then another identifier line, and finally
a line containing the base quality for each inferred nucleotide,
representing the probability that the nucleotide in the corresponding
position has been correctly identified.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>What are potential advantages and disadvantages of paired-end
protocols compared to only sequencing one end of a fragment?</li>
<li>What quality assessment can you think of that would be useful to
perform on the FASTQ files with read sequences?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="experimental-design-considerations">Experimental design considerations<a class="anchor" aria-label="anchor" href="#experimental-design-considerations"></a>
</h2>
<hr class="half-width">
<p>Before starting to collect data, it is essential to take some time to
think about the design of the experiment. Experimental design concerns
the organization of experiments with the purpose of making sure that the
right type of data, and enough of it, is available to answer the
questions of interest as efficiently as possible. Aspects such as which
conditions or sample groups to consider, how many replicates to collect,
and how to plan the data collection in practice are important questions
to consider. Many high-throughput biological experiments (including
RNA-seq) are sensitive to ambient conditions, and it is often difficult
to directly compare measurements that have been done on different days,
by different analysts, in different centers, or using different batches
of reagents. For this reason, it is very important to design experiments
properly, to make it possible to disentangle different types of (primary
and secondary) effects.</p>
<figure><img src="../fig/factors-affecting-measurements.png" alt="A classification of many different factors affecting measurements obtained from an experiment into treatment, biological, technical and error effects" class="figure mx-auto d-block"></figure><p>(figure from <a href="https://www.amazon.com/Experimental-Design-Laboratory-Biologists-Reproducibility/dp/1107424887" class="external-link">Lazic
(2017)</a>).</p>
<div id="challenge-discuss-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss with your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Why is it essential to have replicates?</li>
</ol>
</div>
</div>
</div>
<p>Importantly, not all replicates are equally useful, from a
statistical point of view. One common way to classify the different
types of replicates is as ‘biological’ and ‘technical’, where the latter
are typically used to test the reproducibility of the measurement
device, while biological replicates inform about the variability between
different samples from a population of interest. Another scheme
classifies replicates (or units) into ‘biological’, ‘experimental’ and
‘observational’. Here, biological units are entities we want to make
inferences about (e.g., animals, persons). Replication of biological
units is required to make a general statement of the effect of a
treatment - we can not draw a conclusion about the effect of drug on a
population of mice by studying a single mouse only. Experimental units
are the smallest entities that can be <em>independently assigned to a
treatment</em> (e.g., animal, litter, cage, well). Only replication of
experimental units constitute true replication. Observational units,
finally, are entities at which measurements are made.</p>
<p>To explore the impact of experimental design on the ability to answer
questions of interest, we are going to use an interactive application,
provided in the <a href="https://csoneson.github.io/ConfoundingExplorer/" class="external-link">ConfoundingExplorer</a>
package.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Launch the ConfoundingExplorer application and familiarize yourself
with the interface.</p>
</div>
</div>
</div>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>For a balanced design (equal distribution of replicates between the
two groups in each batch), what is the effect of increasing the strength
of the batch effect? Does it matter whether one adjusts for the batch
effect or not?</li>
<li>For an increasingly unbalanced design (most or all replicates of one
group coming from one batch), what is the effect of increasing the
strength of the batch effect? Does it matter whether one adjusts for the
batch effect or not?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="rna-seq-quantification-from-reads-to-count-matrix">RNA-seq quantification: from reads to count matrix<a class="anchor" aria-label="anchor" href="#rna-seq-quantification-from-reads-to-count-matrix"></a>
</h2>
<hr class="half-width">
<figure><img src="../fig/rna-quant.jpeg" alt="Illustration of a set of reads generated by a sequencer, and genomic and transcriptomic reference sequences" class="figure mx-auto d-block"></figure><p>The read sequences contained in the FASTQ files from the sequencer
are typically not directly useful as they are, since we do not have the
information about which gene or transcript they originate from. Thus,
the first processing step is to attempt to identify the location of
origin for each read, and use this to obtain an estimate of the number
of reads originating from a gene (or another features, such as an
individual transcript). This can then be used as a proxy for the
abundance, or expression level, of the gene. There is a plethora of RNA
quantification pipelines, and the most common approaches can be
categorized into three main types:</p>
<ol style="list-style-type: decimal">
<li><p>Align reads to the genome, and count the number of reads that map
within the exons of each gene. This is the one of simplest methods. For
species for which the transcriptome is poorly annotated, this would be
the preferred approach. Example: <code>STAR</code> alignment to GRCm39 +
<code>Rsubread</code> <a href="https://doi.org/10.1093%2Fnar%2Fgkz114" class="external-link">featureCounts</a></p></li>
<li><p>Align reads to the transcriptome, quantify transcript expression,
and summarize transcript expression into gene expression. This approach
can produce accurate quantification results (<a href="https://doi.org/10.1186/s13059-016-0940-1" class="external-link">independent
benchmarking</a>), particularly for high-quality samples without DNA
contamination. Example: RSEM quantification using
<code>rsem-calculate-expression --star</code> on the GENCODE GRCh38
transcriptome + <code>tximport</code></p></li>
<li><p>Pseudoalign reads against the transcriptome, using the
corresponding genome as a decoy, quantifying transcript expression in
the process, and summarize the transcript-level expression into
gene-level expression. The advantages of this approach include:
computational efficiency, mitigation of the effect of DNA contamination,
and GC bias correction. Example: <code>salmon quant --gcBias</code> +
<code>tximport</code></p></li>
</ol>
<p>At typical sequencing read depth, gene expression quantification is
often more accurate than transcript expression quantification. However,
differential gene expression analyses can be <a href="https://doi.org/10.12688/f1000research.7563.1" class="external-link">improved</a> by
having access also to transcript-level quantifications.</p>
<p>Other tools used in RNA-seq quantification include: TopHat2, bowtie2,
kallisto, HTseq, among many others.</p>
<p>The choice of an appropriate RNA-seq quantification depends on the
quality of the transcriptome annotation, the quality of the RNA-seq
library preparation, the presence of contaminating sequences, among many
factors. Often, it can be informative to compare the quantification
results of multiple approaches.</p>
<p>Because the best quantification method is species- and
experiment-dependent, and often requires large amounts of computing
resources, this workshop will not cover any specifics of how to generate
the counts. Instead, we recommend checking out the references above and
consulting with a local bioinformatics expert if you need help.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Which of the mentioned RNA-Seq quantification tools have you heard
about? Do you know other pros and cons of the methods?</li>
<li>Have you done your own RNA-Seq experiment? If so, what
quantification tool did you use and why did you choose it?</li>
<li>Do you have access to specific tools / local bioinformatics expert /
computational resources for quantification? If you don’t, how might you
gain access?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="finding-the-reference-sequences">Finding the reference sequences<a class="anchor" aria-label="anchor" href="#finding-the-reference-sequences"></a>
</h2>
<hr class="half-width">
<p>In order to quantify abundances of known genes or transcripts from
RNA-seq data, we need a reference database informing us of the sequence
of these features, to which we can then compare our reads. This
information can be obtained from different online repositories. It is
highly recommended to choose one of these for a particular project, and
not mix information from different sources. Depending on the
quantification tool you have chosen, you will need different types of
reference information. If you are aligning your reads to the genome and
investigating the overlap with known annotated features, you will need
the full genome sequence (provided in a fasta file) and a file that
tells you the genomic location of each annotated feature (typically
provided in a gtf file). If you are mapping your reads to the
transcriptome, you will instead need a file with the sequence of each
transcript (again, a fasta file).</p>
<ul>
<li>If you are working with mouse or human samples, the <a href="https://www.gencodegenes.org/" class="external-link">GENCODE project</a> provides
well-curated reference files.</li>
<li>
<a href="https://www.ensembl.org/info/data/ftp/index.html" class="external-link">Ensembl</a>
provides reference files for a large set of organisms, including <a href="https://plants.ensembl.org/info/data/ftp/index.html" class="external-link">plants</a>
and <a href="https://fungi.ensembl.org/info/data/ftp/index.html" class="external-link">fungi</a>.</li>
<li>
<a href="https://hgdownload.soe.ucsc.edu/downloads.html" class="external-link">UCSC</a>
also provides reference files for many organisms.</li>
</ul>
<div id="challenge-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Download the latest mouse transcriptome fasta file from GENCODE. What
do the entries look like? Tip: to read the file into R, consider the
<code>readDNAStringSet()</code> function from the
<code>Biostrings</code> package.</p>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="where-are-we-heading-towards-in-this-workshop">Where are we heading towards in this workshop?<a class="anchor" aria-label="anchor" href="#where-are-we-heading-towards-in-this-workshop"></a>
</h1>
<p>During the coming two days, we will discuss and practice how to
perform differential expression analysis with Bioconductor, and how to
interpret the results. We will start from a count matrix, and thus
assume that the initial quality assessment and quantification of gene
expression have already been done. The outcome of a differential
expression analysis is often represented using graphical
representations, such as MA plots and heatmaps (see below for
examples).</p>
<p><img src="../fig/maPlot-example.png" alt="An example MA plot" class="figure"><img src="../fig/heatmap-example.png" alt="An example heatmap" class="figure"></p>
<p>In the following episodes we will learn, among other things, how to
generate and interpret these plots. It is also common to perform
follow-up analyses to investigate whether there is a functional
relationship among the top-ranked genes, so called gene set (enrichment)
analysis, which will also be covered in a later episode.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>RNA-seq is a technique of measuring the amount of RNA expressed
within a cell/tissue and state at a given time.</li>
<li>Many choices have to be made when planning an RNA-seq experiment,
such as whether to perform poly-A selection or ribosomal depletion,
whether to apply a stranded or an unstranded protocol, and whether to
sequence the reads in a single-end or paired-end fashion. Each of the
choices have consequences for the processing and interpretation of the
data.</li>
<li>Many approaches exist for quantification of RNA-seq data. Some
methods align reads to the genome and count the number of reads
overlapping gene loci. Other methods map reads to the transcriptome and
use a probabilistic approach to estimate the abundance of each gene or
transcript.</li>
<li>Information about annotated genes can be accessed via several
sources, including Ensembl, UCSC and GENCODE.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div></section><section id="aio-01a-from-reads-to-counts"><p>Content from <a href="01a-from-reads-to-counts.html">Downloading and organizing files</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/01a-from-reads-to-counts.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we process raw RNAseq reads to generate count data?</li>
<li>What are the key steps involved in quality control, alignment, and
quantification of RNAseq data?</li>
<li>How do different tools and parameters affect the outcome of RNAseq
data processing?</li>
<li>Why is it important to ensure accurate and consistent read mapping
and quantification?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Participants will learn the essential steps from raw read quality
control to the generation of count data.</li>
<li>Hands-on practice with <code>FastQC</code> to assess the quality of
raw reads.</li>
<li>Learn how to use <code>STAR</code> to map reads to a reference
genome.</li>
<li>Practice using <code>featureCounts</code> from <code>subreads</code>
package to generate count data from aligned reads.</li>
<li>Develop the ability to assess the quality and consistency of the
resulting count data, identifying and addressing common issues.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In RNAseq analysis, transforming raw sequencing data into meaningful
insights requires a series of well-defined steps. This episode will
guide you through the essential process of converting raw RNAseq reads
into count data, which forms the basis for downstream analyses such as
differential expression and functional annotation.</p>
<p>We will start with quality control to ensure that the raw data is
suitable for analysis, followed by the alignment of reads to a reference
genome or transcriptome. Finally, you’ll learn how to quantify gene
expression levels by counting the aligned reads associated with each
gene. Through hands-on exercises, you’ll gain practical experience with
the tools and techniques that are critical for accurate RNAseq data
processing.</p>
<p>Firstly, the reads in FASTQ format are usually stored or transferred
in compressed formats, often with an extension of <code>.fastq.gz</code>
or <code>.fq.gz</code>. This compression helps save storage space and
reduces the time required for data transfer.</p>
<div id="should-i-uncompress-fastq-files" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-uncompress-fastq-files" class="callout-inner">
<h3 class="callout-title">Should I Uncompress fastq files?</h3>
<div class="callout-content">
<p>Uncompressing FASTQ files is often unnecessary since many RNAseq
tools can directly process compressed files, such as those in
<code>.gz</code> format. Keeping FASTQ files compressed not only saves
storage space but also reduces the time and computational resources
required for data handling. It’s generally more efficient to work with
compressed files, especially when dealing with large datasets.</p>
</div>
</div>
</div>
<p>Secondly, library preparation steps of RNAseq usually add adapters
and other technical sequences that may be present at the ends of
reads.</p>
<div id="should-i-trim-adapters-for-rnaseq-analyses" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-trim-adapters-for-rnaseq-analyses" class="callout-inner">
<h3 class="callout-title">Should I trim adapters for RNAseq
analyses?</h3>
<div class="callout-content">
<p>For most RNAseq analyses, trimming adapters is not necessary, as many
popular mapping programs, like HISAT2 and STAR, can effectively
soft-clip adapter sequences during the alignment process. However, if
you are performing more specialized analyses, such as transcript
assembly or other applications where read length consistency is crucial,
trimming adapters may be important. Aggressive trimming might actually
negatively impact expression analyses! <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/" class="external-link">More
information here</a>.</p>
</div>
</div>
</div>
<p>For this lesson, we will use a dataset from the experiment
<fillthissection>. We will start by examining the quality of the reads,
aligning them to a reference genome, and quantifying gene expression
levels to generate count data. Before we begin, let’s ensure that we
have all the data necessary for the analysis.</fillthissection></p>
</section><section><h2 class="section-heading" id="downloading-the-data-and-organization">Downloading the data and organization<a class="anchor" aria-label="anchor" href="#downloading-the-data-and-organization"></a>
</h2>
<hr class="half-width">
<p>We will begin by downloading the raw reads for our experiment from
NCBI GEO database. In real world, you will probably obtain your raw
reads from the sequencing center or from your collaborators, but for
this lesson, we will download the pre-analyzed data from a public
repository.</p>
<div id="what-files-do-i-need-for-rnaseq-analysis" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-files-do-i-need-for-rnaseq-analysis" class="callout-inner">
<h3 class="callout-title">What files do I need for RNAseq analysis?</h3>
<div class="callout-content">
<p>The essential files for RNAseq analysis include raw reads, reference
genome, and annotation files. Raw reads are typically provided in FASTQ
format, which contains the nucleotide sequences and quality scores of
the reads. The reference genome is typically provided in FASTA format,
which contains the nucleotide sequences of the chromosomes or contigs.
The annotation file is usually in GTF (Gene Transfer Format) or GFF
(General Feature Format) format, which provides information about the
genomic features such as genes, transcripts, and exons.</p>
</div>
</div>
</div>
<p>Maintaining this organized structure ensures a streamlined workflow,
reduces the risk of errors, and facilitates easier access to files
throughout the analysis process.</p>
<div id="create-the-directories-needed-for-this-episode" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="create-the-directories-needed-for-this-episode" class="callout-inner">
<h3 class="callout-title">Create the directories needed for this
episode</h3>
<div class="callout-content">
<p>Create a directory to serve as the working directory for the rest of
this episode and lesson (the workshop example uses a directory called
<code>rcac_rnaseq</code>). Then, within this chosen directory, create
the four fundamental directories previously discussed
(<code>data</code>, <code>scripts</code>, <code>mapping</code>, and
<code>counts</code>).</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> rcac_rnaseq/<span class="dt">{data</span><span class="op">,</span><span class="dt">scripts</span><span class="op">,</span><span class="dt">mapping</span><span class="op">,</span><span class="dt">counts}</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<figure style="text-align: center"><img src="../fig/01_fq2counts/dir_org.png" alt="Your working directory should look like this" width="1163" class="figure mx-auto d-block"><figcaption>
Your working directory should look like this
</figcaption></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="downloading-the-data">Downloading the data<a class="anchor" aria-label="anchor" href="#downloading-the-data"></a>
</h2>
<hr class="half-width">
<p>The data files include the raw reads, reference genome, and
annotation files. We will download them to the <code>data</code>
directory in our working directory.</p>
<div id="where-will-i-find-the-data-required-for-my-project" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="where-will-i-find-the-data-required-for-my-project" class="callout-inner">
<h3 class="callout-title">Where will I find the data required for my
project?</h3>
<div class="callout-content">
<p>The raw reads for your RNAseq experiment will typically be provided
by the sequencing facility or your collaborators. The reference genome
and annotation files can be downloaded from public repositories such as
<a href="https://www.ensembl.org/index.html" class="external-link">Ensembl</a>, <a href="https://genome.ucsc.edu/" class="external-link">UCSC Genome Browser</a>, or <a href="https://www.gencodegenes.org/" class="external-link">Gencode</a> or <a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a>. There are also
specialized databases for several organisms, such as <a href="https://flybase.org/" class="external-link">FlyBase</a>, <a href="https://wormbase.org/" class="external-link">WormBase</a>, <a href="https://www.arabidopsis.org/" class="external-link">TAIR</a>, and <a href="https://www.maizegdb.org/" class="external-link">MaizeGDB</a> for other model
organisms.</p>
</div>
</div>
</div>
<p>The raw reads were generated from a mouse model, and we will use the
mouse reference genome and annotation files for mapping and
quantification. We will use <a href="https://www.gencodegenes.org/mouse/" class="external-link">GenCode’s GRCm39</a>
reference genome and annotation files. Once we have the link, we can use
<code>wget</code> to download the files.</p>
<div id="annotation" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="annotation" class="callout-inner">
<h3 class="callout-title">Annotation</h3>
<div class="callout-content">
<p>Take a look at the <a href="https://www.gencodegenes.org/mouse/" class="external-link">GenCode’s GRCm39</a> website.
There are few choices for annotation files (and formats). What
annotation file (and format) will you download?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<figure style="text-align: center"><img src="../fig/01_fq2counts/gencode-gtf.png" alt="Basic gene annotation in GTF format" width="2615" class="figure mx-auto d-block"><figcaption>
Basic gene annotation in GTF format
</figcaption></figure>
</div>
</div>
</div>
</div>
<div id="reference-genome" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="reference-genome" class="callout-inner">
<h3 class="callout-title">Reference genome</h3>
<div class="callout-content">
<p>Similarly, what reference genome file will you download?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<figure style="text-align: center"><img src="../fig/01_fq2counts/gencode-genome.png" alt="Reference genome file in fasta format" width="2224" class="figure mx-auto d-block"><figcaption>
Reference genome file in fasta format
</figcaption></figure>
</div>
</div>
</div>
</div>
<p>Commands to donwload the files are as follows:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">cd</span> rcac_rnaseq</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="va">GTFlink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M35/gencode.vM35.primary_assembly.basic.annotation.gtf.gz"</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="va">GENOMElink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M35/GRCm39.primary_assembly.genome.fa.gz"</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Download the reference genome</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GENOMElink}</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co"># Download the annotation file</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GTFlink}</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Check the files</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="bu">cd</span> data</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># Uncompress the files</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="fu">gunzip</span> GRCm39.primary_assembly.genome.fa.gz</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="fu">gunzip</span> gencode.vM35.primary_assembly.basic.annotation.gtf.gz</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="raw-reads">Raw reads<a class="anchor" aria-label="anchor" href="#raw-reads"></a>
</h2>
<hr class="half-width">
<p>The raw reads for RNA-seq data required for the subsequent episodes
were generated to investigate the impact upper-respiratory infection
have on changes in RNA transcription in the cerebellum and spinal cord
of mice. This dataset was produced as part of the following study:</p>
<blockquote>
<p>Blackmore, Stephen, et al. “Influenza infection triggers disease in a
genetic model of experimental autoimmune encephalomyelitis.” Proceedings
of the National Academy of Sciences 114.30 (2017): E6107-E6116.</p>
</blockquote>
<p>The dataset is available at <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression Omnibus
(GEO)</a>, under the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870" class="external-link">GSE96870</a>.</p>
<figure style="text-align: center"><img src="../fig/01_fq2counts/geo-db.png" alt="FASTQ files from GEO" width="1637" class="figure mx-auto d-block"><figcaption>
FASTQ files from GEO
</figcaption></figure><div id="where-do-you-find-fastq-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="where-do-you-find-fastq-files" class="callout-inner">
<h3 class="callout-title">Where do you find FASTQ files?</h3>
<div class="callout-content">
<p>GEO provides information about the experiment and the raw data files
accessions. This information is needed for understanding the
experimental design and contrasts you need to make to get useful
biological insights. Although GEO provides the metadata, the dataset is
actually hosted in the SRA (Sequence Read Archive) database. How do you
find the raw reads for this experiment?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>Follow the <a href="https://www.ncbi.nlm.nih.gov/sra?term=SRP102241" class="external-link">SRP102241</a>
link and then click on the <code>Send results to Run selector</code> to
get the list of runs. This page will provide you with the list of runs
associated with the experiment. Clicking on the “Accession List” will
download all the SRR ids you need.</p>
<figure style="text-align: center"><img src="../fig/01_fq2counts/srr-runs.png" alt="FASTQ files from GEO" width="2753" class="figure mx-auto d-block"><figcaption>
FASTQ files from GEO
</figcaption></figure>
</div>
</div>
</div>
</div>
<p>For downloading the raw reads, we will use the
<code>fasterq-dump</code> tool from the SRA Toolkit. The following
command will download the raw reads for the experiment:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">module</span> load sratoolkit</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">SRR</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="ex">fasterq-dump</span> <span class="at">-O</span> data <span class="va">${SRR}</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="cf">done</span> <span class="op">&lt;</span> data/SRR_Acc_List.txt</span></code></pre>
</div>
<p>As there are many files to download, and the process may take some
time, we will use the previously downloaded files for the rest of the
episode. The files are located in the <code>depot</code> directory and
can be soft-linked to the <code>data</code> directory.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /depot/data/GSE96870/SRR<span class="pp">*</span> data/</span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Raw reads are typically provided in FASTQ format, which contains the
nucleotide sequences and quality scores of the reads.</li>
<li>Reference genome and annotation files are essential for mapping and
quantification of RNAseq data.</li>
<li>Organizing data files in a structured manner facilitates efficient
data handling and analysis.</li>
<li>Tools such as <code>wget</code> and <code>fasterq-dump</code> can be
used to download data files from public repositories.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-01b-mapping-reads"><p>Content from <a href="01b-mapping-reads.html">Mapping reads to a reference genome</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/01b-mapping-reads.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 45 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we map raw reads to a reference genome?</li>
<li>What programs are available for mapping reads to a reference
genome?</li>
<li>What opitons and parameters need to be considered when mapping?</li>
<li>How do to access the mapping results?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the process of mapping raw reads to a reference
genome.</li>
<li>Fine tune the mapping process by adjusting parameters.</li>
<li>Access the mapping results.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this episode, we will learn how to map raw reads to a reference
genome. Mapping is the process of aligning short reads to a reference
genome. The goal of mapping is to determine the location of the reads in
the reference genome. We will use the RNA-seq data that we downloaded
(soft-linked) in the previous episode and map them to the reference
genome using the STAR aligner. STAR is a fast and accurate RNA-seq
aligner that can align reads to the reference genome with high
sensitivity and specificity.</p>
<div id="what-mapping-programs-should-i-use" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-mapping-programs-should-i-use" class="callout-inner">
<h3 class="callout-title">What mapping programs should I use?</h3>
<div class="callout-content">
<p>One major requirement for a suitable RNAseq mapping program is that
it should be splice-aware. This means the program must be able to handle
reads that span exon-exon junctions, which is crucial for RNAseq data
since the reads are derived from spliced transcripts. <code>STAR</code>
is a widely used splice-aware aligner for RNAseq, known for its speed
and accuracy when mapping reads to a reference genome. Other popular
RNAseq aligners include <code>HISAT2</code> and, for certain
applications, <code>Bowtie2</code>. It’s important not to use DNA
aligners like <code>BWA</code>, which are not designed to handle spliced
reads.</p>
</div>
</div>
</div>
<div id="how-does-read-quality-affect-alignment-performance" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="how-does-read-quality-affect-alignment-performance" class="callout-inner">
<h3 class="callout-title">How does read quality affect alignment
performance?</h3>
<div class="callout-content">
<p>The quality of the reads can have a significant impact on the
alignment performance and thus your downstream analysis.</p>
<ul>
<li>
<strong>Uneven read quality</strong>: poor quality of the reads can
result in partial alignments or truncated read mapping, leaving you with
large portions of the read unmapped.</li>
<li>
<strong>Overrepresented sequences</strong>: such as ribosomal RNA,
or microbial contamination, can lead to too many multi-mapped
reads.</li>
<li>
<strong>Poor library preparation</strong>: can result in RNAseq
reads that consist primarily of adapter sequences without any actual
biological content - this can result in too many unmapped reads.</li>
<li>
<strong>Over-processing</strong> Short reads from over-trimming can
shorten reads to the point where they are too short to align uniquely,
leading to either multimapping or unmapped reads.</li>
</ul>
<p>It is crucial to check the quality of the reads before mapping and
also alignment statistics after mapping.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="preparing-for-read-mapping">Preparing for read mapping<a class="anchor" aria-label="anchor" href="#preparing-for-read-mapping"></a>
</h2>
<hr class="half-width">
<p>Before we start mapping the reads, we need to prepare the reference
genome. Mapping programs require the reference genome to be indexed.
Indexing is the process of creating a data structure that allows the
mapping program to quickly search the reference genome for the best
alignment of the reads. <code>STAR</code> requires the reference genome
to be indexed before mapping. Mapping programs can also use
<code>GTF</code> or <code>GFF</code> files to guide the alignment
process. These files contain information about the gene models in the
reference genome, such as the location of exons and introns, that can
speed up the alignment process.</p>
<p>We will use the reference genome and annotations that we downloaded
in the previous episode and index it using <code>STAR</code>.</p>
</section><section><h2 class="section-heading" id="indexing">Indexing<a class="anchor" aria-label="anchor" href="#indexing"></a>
</h2>
<hr class="half-width">
<p>STAR has <code>--runMode</code> option where you can alter the
function of the program. When <code>genomeGenerate</code> is specified
with this option, it can be used for indexing.</p>
<div id="generating-genome-indexes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="generating-genome-indexes" class="callout-inner">
<h3 class="callout-title">Generating genome indexes</h3>
<div class="callout-content">
<p>There are many options for the <code>STAR</code> program to index the
reference genome. You will need the reference genome fasta file that
needs to indexed, but there are many other options as well. Can you <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" class="external-link">check
the manual</a> and find options that are important for indexing the
reference genome?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In page 5, Section 2, “Generating genome indexes” discuss the options
for indexing the reference genome. You will need to use
<code>--runMode genomeGenerate</code> for indexing. Important options
for indexing are:</p>
<ul>
<li>
<code>runThreadN</code> number of threads to use</li>
<li>
<code>genomeDir</code> output directory for the genome indexes</li>
<li>
<code>genomeFastaFiles</code> reference genome fasta file</li>
<li>
<code>sjdbGTFfile</code> the GTF file</li>
<li>
<code>sjdbOverhang</code> default value of 100 works well, for most
cases this can be omitted</li>
</ul>
</div>
</div>
</div>
</div>
<p>We will also need SLURM directives to submit the job to the cluster.
The number of threads to use can be specified using the
<code>--cpus-per-task</code> directive. The script to index the
reference genome using <code>STAR</code> would look like this:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=4</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#SBATCH --job-name=index_genome</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.out</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="ex">ml</span> biocontainers</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="ex">ml</span> star</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/GRCm39.primary_assembly.genome.fa"</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="va">genomeDir</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/star_index"</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/gencode.vM35.primary_assembly.basic.annotation.gtf"</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>   <span class="at">--runThreadN</span> 4<span class="dt">\</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>   <span class="at">--runMode</span> genomeGenerate <span class="dt">\</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>   <span class="at">--genomeDir</span> <span class="va">${genomeDir}</span> <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>   <span class="at">--genomeFastaFiles</span> <span class="va">${genome}</span> <span class="dt">\</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>   <span class="at">--sjdbGTFfile</span> <span class="va">${gtf}</span></span></code></pre>
</div>
<p>The script can be saved in a file, say <code>index_genome.sh</code>,
within the <code>scripts</code> directory. It should be submitted as a
slurm job after making it executable.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">chmod</span> +x index_genome.sh</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">sbatch</span> index_genome.sh</span></code></pre>
</div>
<p>It takes about 30-45 minutes to complete the indexing process.</p>
<div id="is-gtf-file-necessary-for-indexing" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="is-gtf-file-necessary-for-indexing" class="callout-inner">
<h3 class="callout-title">Is GTF file necessary for indexing?</h3>
<div class="callout-content">
<p>The GTF file is not necessary for indexing the reference genome, but
will be useful for alignment process. The GTF file contains information
about the gene models in the reference genome, such as the location of
exons and introns, that can speed up the alignment process. If you have
a GTF file, it is recommended to use it when indexing the reference
genome.</p>
</div>
</div>
</div>
<div id="can-i-use-gff3-file-instead" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="can-i-use-gff3-file-instead" class="callout-inner">
<h3 class="callout-title">Can I use GFF3 file instead?</h3>
<div class="callout-content">
<p>Yes, you can use GFF3 file instead of GTF file. STAR can handle both
GTF and GFF3 files. In addition to the aforementioned options, for GFF3
formatted annotations you need to use
<code>--sjdbGTFtagExonParentTranscript Parent</code>. But if you have
issues with GFF3, you can convert it to GTF using <code>gffread</code>
from the <code>Cufflinks</code> package.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="mapping-reads-to-the-reference-genome">Mapping reads to the reference genome<a class="anchor" aria-label="anchor" href="#mapping-reads-to-the-reference-genome"></a>
</h2>
<hr class="half-width">
<p>After indexing, our genome is ready for mapping reads. The
<code>STAR</code> program has many options that can be used to fine-tune
the mapping process. You can <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" class="external-link">check
the manual</a> and find options that are important for mapping reads to
the reference genome.</p>
<p>For mapping, you will need to use <code>--runMode alignReads</code>
important options for aligning reads are:</p>
<ul>
<li>
<code>runThreadN</code> number of threads to use</li>
<li>
<code>readFilesIn</code> fastq files to be aligned, can be single (1
file) or paired-end (2 files)</li>
<li>
<code>genomeDir</code> reference genome index directory</li>
</ul>
<p>For output options, you can use:</p>
<ul>
<li>
<code>outFileNamePrefix</code> prefix for all output files</li>
<li>
<code>outSAMtype</code>: output filetype (SAM default), but can be
BAM, CRAM, etc.</li>
<li>
<code>outSAMunmapped</code>: handling of unmapped reads</li>
</ul>
<div id="what-other-mapping-options-should-i-consider" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-other-mapping-options-should-i-consider" class="callout-inner">
<h3 class="callout-title">What other mapping options should I
consider?</h3>
<div class="callout-content">
<p>STAR’s default parameters are optimized for mammalian genomes. If you
are using for other genomes like plant, fungi or any other species, you
may need to fine-tune your mapping options:</p>
<ul>
<li>When using compressed reads, option <code>readFilesCommand</code>
with <em>UncompressionCommand</em> can be used. If using gzipped files
(<code>*.gz</code>) use <code>--readFilesCommand zcat</code> OR
<code>--readFilesCommand gunzip -c</code>. For <code>bzip2</code>
compressed files, use <code>--readFilesCommand bunzip2 -c</code>.</li>
<li>For controling the multi-mapped reads, options like
<code>outFilterMultimapNmax</code> (default 10) and
<code>winAnchorMultimapNmax</code> can be used. Useful when you have
polyploid genomes or repetitive regions.</li>
<li>For controlling mapping sensitivity, options like
<code>outFilterMismatchNmax</code> (default 10) and
<code>outFilterMismatchNoverLmax</code> can be used.</li>
<li>Shorter reads may require lower <code>outFilterMismatchNmax</code>
and <code>outFilterMismatchNoverLmax</code> values.</li>
</ul>
</div>
</div>
</div>
<div id="script-for-mapping-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="script-for-mapping-reads" class="callout-inner">
<h3 class="callout-title">Script for mapping reads</h3>
<div class="callout-content">
<p>Write a script to map the reads to the reference genome using
<code>STAR</code>. You will need the reference genome index directory,
fastq files to be aligned, and output directory. Can you write a mapping
command for <code>STAR</code>?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--runMode</span> alignReads <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>     <span class="at">--runThreadN</span> 12 <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>     <span class="at">--genomeDir</span> /home/aseethar/rcac_rnaseq/data/star_index <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>     <span class="at">--readFilesIn</span> SRR1234567_1.fastq.gz SRR1234567_2.fastq <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>     <span class="at">--outFileNamePrefix</span> SRR1234567 <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>     <span class="at">--outSAMtype</span> BAM SortedByCoordinate <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>     <span class="at">--outSAMunmapped</span> Within</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Since we have 45 sets of fastq files, we can write an array job for
mapping all the reads. The script to map the reads using
<code>STAR</code> would look like this:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=12</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#SBATCH --time=8:00:00</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#SBATCH --job-name=read_mapping</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#SBATCH --array=1-45    </span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.out</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="ex">ml</span> biocontainers</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="ex">ml</span> star</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co"># Define paths</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="va">FASTQ_DIR</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/fastq"</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="va">GENOME_INDEX</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/star_index"</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="va">OUTPUT_DIR</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/mapping"</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co"># Create an array of FASTQ file pairs</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="va">FASTQ_R1</span><span class="op">=</span><span class="va">($FASTQ_DIR</span>/<span class="pp">*</span>_R1_<span class="pp">*</span>.fastq.gz<span class="va">)</span>   <span class="co"># List of all R1 files</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="va">FASTQ_R2</span><span class="op">=</span><span class="va">($FASTQ_DIR</span>/<span class="pp">*</span>_R2_<span class="pp">*</span>.fastq.gz<span class="va">)</span>   <span class="co"># List of all R2 files</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co"># Get the FASTQ file for this array job</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="va">R1</span><span class="op">=</span><span class="va">${FASTQ_R1</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span>-1<span class="op">]</span><span class="va">}</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="va">R2</span><span class="op">=</span><span class="va">${FASTQ_R2</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span>-1<span class="op">]</span><span class="va">}</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co"># Extract sample name from FASTQ file</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$R1</span> _R1_001.fastq.gz<span class="va">)</span></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a><span class="co"># Run STAR alignment for the current pair of FASTQ files</span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--runThreadN</span> 4 <span class="dt">\</span></span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a>     <span class="at">--genomeDir</span> <span class="va">$GENOME_INDEX</span> <span class="dt">\</span></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a>     <span class="at">--readFilesIn</span> <span class="va">$R1</span> <span class="va">$R2</span> <span class="dt">\</span></span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a>     <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a>     <span class="at">--outFileNamePrefix</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE</span> <span class="dt">\</span></span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a>     <span class="at">--outSAMunmapped</span> Within <span class="dt">\</span></span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a>     <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span></code></pre>
</div>
<p>The job can be submitted as an array job using the following
command:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">chmod</span> +x map_reads.sh</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">sbatch</span> <span class="at">--array</span><span class="op">=</span>1-45 map_reads.sh</span></code></pre>
</div>
<p>Since we have already finished mapping the reads, we can skip this
and move on to the next step.</p>
<div id="can-i-use-star-for-counting-reads" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="can-i-use-star-for-counting-reads" class="callout-inner">
<h3 class="callout-title">Can I use STAR for counting reads?</h3>
<div class="callout-content">
<p>Yes, STAR can be used for counting reads. STAR can output the number
of reads that map to each gene in the reference genome. This can be
useful for differential gene expression analysis. You can use the
<code>--quantMode</code> option to enable read counting. The
<code>--quantMode</code> option can be set to <code>GeneCounts</code> to
output the number of reads that map to each gene.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li><p>Mapping Raw Reads: The process of mapping involves aligning short
reads from sequencing data to a reference genome to determine their
locations. STAR, a splice-aware RNA-seq aligner, is commonly used due to
its speed and accuracy in handling exon-exon junctions.</p></li>
<li><p>Impact of Read Quality: Poor read quality, such as
over-represented sequences or low-quality bases, can lead to unmapped or
multi-mapped reads. It is important to check the quality of reads before
mapping and review alignment statistics after mapping to ensure data
quality.</p></li>
<li><p>Reference Genome Indexing: Before mapping, the reference genome
needs to be indexed using tools like STAR. Indexing creates a data
structure to facilitate rapid alignment, and additional files like
GTF/GFF annotations can improve alignment accuracy.</p></li>
<li><p>Optimizing Mapping: Fine-tuning mapping parameters (e.g.,
handling multi-mapped reads, controlling sensitivity) based on the
dataset is essential. The STAR aligner offers numerous options that
allow adjustment for specific genome types and read qualities.</p></li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-01c-count-features"><p>Content from <a href="01c-count-features.html">Quantifying Gene Expression</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/01c-count-features.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we quantify gene expression from aligned reads?</li>
<li>What tools can we use to count reads mapped to genes or other
genomic features?</li>
<li>What options do we need to consider for accurate read counting?</li>
<li>How do we interpret read count outputs for downstream analyses?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the process of read counting and its role in RNA-seq
analysis.</li>
<li>Learn how to use popular tools like <code>featureCounts</code> to
count reads for genes or other genomic features.</li>
<li>Recognize common challenges, such as handling overlapping features
or isoforms, multimapped reads, and learn how to address them.</li>
<li>Interpret read count outputs and prepare data for downstream
analyses like differential expression.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this episode, we will focus on counting the number of reads mapped
to each genomic feature, such as genes or exons, a critical step in
RNA-seq analysis. Accurate read counting provides the basis for
downstream analyses, including differential gene expression. We will
explore tools and methods used to assign reads to features and prepare
the data for meaningful biological insights.</p>
<p>There are quite a few tools available for counting reads in RNA-seq
analysis, each with its strengths and limitations.</p>
<div id="what-programs-should-i-use-for-counting" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-programs-should-i-use-for-counting" class="callout-inner">
<h3 class="callout-title">What programs should I use for counting?</h3>
<div class="callout-content">
<p>Some popular tools include:</p>
<ul>
<li>
<code>featureCounts</code>: A fast and efficient tool for counting
reads aligned to genomic features such as genes, exons, or
transcripts.</li>
<li>
<code>HTSeq-count</code>: Widely used for counting reads mapped to
genes or other genomic features in RNA-seq experiments.</li>
<li>
<code>RSEM</code>: For quantifying transcript-level expression using
RNA-seq data, particularly for isoform-level counts.</li>
<li>
<code>Salmon</code> &amp; <code>Kallisto</code>: Pseudo-alignment
tools that provide transcript-level quantification directly from raw
reads.</li>
<li>
<code>STAR</code>: Includes options like
<code>--quantMode GeneCounts</code> for gene-level read counting during
the alignment process.</li>
</ul>
</div>
</div>
</div>
<p>In this episode, we will focus on using <code>featureCounts</code>, a
popular tool for counting reads mapped to genomic features. We will
explore its features, parameters, and output formats to understand how
read counting influences downstream analyses like differential
expression.</p>
</section><section><h2 class="section-heading" id="organize-your-data">Organize your data<a class="anchor" aria-label="anchor" href="#organize-your-data"></a>
</h2>
<hr class="half-width">
<p>We will need <code>bam</code> files for read counting. These were
generated in the previous episode, by mapping RNAseq reads to the
genome. We will softlink them from the <code>mapping</code> directory to
the <code>counting</code> directory before proceeding.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> rcac_rnaseq</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../mapping/<span class="pp">*</span>.bam counts/</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="counting-reads-with-featurecounts">Counting reads with <code>featureCounts</code>
<a class="anchor" aria-label="anchor" href="#counting-reads-with-featurecounts"></a>
</h2>
<hr class="half-width">
<p><code>featureCounts</code> is a widely used tool for counting reads
mapped to genomic features, such as genes, exons, or transcripts. It is
part of the <a href="https://subread.sourceforge.net/" class="external-link">Subread</a>
package and provides a fast and efficient way to quantify gene
expression from RNA-seq data. Genomic features are defined in a GTF
(Gene Transfer Format) file, which contains the genomic coordinates of
genes, exons, and other features. <code>bam</code> files provide the
read alignments to the reference genome. Using the location information
from the GTF file, <code>featureCounts</code> assigns read counts to
features and generates count matrices for downstream analyses.</p>
<div id="what-type-of-features-should-i-count-reads-for-genes-exons-transcripts" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-type-of-features-should-i-count-reads-for-genes-exons-transcripts" class="callout-inner">
<h3 class="callout-title">What type of features should I count reads for
(genes, exons, transcripts)?</h3>
<div class="callout-content">
<p>When counting RNA-seq reads, the most common feature to count is
genes, as it provides gene-level expression data for downstream
analysis. However, if your study focuses on alternative splicing or
transcript isoforms, counting exons or transcripts might be more
appropriate. The choice depends on the biological question: gene-level
counts for overall expression, exon-level for splicing events, or
transcript-level for isoform-specific expression.</p>
</div>
</div>
</div>
<div id="what-options-do-you-need-to-specify-for-counting-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="what-options-do-you-need-to-specify-for-counting-reads" class="callout-inner">
<h3 class="callout-title">What options do you need to specify for
counting reads?</h3>
<div class="callout-content">
<p>When using <code>featureCounts</code>, you need to specify several
options to ensure accurate read counting. Take a look at the <a href="https://subread.sourceforge.net/featureCounts.html" class="external-link">featureCounts
manual</a> and identify the key parameters you need to provide for
counting reads. There are also example commands specified in the manual
that you can use as a reference.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>The key parameters you need to specify for counting reads with
<code>featureCounts</code> include:</p>
<ul>
<li>
<code>-a</code>: Annotation file in GTF format containing genomic
features.</li>
<li>
<code>-o</code>: Output file to store read counts.</li>
<li>
<code>-t</code>: Feature type to count (e.g., <code>gene</code>,
<code>exon</code>, <code>transcript</code>).</li>
<li>
<code>-g</code>: Attribute type to group features (e.g.,
<code>gene_id</code>, <code>transcript_id</code>).</li>
<li>
<code>-T</code>: Number of threads to use for parallel
processing.</li>
<li>
<code>-p</code>: Count read pairs instead of individual reads (for
paired-end data).</li>
<li>
<code>--countReadPairs</code>: Count read pairs instead of
individual reads (when using paired-end data).</li>
<li>
<code>--primary</code>: Count only primary alignments.</li>
</ul>
</div>
</div>
</div>
</div>
<p>Putting it all together, the <code>featureCounts</code> command to
count reads for genes in a BAM file would look like this:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">module</span> load subread</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>   <span class="at">-p</span> <span class="dt">\ </span># input is paired-end data</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>   <span class="ex">--countReadPairs</span> <span class="dt">\ </span># count read pairs instead of individual reads</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>   <span class="ex">-t</span> exon <span class="dt">\ </span># feature type to count</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>   <span class="ex">-g</span> gene_id <span class="dt">\ </span># attribute type to group features</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>   <span class="ex">-a</span> data/annotation.gtf <span class="dt">\ </span># annotation file in GTF format</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>   <span class="ex">-o</span> counts/counts.txt <span class="dt">\ </span># output file for read counts</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>   <span class="ex">counts/*.bam</span> <span class="co"># input BAM files</span></span></code></pre>
</div>
<p>You can do this as a job file and submit it to the cluster for
processing. The output will be a tab-delimited file with read counts for
each gene in each sample.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=16</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#SBATCH --job-name=count-features</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.out</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="ex">ml</span> biocontainers</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="ex">ml</span> subread</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span><span class="st">"/home/aseethar/rcac_rnaseq/data/gencode.vM35.primary_assembly.basic.annotation.gtf"</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="dt">\</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>   <span class="at">-p</span> <span class="dt">\ </span> </span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>   <span class="ex">--countReadPairs</span> <span class="dt">\</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>   <span class="at">-t</span> exon <span class="dt">\</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>   <span class="at">-T</span> 16 <span class="dt">\</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>   <span class="at">-g</span> gene_id <span class="dt">\</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>   <span class="at">-a</span> <span class="va">${gtf}</span> <span class="dt">\</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>   <span class="at">-o</span> counts/counts.txt <span class="dt">\</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>   counts/<span class="pp">*</span>.bam</span></code></pre>
</div>
<p>You can submit this job file to the cluster using the
<code>sbatch</code> command:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">sbatch</span> count_features.sh</span></code></pre>
</div>
<p>This will run the <code>featureCounts</code> command on the cluster,
counting reads for each gene in the BAM files provided.</p>
</section><section><h2 class="section-heading" id="understanding-the-output">Understanding the output<a class="anchor" aria-label="anchor" href="#understanding-the-output"></a>
</h2>
<hr class="half-width">
<p>The output of <code>featureCounts</code> is a tab-delimited file
containing read counts for each feature in each sample. The first few
lines of the output file look like this:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">Geneid</span>  Chr Start   End Strand  Length  Sample1 Sample2 Sample3</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">ENSMUSG00000000001</span>  1   3073253 3074322 +   1069    0   0   0</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">ENSMUSG00000000003</span>  1   3102016 3102125 <span class="at">-</span>   110    0   0   0</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">ENSMUSG00000000028</span>  1   3491927 3513553 <span class="at">-</span>   21627   0   0   0</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="ex">ENSMUSG00000000031</span>  1   3461304 3513553 <span class="at">-</span>   52250   0   0   0</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="ex">ENSMUSG00000000037</span>  1   1442067 1467620 +   25554   0   0   0</span></code></pre>
</div>
<p>The columns in the output file include:</p>
<ul>
<li>
<code>Geneid</code>: Gene identifier from the GTF file.</li>
<li>
<code>Chr</code>: Chromosome where the gene is located.</li>
<li>
<code>Start</code>: Start position of the gene.</li>
<li>
<code>End</code>: End position of the gene.</li>
<li>
<code>Strand</code>: Strand where the gene is located.</li>
<li>
<code>Length</code>: Length of the gene.</li>
<li>
<code>Sample1</code>, <code>Sample2</code>, <code>Sample3</code>,
etc.: Read counts for each sample.</li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li><p>Quantifying gene expression involves counting the number of reads
mapped to genomic features like genes or exons, which provides data for
downstream analyses, such as differential gene expression.</p></li>
<li><p>Tools like <code>featureCounts</code>, <code>RSEM</code>,
<code>HTSeq-count</code>, <code>Salmon</code>, and <code>Kallisto</code>
are commonly used for counting reads, each offering different advantages
based on feature type and data structure.</p></li>
<li><p>Depending on the research question, reads can be counted at
various feature levels, such as genes, exons, or transcripts, to capture
overall gene expression or more granular details like splicing
events.</p></li>
<li><p>The output of a read counting tool typically includes read counts
for each feature across all samples, serving as input for further
analysis, such as identifying differentially expressed genes.</p></li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-02-setup"><p>Content from <a href="02-setup.html">RStudio Project and Experimental Data</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/02-setup.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you use RStudio project to manage your analysis project?</li>
<li>What is the most effective way to organize directories for an
analysis project?</li>
<li>How to download a dataset from the internet and save it as a
file.</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create an RStudio project and the directories required for storing
the files pertinent to an analysis project.</li>
<li>Download the data set that will be used for the subsequent
episodes.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Typically, an analysis project begins with dataset files, a handful
of R scripts and output files in a directory. As the project advances,
complexity inevitably rises with the addition of more scripts, output
files, and possibly new datasets. The complexity is further amplified
when dealing with multiple versions of scripts and output files,
necessitating efficient organisation. If these are not well-managed from
the beginning, resuming the project after a break, or sharing the
project with someone else becomes challenging and time-consuming, as we
struggle to recall the project’s status and navigate the directory tree.
Additionally, without proper organisation, the project’s complexity
could lead to frequent use of the <code>setwd</code> function to switch
between different working directories, resulting in a disorganised
workspace.</p>
<p>In this lesson, we will first focus on an effective strategy for
managing the files, both used and generated by our data analysis
project, within a <em>working directory</em>.</p>
<div id="what-is-a-working-directory" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-is-a-working-directory" class="callout-inner">
<h3 class="callout-title">What is a working directory?</h3>
<div class="callout-content">
<p>A working directory in R is the default location on your computer
where R looks for files to load or store any data you wish to save. More
information can be found in our <a href="https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html#the-working-directory" class="external-link">Introduction
to data analysis with R and Bioconductor</a> lesson.</p>
</div>
</div>
</div>
<p>Secondly, we will also learn how to leverage RStudio projects, a
feature built-in to RStudio for managing our analysis project.</p>
<div id="what-is-rstudio" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-is-rstudio" class="callout-inner">
<h3 class="callout-title">What is RStudio?</h3>
<div class="callout-content">
<p>RStudio is a freely available integrated development environment
(IDE), widely used by scientists and software developers for developing
software or analysing datasets in R. If you require assistance with
RStudio or its general usage, please refer to our <a href="https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html" class="external-link">Introduction
to data analysis with R and Bioconductor</a> lesson.</p>
</div>
</div>
</div>
<p>Finally, in this lesson, we will also learn to use the R function
<code>download.file</code> to download the data for the subsequent
episodes.</p>
</section><section><h2 class="section-heading" id="structuring-your-working-directory">Structuring your working directory<a class="anchor" aria-label="anchor" href="#structuring-your-working-directory"></a>
</h2>
<hr class="half-width">
<p>For a more streamlined workflow, we suggest storing all files
associated with an analysis in a specific directory, which will serve as
your project’s working directory. Initially, this working directory
should contain four distinct directories:</p>
<ul>
<li>
<code>data</code>: dedicated to storing raw data. This folder should
ideally only house raw data and not be modified unless you receive a new
dataset (even then, if you have the storage capacity, we suggest you
retain the previous dataset as well in case you need it again in the
future). For RNA-seq data analysis, this directory will typically
contain <code>*.fastq</code> files and any related metadata files for
the experiment.</li>
<li>
<code>scripts</code>: for storing the R scripts you’ve written and
utilised for analysing the data.</li>
<li>
<code>documents</code>: for storing documents related to your
analysis, such as a manuscript outline or meeting notes with your
team.</li>
<li>
<code>output</code>: for storing intermediate or final results
generated by the R scripts in the <code>scripts</code> directory.
Importantly, if you carry out data cleaning or pre-processing, the
output should ideally be stored in this directory, as these no longer
represent raw data.</li>
</ul>
<p>As your project grows in complexity, you might find it necessary to
create more directories or sub-directories. Nevertheless, the
aforementioned four directories should serve as the foundation of your
working directory.</p>
<div id="create-the-directories-for-subsequent-episodes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="create-the-directories-for-subsequent-episodes" class="callout-inner">
<h3 class="callout-title">Create the directories for subsequent
episodes</h3>
<div class="callout-content">
<p>Create a directory on your computer to serve as the working directory
for the rest of this episode and lesson (the workshop example uses a
directory called <code>bio_rnaseq</code>). Then, within this chosen
directory, create the four fundamental directories previously discussed
(<code>data</code>, <code>scripts</code>, <code>documents</code>, and
<code>output</code>).</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<figure style="text-align: center"><img src="../fig/02_setup/folders_working_dir.png" alt="Your working directory should look like this" width="1221" class="figure mx-auto d-block"><figcaption>
Your working directory should look like this
</figcaption></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="using-rstudio-project-to-manage-your-working-directory">Using RStudio project to manage your working directory<a class="anchor" aria-label="anchor" href="#using-rstudio-project-to-manage-your-working-directory"></a>
</h2>
<hr class="half-width">
<p>As previously highlighted, RStudio project is a feature built-into
RStudio for managing your analysis project. It does so by storing
project-specific settings in an <code>.Rproj</code> file stored in your
project’s working directory. Loading these settings up into RStudio by
either opening the <code>.Rproj</code> file directly or through
RStudio’s open project option (from the menu bar, select
<code>File</code> &gt; <code>Open Project...</code>) will automatically
set your working directory in R to the location of the
<code>.Rproj</code> file, essentially your project’s working
directory.</p>
<p>To create an RStudio project:</p>
<ol style="list-style-type: decimal">
<li>Start RStudio.</li>
<li>Navigate to the menu bar and select <code>File</code> &gt;
<code>New Project...</code>.</li>
<li>Choose <code>Existing Directory</code>.</li>
<li>Click <code>Browse...</code> button, and select the directory you
have previously chosen as the working directory for the analysis (i.e.,
the directory where the 4 essential directories reside).</li>
<li>Click <code>Create Project</code> at the bottom right of the
window.</li>
</ol>
<p>Upon completion of the steps above, you will find a
<code>.Rproj</code> file within your project’s working directory.</p>
<figure style="text-align: center"><img src="../fig/02_setup/rstudio_project.png" alt="A new .Rproj file should be created in your chosen working directory." width="1621" class="figure mx-auto d-block"><figcaption>
A new .Rproj file should be created in your chosen working directory.
</figcaption></figure><p>Moreover, the heading of your RStudio console should now also display
the absolute path of your project’s working directory, i.e., where the
<code>.Rproj</code> file resides, indicating that RStudio has set this
directory as your working directory in R.</p>
<figure style="text-align: center"><img src="../fig/02_setup/rstudio_console.png" alt="Your R working directory should now be set to where the .Rproj file resides." width="1211" class="figure mx-auto d-block"><figcaption>
Your R working directory should now be set to where the .Rproj file
resides.
</figcaption></figure><p>From this point forward, any R code that you execute that involves
reading data from a file or saving data in a file will, by default, be
directed to a path relative to your project’s working directory.</p>
<p>If you wish to close the project, perhaps to open another project,
create a new one, or take a break from the project, you can do so by
using to <code>File</code> &gt; <code>Close Project</code> option
located in the menu bar. To open the project back up, either
double-click on the .Rproj file in the working directory, or open up
RStudio and using the <code>File</code> &gt; <code>Open Project</code>
option in the menu bar.</p>
</section><section><h2 class="section-heading" id="download-the-rna-seq-data-for-subsequent-episodes">Download the RNA-seq data for subsequent episodes<a class="anchor" aria-label="anchor" href="#download-the-rna-seq-data-for-subsequent-episodes"></a>
</h2>
<hr class="half-width">
<p>Finally, we will learn how to use R to download the RNA-seq data
required for the subsequent episodes of the lesson. The dataset we will
be using was generated to investigate the impact upper-respiratory
infection have on changes in RNA transcription in the cerebellum and
spinal cord of mice. This dataset was produced as part of the following
study:</p>
<blockquote>
<p>Blackmore, Stephen, et al. “Influenza infection triggers disease in a
genetic model of experimental autoimmune encephalomyelitis.” Proceedings
of the National Academy of Sciences 114.30 (2017): E6107-E6116.</p>
</blockquote>
<p>The dataset is available at <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression Omnibus
(GEO)</a>, under the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870" class="external-link">GSE96870</a>.
Downloading data from GEO is not straightforward (and won’t be covered
in this lesson). Hence, we have made the data available on our GitHub
repository for easier access.</p>
<p>To download the files, we will use the R function
<code>download.file</code>, which necessitates at least two parameters:
<code>url</code> and <code>destfile</code>. The <code>url</code>
parameter is used to specify the address on the internet to download the
data from. The <code>destfile</code> parameter indicates where to save
the downloaded file and what the downloaded file should be named.</p>
<p>Let’s download one of the four data files needed for the remainder of
this lesson. The data file is located at <a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv</a>.
We shall save the downloaded file in the <code>data</code> folder of our
working directory with the name
<code>GSE96870_counts_cerebellum.csv</code>.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_counts_cerebellum.csv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>If you navigate to the <code>data</code> folder in your working
directory, you should now find a file named
<code>GSE96870_counts_cerebellum.csv</code>.</p>
<figure style="text-align: center"><img src="../fig/02_setup/data_folder.png" alt="A file named GSE96870_counts_cerebellum.csv should now reside in the data folder." width="1632" class="figure mx-auto d-block"><figcaption>
A file named GSE96870_counts_cerebellum.csv should now reside in the
data folder.
</figcaption></figure><div id="download-the-remaining-data-set-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="download-the-remaining-data-set-files" class="callout-inner">
<h3 class="callout-title">Download the remaining data set files</h3>
<div class="callout-content">
<p>There are three more data set files we need to download for the
remainder of this lesson.</p>
<table class="table">
<colgroup>
<col width="77%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th>URL</th>
<th>Filename</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv</a></td>
<td>GSE96870_coldata_cerebellum.csv</td>
</tr>
<tr class="even">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv</a></td>
<td>GSE96870_coldata_all.csv</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv</a></td>
<td>GSE96870_rowranges.tsv</td>
</tr>
</tbody>
</table>
<p>Use the <code>download.file</code> function to download the files
into the <code>data</code> folder in your working directory.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_coldata_cerebellum.csv"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_coldata_all.csv"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_rowranges.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Proper organisation of the files required for your project in a
working directory is crucial for maintaining order and ensuring easy
access in the future.</li>
<li>RStudio project serves as a valuable tool for managing your
project’s working directory and facilitating analysis.</li>
<li>The <code>download.file</code> function in R can be used for
downloading datasets from the internet.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-03-import-annotate"><p>Content from <a href="03-import-annotate.html">Importing and annotating quantified data into R</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/03-import-annotate.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 120 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can one import quantified gene expression data into an object
suitable for downstream statistical analysis in R?</li>
<li>What types of gene identifiers are typically used, and how are
mappings between them done?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn how to import the quantifications into a SummarizedExperiment
object.</li>
<li>Learn how to add additional gene annotations to the object.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<hr class="half-width">
<p>In this episode we will use some functions from add-on R packages. In
order to use them, we need to load them from our
<code>library</code>:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">AnnotationDbi</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">hgu95av2.db</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>If you get any error messages about
<code>there is no package called 'XXXX'</code> it means you have not
installed the package/s yet for this version of R. See the bottom of the
<a href="https://carpentries-incubator.github.io/bioc-rnaseq/index.html" class="external-link">Summary
and Setup</a> to install all the necessary packages for this workshop.
If you have to install, remember to re-run the <code>library</code>
commands above to load them.</p>
</section><section><h2 class="section-heading" id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<hr class="half-width">
<p>In the last episode, we used R to download 4 files from the internet
and saved them on our computer. But we do not have these files loaded
into R yet so that we can work with them. The original experimental
design in <a href="https://pubmed.ncbi.nlm.nih.gov/28696309/" class="external-link">Blackmore
et al. 2017</a> was fairly complex: 8 week old male and female C57BL/6
mice were collected at Day 0 (before influenza infection), Day 4 and Day
8 after influenza infection. From each mouse, cerebellum and spinal cord
tissues were taken for RNA-seq. There were originally 4 mice per ‘Sex x
Time x Tissue’ group, but a few were lost along the way resulting in a
total of 45 samples. For this workshop, we are going to simplify the
analysis by only using the 22 cerebellum samples. Expression
quantification was done using STAR to align to the mouse genome and then
counting reads that map to genes. In addition to the counts per gene per
sample, we also need information on which sample belongs to which
Sex/Time point/Replicate. And for the genes, it is helpful to have extra
information called annotation. Let’s read in the data files that we
downloaded in the last episode and start to explore them:</p>
<div class="section level3">
<h3 id="counts">Counts<a class="anchor" aria-label="anchor" href="#counts"></a>
</h3>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_counts_cerebellum.csv"</span>, </span>
<span>                   row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 41786    22</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(counts)</span></span></code></pre>
</div>
<p>Genes are in rows and samples are in columns, so we have counts for
41,786 genes and 22 samples. The <code>View()</code> command has been
commented out for the website, but running it will open a tab in RStudio
that lets us look at the data and even sort the table by a particular
column. However, the viewer cannot change the data inside the
<code>counts</code> object, so we can only look, not permanently sort
nor edit the entries. When finished, close the viewer using the X in the
tab. Looks like the rownames are gene symbols and the column names are
the GEO sample IDs, which are not very informative for telling us which
sample is what.</p>
</div>
<div class="section level3">
<h3 id="sample-annotations">Sample annotations<a class="anchor" aria-label="anchor" href="#sample-annotations"></a>
</h3>
<p>Next read in the sample annotations. Because samples are in columns
in the count matrix, we will name the object <code>coldata</code>:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_coldata_cerebellum.csv"</span>,</span>
<span>                    row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22 10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(coldata)</span></span></code></pre>
</div>
<p>Now samples are in rows with the GEO sample IDs as the rownames, and
we have 10 columns of information. The columns that are the most useful
for this workshop are <code>geo_accession</code> (GEO sample IDs again),
<code>sex</code> and <code>time</code>.</p>
</div>
<div class="section level3">
<h3 id="gene-annotations">Gene annotations<a class="anchor" aria-label="anchor" href="#gene-annotations"></a>
</h3>
<p>The counts only have gene symbols, which while short and somewhat
recognizable to the human brain, are not always good absolute
identifiers for exactly what gene was measured. For this we need
additional gene annotations that were provided by the authors. The
<code>count</code> and <code>coldata</code> files were in comma
separated value (.csv) format, but we cannot use that for our gene
annotation file because the descriptions can contain commas that would
prevent a .csv file from being read in correctly. Instead the gene
annotation file is in tab separated value (.tsv) format. Likewise, the
descriptions can contain the single quote <code>'</code> (e.g., 5’),
which by default R assumes indicates a character entry. So we have to
use a more generic function <code>read.delim()</code> with extra
arguments to specify that we have tab-separated data
(<code>sep = "\t"</code>) with no quotes used (<code>quote = ""</code>).
We also put in other arguments to specify that the first row contains
our column names (<code>header = TRUE</code>), the gene symbols that
should be our <code>row.names</code> are in the 5th column
(<code>row.names = 5</code>), and that NCBI’s species-specific gene ID
(i.e., ENTREZID) should be read in as character data even though they
look like numbers (<code>colClasses</code> argument). You can look up
this details on available arguments by simply entering the function name
starting with question mark. (e.g., <code>?read.delim</code>)</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rowranges</span> <span class="op">&lt;-</span> <span class="fu">read.delim</span><span class="op">(</span><span class="st">"data/GSE96870_rowranges.tsv"</span>, </span>
<span>                        sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                        colClasses <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>ENTREZID <span class="op">=</span> <span class="st">"character"</span><span class="op">)</span>,</span>
<span>                        header <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                        quote <span class="op">=</span> <span class="st">""</span>, </span>
<span>                        row.names <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 41786     7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(rowranges)</span></span></code></pre>
</div>
<p>For each of the 41,786 genes, we have the <code>seqnames</code>
(e.g., chromosome number), <code>start</code> and <code>end</code>
positions, <code>strand</code>, <code>ENTREZID</code>, gene product
description (<code>product</code>) and the feature type
(<code>gbkey</code>). These gene-level metadata are useful for the
downstream analysis. For example, from the <code>gbkey</code> column, we
can check what types of genes and how many of them are in our
dataset:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">rowranges</span><span class="op">$</span><span class="va">gbkey</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
     C_region     D_segment          exon     J_segment      misc_RNA
           20            23          4008            94          1988
         mRNA         ncRNA precursor_RNA          rRNA          tRNA
        21198         12285          1187            35           413
    V_segment
          535 </code></pre>
</div>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>How are the 3 objects <code>counts</code>, <code>coldata</code> and
<code>rowranges</code> related to each other in terms of their rows and
columns?</li>
<li>If you only wanted to analyse the mRNA genes, what would you have to
do keep just those (generally speaking, not exact codes)?</li>
<li>If you decided the first two samples were outliers, what would you
have to do to remove those (generally speaking, not exact codes)?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>In <code>counts</code>, the rows are genes just like the rows in
<code>rowranges</code>. The columns in <code>counts</code> are the
samples, but this corresponds to the rows in <code>coldata</code>.</li>
<li>I would have to remember subset both the rows of <code>counts</code>
and the rows of <code>rowranges</code> to just the mRNA genes.</li>
<li>I would have to remember to subset both the columns of
<code>counts</code> but the rows of <code>coldata</code> to exclude the
first two samples.</li>
</ol>
</div>
</div>
</div>
</div>
<p>You can see how keeping related information in separate objects could
easily lead to mis-matches between our counts, gene annotations and
sample annotations. This is why Bioconductor has created a specialized
S4 class called a <code>SummarizedExperiment</code>. The details of a
<code>SummarizedExperiment</code> were covered extensively at the end of
the <a href="https://carpentries-incubator.github.io/bioc-intro/60-next-steps.html#next-steps" class="external-link">Introduction
to data analysis with R and Bioconductor</a> workshop. As a reminder,
let’s take a look at the figure below representing the anatomy of the
<code>SummarizedExperiment</code> class:</p>
<figure><img src="https://uclouvain-cbio.github.io/WSBIM1322/figs/SE.svg" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It is designed to hold any type of quantitative ’omics data
(<code>assays</code>) along with linked sample annotations
(<code>colData</code>) and feature annotations with
(<code>rowRanges</code>) or without (<code>rowData</code>) chromosome,
start and stop positions. Once these three tables are (correctly!)
linked, subsetting either samples and/or features will correctly subset
the <code>assay</code>, <code>colData</code> and <code>rowRanges</code>.
Additionally, most Bioconductor packages are built around the same core
data infrastructure so they will recognize and be able to manipulate
<code>SummarizedExperiment</code> objects. Two of the most popular
RNA-seq statistical analysis packages have their own extended S4 classes
similar to a <code>SummarizedExperiment</code> with the additional slots
for statistical results: <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseqdataset" class="external-link">DESeq2</a>’s
<code>DESeqDataSet</code> and <a href="https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/DGEList-class" class="external-link">edgeR</a>’s
<code>DGEList</code>. No matter which one you end up using for
statistical analysis, you can start by putting your data in a
<code>SummarizedExperiment</code>.</p>
</div>
</section><section><h2 class="section-heading" id="assemble-summarizedexperiment">Assemble SummarizedExperiment<a class="anchor" aria-label="anchor" href="#assemble-summarizedexperiment"></a>
</h2>
<hr class="half-width">
<p>We will create a <code>SummarizedExperiment</code> from these
objects:</p>
<ul>
<li>The <code>count</code> object will be saved in <code>assays</code>
slot<br>
</li>
<li>The <code>coldata</code> object with sample information will be
stored in <code>colData</code> slot (<em><strong>sample
metadata</strong></em>)<br>
</li>
<li>The <code>rowranges</code> object describing the genes will be
stored in <code>rowRanges</code> slot (<em><strong>features
metadata</strong></em>)</li>
</ul>
<p>Before we put them together, you ABSOLUTELY MUST MAKE SURE THE
SAMPLES AND GENES ARE IN THE SAME ORDER! Even though we saw that
<code>count</code> and <code>coldata</code> had the same number of
samples and <code>count</code> and <code>rowranges</code> had the same
number of genes, we never explicitly checked to see if they were in the
same order. One quick way to check:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">all.equal</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span> <span class="co"># samples</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">all.equal</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span><span class="op">)</span> <span class="co"># genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If the first is not TRUE, you can match up the samples/columns in</span></span>
<span><span class="co"># counts with the samples/rows in coldata like this (which is fine</span></span>
<span><span class="co"># to run even if the first was TRUE):</span></span>
<span></span>
<span><span class="va">tempindex</span> <span class="op">&lt;-</span> <span class="fu">match</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span><span class="va">tempindex</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Check again:</span></span>
<span><span class="fu">all.equal</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>If the features (i.e., genes) in the assay (e.g.,
<code>counts</code>) and the gene annotation table (e.g.,
<code>rowranges</code>) are different, how can we fix them? Write the
codes.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tempindex</span> <span class="op">&lt;-</span> <span class="fu">match</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rowranges</span> <span class="op">&lt;-</span> <span class="va">rowranges</span><span class="op">[</span><span class="va">tempindex</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu">all.equal</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Once we have verified that samples and genes are in the same order,
we can then create our <code>SummarizedExperiment</code> object.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One final check:</span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="co"># features</span></span>
<span>          <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">==</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="co"># samples</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    rowRanges <span class="op">=</span> <span class="fu">as</span><span class="op">(</span><span class="va">rowranges</span>, <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">coldata</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Because matching the genes and samples is so important, the
<code>SummarizedExperiment()</code> constructor does some internal check
to make sure they contain the same number of genes/samples and the
sample/row names match. If not, you will get some error messages:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># wrong number of samples:</span></span>
<span></span>
<span><span class="va">bad1</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    rowRanges <span class="op">=</span> <span class="fu">as</span><span class="op">(</span><span class="va">rowranges</span>, <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">coldata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in validObject(.Object): invalid class "SummarizedExperiment" object:
    nb of cols in 'assay' (22) must equal nb of rows in 'colData' (3)</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># same number of genes but in different order:</span></span>
<span></span>
<span><span class="va">bad2</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowRanges <span class="op">=</span> <span class="fu">as</span><span class="op">(</span><span class="va">rowranges</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span>,<span class="op">]</span>, <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">coldata</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in SummarizedExperiment(assays = list(counts = as.matrix(counts)), : the rownames and colnames of the supplied assay(s) must be NULL or identical
  to those of the RangedSummarizedExperiment object (or derivative) to
  construct</code></pre>
</div>
<p>A brief recap of how to access the various data slots in a
<code>SummarizedExperiment</code> and how to make some
manipulations:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the counts</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
Xkr4               1891       2410       2159       1980       1977       1945
LOC105243853          0          0          1          4          0          0
LOC105242387        204        121        110        120        172        173
LOC105242467         12          5          5          5          2          6
Rp1                   2          2          0          3          2          1
Sox17               251        239        218        220        261        232
             GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
Xkr4               1757       2235       1779       1528       1644       1585
LOC105243853          1          3          3          0          1          3
LOC105242387        177        130        131        160        180        176
LOC105242467          3          2          2          2          1          2
Rp1                   3          1          1          2          2          2
Sox17               179        296        233        271        205        230
             GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
Xkr4               2275       1881       2584       1837       1890       1910
LOC105243853          1          0          0          1          1          0
LOC105242387        161        154        124        221        272        214
LOC105242467          2          4          7          1          3          1
Rp1                   3          6          5          3          5          1
Sox17               302        286        325        201        267        322
             GSM2545354 GSM2545362 GSM2545363 GSM2545380
Xkr4               1771       2315       1645       1723
LOC105243853          0          1          0          1
LOC105242387        124        189        223        251
LOC105242467          4          2          1          4
Rp1                   3          3          1          0
Sox17               273        197        310        246</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 41786    22</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The above works now because we only have one assay, "counts"</span></span>
<span><span class="co"># But if there were more than one assay, we would have to specify</span></span>
<span><span class="co"># which one like so:</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
Xkr4               1891       2410       2159       1980       1977       1945
LOC105243853          0          0          1          4          0          0
LOC105242387        204        121        110        120        172        173
LOC105242467         12          5          5          5          2          6
Rp1                   2          2          0          3          2          1
Sox17               251        239        218        220        261        232
             GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
Xkr4               1757       2235       1779       1528       1644       1585
LOC105243853          1          3          3          0          1          3
LOC105242387        177        130        131        160        180        176
LOC105242467          3          2          2          2          1          2
Rp1                   3          1          1          2          2          2
Sox17               179        296        233        271        205        230
             GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
Xkr4               2275       1881       2584       1837       1890       1910
LOC105243853          1          0          0          1          1          0
LOC105242387        161        154        124        221        272        214
LOC105242467          2          4          7          1          3          1
Rp1                   3          6          5          3          5          1
Sox17               302        286        325        201        267        322
             GSM2545354 GSM2545362 GSM2545363 GSM2545380
Xkr4               1771       2315       1645       1723
LOC105243853          0          1          0          1
LOC105242387        124        189        223        251
LOC105242467          4          2          1          4
Rp1                   3          3          1          0
Sox17               273        197        310        246</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the sample annotations</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 22 rows and 10 columns
                     title geo_accession     organism         age         sex
               &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt; &lt;character&gt;
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus     8 weeks      Female
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus     8 weeks      Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus     8 weeks      Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus     8 weeks      Female
GSM2545340 CNS_RNA-seq_14C    GSM2545340 Mus musculus     8 weeks        Male
...                    ...           ...          ...         ...         ...
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus     8 weeks      Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus     8 weeks        Male
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus     8 weeks      Female
GSM2545363  CNS_RNA-seq_6C    GSM2545363 Mus musculus     8 weeks        Male
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus     8 weeks      Female
             infection      strain        time      tissue     mouse
           &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;
GSM2545336  InfluenzaA     C57BL/6        Day8  Cerebellum        14
GSM2545337 NonInfected     C57BL/6        Day0  Cerebellum         9
GSM2545338 NonInfected     C57BL/6        Day0  Cerebellum        10
GSM2545339  InfluenzaA     C57BL/6        Day4  Cerebellum        15
GSM2545340  InfluenzaA     C57BL/6        Day4  Cerebellum        18
...                ...         ...         ...         ...       ...
GSM2545353 NonInfected     C57BL/6        Day0  Cerebellum         4
GSM2545354 NonInfected     C57BL/6        Day0  Cerebellum         2
GSM2545362  InfluenzaA     C57BL/6        Day4  Cerebellum        20
GSM2545363  InfluenzaA     C57BL/6        Day4  Cerebellum        12
GSM2545380  InfluenzaA     C57BL/6        Day8  Cerebellum        19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22 10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the gene annotations</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 6 rows and 3 columns
                ENTREZID                product       gbkey
             &lt;character&gt;            &lt;character&gt; &lt;character&gt;
Xkr4              497097 X Kell blood group p..        mRNA
LOC105243853   105243853 uncharacterized LOC1..       ncRNA
LOC105242387   105242387 uncharacterized LOC1..       ncRNA
LOC105242467   105242467 lipoxygenase homolog..        mRNA
Rp1                19888 retinitis pigmentosa..        mRNA
Sox17              20671 SRY (sex determining..        mRNA</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 41786     3</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make better sample IDs that show sex, time and mouse ID:</span></span>
<span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Label</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">sex</span>, <span class="va">se</span><span class="op">$</span><span class="va">time</span>, <span class="va">se</span><span class="op">$</span><span class="va">mouse</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Label</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Female_Day8_14" "Female_Day0_9"  "Female_Day0_10" "Female_Day4_15"
 [5] "Male_Day4_18"   "Male_Day8_6"    "Female_Day8_5"  "Male_Day0_11"
 [9] "Female_Day4_22" "Male_Day4_13"   "Male_Day8_23"   "Male_Day8_24"
[13] "Female_Day0_8"  "Male_Day0_7"    "Male_Day4_1"    "Female_Day8_16"
[17] "Female_Day4_21" "Female_Day0_4"  "Male_Day0_2"    "Female_Day4_20"
[21] "Male_Day4_12"   "Female_Day8_19"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">$</span><span class="va">Label</span></span>
<span></span>
<span><span class="co"># Our samples are not in order based on sex and time</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Group</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">sex</span>, <span class="va">se</span><span class="op">$</span><span class="va">time</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Group</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Female_Day8" "Female_Day0" "Female_Day0" "Female_Day4" "Male_Day4"
 [6] "Male_Day8"   "Female_Day8" "Male_Day0"   "Female_Day4" "Male_Day4"
[11] "Male_Day8"   "Male_Day8"   "Female_Day0" "Male_Day0"   "Male_Day4"
[16] "Female_Day8" "Female_Day4" "Female_Day0" "Male_Day0"   "Female_Day4"
[21] "Male_Day4"   "Female_Day8"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change this to factor data with the levels in order </span></span>
<span><span class="co"># that we want, then rearrange the se object:</span></span>
<span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Group</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">Group</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Female_Day0"</span>,<span class="st">"Male_Day0"</span>, </span>
<span>                                        <span class="st">"Female_Day4"</span>,<span class="st">"Male_Day4"</span>,</span>
<span>                                        <span class="st">"Female_Day8"</span>,<span class="st">"Male_Day8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="fu">order</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 22 rows and 12 columns
                         title geo_accession     organism         age
                   &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt;
Female_Day0_9  CNS_RNA-seq_11C    GSM2545337 Mus musculus     8 weeks
Female_Day0_10 CNS_RNA-seq_12C    GSM2545338 Mus musculus     8 weeks
Female_Day0_8  CNS_RNA-seq_27C    GSM2545348 Mus musculus     8 weeks
Female_Day0_4   CNS_RNA-seq_3C    GSM2545353 Mus musculus     8 weeks
Male_Day0_11   CNS_RNA-seq_20C    GSM2545343 Mus musculus     8 weeks
...                        ...           ...          ...         ...
Female_Day8_16  CNS_RNA-seq_2C    GSM2545351 Mus musculus     8 weeks
Female_Day8_19  CNS_RNA-seq_9C    GSM2545380 Mus musculus     8 weeks
Male_Day8_6    CNS_RNA-seq_17C    GSM2545341 Mus musculus     8 weeks
Male_Day8_23   CNS_RNA-seq_25C    GSM2545346 Mus musculus     8 weeks
Male_Day8_24   CNS_RNA-seq_26C    GSM2545347 Mus musculus     8 weeks
                       sex   infection      strain        time      tissue
               &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
Female_Day0_9       Female NonInfected     C57BL/6        Day0  Cerebellum
Female_Day0_10      Female NonInfected     C57BL/6        Day0  Cerebellum
Female_Day0_8       Female NonInfected     C57BL/6        Day0  Cerebellum
Female_Day0_4       Female NonInfected     C57BL/6        Day0  Cerebellum
Male_Day0_11          Male NonInfected     C57BL/6        Day0  Cerebellum
...                    ...         ...         ...         ...         ...
Female_Day8_16      Female  InfluenzaA     C57BL/6        Day8  Cerebellum
Female_Day8_19      Female  InfluenzaA     C57BL/6        Day8  Cerebellum
Male_Day8_6           Male  InfluenzaA     C57BL/6        Day8  Cerebellum
Male_Day8_23          Male  InfluenzaA     C57BL/6        Day8  Cerebellum
Male_Day8_24          Male  InfluenzaA     C57BL/6        Day8  Cerebellum
                   mouse          Label       Group
               &lt;integer&gt;    &lt;character&gt;    &lt;factor&gt;
Female_Day0_9          9  Female_Day0_9 Female_Day0
Female_Day0_10        10 Female_Day0_10 Female_Day0
Female_Day0_8          8  Female_Day0_8 Female_Day0
Female_Day0_4          4  Female_Day0_4 Female_Day0
Male_Day0_11          11   Male_Day0_11 Male_Day0
...                  ...            ...         ...
Female_Day8_16        16 Female_Day8_16 Female_Day8
Female_Day8_19        19 Female_Day8_19 Female_Day8
Male_Day8_6            6    Male_Day8_6 Male_Day8
Male_Day8_23          23   Male_Day8_23 Male_Day8
Male_Day8_24          24   Male_Day8_24 Male_Day8  </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Finally, also factor the Label column to keep in order in plots:</span></span>
<span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Label</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">Label</span>, levels <span class="op">=</span> <span class="va">se</span><span class="op">$</span><span class="va">Label</span><span class="op">)</span></span></code></pre>
</div>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>How many samples are there for each level of the
<code>Infection</code> variable?</li>
<li>Create 2 objects named <code>se_infected</code> and
<code>se_noninfected</code> containing a subset of <code>se</code> with
only infected and non-infected samples, respectively. Then, calculate
the mean expression levels of the first 500 genes for each object, and
use the <code>summary()</code> function to explore the distribution of
expression levels for infected and non-infected samples based on these
genes.</li>
<li>How many samples represent female mice infected with Influenza A on
day 8?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
 InfluenzaA NonInfected
         15           7 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 2</span></span>
<span><span class="va">se_infected</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">infection</span> <span class="op">==</span> <span class="st">"InfluenzaA"</span><span class="op">]</span></span>
<span><span class="va">se_noninfected</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">infection</span> <span class="op">==</span> <span class="st">"NonInfected"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">means_infected</span> <span class="op">&lt;-</span> <span class="fu">rowMeans</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_infected</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">means_noninfected</span> <span class="op">&lt;-</span> <span class="fu">rowMeans</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_noninfected</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">means_infected</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max.
0.000e+00 1.333e-01 2.867e+00 7.641e+02 3.374e+02 1.890e+04 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">means_noninfected</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max.
0.000e+00 1.429e-01 3.143e+00 7.710e+02 3.666e+02 2.001e+04 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 3</span></span>
<span><span class="fu">ncol</span><span class="op">(</span><span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> <span class="va">se</span><span class="op">$</span><span class="va">infection</span> <span class="op">==</span> <span class="st">"InfluenzaA"</span> <span class="op">&amp;</span> <span class="va">se</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day8"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="save-summarizedexperiment">Save SummarizedExperiment<a class="anchor" aria-label="anchor" href="#save-summarizedexperiment"></a>
</h2>
<hr class="half-width">
<p>This was a bit of code and time to create our
<code>SummarizedExperiment</code> object. We will need to keep using it
throughout the workshop, so it can be useful to save it as an actual
single file on our computer to read it back in to R’s memory if we have
to shut down RStudio. To save an R-specific file we can use the
<code>saveRDS()</code> function and later read it back into R using the
<code>readRDS()</code> function.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">se</span>, <span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="fu">rm</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="co"># remove the object!</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="data-provenance-and-reproducibility">Data provenance and reproducibility<a class="anchor" aria-label="anchor" href="#data-provenance-and-reproducibility"></a>
</h2>
<hr class="half-width">
<p>We have now created an external .rds file that represents our RNA-Seq
data in a format that can be read into R and used by various packages
for our analyses. But we should still keep a record of the codes that
created the .rds file from the 3 files we downloaded from the internet.
But what is the provenance of those files - i.e, where did they come
from and how were they made? The original counts and gene information
were deposited in the GEO public database, accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870" class="external-link">GSE96870</a>.
But these counts were generated by running alignment/quantification
programs on the also-deposited fastq files that hold the sequence base
calls and quality scores, which in turn were generated by a specific
sequencing machine using some library preparation method on RNA
extracted from samples collected in a particular experiment. Whew!</p>
<p>If you conducted the original experiment ideally you would have the
complete record of where and how the data were generated. But you might
use publicly-available data sets so the best you can do is to keep track
of what original files you got from where and what manipulations you
have done to them. Using R codes to keep track of everything is an
excellent way to be able to reproduce the entire analysis from the
original input files. The exact results you get can differ depending on
the R version, add-on package versions and even what operating system
you use, so make sure to keep track of all this information as well by
running <code>sessionInfo()</code> and recording the output (see example
at end of lesson).</p>
<div id="challenge-how-to-subset-to-mrna-genes" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-how-to-subset-to-mrna-genes" class="callout-inner">
<h3 class="callout-title">Challenge: How to subset to mRNA genes</h3>
<div class="callout-content">
<p>Before, we conceptually discussed subsetting to only the mRNA genes.
Now that we have our <code>SummarizedExperiment</code> object, it
becomes much easier to write the codes to subset <code>se</code> to a
new object called <code>se_mRNA</code> that contains only the genes/rows
where the <code>rowData(se)$gbkey</code> is equal to mRNA. Write the
codes and then check you correctly got the 21,198 mRNA genes:</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mRNA</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">gbkey</span> <span class="op">==</span> <span class="st">"mRNA"</span> , <span class="op">]</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">se_mRNA</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 21198    22</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="gene-annotations-1">Gene Annotations<a class="anchor" aria-label="anchor" href="#gene-annotations-1"></a>
</h2>
<hr class="half-width">
<p>Depending on who generates your count data, you might not have a nice
file of additional gene annotations. There may only be the count row
names, which could be gene symbols or ENTREZIDs or another database’s
ID. Characteristics of gene annotations differ based on their annotation
strategies and information sources. For example, RefSeq human gene
models (i.e., Entrez from NCBI) are well supported and broadly used in
various studies. The UCSC Known Genes dataset is based on protein data
from Swiss-Prot/TrEMBL (UniProt) and the associated mRNA data from
GenBank, and serves as a foundation for the UCSC Genome Browser. Ensembl
genes contain both automated genome annotation and manual curation.</p>
<p>You can find more information in Bioconductor <a href="https://jmacdon.github.io/Bioc2022Anno/articles/AnnotationWorkshop.html" class="external-link">Annotation
Workshop</a> material.</p>
<p>Bioconductor has many packages and functions that can help you to get
additional annotation information for your genes. The available
resources are covered in more detail in <a href="https://carpentries-incubator.github.io/bioc-rnaseq/07-gene-set-analysis.html#gene-set-resources" class="external-link">Episode
7 Gene set enrichment analysis</a>.</p>
<p>Here, we will introduce one of the gene ID mapping functions,
<code>mapIds</code>:</p>
<pre><code><span><span class="fu">mapIds</span><span class="op">(</span><span class="va">annopkg</span>, <span class="va">keys</span>, <span class="va">column</span>, <span class="va">keytype</span>, <span class="va">...</span>, <span class="va">multiVals</span><span class="op">)</span></span></code></pre>
<p>Where</p>
<ul>
<li>
<em>annopkg</em> is the annotation package<br>
</li>
<li>
<em>keys</em> are the IDs that we <strong>know</strong><br>
</li>
<li>
<em>column</em> is the value we <strong>want</strong><br>
</li>
<li>
<em>keytype</em> is the type of key used</li>
</ul>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keys <span class="op">=</span> <span class="st">"497097"</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">497097</span></span>
<span><span class="st">"Xkr4"</span> </span></code></pre>
</div>
<p>Different from the <code>select()</code> function,
<code>mapIds()</code> function handles 1:many mapping between keys and
columns through an additional argument, <code>multiVals</code>. The
below example demonstrate this functionality using the
<code>hgu95av2.db</code> package, an Affymetrix Human Genome U95 Set
annotation data.</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keys</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="fu">keys</span><span class="op">(</span><span class="va">hgu95av2.db</span>, <span class="st">"ENTREZID"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">last</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">[[</span><span class="fu">length</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">}</span></span>
<span></span>
<span><span class="fu">mapIds</span><span class="op">(</span><span class="va">hgu95av2.db</span>, keys <span class="op">=</span> <span class="va">keys</span>, column <span class="op">=</span> <span class="st">"ALIAS"</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       10       100      1000     10000 100008586     10001
   "AAC2"    "ADA1"   "ACOGS"    "MPPH"     "AL4"   "ARC33" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># When there is 1:many mapping, the default behavior was </span></span>
<span><span class="co"># to output the first match. This can be changed to a function,</span></span>
<span><span class="co"># which we defined above to give us the last match:</span></span>
<span></span>
<span><span class="fu">mapIds</span><span class="op">(</span><span class="va">hgu95av2.db</span>, keys <span class="op">=</span> <span class="va">keys</span>, column <span class="op">=</span> <span class="st">"ALIAS"</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="va">last</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       10       100      1000     10000 100008586     10001
   "NAT2"     "ADA"    "CDH2"    "AKT3" "GAGE12F"    "MED6" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or we can get back all the many mappings:</span></span>
<span></span>
<span><span class="fu">mapIds</span><span class="op">(</span><span class="va">hgu95av2.db</span>, keys <span class="op">=</span> <span class="va">keys</span>, column <span class="op">=</span> <span class="st">"ALIAS"</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`10`
[1] "AAC2"  "NAT-2" "PNAT"  "NAT2"

$`100`
[1] "ADA1" "ADA"

$`1000`
[1] "ACOGS"  "ADHD8"  "ARVD14" "CD325"  "CDHN"   "CDw325" "NCAD"   "CDH2"

$`10000`
[1] "MPPH"         "MPPH2"        "PKB-GAMMA"    "PKBG"         "PRKBG"
[6] "RAC-PK-gamma" "RAC-gamma"    "STK-2"        "AKT3"

$`100008586`
[1] "AL4"     "CT4.7"   "GAGE-7"  "GAGE-7B" "GAGE-8"  "GAGE7"   "GAGE7B"
[8] "GAGE12F"

$`10001`
[1] "ARC33"     "NY-REN-28" "MED6"     </code></pre>
</div>
</section><section><h2 class="section-heading" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] hgu95av2.db_3.13.0          org.Hs.eg.db_3.22.0
 [3] org.Mm.eg.db_3.22.0         AnnotationDbi_1.72.0
 [5] SummarizedExperiment_1.40.0 Biobase_2.70.0
 [7] MatrixGenerics_1.22.0       matrixStats_1.5.0
 [9] GenomicRanges_1.62.0        Seqinfo_1.0.0
[11] IRanges_2.44.0              S4Vectors_0.48.0
[13] BiocGenerics_0.56.0         generics_0.1.4
[15] knitr_1.50

loaded via a namespace (and not attached):
 [1] Matrix_1.7-4        bit_4.6.0           crayon_1.5.3
 [4] compiler_4.5.2      BiocManager_1.30.26 renv_1.1.5
 [7] blob_1.2.4          Biostrings_2.78.0   png_0.1-8
[10] yaml_2.3.10         fastmap_1.2.0       lattice_0.22-5
[13] R6_2.6.1            XVector_0.50.0      S4Arrays_1.10.0
[16] DelayedArray_0.36.0 DBI_1.2.3           rlang_1.1.6
[19] KEGGREST_1.50.0     cachem_1.1.0        xfun_0.54
[22] bit64_4.6.0-1       SparseArray_1.10.1  RSQLite_2.4.4
[25] memoise_2.0.1       cli_3.6.5           grid_4.5.2
[28] vctrs_0.6.5         evaluate_1.0.5      abind_1.4-8
[31] httr_1.4.7          pkgconfig_2.0.3     tools_4.5.2        </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Depending on the gene expression quantification tool used, there are
different ways (often distributed in Bioconductor packages) to read the
output into a <code>SummarizedExperiment</code> or <code>DGEList</code>
object for further processing in R.</li>
<li>Stable gene identifiers such as Ensembl or Entrez IDs should
preferably be used as the main identifiers throughout an RNA-seq
analysis, with gene symbols added for easier interpretation.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-04-exploratory-qc"><p>Content from <a href="04-exploratory-qc.html">Exploratory analysis and quality control</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/04-exploratory-qc.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 180 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why is exploratory analysis an essential part of an RNA-seq
analysis?</li>
<li>How should one preprocess the raw count matrix for exploratory
analysis?<br>
</li>
<li>Are two dimensions sufficient to represent your data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn how to explore the gene expression matrix and perform common
quality control steps.</li>
<li>Learn how to set up an interactive application for exploratory
analysis.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<hr class="half-width">
<p>Assuming you just started RStudio again, load some packages we will
use in this lesson along with the <code>SummarizedExperiment</code>
object we created in the last lesson.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">vsn</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">hexbin</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">iSEE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="remove-unexpressed-genes">Remove unexpressed genes<a class="anchor" aria-label="anchor" href="#remove-unexpressed-genes"></a>
</h2>
<hr class="half-width">
<p>Exploratory analysis is crucial for quality control and to get to
know our data. It can help us detect quality problems, sample swaps and
contamination, as well as give us a sense of the most salient patterns
present in the data. In this episode, we will learn about two common
ways of performing exploratory analysis for RNA-seq data; namely
clustering and principal component analysis (PCA). These tools are in no
way limited to (or developed for) analysis of RNA-seq data. However,
there are certain characteristics of count assays that need to be taken
into account when they are applied to this type of data. First of all,
not all mouse genes in the genome will be expressed in our Cerebellum
samples. There are many different threshold you could use to say whether
a gene’s expression was detectable or not; here we are going to use a
very minimal one that if a gene does not have more than 5 counts total
across all samples, there is simply not enough data to be able to do
anything with it anyway.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 41786</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove genes/rows that do not have &gt; 5 total counts </span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 27430</code></pre>
</div>
<div id="challenge-what-kind-of-genes-survived-this-filtering" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-what-kind-of-genes-survived-this-filtering" class="callout-inner">
<h3 class="callout-title">Challenge: What kind of genes survived this
filtering?</h3>
<div class="callout-content">
<p>Last episode we discussed subsetting down to only mRNA genes. Here we
subsetted based on a minimal expression level.</p>
<ol style="list-style-type: decimal">
<li>How many of each type of gene survived the filtering?</li>
<li>Compare the number of genes that survived filtering using different
thresholds.</li>
<li>What are pros and cons of more aggressive filtering? What are
important considerations?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">gbkey</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
     C_region          exon     J_segment      misc_RNA          mRNA
           14          1765            14          1539         16859
        ncRNA precursor_RNA          rRNA          tRNA     V_segment
         6789           362             2            64            22 </code></pre>
</div>
<ol start="2" style="list-style-type: decimal"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>  <span class="co"># represents the number of genes using 5 as filtering threshold</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 27430</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25736</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23860</code></pre>
</div>
<ol start="3" style="list-style-type: decimal">
<li>Cons: Risk of removing interesting information Pros:</li>
</ol>
<ul>
<li>Not or lowly expressed genes are unlikely to be biological
meaningful.</li>
<li>Reduces number of statistical tests (multiple testing).</li>
<li>More reliable estimation of mean-variance relationship</li>
</ul>
<p>Potential considerations: - Is a gene expressed in both groups? - How
many samples of each group express a gene?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="library-size-differences">Library size differences<a class="anchor" aria-label="anchor" href="#library-size-differences"></a>
</h2>
<hr class="half-width">
<p>Differences in the total number of reads assigned to genes between
samples typically occur for technical reasons. In practice, it means
that we can not simply compare a gene’s raw read count directly between
samples and conclude that a sample with a higher read count also
expresses the gene more strongly - the higher count may be caused by an
overall higher number of reads in that sample. In the rest of this
section, we will use the term <em>library size</em> to refer to the
total number of reads assigned to genes for a sample. First we should
compare the library sizes of all samples.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add in the sum of all counts</span></span>
<span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">libSize</span> <span class="op">&lt;-</span>  <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the libSize by using R's native pipe |&gt;</span></span>
<span><span class="co"># to extract the colData, turn it into a regular</span></span>
<span><span class="co"># data frame then send to ggplot:</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">libSize</span> <span class="op">/</span> <span class="fl">1e6</span>, fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>         <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>         <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sample"</span>, y <span class="op">=</span> <span class="st">"Total count in millions"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>         <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-lib-size-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We need to adjust for the differences in library size between
samples, to avoid drawing incorrect conclusions. The way this is
typically done for RNA-seq data can be described as a two-step
procedure. First, we estimate <em>size factors</em> - sample-specific
correction factors such that if the raw counts were to be divided by
these factors, the resulting values would be more comparable across
samples. Next, these size factors are incorporated into the statistical
analysis of the data. It is important to pay close attention to how this
is done in practice for a given analysis method. Sometimes the division
of the counts by the size factors needs to be done explicitly by the
analyst. Other times (as we will see for the differential expression
analysis) it is important that they are provided separately to the
analysis tool, which will then use them appropriately in the statistical
model.</p>
<p>With <code>DESeq2</code>, size factors are calculated using the
<code>estimateSizeFactors()</code> function. The size factors estimated
by this function combines an adjustment for differences in library sizes
with an adjustment for differences in the RNA composition of the
samples. The latter is important due to the compositional nature of
RNA-seq data. There is a fixed number of reads to distribute between the
genes, and if a single (or a few) very highly expressed gene consume a
large part of the reads, all other genes will consequently receive very
low counts. We now switch our <code>SummarizedExperiment</code> object
over to a <code>DESeqDataSet</code> as it has the internal structure to
store these size factors. We also need to tell it our main experiment
design, which is sex and time:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in DESeq2::DESeqDataSet(se, design = ~sex + time): some variables in
design formula are characters, converting to factors</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the size factors against library size</span></span>
<span><span class="co"># and look for any patterns by group:</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span>libSize <span class="op">=</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  sizeFactor <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,</span>
<span>                  Group <span class="op">=</span> <span class="va">dds</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>,</span>
<span>       <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">libSize</span>, y <span class="op">=</span> <span class="va">sizeFactor</span>, col <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size"</span>, y <span class="op">=</span> <span class="st">"Size factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-est-size-factors-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="transform-data">Transform data<a class="anchor" aria-label="anchor" href="#transform-data"></a>
</h2>
<hr class="half-width">
<p>There is a rich literature on methods for exploratory analysis. Most
of these work best in situations where the variance of the input data
(here, each gene) is relatively independent of the average value. For
read count data such as RNA-seq, this is not the case. In fact, the
variance increases with the average read count.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the vsn package.
  Please report the issue to the authors.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-mean-sd-plot-raw-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are two ways around this: either we develop methods
specifically adapted to count data, or we adapt (transform) the count
data so that the existing methods are applicable. Both ways have been
explored; however, at the moment the second approach is arguably more
widely applied in practice. We can transform our data using DESeq2’s
variance stabilizing transformation and then verify that it has removed
the correlation between average read count and variance.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-mean-sd-plot-vst-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="heatmaps-and-clustering">Heatmaps and clustering<a class="anchor" aria-label="anchor" href="#heatmaps-and-clustering"></a>
</h2>
<hr class="half-width">
<p>There are many ways to cluster samples based on their similarity of
expression patterns. One simple way is to calculate Euclidean distances
between all pairs of samples (longer distance = more different) and then
display the results with both a branching dendrogram and a heatmap to
visualize the distances in color. From this, we infer that the Day 8
samples are more similar to each other than the rest of the samples,
although Day 4 and Day 0 do not separate distinctly. Instead, males and
females reliably separate.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dst</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Blues"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="fu">as.matrix</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>, </span>
<span>    col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"Euclidean\ndistance"</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>,</span>
<span>    cluster_columns <span class="op">=</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>,</span>
<span>    bottom_annotation <span class="op">=</span> <span class="fu">columnAnnotation</span><span class="op">(</span></span>
<span>        sex <span class="op">=</span> <span class="va">vsd</span><span class="op">$</span><span class="va">sex</span>,</span>
<span>        time <span class="op">=</span> <span class="va">vsd</span><span class="op">$</span><span class="va">time</span>,</span>
<span>        col <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>sex <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Female <span class="op">=</span> <span class="st">"red"</span>, Male <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                   time <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Day0 <span class="op">=</span> <span class="st">"yellow"</span>, Day4 <span class="op">=</span> <span class="st">"forestgreen"</span>, Day8 <span class="op">=</span> <span class="st">"purple"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-heatmap-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<hr class="half-width">
<p>Principal component analysis is a dimensionality reduction method,
which projects the samples into a lower-dimensional space. This
lower-dimensional representation can be used for visualization, or as
the input for other analysis methods. The principal components are
defined in such a way that they are orthogonal, and that the projection
of the samples into the space they span contains as much variance as
possible. It is an <em>unsupervised</em> method in the sense that no
external information about the samples (e.g., the treatment condition)
is taken into account. In the plot below we represent the samples in a
two-dimensional principal component space. For each of the two
dimensions, we indicate the fraction of the total variance that is
represented by that component. By definition, the first principal
component will always represent more of the variance than the subsequent
ones. The fraction of explained variance is a measure of how much of the
‘signal’ in the data that is retained when we project the samples from
the original, high-dimensional space to the low-dimensional space for
visualization.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaData</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">vsd</span>, intgroup <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"time"</span><span class="op">)</span>,</span>
<span>                           returnData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using ntop=500 top features by variance</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaData</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaData</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">sex</span>, shape <span class="op">=</span> <span class="va">time</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Male <span class="op">=</span> <span class="st">"blue"</span>, Female <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-pca-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-discuss-the-following-points-with-your-neighbour" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbour" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbour</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Assume you are mainly interested in expression changes associated
with the time after infection (Reminder Day0 -&gt; before infection).
What do you need to consider in downstream analysis?</p></li>
<li><p>Consider an experimental design where you have multiple samples
from the same donor. You are still interested in differences by time and
observe the following PCA plot. What does this PCA plot
suggest?</p></li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using ntop=500 top features by variance</code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-pca-exercise-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li><p>The major signal in this data (37% variance) is associated with
sex. As we are not interested in sex-specific changes over time, we need
to adjust for this in downstream analysis (see <a href="05-differential-expression.html">next episodes</a>) and keep it in
mind for further exploratory downstream analysis. A possible way to do
so is to remove genes on sex chromosomes.</p></li>
<li>
</li>
</ol>
<ul>
<li>A strong donor effect, that needs to be accounted for.</li>
<li>What does PC1 (37% variance) represent? Looks like 2 donor
groups?</li>
<li>No association of PC1 and PC2 with time –&gt; no or weak
transcriptional effect of time –&gt; Check association with higher PCs
(e.g., PC3,PC4, ..)</li>
</ul>
</div>
</div>
</div>
</div>
<div id="challenge-plot-the-pca-colored-by-library-sizes." class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-plot-the-pca-colored-by-library-sizes." class="callout-inner">
<h3 class="callout-title">Challenge: Plot the PCA colored by library
sizes.</h3>
<div class="callout-content">
<p>Compare before and after variance stabilizing transformation.</p>
<p><em>Hint: The <code>DESeq2::plotPCA</code> expect an object of the
class <code>DESeqTransform</code> as input. You can transform a
<code>SummarizedExperiment</code> object using
<code>plotPCA(DESeqTransform(se))</code></em></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaDataVst</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">vsd</span>, intgroup <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"libSize"</span><span class="op">)</span>,</span>
<span>                              returnData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using ntop=500 top features by variance</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaDataVst</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaDataVst</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">libSize</span> <span class="op">/</span> <span class="fl">1e6</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_color_continuous</span><span class="op">(</span><span class="st">"Total count in millions"</span>, type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-pca-lib-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaDataCts</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span><span class="fu">DESeqTransform</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, intgroup <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"libSize"</span><span class="op">)</span>,</span>
<span>                              returnData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using ntop=500 top features by variance</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaDataCts</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaDataCts</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">libSize</span> <span class="op">/</span> <span class="fl">1e6</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_color_continuous</span><span class="op">(</span><span class="st">"Total count in millions"</span>, type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/04-exploratory-qc-rendered-pca-lib-vst-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="interactive-exploratory-data-analysis">Interactive exploratory data analysis<a class="anchor" aria-label="anchor" href="#interactive-exploratory-data-analysis"></a>
</h2>
<hr class="half-width">
<p>Often it is useful to look at QC plots in an interactive way to
directly explore different experimental factors or get insides from
someone without coding experience. Useful tools for interactive
exploratory data analysis for RNA-seq are <a href="https://bioconductor.org/packages/release/bioc/html/Glimma.html" class="external-link">Glimma</a>
and <a href="https://bioconductor.org/packages/release/bioc/html/iSEE.html" class="external-link">iSEE</a></p>
<div id="challenge-interactively-explore-our-data-using-isee" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-interactively-explore-our-data-using-isee" class="callout-inner">
<h3 class="callout-title">Challenge: Interactively explore our data
using iSEE</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Convert DESeqDataSet object to a SingleCellExperiment object, in order to </span></span>
<span><span class="co">## be able to store the PCA representation</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add PCA to the 'reducedDim' slot</span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">pcaData</span><span class="op">)</span> <span class="op">==</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">pcaData</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add variance-stabilized data as a new assay</span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span> <span class="op">==</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"vsd"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span></span>
<span><span class="va">app</span> <span class="op">&lt;-</span> <span class="fu">iSEE</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu">runApp</span><span class="op">(</span><span class="va">app</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
 [1] iSEE_2.22.0                 SingleCellExperiment_1.32.0
 [3] hexbin_1.28.5               RColorBrewer_1.1-3
 [5] ComplexHeatmap_2.26.0       ggplot2_4.0.0
 [7] vsn_3.78.0                  DESeq2_1.50.1
 [9] SummarizedExperiment_1.40.0 Biobase_2.70.0
[11] MatrixGenerics_1.22.0       matrixStats_1.5.0
[13] GenomicRanges_1.62.0        Seqinfo_1.0.0
[15] IRanges_2.44.0              S4Vectors_0.48.0
[17] BiocGenerics_0.56.0         generics_0.1.4

loaded via a namespace (and not attached):
 [1] rlang_1.1.6           magrittr_2.0.4        shinydashboard_0.7.3
 [4] clue_0.3-66           GetoptLong_1.0.5      otel_0.2.0
 [7] compiler_4.5.2        mgcv_1.9-1            png_0.1-8
[10] vctrs_0.6.5           pkgconfig_2.0.3       shape_1.4.6.1
[13] crayon_1.5.3          fastmap_1.2.0         XVector_0.50.0
[16] labeling_0.4.3        promises_1.5.0        preprocessCore_1.72.0
[19] shinyAce_0.4.4        xfun_0.54             cachem_1.1.0
[22] jsonlite_2.0.0        listviewer_4.0.0      later_1.4.4
[25] DelayedArray_0.36.0   BiocParallel_1.44.0   parallel_4.5.2
[28] cluster_2.1.8.1       R6_2.6.1              bslib_0.9.0
[31] limma_3.66.0          jquerylib_0.1.4       Rcpp_1.1.0
[34] iterators_1.0.14      knitr_1.50            httpuv_1.6.16
[37] Matrix_1.7-4          splines_4.5.2         igraph_2.2.1
[40] tidyselect_1.2.1      abind_1.4-8           yaml_2.3.10
[43] doParallel_1.0.17     codetools_0.2-19      affy_1.88.0
[46] miniUI_0.1.2          lattice_0.22-5        tibble_3.3.0
[49] shiny_1.11.1          withr_3.0.2           S7_0.2.0
[52] evaluate_1.0.5        circlize_0.4.16       pillar_1.11.1
[55] affyio_1.80.0         BiocManager_1.30.26   renv_1.1.5
[58] DT_0.34.0             foreach_1.5.2         shinyjs_2.1.0
[61] scales_1.4.0          xtable_1.8-4          glue_1.8.0
[64] tools_4.5.2           colourpicker_1.3.0    locfit_1.5-9.12
[67] colorspace_2.1-2      nlme_3.1-168          vipor_0.4.7
[70] cli_3.6.5             viridisLite_0.4.2     S4Arrays_1.10.0
[73] dplyr_1.1.4           gtable_0.3.6          rintrojs_0.3.4
[76] sass_0.4.10           digest_0.6.38         SparseArray_1.10.1
[79] ggrepel_0.9.6         rjson_0.2.23          htmlwidgets_1.6.4
[82] farver_2.1.2          htmltools_0.5.8.1     lifecycle_1.0.4
[85] shinyWidgets_0.9.0    GlobalOptions_0.1.2   statmod_1.5.1
[88] mime_0.13            </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Exploratory analysis is essential for quality control and to detect
potential problems with a data set.</li>
<li>Different classes of exploratory analysis methods expect differently
preprocessed data. The most commonly used methods expect counts to be
normalized and log-transformed (or similar- more
sensitive/sophisticated), to be closer to homoskedastic. Other methods
work directly on the raw counts.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-05-differential-expression"><p>Content from <a href="05-differential-expression.html">Differential expression analysis</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05-differential-expression.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 105 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the steps performed in a typical differential expression
analysis?</li>
<li>How does one interpret the output of DESeq2?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the steps involved in a differential expression
analysis.</li>
<li>Explain how to perform these steps in R, using DESeq2.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="differential-expression-inference">Differential expression inference<a class="anchor" aria-label="anchor" href="#differential-expression-inference"></a>
</h2>
<hr class="half-width">
<p>A major goal of RNA-seq data analysis is the quantification and
statistical inference of systematic changes between experimental groups
or conditions (e.g., treatment vs. control, timepoints, tissues). This
is typically performed by identifying genes with differential expression
pattern using between- and within-condition variability and thus
requires biological replicates (multiple sample of the same condition).
Multiple software packages exist to perform differential expression
analysis. Comparative studies have shown some concordance of
differentially expressed (DE) genes, but also variability between tools
with no tool consistently outperforming all others (see <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91" class="external-link">Soneson
and Delorenzi, 2013</a>). In the following we will explain and conduct
differential expression analysis using the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" class="external-link">DESeq2</a>
software package. The <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" class="external-link">edgeR</a>
package implements similar methods following the same main assumptions
about count data. Both packages show a general good and stable
performance with comparable results.</p>
</section><section><h2 class="section-heading" id="the-deseqdataset">The DESeqDataSet<a class="anchor" aria-label="anchor" href="#the-deseqdataset"></a>
</h2>
<hr class="half-width">
<p>To run <code>DESeq2</code> we need to represent our count data as
object of the <code>DESeqDataSet</code> class. The
<code>DESeqDataSet</code> is an extension of the
<code>SummarizedExperiment</code> class (see section <a href="03-import-annotate.html">Importing and annotating quantified data
into R</a> ) that stores a <em>design formula</em> in addition to the
count assay(s) and feature (here gene) and sample metadata. The
<em>design formula</em> expresses the variables which will be used in
modeling. These are typically the variable of interest (group variable)
and other variables you want to account for (e.g., batch effect
variables). A detailed explanation of <em>design formulas</em> and
related <em>design matrices</em> will follow in the section about <a href="06-extra-design.html">extra exploration of design matrices</a>.
Objects of the <code>DESeqDataSet</code> class can be build from <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat" class="external-link">count
matrices</a>, <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#se" class="external-link">SummarizedExperiment
objects</a>, <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#tximport" class="external-link">transcript
abundance files</a> or <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#htseq" class="external-link">htseq
count files</a>.</p>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ExploreModelMatrix</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">cowplot</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">apeglm</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>Let’s load in our <code>SummarizedExperiment</code> object again. In
the last episode for quality control exploration, we removed ~35% genes
that had 5 or fewer counts because they had too little information in
them. For DESeq2 statistical analysis, we do not technically have to
remove these genes because by default it will do some independent
filtering, but it can reduce the memory size of the
<code>DESeqDataSet</code> object resulting in faster computation. Plus,
we do not want these genes cluttering up some of the visualizations.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="create-deseqdataset">Create DESeqDataSet<a class="anchor" aria-label="anchor" href="#create-deseqdataset"></a>
</h3>
<p>The design matrix we will use in this example is
<code>~ sex + time</code>. This will allow us test the difference
between males and females (averaged over time point) and the difference
between day 0, 4 and 8 (averaged over males and females). If we wanted
to test other comparisons (e.g., Female.Day8 vs. Female.Day0 and also
Male.Day8 vs. Male.Day0) we could use a different design matrix to more
easily extract those pairwise comparisons.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in DESeq2::DESeqDataSet(se, design = ~sex + time): some variables in
design formula are characters, converting to factors</code></pre>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>The function to generate a <code>DESeqDataSet</code> needs to be
adapted depending on the input type, e.g,</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#From SummarizedExperiment object</span></span>
<span><span class="va">ddsSE</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#From count matrix</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="fu">assays</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>                              colData <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<hr class="half-width">
<p><code>DESeq2</code> and <code>edgeR</code> make the following
assumptions:</p>
<ul>
<li>most genes are not differentially expressed</li>
<li>the probability of a read mapping to a specific gene is the same for
all samples within the same group</li>
</ul>
<p>As shown in the <a href="04-exploratory-qc.html">previous section</a>
on exploratory data analysis the total counts of a sample (even from the
same condition) depends on the library size (total number of reads
sequenced). To compare the variability of counts from a specific gene
between and within groups we first need to account for library sizes and
compositional effects. Recall the <code>estimateSizeFactors()</code>
function from the previous section:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" aria-labelledby="headingInstructor2" data-bs-parent="#accordionInstructor2">
<div class="accordion-body">
<p><em>DESeq2</em> uses the <strong>“Relative Log Expression”
(RLE)</strong> method to calculate sample-wise <em>size factors</em>
tĥat account for read depth and library composition. <em>edgeR</em> uses
the <strong>“Trimmed Mean of M-Values” (TMM)</strong> method to account
for library size differences and compositional effects. <em>edgeR</em>’s
<em>normalization factors</em> and <em>DESeq2</em>’s <em>size
factors</em> yield similar results, but are not equivalent theoretical
parameters.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="statistical-modeling">Statistical modeling<a class="anchor" aria-label="anchor" href="#statistical-modeling"></a>
</h2>
<hr class="half-width">
<p><code>DESeq2</code> and <code>edgeR</code> model RNA-seq counts as
<strong>negative binomial</strong> distribution to account for a limited
number of replicates per group, a mean-variance dependency (see <a href="04-exploratory-qc.html">exploratory data analysis</a>) and a
skewed count distribution.</p>
<div class="section level3">
<h3 id="dispersion">Dispersion<a class="anchor" aria-label="anchor" href="#dispersion"></a>
</h3>
<p>The within-group variance of the counts for a gene following a
negative binomial distribution with mean <span class="math inline">\(\mu\)</span> can be modeled as:</p>
<p><span class="math inline">\(var = \mu + \theta \mu^2\)</span></p>
<p><span class="math inline">\(\theta\)</span> represents the
gene-specific <strong>dispersion</strong>, a measure of variability or
spread in the data. As a second step, we need to estimate gene-wise
dispersions to get the expected within-group variance and test for group
differences. Good dispersion estimates are challenging with a few
samples per group only. Thus, information from genes with similar
expression pattern are “borrowed”. Gene-wise dispersion estimates are
<em>shrinked</em> towards center values of the observed distribution of
dispersions. With <code>DESeq2</code> we can get dispersion estimates
using the <code>estimateDispersions()</code> function. We can visualize
the effect of <em>shrinkage</em> using <code>plotDispEsts()</code>:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateDispersions</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/05-differential-expression-rendered-estimate-dispersions-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level3">
<h3 id="testing">Testing<a class="anchor" aria-label="anchor" href="#testing"></a>
</h3>
<p>We can use the <code>nbinomWaldTest()</code>function of
<code>DESeq2</code> to fit a <em>generalized linear model (GLM)</em> and
compute <em>log2 fold changes</em> (synonymous with “GLM coefficients”,
“beta coefficients” or “effect size”) corresponding to the variables of
the <em>design matrix</em>. The <em>design matrix</em> is directly
related to the <em>design formula</em> and automatically derived from
it. Assume a design formula with one variable (<code>~ treatment</code>)
and two factor levels (treatment and control). The mean expression <span class="math inline">\(\mu_{j}\)</span> of a specific gene in sample
<span class="math inline">\(j\)</span> will be modeled as following:</p>
<p><span class="math inline">\(log(μ_j) = β_0 + x_j β_T\)</span>,</p>
<p>with <span class="math inline">\(β_T\)</span> corresponding to the
log2 fold change of the treatment groups, <span class="math inline">\(x_j\)</span> = 1, if <span class="math inline">\(j\)</span> belongs to the treatment group and
<span class="math inline">\(x_j\)</span> = 0, if <span class="math inline">\(j\)</span> belongs to the control group.</p>
<p>Finally, the estimated log2 fold changes are scaled by their standard
error and tested for being significantly different from 0 using the
<em>Wald test</em>.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">nbinomWaldTest</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>Standard differential expression analysis as performed above is
wrapped into a single function, <code>DESeq()</code>. Running the first
code chunk is equivalent to running the second one:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateDispersions</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">nbinomWaldTest</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="explore-results-for-specific-contrasts">Explore results for specific contrasts<a class="anchor" aria-label="anchor" href="#explore-results-for-specific-contrasts"></a>
</h2>
<hr class="half-width">
<p>The <code>results()</code> function can be used to extract gene-wise
test statistics, such as log2 fold changes and (adjusted) p-values. The
comparison of interest can be defined using contrasts, which are linear
combinations of the model coefficients (equivalent to combinations of
columns within the <em>design matrix</em>) and thus directly related to
the design formula. A detailed explanation of design matrices and how to
use them to specify different contrasts of interest can be found in the
section on the <a href="06-extra-design.html">exploration of design
matrices</a>. In the <code>results()</code> function a contrast can be
represented by the variable of interest (reference variable) and the
related level to compare using the <code>contrast</code> argument. By
default the reference variable will be the <strong>last
variable</strong> of the design formula, the <em>reference level</em>
will be the first factor level and the <em>last level</em> will be used
for comparison. You can also explicitly specify a contrast by the
<code>name</code> argument of the <code>results()</code> function. Names
of all available contrasts can be accessed using
<code>resultsNames()</code>.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>What will be the default <strong>contrast</strong>, <strong>reference
level</strong> and <strong>“last level”</strong> for comparisons when
running <code>results(dds)</code> for the example used in this
lesson?</p>
<p><em>Hint: Check the design formula used to build the object.</em></p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>In the lesson example the last variable of the design formula is
<code>time</code>. The <strong>reference level</strong> (first in
alphabetical order) is <code>Day0</code> and the <strong>last
level</strong> is <code>Day8</code></p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">dds</span><span class="op">$</span><span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Day0" "Day4" "Day8"</code></pre>
</div>
<p>No worries, if you had difficulties to identify the default contrast
the output of the <code>results()</code> function explicitly states the
contrast it is referring to (see below)!</p>
</div>
</div>
</div>
</div>
<p>To explore the output of the <code>results()</code> function we can
use the <code>summary()</code> function and order results by
significance (p-value). Here we assume that we are interested in changes
over <code>time</code> (“variable of interest”), more specifically genes
with differential expression between <code>Day0</code> (“reference
level”) and <code>Day8</code> (“level to compare”). The model we used
included the <code>sex</code> variable (see above). Thus our results
will be “corrected” for sex-related differences.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Day 8 vs Day 0</span></span>
<span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 16%
LFC &lt; 0 (down)     : 4282, 16%
outliers [1]       : 10, 0.036%
low counts [2]     : 3723, 14%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(resTime)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTime</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>log2 fold change (MLE): time Day8 vs Day0
Wald test p-value: time Day8 vs Day0
DataFrame with 6 rows and 6 columns
               baseMean log2FoldChange     lfcSE      stat      pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
Asl             701.343       1.117332 0.0594128   18.8062 6.71212e-79
Apod          18765.146       1.446981 0.0805056   17.9737 3.13229e-72
Cyp2d22        2550.480       0.910202 0.0556002   16.3705 3.10712e-60
Klk6            546.503      -1.671897 0.1057395  -15.8115 2.59339e-56
Fcrls           184.235      -1.947016 0.1277235  -15.2440 1.80488e-52
A330076C08Rik   107.250      -1.749957 0.1155125  -15.1495 7.63434e-52
                     padj
                &lt;numeric&gt;
Asl           1.59057e-74
Apod          3.71130e-68
Cyp2d22       2.45431e-56
Klk6          1.53639e-52
Fcrls         8.55406e-49
A330076C08Rik 3.01518e-48</code></pre>
</div>
<div id="accordionInstructor3" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor3" aria-expanded="false" aria-controls="collapseInstructor3">
  <h3 class="accordion-header" id="headingInstructor3">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor3" class="accordion-collapse collapse" aria-labelledby="headingInstructor3" data-bs-parent="#accordionInstructor3">
<div class="accordion-body">
<p>Both of the below ways of specifying the contrast are essentially
equivalent. The <code>name</code> parameter can be accessed using
<code>resultsNames()</code>.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, name <span class="op">=</span> <span class="st">"time_Day8_vs_Day0"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Explore the DE genes between males and females independent of
time.</p>
<p><em>Hint: You don’t need to fit the GLM again. Use
<code>resultsNames()</code> to get the correct contrast.</em></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Male vs Female</span></span>
<span><span class="va">resSex</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resSex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 51, 0.19%
LFC &lt; 0 (down)     : 70, 0.26%
outliers [1]       : 10, 0.036%
low counts [2]     : 8504, 31%
(mean count &lt; 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">resSex</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resSex</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>log2 fold change (MLE): sex Male vs Female
Wald test p-value: sex Male vs Female
DataFrame with 6 rows and 6 columns
               baseMean log2FoldChange     lfcSE      stat       pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;
Xist         22603.0359      -11.60429  0.336282  -34.5076 6.16852e-261
Ddx3y         2072.9436       11.87241  0.397493   29.8683 5.08722e-196
Eif2s3y       1410.8750       12.62513  0.565194   22.3377 1.58997e-110
Kdm5d          692.1672       12.55386  0.593607   21.1484  2.85293e-99
Uty            667.4375       12.01728  0.593573   20.2457  3.87772e-91
LOC105243748    52.9669        9.08325  0.597575   15.2002  3.52699e-52
                     padj
                &lt;numeric&gt;
Xist         1.16684e-256
Ddx3y        4.81149e-192
Eif2s3y      1.00253e-106
Kdm5d         1.34915e-95
Uty           1.46702e-87
LOC105243748  1.11194e-48</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="multiple-testing-correction" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="multiple-testing-correction" class="callout-inner">
<h3 class="callout-title">Multiple testing correction</h3>
<div class="callout-content">
<p>Due to the high number of tests (one per gene) our DE results will
contain a substantial number of <strong>false positives</strong>. For
example, if we tested 20,000 genes at a threshold of <span class="math inline">\(\alpha = 0.05\)</span> we would expect 1,000
significant DE genes with no differential expression.</p>
<p>To account for this expected high number of false positives, we can
correct our results for <strong>multiple testing</strong>. By default
<code>DESeq2</code> uses the <a href="https://link.springer.com/referenceworkentry/10.1007/978-1-4419-9863-7_1215" class="external-link">Benjamini-Hochberg
procedure</a> to calculate <strong>adjusted p-values</strong> (padj) for
DE results.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="independent-filtering-and-log-fold-shrinkage">Independent Filtering and log-fold shrinkage<a class="anchor" aria-label="anchor" href="#independent-filtering-and-log-fold-shrinkage"></a>
</h2>
<hr class="half-width">
<p>We can visualize the results in many ways. A good check is to explore
the relationship between <em>log2fold changes</em>, <em>significant DE
genes</em> and the <em>genes mean count</em>. <code>DESeq2</code>
provides a useful function to do so, <code>plotMA()</code>.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotMA</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/05-differential-expression-rendered-plot-ma-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can see that genes with a low mean count tend to have larger log
fold changes. This is caused by counts from lowly expressed genes
tending to be very noisy. We can <em>shrink</em> the log fold changes of
these genes with low mean and high dispersion, as they contain little
information.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeLfc</span> <span class="op">&lt;-</span> <span class="fu">lfcShrink</span><span class="op">(</span><span class="va">dds</span>, coef <span class="op">=</span> <span class="st">"time_Day8_vs_Day0"</span>, res <span class="op">=</span> <span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotMA</span><span class="op">(</span><span class="va">resTimeLfc</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="../fig/05-differential-expression-rendered-res-time-lfc-1.png" style="display: block; margin: auto;" class="figure">
Shrinkage of log fold changes is useful for visualization and ranking of
genes, but for result exploration typically the
<code>independentFiltering</code> argument is used to remove lowly
expressed genes.</p>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>By default <code>independentFiltering</code> is set to
<code>TRUE</code>. What happens without filtering lowly expressed genes?
Use the <code>summary()</code> function to compare the results. Most of
the lowly expressed genes are not significantly differential expressed
(blue in the above MA plots). What could cause the difference in the
results then?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeNotFiltered</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                              contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span>, </span>
<span>                              independentFiltering <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 16%
LFC &lt; 0 (down)     : 4282, 16%
outliers [1]       : 10, 0.036%
low counts [2]     : 3723, 14%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeNotFiltered</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4324, 16%
LFC &lt; 0 (down)     : 4129, 15%
outliers [1]       : 10, 0.036%
low counts [2]     : 0, 0%
(mean count &lt; 0)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<p>Genes with very low counts are not likely to see significant
differences typically due to high dispersion. Filtering of lowly
expressed genes thus increased detection power at the same
experiment-wide false positive rate.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="visualize-selected-set-of-genes">Visualize selected set of genes<a class="anchor" aria-label="anchor" href="#visualize-selected-set-of-genes"></a>
</h2>
<hr class="half-width">
<p>The amount of DE genes can be overwhelming and a ranked list of genes
can still be hard to interpret with regards to an experimental question.
Visualizing gene expression can help to detect expression pattern or
group of genes with related functions. We will perform systematic
detection of over represented groups of genes in a <a href="07-gene-set-analysis.html">later section</a>. Before this
visualization can already help us to get a good intuition about what to
expect.</p>
<p>We will use transformed data (see <a href="04-exploratory-qc.html">exploratory data analysis</a>) and the top
differentially expressed genes for visualization. A heatmap can reveal
expression pattern across sample groups (columns) and automatically
orders genes (rows) according to their similarity.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transform counts</span></span>
<span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get top DE genes</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">resTime</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>         <span class="fu">head</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>         <span class="fu">rownames</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">heatmapData</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Scale counts for visualization</span></span>
<span><span class="va">heatmapData</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">scale</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">heatmapData</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add annotation</span></span>
<span><span class="va">heatmapColAnnot</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">heatmapColAnnot</span> <span class="op">&lt;-</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span>df <span class="op">=</span> <span class="va">heatmapColAnnot</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Plot as heatmap</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu">Heatmap</span><span class="op">(</span><span class="va">heatmapData</span>,</span>
<span>                        top_annotation <span class="op">=</span> <span class="va">heatmapColAnnot</span>,</span>
<span>                        cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/05-differential-expression-rendered-heatmap-time-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Check the heatmap and top DE genes. Do you find something
expected/unexpected in terms of change across all 3 time points?</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="output-results">Output results<a class="anchor" aria-label="anchor" href="#output-results"></a>
</h2>
<hr class="half-width">
<p>We may want to to output our results out of R to have a stand-alone
file. The format of <code>resTime</code> only has the gene symbols as
rownames, so let us join the gene annotation information, and then write
out as .csv file:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write.csv</span><span class="op">(</span><span class="va">temp</span>, file <span class="op">=</span> <span class="st">"output/Day8vsDay0.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>With DESeq2, the main steps of a differential expression analysis
(size factor estimation, dispersion estimation, calculation of test
statistics) are wrapped in a single function: DESeq().</li>
<li>Independent filtering of lowly expressed genes is often
beneficial.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-06-extra-design"><p>Content from <a href="06-extra-design.html">Extra exploration of design matrices</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/06-extra-design.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can one translate biological questions and comparisons to
statistical terms suitable for use with RNA-seq analysis packages?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the formula notation and design matrices.</li>
<li>Explore different designs and learn how to interpret
coefficients.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="loading-required-packages-and-reading-data">Loading required packages and reading data<a class="anchor" aria-label="anchor" href="#loading-required-packages-and-reading-data"></a>
</h2>
<hr class="half-width">
<p>We start by loading a few packages that will be needed in this
episode. In particular, the <a href="https://bioconductor.org/packages/ExploreModelMatrix/" class="external-link">ExploreModelMatrix</a>
package provides resources for exploring design matrices in a graphical
fashion, for easier interpretation.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ExploreModelMatrix</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Next, we read the metadata table for our data set. Because we want to
explore many different design matrices, we will read in the 4th file we
downloaded but haven’t used yet: that for both Cerebellum and Spinal
Cord samples (45 samples total). As seen in previous episodes, the
metadata contains information about the age, sex, infection status, time
of measurement and tissue of the collected samples. Note that Day0
always corresponds to non-infected samples, and that infected samples
are collected on days 4 and 8. Moreover, all mice have the same age (8
weeks). Hence, in the first part of this episode we consider only the
sex, tissue and time variables further.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_coldata_all.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">meta</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus 8 weeks Female
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus 8 weeks Female
GSM2545340 CNS_RNA-seq_14C    GSM2545340 Mus musculus 8 weeks   Male
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus 8 weeks   Male
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545344 CNS_RNA-seq_21C    GSM2545344 Mus musculus 8 weeks Female
GSM2545345 CNS_RNA-seq_22C    GSM2545345 Mus musculus 8 weeks   Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus 8 weeks   Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545350 CNS_RNA-seq_29C    GSM2545350 Mus musculus 8 weeks   Male
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus 8 weeks Female
GSM2545352 CNS_RNA-seq_30C    GSM2545352 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545355 CNS_RNA-seq_571    GSM2545355 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545359 CNS_RNA-seq_585    GSM2545359 Mus musculus 8 weeks Female
GSM2545360 CNS_RNA-seq_589    GSM2545360 Mus musculus 8 weeks   Male
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus 8 weeks Female
GSM2545363  CNS_RNA-seq_6C    GSM2545363 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
GSM2545368 CNS_RNA-seq_728    GSM2545368 Mus musculus 8 weeks   Male
GSM2545369 CNS_RNA-seq_729    GSM2545369 Mus musculus 8 weeks   Male
GSM2545370 CNS_RNA-seq_730    GSM2545370 Mus musculus 8 weeks Female
GSM2545371 CNS_RNA-seq_731    GSM2545371 Mus musculus 8 weeks Female
GSM2545372 CNS_RNA-seq_733    GSM2545372 Mus musculus 8 weeks   Male
GSM2545373 CNS_RNA-seq_735    GSM2545373 Mus musculus 8 weeks   Male
GSM2545374 CNS_RNA-seq_736    GSM2545374 Mus musculus 8 weeks Female
GSM2545375 CNS_RNA-seq_738    GSM2545375 Mus musculus 8 weeks Female
GSM2545376 CNS_RNA-seq_740    GSM2545376 Mus musculus 8 weeks Female
GSM2545377 CNS_RNA-seq_741    GSM2545377 Mus musculus 8 weeks Female
GSM2545378 CNS_RNA-seq_742    GSM2545378 Mus musculus 8 weeks   Male
GSM2545379 CNS_RNA-seq_743    GSM2545379 Mus musculus 8 weeks   Male
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545336  InfluenzaA C57BL/6 Day8 Cerebellum    14
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545339  InfluenzaA C57BL/6 Day4 Cerebellum    15
GSM2545340  InfluenzaA C57BL/6 Day4 Cerebellum    18
GSM2545341  InfluenzaA C57BL/6 Day8 Cerebellum     6
GSM2545342  InfluenzaA C57BL/6 Day8 Cerebellum     5
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545344  InfluenzaA C57BL/6 Day4 Cerebellum    22
GSM2545345  InfluenzaA C57BL/6 Day4 Cerebellum    13
GSM2545346  InfluenzaA C57BL/6 Day8 Cerebellum    23
GSM2545347  InfluenzaA C57BL/6 Day8 Cerebellum    24
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545350  InfluenzaA C57BL/6 Day4 Cerebellum     1
GSM2545351  InfluenzaA C57BL/6 Day8 Cerebellum    16
GSM2545352  InfluenzaA C57BL/6 Day4 Cerebellum    21
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545355  InfluenzaA C57BL/6 Day4 Spinalcord     1
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545359  InfluenzaA C57BL/6 Day8 Spinalcord     5
GSM2545360  InfluenzaA C57BL/6 Day8 Spinalcord     6
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545362  InfluenzaA C57BL/6 Day4 Cerebellum    20
GSM2545363  InfluenzaA C57BL/6 Day4 Cerebellum    12
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11
GSM2545368  InfluenzaA C57BL/6 Day4 Spinalcord    12
GSM2545369  InfluenzaA C57BL/6 Day4 Spinalcord    13
GSM2545370  InfluenzaA C57BL/6 Day8 Spinalcord    14
GSM2545371  InfluenzaA C57BL/6 Day4 Spinalcord    15
GSM2545372  InfluenzaA C57BL/6 Day8 Spinalcord    17
GSM2545373  InfluenzaA C57BL/6 Day4 Spinalcord    18
GSM2545374  InfluenzaA C57BL/6 Day8 Spinalcord    19
GSM2545375  InfluenzaA C57BL/6 Day4 Spinalcord    20
GSM2545376  InfluenzaA C57BL/6 Day4 Spinalcord    21
GSM2545377  InfluenzaA C57BL/6 Day4 Spinalcord    22
GSM2545378  InfluenzaA C57BL/6 Day8 Spinalcord    23
GSM2545379  InfluenzaA C57BL/6 Day8 Spinalcord    24
GSM2545380  InfluenzaA C57BL/6 Day8 Cerebellum    19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">time</span>, <span class="va">meta</span><span class="op">$</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
       InfluenzaA NonInfected
  Day0          0          15
  Day4         16           0
  Day8         14           0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
8 weeks
     45 </code></pre>
</div>
<p>We can start by visualizing the number of observations for each
combination of the three predictor variables.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">tissue</span> <span class="op">+</span> <span class="va">time</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">cooccurrenceplots</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`tissue = Cerebellum`</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-cooccplots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
$`tissue = Spinalcord`</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-cooccplots-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Based on this visualization, would you say that the data set is
balanced, or are there combinations of predictor variables that are
severely over- or underrepresented?</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="compare-males-and-females-non-infected-spinal-cord">Compare males and females, non-infected spinal cord<a class="anchor" aria-label="anchor" href="#compare-males-and-females-non-infected-spinal-cord"></a>
</h2>
<hr class="half-width">
<p>Next, we will set up our first design matrix. Here, we will focus on
the uninfected (Day0) spinal cord samples, and our aim is to compare the
male and female mice. Thus, we first subset the metadata to only the
samples of interest, and next set up and visualize the design matrix
with a single predictor variable (sex). By defining the design formula
as <code>~ sex</code>, we tell R to include an intercept in the design.
This intercept will represent the ‘baseline’ level of the predictor
variable, which in this case is selected to be the Female mice. If not
explicitly specified, R will order the values of the predictor in
alphabetical order and select the first one as the reference or baseline
level.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset metadata</span></span>
<span><span class="va">meta_noninf_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span> <span class="op">&amp;</span> </span>
<span>                                       <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use ExploreModelMatrix to create a design matrix and visualizations, given </span></span>
<span><span class="co">## the desired design formula. </span></span>
<span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf_spc</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale
GSM2545356           1       1
GSM2545357           1       1
GSM2545358           1       0
GSM2545361           1       1
GSM2545364           1       0
GSM2545365           1       0
GSM2545366           1       0
GSM2545367           1       1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-male-vs-female-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Note that we can also generate the design matrix like this</span></span>
<span><span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span> <span class="va">sex</span>, data <span class="op">=</span> <span class="va">meta_noninf_spc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale
GSM2545356           1       1
GSM2545357           1       1
GSM2545358           1       0
GSM2545361           1       1
GSM2545364           1       0
GSM2545365           1       0
GSM2545366           1       0
GSM2545367           1       1
attr(,"assign")
[1] 0 1
attr(,"contrasts")
attr(,"contrasts")$sex
[1] "contr.treatment"</code></pre>
</div>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>With this design, what is the interpretation of the
<code>sexMale</code> coefficient?</p>
</div>
</div>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Set up the design formula to compare male and female spinal cord
samples from Day0 as above, but instruct R to not include an intercept
in the model. How does this change the interpretation of the
coefficients? What contrast would have to be specified to compare the
mean expression of a gene between male and female mice?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span> <span class="op">&amp;</span> </span>
<span>                                       <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf_spc</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           sexFemale sexMale
GSM2545356         0       1
GSM2545357         0       1
GSM2545358         1       0
GSM2545361         0       1
GSM2545364         1       0
GSM2545365         1       0
GSM2545366         1       0
GSM2545367         0       1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-male-vs-female-noint-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Set up the design formula to compare the three time points (Day0,
Day4, Day8) in the male spinal cord samples, and visualize it using
<code>ExploreModelMatrix</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_male_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span> <span class="op">&amp;</span> <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_male_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age  sex   infection
GSM2545355 CNS_RNA-seq_571    GSM2545355 Mus musculus 8 weeks Male  InfluenzaA
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks Male NonInfected
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks Male NonInfected
GSM2545360 CNS_RNA-seq_589    GSM2545360 Mus musculus 8 weeks Male  InfluenzaA
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks Male NonInfected
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks Male NonInfected
GSM2545368 CNS_RNA-seq_728    GSM2545368 Mus musculus 8 weeks Male  InfluenzaA
GSM2545369 CNS_RNA-seq_729    GSM2545369 Mus musculus 8 weeks Male  InfluenzaA
GSM2545372 CNS_RNA-seq_733    GSM2545372 Mus musculus 8 weeks Male  InfluenzaA
GSM2545373 CNS_RNA-seq_735    GSM2545373 Mus musculus 8 weeks Male  InfluenzaA
GSM2545378 CNS_RNA-seq_742    GSM2545378 Mus musculus 8 weeks Male  InfluenzaA
GSM2545379 CNS_RNA-seq_743    GSM2545379 Mus musculus 8 weeks Male  InfluenzaA
            strain time     tissue mouse
GSM2545355 C57BL/6 Day4 Spinalcord     1
GSM2545356 C57BL/6 Day0 Spinalcord     2
GSM2545357 C57BL/6 Day0 Spinalcord     3
GSM2545360 C57BL/6 Day8 Spinalcord     6
GSM2545361 C57BL/6 Day0 Spinalcord     7
GSM2545367 C57BL/6 Day0 Spinalcord    11
GSM2545368 C57BL/6 Day4 Spinalcord    12
GSM2545369 C57BL/6 Day4 Spinalcord    13
GSM2545372 C57BL/6 Day8 Spinalcord    17
GSM2545373 C57BL/6 Day4 Spinalcord    18
GSM2545378 C57BL/6 Day8 Spinalcord    23
GSM2545379 C57BL/6 Day8 Spinalcord    24</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_male_spc</span>, designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) timeDay4 timeDay8
GSM2545355           1        1        0
GSM2545356           1        0        0
GSM2545357           1        0        0
GSM2545360           1        0        1
GSM2545361           1        0        0
GSM2545367           1        0        0
GSM2545368           1        1        0
GSM2545369           1        1        0
GSM2545372           1        0        1
GSM2545373           1        1        0
GSM2545378           1        0        1
GSM2545379           1        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-all-times-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="factorial-design-without-interactions">Factorial design without interactions<a class="anchor" aria-label="anchor" href="#factorial-design-without-interactions"></a>
</h2>
<hr class="half-width">
<p>Next, we again consider only non-infected mice, but fit a model
incorporating both sex and tissue as predictors. We assume that the
tissue differences are the same for both male and female mice, and
consequently fit an additive model, without interaction terms.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale tissueSpinalcord
GSM2545337           1       0                0
GSM2545338           1       0                0
GSM2545343           1       1                0
GSM2545348           1       0                0
GSM2545349           1       1                0
GSM2545353           1       0                0
GSM2545354           1       1                0
GSM2545356           1       1                1
GSM2545357           1       1                1
GSM2545358           1       0                1
GSM2545361           1       1                1
GSM2545364           1       0                1
GSM2545365           1       0                1
GSM2545366           1       0                1
GSM2545367           1       1                1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-factorial-noint-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="factorial-design-with-interactions">Factorial design with interactions<a class="anchor" aria-label="anchor" href="#factorial-design-with-interactions"></a>
</h2>
<hr class="half-width">
<p>In the previous model, we assumed that the tissue differences were
the same for both male and female mice. To allow for the estimation of
sex-specific tissue differences (at the expense of having one additional
coefficient to estimate from the data), we can include an interaction
term in the model.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define a design including an interaction term</span></span>
<span><span class="co">## Note that ~ sex * tissue is equivalent to </span></span>
<span><span class="co">## ~ sex + tissue + sex:tissue</span></span>
<span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">*</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale tissueSpinalcord sexMale:tissueSpinalcord
GSM2545337           1       0                0                        0
GSM2545338           1       0                0                        0
GSM2545343           1       1                0                        0
GSM2545348           1       0                0                        0
GSM2545349           1       1                0                        0
GSM2545353           1       0                0                        0
GSM2545354           1       1                0                        0
GSM2545356           1       1                1                        1
GSM2545357           1       1                1                        1
GSM2545358           1       0                1                        0
GSM2545361           1       1                1                        1
GSM2545364           1       0                1                        0
GSM2545365           1       0                1                        0
GSM2545366           1       0                1                        0
GSM2545367           1       1                1                        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-factorial-withint-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="combining-multiple-factors-into-one">Combining multiple factors into one<a class="anchor" aria-label="anchor" href="#combining-multiple-factors-into-one"></a>
</h2>
<hr class="half-width">
<p>Sometimes, for experiments with multiple factors, it is easier to
interpret coefficients and set up contrasts of interest if the factors
are combined into one. Let’s consider the previous example again, using
this approach:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span><span class="op">$</span><span class="va">sex_tissue</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">meta_noninf</span><span class="op">$</span><span class="va">sex</span>, <span class="st">"_"</span>, <span class="va">meta_noninf</span><span class="op">$</span><span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse        sex_tissue
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9 Female_Cerebellum
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10 Female_Cerebellum
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11   Male_Cerebellum
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8 Female_Cerebellum
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7   Male_Cerebellum
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4 Female_Cerebellum
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2   Male_Cerebellum
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2   Male_Spinalcord
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3   Male_Spinalcord
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4 Female_Spinalcord
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7   Male_Spinalcord
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8 Female_Spinalcord
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9 Female_Spinalcord
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10 Female_Spinalcord
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11   Male_Spinalcord</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sex_tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           sex_tissueFemale_Cerebellum sex_tissueFemale_Spinalcord
GSM2545337                           1                           0
GSM2545338                           1                           0
GSM2545343                           0                           0
GSM2545348                           1                           0
GSM2545349                           0                           0
GSM2545353                           1                           0
GSM2545354                           0                           0
GSM2545356                           0                           0
GSM2545357                           0                           0
GSM2545358                           0                           1
GSM2545361                           0                           0
GSM2545364                           0                           1
GSM2545365                           0                           1
GSM2545366                           0                           1
GSM2545367                           0                           0
           sex_tissueMale_Cerebellum sex_tissueMale_Spinalcord
GSM2545337                         0                         0
GSM2545338                         0                         0
GSM2545343                         1                         0
GSM2545348                         0                         0
GSM2545349                         1                         0
GSM2545353                         0                         0
GSM2545354                         1                         0
GSM2545356                         0                         1
GSM2545357                         0                         1
GSM2545358                         0                         0
GSM2545361                         0                         1
GSM2545364                         0                         0
GSM2545365                         0                         0
GSM2545366                         0                         0
GSM2545367                         0                         1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-factorial-combine-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="paired-design">Paired design<a class="anchor" aria-label="anchor" href="#paired-design"></a>
</h2>
<hr class="half-width">
<p>In this particular data set the samples are paired - the same mice
have contributed both the cerebellum and spinal cord samples. This
information was not included in the previous models. However, accounting
for it can increase power to detect tissue differences by eliminating
variability in baseline expression levels between mice. Here, we define
a paired design for the female non-infected mice, aimed at testing for
differences between tissues after accounting for baseline differences
between mice.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_fem_day0</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> </span>
<span>                                     <span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ensure that mouse is treated as a categorical variable</span></span>
<span><span class="va">meta_fem_day0</span><span class="op">$</span><span class="va">mouse</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">meta_fem_day0</span><span class="op">$</span><span class="va">mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_fem_day0</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_fem_day0</span>,</span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">mouse</span> <span class="op">+</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 tissueSpinalcord
GSM2545337           1      0      1       0                0
GSM2545338           1      0      0       1                0
GSM2545348           1      1      0       0                0
GSM2545353           1      0      0       0                0
GSM2545358           1      0      0       0                1
GSM2545364           1      1      0       0                1
GSM2545365           1      0      1       0                1
GSM2545366           1      0      0       1                1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-paired-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="within--and-between-subject-comparisons">Within- and between-subject comparisons<a class="anchor" aria-label="anchor" href="#within--and-between-subject-comparisons"></a>
</h2>
<hr class="half-width">
<p>In some situations, we need to combine the types of models considered
above. For example, let’s say that we want to investigate if the tissue
differences are different for infected and non-infected female mice. In
this case, each mice only contributes to one of the infection groups
(each mice is either infected or non-infected), but contributes both a
cerebellum and a spinal cord sample. One way to view this type of design
is as two paired experiments, one for each infection group (see the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf" class="external-link">edgeR
user guide section 3.5</a>). We can then easily compare the two tissues
in each infection group, and contrast the tissue differences between the
infection groups.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_fem_day04</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> </span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> </span>
<span>               <span class="va">time</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Day0"</span>, <span class="st">"Day4"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">droplevels</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># ensure that mouse is treated as a categorical variable</span></span>
<span><span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">mouse</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_fem_day04</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus 8 weeks Female
GSM2545344 CNS_RNA-seq_21C    GSM2545344 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545352 CNS_RNA-seq_30C    GSM2545352 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus 8 weeks Female
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545371 CNS_RNA-seq_731    GSM2545371 Mus musculus 8 weeks Female
GSM2545375 CNS_RNA-seq_738    GSM2545375 Mus musculus 8 weeks Female
GSM2545376 CNS_RNA-seq_740    GSM2545376 Mus musculus 8 weeks Female
GSM2545377 CNS_RNA-seq_741    GSM2545377 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545339  InfluenzaA C57BL/6 Day4 Cerebellum    15
GSM2545344  InfluenzaA C57BL/6 Day4 Cerebellum    22
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545352  InfluenzaA C57BL/6 Day4 Cerebellum    21
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545362  InfluenzaA C57BL/6 Day4 Cerebellum    20
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545371  InfluenzaA C57BL/6 Day4 Spinalcord    15
GSM2545375  InfluenzaA C57BL/6 Day4 Spinalcord    20
GSM2545376  InfluenzaA C57BL/6 Day4 Spinalcord    21
GSM2545377  InfluenzaA C57BL/6 Day4 Spinalcord    22</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span> <span class="va">mouse</span>, data <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">design</span>, </span>
<span>                Spc.Day0 <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span>,</span>
<span>                Spc.Day4 <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day4"</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">meta_fem_day04</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 mouse15 mouse20 mouse21 mouse22
GSM2545337           1      0      1       0       0       0       0       0
GSM2545338           1      0      0       1       0       0       0       0
GSM2545339           1      0      0       0       1       0       0       0
GSM2545344           1      0      0       0       0       0       0       1
GSM2545348           1      1      0       0       0       0       0       0
GSM2545352           1      0      0       0       0       0       1       0
GSM2545353           1      0      0       0       0       0       0       0
GSM2545358           1      0      0       0       0       0       0       0
GSM2545362           1      0      0       0       0       1       0       0
GSM2545364           1      1      0       0       0       0       0       0
GSM2545365           1      0      1       0       0       0       0       0
GSM2545366           1      0      0       1       0       0       0       0
GSM2545371           1      0      0       0       1       0       0       0
GSM2545375           1      0      0       0       0       1       0       0
GSM2545376           1      0      0       0       0       0       1       0
GSM2545377           1      0      0       0       0       0       0       1
           Spc.Day0 Spc.Day4
GSM2545337        0        0
GSM2545338        0        0
GSM2545339        0        0
GSM2545344        0        0
GSM2545348        0        0
GSM2545352        0        0
GSM2545353        0        0
GSM2545358        1        0
GSM2545362        0        0
GSM2545364        1        0
GSM2545365        1        0
GSM2545366        1        0
GSM2545371        0        1
GSM2545375        0        1
GSM2545376        0        1
GSM2545377        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_fem_day04</span> <span class="op">%&gt;%</span></span>
<span>                          <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">tissue</span>, <span class="va">mouse</span><span class="op">)</span>,</span>
<span>                      designFormula <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                      designMatrix <span class="op">=</span> <span class="va">design</span>, flipCoordFitted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 mouse15 mouse20 mouse21 mouse22
GSM2545337           1      0      1       0       0       0       0       0
GSM2545338           1      0      0       1       0       0       0       0
GSM2545339           1      0      0       0       1       0       0       0
GSM2545344           1      0      0       0       0       0       0       1
GSM2545348           1      1      0       0       0       0       0       0
GSM2545352           1      0      0       0       0       0       1       0
GSM2545353           1      0      0       0       0       0       0       0
GSM2545358           1      0      0       0       0       0       0       0
GSM2545362           1      0      0       0       0       1       0       0
GSM2545364           1      1      0       0       0       0       0       0
GSM2545365           1      0      1       0       0       0       0       0
GSM2545366           1      0      0       1       0       0       0       0
GSM2545371           1      0      0       0       1       0       0       0
GSM2545375           1      0      0       0       0       1       0       0
GSM2545376           1      0      0       0       0       0       1       0
GSM2545377           1      0      0       0       0       0       0       1
           Spc.Day0 Spc.Day4
GSM2545337        0        0
GSM2545338        0        0
GSM2545339        0        0
GSM2545344        0        0
GSM2545348        0        0
GSM2545352        0        0
GSM2545353        0        0
GSM2545358        1        0
GSM2545362        0        0
GSM2545364        1        0
GSM2545365        1        0
GSM2545366        1        0
GSM2545371        0        1
GSM2545375        0        1
GSM2545376        0        1
GSM2545377        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`time = Day0`</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-within-and-between-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
$`time = Day4`</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-within-and-between-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="how-does-this-relate-to-the-deseq2-analysis-we-did-in-the-previous-episode">How does this relate to the DESeq2 analysis we did in the previous
episode?<a class="anchor" aria-label="anchor" href="#how-does-this-relate-to-the-deseq2-analysis-we-did-in-the-previous-episode"></a>
</h2>
<hr class="half-width">
<p>Now that we have learnt more about interpreting design matrices,
let’s look back to the differential expression analysis we performed in
the previous episode. We will repeat the main lines of code here.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating size factors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<p><code>DESeq2</code> stores the design matrix in the object:</p>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               Intercept sex_Male_vs_Female time_Day4_vs_Day0 time_Day8_vs_Day0
Female_Day0_9          1                  0                 0                 0
Female_Day0_10         1                  0                 0                 0
Female_Day0_8          1                  0                 0                 0
Female_Day0_4          1                  0                 0                 0
Male_Day0_11           1                  1                 0                 0
Male_Day0_7            1                  1                 0                 0
Male_Day0_2            1                  1                 0                 0
Female_Day4_15         1                  0                 1                 0
Female_Day4_22         1                  0                 1                 0
Female_Day4_21         1                  0                 1                 0
Female_Day4_20         1                  0                 1                 0
Male_Day4_18           1                  1                 1                 0
Male_Day4_13           1                  1                 1                 0
Male_Day4_1            1                  1                 1                 0
Male_Day4_12           1                  1                 1                 0
Female_Day8_14         1                  0                 0                 1
Female_Day8_5          1                  0                 0                 1
Female_Day8_16         1                  0                 0                 1
Female_Day8_19         1                  0                 0                 1
Male_Day8_6            1                  1                 0                 1
Male_Day8_23           1                  1                 0                 1
Male_Day8_24           1                  1                 0                 1
attr(,"assign")
[1] 0 1 2 2
attr(,"contrasts")
attr(,"contrasts")$sex
[1] "contr.treatment"

attr(,"contrasts")$time
[1] "contr.treatment"</code></pre>
</div>
<p>The column names can be obtained via the <code>resultsNames</code>
function:</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">resultsNames</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Intercept"          "sex_Male_vs_Female" "time_Day4_vs_Day0"
[4] "time_Day8_vs_Day0" </code></pre>
</div>
<p>Let’s visualize this design:</p>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"time"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                      designMatrix <span class="op">=</span> <span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span>, </span>
<span>                      flipCoordFitted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-vis-sex-time-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the previous episode, we performed a test comparing Day8 samples
to Day0 samples:</p>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>From the figure above, we see that this comparison is represented by
the <code>time_Day8_vs_Day0</code> coefficient, which corresponds to the
fourth column in the design matrix. Thus, an alternative way of
specifying the contrast for the test would be:</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeNum</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s check if the results are comparable:</p>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 16%
LFC &lt; 0 (down)     : 4282, 16%
outliers [1]       : 10, 0.036%
low counts [2]     : 3723, 14%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeNum</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 16%
LFC &lt; 0 (down)     : 4282, 16%
outliers [1]       : 10, 0.036%
low counts [2]     : 3723, 14%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## logFC</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">log2FoldChange</span>, <span class="va">resTimeNum</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-compare-tests-sex-time-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## -log10(p-value)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTimeNum</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-compare-tests-sex-time-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="redo-deseq2-analysis-with-interaction">Redo DESeq2 analysis with interaction<a class="anchor" aria-label="anchor" href="#redo-deseq2-analysis-with-interaction"></a>
</h2>
<hr class="half-width">
<p>Next, let’s look at a different setup. We still consider the sex and
time predictors, but now we allow an interaction between them. In other
words, we allow the time effect to be different for males and
females.</p>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">*</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating size factors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               Intercept sex_Male_vs_Female time_Day4_vs_Day0 time_Day8_vs_Day0
Female_Day0_9          1                  0                 0                 0
Female_Day0_10         1                  0                 0                 0
Female_Day0_8          1                  0                 0                 0
Female_Day0_4          1                  0                 0                 0
Male_Day0_11           1                  1                 0                 0
Male_Day0_7            1                  1                 0                 0
Male_Day0_2            1                  1                 0                 0
Female_Day4_15         1                  0                 1                 0
Female_Day4_22         1                  0                 1                 0
Female_Day4_21         1                  0                 1                 0
Female_Day4_20         1                  0                 1                 0
Male_Day4_18           1                  1                 1                 0
Male_Day4_13           1                  1                 1                 0
Male_Day4_1            1                  1                 1                 0
Male_Day4_12           1                  1                 1                 0
Female_Day8_14         1                  0                 0                 1
Female_Day8_5          1                  0                 0                 1
Female_Day8_16         1                  0                 0                 1
Female_Day8_19         1                  0                 0                 1
Male_Day8_6            1                  1                 0                 1
Male_Day8_23           1                  1                 0                 1
Male_Day8_24           1                  1                 0                 1
               sexMale.timeDay4 sexMale.timeDay8
Female_Day0_9                 0                0
Female_Day0_10                0                0
Female_Day0_8                 0                0
Female_Day0_4                 0                0
Male_Day0_11                  0                0
Male_Day0_7                   0                0
Male_Day0_2                   0                0
Female_Day4_15                0                0
Female_Day4_22                0                0
Female_Day4_21                0                0
Female_Day4_20                0                0
Male_Day4_18                  1                0
Male_Day4_13                  1                0
Male_Day4_1                   1                0
Male_Day4_12                  1                0
Female_Day8_14                0                0
Female_Day8_5                 0                0
Female_Day8_16                0                0
Female_Day8_19                0                0
Male_Day8_6                   0                1
Male_Day8_23                  0                1
Male_Day8_24                  0                1
attr(,"assign")
[1] 0 1 2 2 3 3
attr(,"contrasts")
attr(,"contrasts")$sex
[1] "contr.treatment"

attr(,"contrasts")$time
[1] "contr.treatment"</code></pre>
</div>
<p>Let’s visualize this design:</p>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"time"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                      designMatrix <span class="op">=</span> <span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span>, </span>
<span>                      flipCoordFitted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-vis-sex-time-int-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that now, the <code>time_Day8_vs_Day0</code> coefficient
represents the difference between Day8 and Day0 <strong>for the Female
samples</strong>. To get the corresponding difference for the male
samples, we need to also add the interaction effect
(<code>sexMale.timeDay8</code>).</p>
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Day8 vs Day0, female</span></span>
<span><span class="va">resTimeFemale</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Interaction effect (difference in Day8-Day0 effect between Male and Female)</span></span>
<span><span class="va">resTimeInt</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, name <span class="op">=</span> <span class="st">"sexMale.timeDay8"</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s try to fit this model with the second approach mentioned above,
namely to create a single factor.</p>
<div class="codewrapper sourceCode" id="cb97">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">sex_time</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">sex</span>, <span class="st">"_"</span>, <span class="va">se</span><span class="op">$</span><span class="va">time</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sex_time</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating size factors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb104">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               sex_timeFemale_Day0 sex_timeFemale_Day4 sex_timeFemale_Day8
Female_Day0_9                    1                   0                   0
Female_Day0_10                   1                   0                   0
Female_Day0_8                    1                   0                   0
Female_Day0_4                    1                   0                   0
Male_Day0_11                     0                   0                   0
Male_Day0_7                      0                   0                   0
Male_Day0_2                      0                   0                   0
Female_Day4_15                   0                   1                   0
Female_Day4_22                   0                   1                   0
Female_Day4_21                   0                   1                   0
Female_Day4_20                   0                   1                   0
Male_Day4_18                     0                   0                   0
Male_Day4_13                     0                   0                   0
Male_Day4_1                      0                   0                   0
Male_Day4_12                     0                   0                   0
Female_Day8_14                   0                   0                   1
Female_Day8_5                    0                   0                   1
Female_Day8_16                   0                   0                   1
Female_Day8_19                   0                   0                   1
Male_Day8_6                      0                   0                   0
Male_Day8_23                     0                   0                   0
Male_Day8_24                     0                   0                   0
               sex_timeMale_Day0 sex_timeMale_Day4 sex_timeMale_Day8
Female_Day0_9                  0                 0                 0
Female_Day0_10                 0                 0                 0
Female_Day0_8                  0                 0                 0
Female_Day0_4                  0                 0                 0
Male_Day0_11                   1                 0                 0
Male_Day0_7                    1                 0                 0
Male_Day0_2                    1                 0                 0
Female_Day4_15                 0                 0                 0
Female_Day4_22                 0                 0                 0
Female_Day4_21                 0                 0                 0
Female_Day4_20                 0                 0                 0
Male_Day4_18                   0                 1                 0
Male_Day4_13                   0                 1                 0
Male_Day4_1                    0                 1                 0
Male_Day4_12                   0                 1                 0
Female_Day8_14                 0                 0                 0
Female_Day8_5                  0                 0                 0
Female_Day8_16                 0                 0                 0
Female_Day8_19                 0                 0                 0
Male_Day8_6                    0                 0                 1
Male_Day8_23                   0                 0                 1
Male_Day8_24                   0                 0                 1
attr(,"assign")
[1] 1 1 1 1 1 1
attr(,"contrasts")
attr(,"contrasts")$sex_time
[1] "contr.treatment"</code></pre>
</div>
<p>We again visualize this design:</p>
<div class="codewrapper sourceCode" id="cb106">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"time"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                      designMatrix <span class="op">=</span> <span class="fu">attr</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"modelMatrix"</span><span class="op">)</span>, </span>
<span>                      flipCoordFitted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-vis-sex-time-int-single-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then set up the same contrasts as above</p>
<div class="codewrapper sourceCode" id="cb108">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Day8 vs Day0, female</span></span>
<span><span class="va">resTimeFemaleSingle</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex_time"</span>, <span class="st">"Female_Day8"</span>, <span class="st">"Female_Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Interaction effect (difference in Day8-Day0 effect between Male and Female)</span></span>
<span><span class="fu">resultsNames</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "sex_timeFemale_Day0" "sex_timeFemale_Day4" "sex_timeFemale_Day8"
[4] "sex_timeMale_Day0"   "sex_timeMale_Day4"   "sex_timeMale_Day8"  </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb110">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeIntSingle</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Check that these results agree with the ones obtained by fitting the
model with the two factors and the interaction term.</p>
<div class="codewrapper sourceCode" id="cb111">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeFemale</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 2969, 11%
LFC &lt; 0 (down)     : 3218, 12%
outliers [1]       : 6, 0.022%
low counts [2]     : 6382, 23%
(mean count &lt; 3)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeFemaleSingle</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 2969, 11%
LFC &lt; 0 (down)     : 3218, 12%
outliers [1]       : 6, 0.022%
low counts [2]     : 6382, 23%
(mean count &lt; 3)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb115">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTimeFemale</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTimeFemaleSingle</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-compare-sex-time-int-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb116">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeInt</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 0, 0%
LFC &lt; 0 (down)     : 0, 0%
outliers [1]       : 6, 0.022%
low counts [2]     : 0, 0%
(mean count &lt; 0)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb118">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">resTimeIntSingle</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 27430 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 0, 0%
LFC &lt; 0 (down)     : 0, 0%
outliers [1]       : 6, 0.022%
low counts [2]     : 0, 0%
(mean count &lt; 0)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb120">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTimeInt</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">resTimeIntSingle</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/06-extra-design-rendered-compare-sex-time-int-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The formula framework in R allows creation of design matrices, which
details the variables expected to be associated with systematic
differences in gene expression levels.</li>
<li>Comparisons of interest can be defined using contrasts, which are
linear combinations of the model coefficients.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-07-gene-set-analysis"><p>Content from <a href="07-gene-set-analysis.html">Gene set enrichment analysis</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/07-gene-set-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 105 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<style>
body h3 {
    margin-top: 50px;
}
</style>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the aim of performing gene set enrichment analysis?</li>
<li>What is the method of over-representation analysis?</li>
<li>What are the commonly-used gene set databases?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the method of gene set enrichment analysis.</li>
<li>Learn how to obtain gene sets from various resources in R.</li>
<li>Learn how to perform gene set enrichment analysis and how to
visualize enrichment results.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>After we have obtained a list of differentially expressed (DE) genes,
the next question naturally to ask is what biological functions these DE
genes may affect. Gene set enrichment analysis (GSEA) evaluates the
associations of a list of DE genes to a collection of pre-defined gene
sets, where each gene set has a specific biological meaning. Once DE
genes are significantly enriched in a gene set, the conclusion is made
that the corresponding biological meaning (e.g. a biological process or
a pathway) is significantly affected.</p>
<p>The definition of a gene set is very flexible and the construction of
gene sets is straightforward. In most cases, gene sets are from public
databases where huge efforts from scientific curators have already been
made to carefully categorize genes into gene sets with clear biological
meanings. Nevertheless, gene sets can also be self-defined from
individual studies, such as a set of genes in a network module from a
co-expression network analysis, or a set of genes that are up-regulated
in a certain disease.</p>
<p>There are a huge amount of methods available for GSEA analysis. In
this episode, we will learn the simplest but the mostly used one: the
over-representation analysis (ORA). ORA is directly applied to the list
of DE genes and it evaluates the association of the DE genes and the
gene set by the numbers of genes in different categories.</p>
<p>Please note, ORA is a universal method that it can not only be
applied to the DE gene list, but also any type of gene list of interest
to look for their statistically associated biological meanings.</p>
<p>In this episode, we will start with a tiny example to illustrate the
statistical method of ORA. Next we will go through several commonly-used
gene set databases and how to access them in R. Then, we will learn how
to perform ORA analysis with the Bioconductor package
<strong>clusterProfiler</strong>. And in the end, we will learn several
visualization methods on the GSEA results.</p>
<p>Following is a list of packages that will be used in this
episode:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">microbenchmark</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">simplifyEnrichment</span><span class="op">)</span></span></code></pre>
</div>
<section><h2 class="section-heading" id="the-statistical-method">The statistical method<a class="anchor" aria-label="anchor" href="#the-statistical-method"></a>
</h2>
<hr class="half-width">
<p>To demonstrate the ORA analysis, we use a list of DE genes from a
comparison between genders. The following code performs
<strong>DESeq2</strong> analysis which you should have already learnt in
the previous episode. In the end, we have a list of DE genes filtered by
FDR &lt; 0.05, and save it in the object <code>sexDEgenes</code>.</p>
<p>The file <code>data/GSE96870_se.rds</code> contains a
<code>RangedSummarizedExperiment</code> that contains RNA-Seq counts
that were downloaded in <a href="02-setup.html">Episode 2</a> and
constructed in <a href="03-import-annotate.html">Episode 3</a> (minimal
codes for downloading and constructing in the script <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/episodes/download_data.R" class="external-link"><code>download_data.R</code></a>.
In the following code, there are also comments that explain every step
of the analysis.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read the example dataset which is a `RangedSummarizedExperiment` object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># only restrict to mRNA (protein-coding genes)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">gbkey</span> <span class="op">==</span> <span class="st">"mRNA"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># construct a `DESeqDataSet` object where we also specify the experimental design</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="co"># perform DESeq2 analysis</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co"># obtain DESeq2 results, here we only want Male vs Female in the "sex" variable</span></span>
<span><span class="va">resSex</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># extract DE genes with padj &lt; 0.05</span></span>
<span><span class="va">sexDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resSex</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># the list of DE genes</span></span>
<span><span class="va">sexDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">sexDE</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s check the number of DE genes and how they look like. It seems
the number of DE genes is very small, but it is OK for this example.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 54</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Lgr6"   "Myoc"   "Fibcd1" "Kcna4"  "Ctxn2"  "S100a9"</code></pre>
</div>
<p>Next we construct a gene set which contains genes on sex chromosomes
(let’s call it the “<em>XY gene set</em>”). Recall the
<code>RangedSummarizedExperiment</code> object also includes genomic
locations of genes, thus we can simply obtain sex genes by filtering the
chromosome names.</p>
<p>In the following code, <code>geneGR</code> is a <code>GRanges</code>
object on which <code>seqnames()</code> is applied to extract chromosome
names. <code>seqnames()</code> returns a special data format and we need
to explicitly convert it to a normal vector by
<code>as.vector()</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneGR</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">totalGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">XYGeneSet</span> <span class="op">&lt;-</span> <span class="va">totalGenes</span><span class="op">[</span><span class="fu">as.vector</span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">geneGR</span><span class="op">)</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Gm21950"   "Gm14346"   "Gm14345"   "Gm14351"   "Spin2-ps1" "Gm3701"   </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1134</code></pre>
</div>
<p>The format of a single gene set is very straightforward, which is
simply a vector. The ORA analysis is applied on the DE gene vector and
gene set vector.</p>
<p>Before we move on, one thing worth to mention is that ORA deals with
two gene vectors. To correctly map between them, gene ID types must be
consistent in the two vectors. In this tiny example, since both DE genes
and the <em>XY gene set</em> are from the same object <code>se</code>,
they are ensured to be in the same gene ID types (the gene symbol). But
in general, DE genes and gene sets are from two different sources
(e.g. DE genes are from researcher’s experiment and gene sets are from
public databases), it is very possible that gene IDs are not consistent
in the two. Later in this episode, we will learn how to perform gene ID
conversion in the ORA analysis.</p>
<p>Since the DE genes and the gene set can be mathematically thought of
as two sets, a natural way is to first visualize them with a Venn
diagram.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">venn</span><span class="op">(</span><span class="fu">list</span><span class="op">(</span><span class="st">"sexDEgenes"</span>  <span class="op">=</span> <span class="va">sexDEgenes</span>, </span>
<span>               <span class="st">"XY gene set"</span> <span class="op">=</span> <span class="va">XYGeneSet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">title</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"|universe| = "</span>, <span class="fu">length</span><span class="op">(</span><span class="va">totalGenes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-venn-diagram-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the Venn diagram, we can observe that around 1.1% (13/1134) of
genes in the <em>XY gene set</em> are DE. Compared to the global
fraction of DE genes (54/21198 = 0.25%), it seems there is a strong
relations between DE genes and the gene set. We can also compare the
fraction of DE genes that belong to the gene set (13/54 = 24.1%) and the
global fraction of <em>XY gene set</em> in the genome (1134/21198 =
5.3%). On the other hand, it is quite expected because the two events
are actually biologically relevant where one is from a comparison
between genders and the other is the set of gender-related genes.</p>
<p>Then, how to statistically measure the enrichment or
over-representation? Let’s go to the next section.</p>
<div class="section level3">
<h3 id="fishers-exact-test">Fisher’s exact test<a class="anchor" aria-label="anchor" href="#fishers-exact-test"></a>
</h3>
<p>To statistically measure the enrichment, the relationship of DE genes
and the gene set is normally formatted into the following 2x2
contingency table, where in the table are the numbers of genes in
different categories. <span class="math inline">\(n_{+1}\)</span> is the
size of the <em>XY gene set</em> (i.e. the number of member genes),
<span class="math inline">\(n_{1+}\)</span> is the number of DE genes,
<span class="math inline">\(n\)</span> is the number of total genes.</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td><span class="math inline">\(n_{11}\)</span></td>
<td><span class="math inline">\(n_{12}\)</span></td>
<td><span class="math inline">\(n_{1+}\)</span></td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td><span class="math inline">\(n_{21}\)</span></td>
<td><span class="math inline">\(n_{22}\)</span></td>
<td><span class="math inline">\(n_{2+}\)</span></td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td><span class="math inline">\(n_{+1}\)</span></td>
<td><span class="math inline">\(n_{+2}\)</span></td>
<td><span class="math inline">\(n\)</span></td>
</tr>
</tbody>
</table>
</center>
<p>These numbers can be obtained as in the following code<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Note we replace
<code>+</code> with <code>0</code> in the R variable names.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span>    <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span>
<span><span class="va">n_10</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">sexDEgenes</span>, <span class="va">XYGeneSet</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Other values can be obtained by:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_12</span> <span class="op">&lt;-</span> <span class="va">n_10</span> <span class="op">-</span> <span class="va">n_11</span></span>
<span><span class="va">n_21</span> <span class="op">&lt;-</span> <span class="va">n_01</span> <span class="op">-</span> <span class="va">n_11</span></span>
<span><span class="va">n_20</span> <span class="op">&lt;-</span> <span class="va">n</span>    <span class="op">-</span> <span class="va">n_10</span></span>
<span><span class="va">n_02</span> <span class="op">&lt;-</span> <span class="va">n</span>    <span class="op">-</span> <span class="va">n_01</span></span>
<span><span class="va">n_22</span> <span class="op">&lt;-</span> <span class="va">n_02</span> <span class="op">-</span> <span class="va">n_12</span></span></code></pre>
</div>
<p>All the values are:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_10</span>, <span class="va">n_21</span>, <span class="va">n_22</span>, <span class="va">n_20</span>, <span class="va">n_01</span>, <span class="va">n_02</span>, <span class="va">n</span><span class="op">)</span>,</span>
<span>    nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     [,1]  [,2]  [,3]
[1,]   13    41    54
[2,] 1121 20023 21144
[3,] 1134 20064 21198</code></pre>
</div>
<p>And we fill these numbers into the 2x2 contingency table:</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td>13</td>
<td>41</td>
<td>54</td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td>1121</td>
<td>20023</td>
<td>21144</td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td>1134</td>
<td>20064</td>
<td>21198</td>
</tr>
</tbody>
</table>
</center>
<p>Fisher’s exact test can be used to test the associations of the two
marginal attributes, i.e. is there a dependency of a gene to be a DE
gene and to be in the <em>XY gene set</em>? In R, we can use the
function <code>fisher.test()</code> to perform the test. The input is
the top-left 2x2 sub-matrix. We specify
<code>alternative = "greater"</code> in the function because we are only
interested in over-representation.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  matrix(c(n_11, n_12, n_21, n_22), nrow = 2, byrow = TRUE)
p-value = 3.906e-06
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 3.110607      Inf
sample estimates:
odds ratio
  5.662486 </code></pre>
</div>
<p>In the output, we can see the <em>p</em>-value is very small
(<code>3.906e-06</code>), then we can conclude DE genes have a very
strong enrichment in the <em>XY gene set</em>.</p>
<p>Results of the Fisher’s Exact test can be saved into an object
<code>t</code>, which is a simple list, and the <em>p</em>-value can be
obtained by <code>t$p.value</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span>
<span><span class="va">t</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>Odds ratio from the Fisher’s exact test is defined as follows:</p>
<p><span class="math display">\[
\mathrm{Odds\_ratio} = \frac{n_{11}/n_{21}}{n_{12}/n_{22}} =
\frac{n_{11}/n_{12}}{n_{21}/n_{22}} = \frac{n_{11} * n_{22}}{n_{12} *
n_{21}}
\]</span></p>
<p>If there is no association between DE genes and the gene set, odds
ratio is expected to be 1. And it is larger than 1 if there is an
over-representation of DE genes on the gene set.</p>
<div id="further-reading" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>The 2x2 contingency table can be transposed and it does not affect
the Fisher’s exact test. E.g. let’s put whether genes are in the gene
sets on rows, and put whether genes are DE on columns.</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>DE</th>
<th>Not DE</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>In the gene set</strong></td>
<td>13</td>
<td>1121</td>
<td>1134</td>
</tr>
<tr class="even">
<td><strong>Not in the gene set</strong></td>
<td>41</td>
<td>20023</td>
<td>20064</td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td>54</td>
<td>21144</td>
<td>21198</td>
</tr>
</tbody>
</table>
</center>
<p>And the corresponding <code>fisher.test()</code> is:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">13</span>, <span class="fl">1121</span>, <span class="fl">41</span>, <span class="fl">20023</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  matrix(c(13, 1121, 41, 20023), nrow = 2, byrow = TRUE)
p-value = 3.906e-06
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 3.110607      Inf
sample estimates:
odds ratio
  5.662486 </code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="the-hypergeometric-distribution">The hypergeometric distribution<a class="anchor" aria-label="anchor" href="#the-hypergeometric-distribution"></a>
</h3>
<p>We can look at the problem from another aspect. This time we treat
all genes as balls in a big box where all the genes have the same
probability to be picked up. Some genes are marked as DE genes (in red
in the figure) and other genes are marked as non-DE genes (in blue). We
grab <span class="math inline">\(n_{+1}\)</span> genes (the size of the
gene set) from the box and we want to ask <strong>what is the
probability of having <span class="math inline">\(n_{11}\)</span> DE
genes in our hand?</strong></p>
<figure><img src="../fig/07-gene-set-analysis-rendered-hypergeom-1.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We first calculate the total number of ways of picking <span class="math inline">\(n_{+1}\)</span> genes from total <span class="math inline">\(n\)</span> genes, without distinguishing whether
they are DE or not: <span class="math inline">\(\binom{n}{n_{+1}}\)</span>.</p>
<p>Next, in the <span class="math inline">\(n_{+1}\)</span> genes that
have been picked, there are <span class="math inline">\(n_{11}\)</span>
DE genes which can only be from the total <span class="math inline">\(n_{1+}\)</span> DE genes. Then the number of ways
of picking <span class="math inline">\(n_{11}\)</span> DE genes from
<span class="math inline">\(n_{1+}\)</span> total DE genes is <span class="math inline">\(\binom{n_{1+}}{n_{11}}\)</span>.</p>
<p>Similarly, there are still <span class="math inline">\(n_{21}\)</span> non-DE genes in our hand, which
can only be from the total <span class="math inline">\(n_{2+}\)</span>
non-DE genes. Then the number of ways of picking <span class="math inline">\(n_{21}\)</span> non-DE genes from <span class="math inline">\(n_{2+}\)</span> total non-DE genes: <span class="math inline">\(\binom{n_{2+}}{n_{21}}\)</span>.</p>
<p>Since picking DE genes and picking non-DE genes are independent, the
number of ways of picking <span class="math inline">\(n_{+1}\)</span>
genes which contain <span class="math inline">\(n_{11}\)</span> DE genes
and <span class="math inline">\(n_{21}\)</span> non-DE genes is their
multiplication: <span class="math inline">\(\binom{n_{1+}}{n_{11}}
\binom{n_{2+}}{n_{21}}\)</span>.</p>
<p>And the probability <span class="math inline">\(P\)</span> is:</p>
<p><span class="math display">\[P = \frac{\binom{n_{1+}}{n_{11}}
\binom{n_{2+}}{n_{21}}}{\binom{n}{n_{+1}}}
=  \frac{\binom{n_{1+}}{n_{11}} \binom{n - n_{1+}}{n_{+1}
-n_{11}}}{\binom{n}{n_{+1}}} \]</span></p>
<p>where in the denominator is the number of ways of picking <span class="math inline">\(n_{+1}\)</span> genes without distinguishing
whether they are DE or not.</p>
<p>If <span class="math inline">\(n\)</span> (number of total genes),
<span class="math inline">\(n_{1+}\)</span> (number of DE genes) and
<span class="math inline">\(n_{+1}\)</span> (size of gene set) are all
fixed values, the number of DE genes that are picked can be denoted as a
random variable <span class="math inline">\(X\)</span>. Then <span class="math inline">\(X\)</span> follows the hypergeometric distribution
with parameters <span class="math inline">\(n\)</span>, <span class="math inline">\(n_{1+}\)</span> and <span class="math inline">\(n_{+1}\)</span>, written as:</p>
<p><span class="math display">\[ X \sim \mathrm{Hyper}(n, n_{1+},
n_{+1})\]</span></p>
<p>The <em>p</em>-value of the enrichment is calculated as the
probability of having an observation equal to or larger than <span class="math inline">\(n_{11}\)</span> under the assumption of
independence:</p>
<p><span class="math display">\[
\mathrm{Pr}( X \geqslant  n_{11} ) = \sum_{x \in \{ {n_{11}, n_{11}+1,
..., \min\{n_{1+}, n_{+1}\} \}}} \mathrm{Pr}(X = x)
\]</span></p>
<p>In R, the function <code>phyper()</code> calculates <em>p</em>-values
from the hypergeometric distribution. There are four arguments:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phyper</span><span class="op">(</span><span class="va">q</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span><span class="op">)</span></span></code></pre>
</div>
<p>which are:</p>
<ul>
<li>
<code>q</code>: the observation,</li>
<li>
<code>m</code>: number of DE genes,</li>
<li>
<code>n</code>: number of non-DE genes,</li>
<li>
<code>k</code>: size of the gene set.</li>
</ul>
<p><code>phyper()</code> calculates <span class="math inline">\(\mathrm{Pr}(X \leqslant q)\)</span>. To calculate
<span class="math inline">\(\mathrm{Pr}(X \geqslant q)\)</span>, we need
to transform it a little bit:</p>
<p><span class="math display">\[ \mathrm{Pr}(X \geqslant q) = 1 -
\mathrm{Pr}(X &lt; q) = 1 - \mathrm{Pr}(X \leqslant q-1)\]</span></p>
<p>Then, the correct use of <code>phyper()</code> is:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">q</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s plugin our variables:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>Optionally, <code>lower.tail</code> argument can be specified which
directly calculates <em>p</em>-values from the upper tail of the
distribution.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>If we switch <code>n_01</code> and <code>n_10</code>, the
<em>p</em>-values are identical:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_01</span>, <span class="va">n_02</span>, <span class="va">n_10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p><code>fisher.test()</code> and <code>phyper()</code> give the same
<em>p</em>-value. Actually the two methods are identical because in
Fisher’s exact test, hypergeometric distribution is the exact
distribution of its statistic.</p>
<p>Let’s test the runtime of the two functions:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbenchmark</span><span class="op">)</span></span>
<span><span class="fu">microbenchmark</span><span class="op">(</span></span>
<span>    fisher <span class="op">=</span> <span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>        alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span>,</span>
<span>    hyper <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Unit: microseconds
   expr     min       lq      mean   median      uq     max neval
 fisher 249.916 255.6260 266.44069 259.4835 272.503 482.399   100
  hyper   1.563   1.7485   2.63498   2.0490   3.091  19.476   100</code></pre>
</div>
<p>It is very astonishing that <code>phyper()</code> is hundreds of
times faster than <code>fisher.test()</code>. Main reason is in
<code>fisher.test()</code>, there are many additional calculations
besides calculating the <em>p</em>-value. So if you want to implement
ORA analysis by yourself, always consider to use <code>phyper()</code><a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<div id="further-reading-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading-1" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>Current tools also use Binomial distribution or chi-square test for
ORA analysis. These two are just approximations. Please refer to <a href="https://doi.org/10.1093/bioinformatics/btl633" class="external-link">Rivals et al.,
Enrichment or depletion of a GO category within a class of genes: which
test? Bioinformatics 2007</a> which gives an overview of statistical
methods used in ORA analysis.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="gene-set-resources">Gene set resources<a class="anchor" aria-label="anchor" href="#gene-set-resources"></a>
</h2>
<hr class="half-width">
<p>We have learnt the basic methods of ORA analysis. Now we go to the
second component of the analysis: the gene sets.</p>
<p>Gene sets represent prior knowledge of what is the general shared
biological attribute of genes in the gene set. For example, in a “cell
cycle” gene set, all the genes are involved in the cell cycle process.
Thus, if DE genes are significantly enriched in the “cell cycle” gene
set, which means there are significantly more cell cycle genes
differentially expressed than expected, we can conclude that the normal
function of cell cycle process may be affected.</p>
<p>As we have mentioned, genes in the gene set share the same
“biological attribute” where “the attribute” will be used for making
conclusions. The definition of “biological attribute” is very flexible.
It can be a biological process such as “cell cycle”. It can also be from
a wide range of other definitions, to name a few:</p>
<ul>
<li>Locations in the cell, e.g. cell membrane or cell nucleus.</li>
<li>Positions on chromosomes, e.g. sex chromosomes or the cytogenetic
band p13 on chromome 10.</li>
<li>Target genes of a transcription factor or a microRNA, e.g. all genes
that are transcriptionally regulationed by NF-κB.</li>
<li>Signature genes in a certain tumor type, i.e. genes that are
uniquely highly expressed in a tumor type.</li>
</ul>
<p>The <a href="https://www.gsea-msigdb.org/gsea/msigdb" class="external-link">MSigDB
database</a> contains gene sets in many topics. We will introduce it
later in this section.</p>
<p>You may have encountered many different ways to name gene sets: “gene
sets”, “biological terms”, “GO terms”, “GO gene sets”, “pathways”, and
so on. They basically refer to the same thing, but from different
aspects. As shown in the following figure, “gene set” corresponds to a
vector of genes and it is the representation of the data for
computation. “Biological term” is a textual entity that contains
description of its biological meaning; It corresponds to the knowledge
of the gene set and is for the inference of the analysis. “GO gene sets”
and “pathways” specifically refer to the enrichment analysis using GO
gene sets and pahtway gene sets.</p>
<figure><img src="../fig/geneset.svg" alt="Gene set, biological term and GO gene set" class="figure mx-auto d-block"></figure><p>Before we touch the gene set databases, we first summarize the
general formats of gene sets in R. In most analysis, a gene set is
simply treated as a vector of genes. Thus, naturally, a collection of
gene sets can be represented as a list of vectors. In the following
example, there are three gene sets with 3, 5 and 2 genes. Some genes
exist in multiple gene sets.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>gene_set_1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span><span class="op">)</span>,</span>
<span>           gene_set_2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>, <span class="st">"gene_5"</span>, <span class="st">"gene_6"</span><span class="op">)</span>,</span>
<span>           gene_set_3 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_4"</span>, <span class="st">"gene_7"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lt</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$gene_set_1
[1] "gene_1" "gene_2" "gene_3"

$gene_set_2
[1] "gene_1" "gene_3" "gene_4" "gene_5" "gene_6"

$gene_set_3
[1] "gene_4" "gene_7"</code></pre>
</div>
<p>It is also very common to store the relations of gene sets and genes
as a two-column data frame. The order of the gene set column and the
gene column, i.e. which column locates as the first column, are quite
arbitrary. Different tools may require differently.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data.frame</span><span class="op">(</span>gene_set <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">lt</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           gene <span class="op">=</span> <span class="fu">unname</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene_set   gene
1  gene_set_1 gene_1
2  gene_set_1 gene_2
3  gene_set_1 gene_3
4  gene_set_2 gene_1
5  gene_set_2 gene_3
6  gene_set_2 gene_4
7  gene_set_2 gene_5
8  gene_set_2 gene_6
9  gene_set_3 gene_4
10 gene_set_3 gene_7</code></pre>
</div>
<p>Or genes be in the first column:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene   gene_set
1  gene_1 gene_set_1
2  gene_2 gene_set_1
3  gene_3 gene_set_1
4  gene_1 gene_set_2
5  gene_3 gene_set_2
6  gene_4 gene_set_2
7  gene_5 gene_set_2
8  gene_6 gene_set_2
9  gene_4 gene_set_3
10 gene_7 gene_set_3</code></pre>
</div>
<p>Not very often, gene sets are represented as a matrix where one
dimension corresponds to gene sets and the other dimension corresponds
to genes. The values in the matix are binary where a value of 1
represents the gene is a member of the corresponding gene sets. In some
methods, 1 is replaced by <span class="math inline">\(w_{ij}\)</span> to
weight the effect of the genes in the gene set.</p>
<pre><code><span><span class="co">#            gene_1 gene_2 gene_3 gene_4</span></span>
<span><span class="co"># gene_set_1      1      1      0      0</span></span>
<span><span class="co"># gene_set_2      1      0      1      1</span></span></code></pre>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Can you convert between different gene set representations? E.g.
convert a list to a two-column data frame?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>gene_set_1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span><span class="op">)</span>,</span>
<span>           gene_set_2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>, <span class="st">"gene_5"</span>, <span class="st">"gene_6"</span><span class="op">)</span>,</span>
<span>           gene_set_3 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_4"</span>, <span class="st">"gene_7"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>To convert <code>lt</code> to a data frame (e.g. let’s put gene sets
in the first column):</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span>gene_set <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">lt</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                gene <span class="op">=</span> <span class="fu">unname</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene_set   gene
1  gene_set_1 gene_1
2  gene_set_1 gene_2
3  gene_set_1 gene_3
4  gene_set_2 gene_1
5  gene_set_2 gene_3
6  gene_set_2 gene_4
7  gene_set_2 gene_5
8  gene_set_2 gene_6
9  gene_set_3 gene_4
10 gene_set_3 gene_7</code></pre>
</div>
<p>To convert <code>df</code> back to the list:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">split</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">gene</span>, <span class="va">df</span><span class="op">$</span><span class="va">gene_set</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$gene_set_1
[1] "gene_1" "gene_2" "gene_3"

$gene_set_2
[1] "gene_1" "gene_3" "gene_4" "gene_5" "gene_6"

$gene_set_3
[1] "gene_4" "gene_7"</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Next, let’s go through gene sets from several major databases: the
GO, KEGG and MSigDB databases.</p>
<div class="section level3">
<h3 id="gene-ontology-gene-sets">Gene Ontology gene sets<a class="anchor" aria-label="anchor" href="#gene-ontology-gene-sets"></a>
</h3>
<p>Gene Ontology (GO) is the standard source for gene set enrichment
analysis. GO contains three namespaces of biological process (BP),
cellular components (CC) and molecular function (MF) which describe a
biological entity from different aspect. The associations between GO
terms and genes are integrated in the Bioconductor standard packages:
<em>the organism annotation packages</em>. In the current Bioconductor
release (3.17), there are the following organism packages:</p>
<center>
<table class="table">
<colgroup>
<col width="27%">
<col width="25%">
<col width="36%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Organism</th>
<th>Package</th>
<th>Organism</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>org.Hs.eg.db</td>
<td>Human</td>
<td>org.Mm.eg.db</td>
<td>Mouse</td>
</tr>
<tr class="even">
<td>org.Rn.eg.db</td>
<td>Rat</td>
<td>org.Dm.eg.db</td>
<td>Fly</td>
</tr>
<tr class="odd">
<td>org.At.tair.db</td>
<td>Arabidopsis</td>
<td>org.Dr.eg.db</td>
<td>Zebrafish</td>
</tr>
<tr class="even">
<td>org.Sc.sgd.db</td>
<td>Yeast</td>
<td>org.Ce.eg.db</td>
<td>Worm</td>
</tr>
<tr class="odd">
<td>org.Bt.eg.db</td>
<td>Bovine</td>
<td>org.Gg.eg.db</td>
<td>Chicken</td>
</tr>
<tr class="even">
<td>org.Ss.eg.db</td>
<td>Pig</td>
<td>org.Mmu.eg.db</td>
<td>Rhesus</td>
</tr>
<tr class="odd">
<td>org.Cf.eg.db</td>
<td>Canine</td>
<td>org.EcK12.eg.db</td>
<td>E coli strain K12</td>
</tr>
<tr class="even">
<td>org.Xl.eg.db</td>
<td>Xenopus</td>
<td>org.Pt.eg.db</td>
<td>Chimp</td>
</tr>
<tr class="odd">
<td>org.Ag.eg.db</td>
<td>Anopheles</td>
<td>org.EcSakai.eg.db</td>
<td>E coli strain Sakai</td>
</tr>
</tbody>
</table>
</center>
<p>There are four sections in the name of an organism package. The
naming convention is: <code>org</code> simply means “organism”. The
second section corresponds to a specific organism, e.g. <code>Hs</code>
for human and <code>Mm</code> for mouse. The third section corresponds
to the primary gene ID type used in the package, where normally
<code>eg</code> is used which means “Entrez genes” because data is
mostly retrieved from the NCBI database. However, for some organisms,
the primary ID can be from its own primary database,
e.g. <code>sgd</code> for Yeast which corresponds to the <a href="https://www.yeastgenome.org/" class="external-link">Saccharomyces Genome Database</a>,
the primary database for yeast. The last section is always “db”, which
simply implies it is a database package.</p>
<p>Taking the <strong>org.Hs.eg.db</strong> package as an example, all
the data is stored in a database object <code>org.Hs.eg.db</code> in the
<code>OrgDb</code> class. The object contains a connection to a local
SQLite database. Users can simply think <code>org.Hs.eg.db</code> as a
huge table that contains ID mappings between various databases. GO gene
sets are essentially mappings between GO terms and genes. Let’s try to
extract it from the <code>org.Hs.eg.db</code> object.</p>
<p>All the columns (the key column or the source column) can be obtained
by <code>keytypes()</code>:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="fu">keytypes</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
<p>To get the GO gene sets, we first obtain all GO IDs under the BP
(biological process) namespace. As shown in the output from
<code>keytype()</code>, <code>"ONTOLOGY"</code> is also a valid “key
column”, thus we can query “<em>select all GO IDs where the
corresponding ONTOLOGY is BP</em>”, which is translated into the
following code:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BP_Id</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="st">"BP"</span>, keytype <span class="op">=</span> <span class="st">"ONTOLOGY"</span>, </span>
<span>               column <span class="op">=</span> <span class="st">"GO"</span>, multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">BP_Id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "GO:0002764" "GO:0001869" "GO:0048863" "GO:0006805" "GO:0006953"
[6] "GO:0006954"</code></pre>
</div>
<p><code>mapIds()</code> maps IDs between two sources. Since a GO
namespace have more than one GO terms, we have to set
<code>multiVals = "list"</code> to obtain all GO terms under that
namespace. And since we only query for one GO “ONTOLOGY”, we directly
take the first element from the list returned by
<code>mapIds()</code>.</p>
<p>Next we do mapping from GO IDs to gene Entrez IDs. Now the query
becomes “<em>providing a vector of GO IDs, select ENTREZIDs which
correspond to every one of them</em>”.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BPGeneSets</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="va">BP_Id</span>, keytype <span class="op">=</span> <span class="st">"GOALL"</span>, </span>
<span>                    column <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span></span></code></pre>
</div>
<p>You may have noticed there is a “GO” key column as well a “GOALL”
column in the database. As GO has a hierarchical structure where a child
term is a sub-class of a parent term. All the genes annotated to a child
term are also annotated to its parent terms. To reduce the duplicated
information when annotating genes to GO terms, genes are normally
annotated to the most specific offspring terms in the GO hierarchy.
Upstream merging of gene annotations should be done by the tools which
perform analysis. In this way, the mapping between <code>"GO"</code> and
<code>"ENTREZID"</code> only contains “primary” annotations which is not
complete, and mapping between <code>"GOALL"</code> and
<code>"ENTREZID"</code> is the correct one to use.</p>
<p>We filter out GO gene sets with no gene annotated.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BPGeneSets</span> <span class="op">=</span> <span class="va">BPGeneSets</span><span class="op">[</span><span class="fu">sapply</span><span class="op">(</span><span class="va">BPGeneSets</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">BPGeneSets</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="co"># BPGeneSets[[1]] is too long</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`GO:0001869`
[1] "2"   "710"

$`GO:0048863`
  [1] "2"         "25"        "90"        "182"       "182"       "182"
  [7] "596"       "632"       "652"       "652"       "655"       "657"
 [13] "657"       "659"       "678"       "678"       "860"       "998"
 [19] "1000"      "1000"      "1021"      "1045"      "1072"      "1106"
 [25] "1107"      "1108"      "1432"      "1482"      "1482"      "1499"
 [31] "1499"      "1636"      "1906"      "1908"      "1909"      "1910"
 [37] "1947"      "2066"      "2066"      "2068"      "2099"      "2103"
 [43] "2131"      "2146"      "2247"      "2260"      "2263"      "2296"
 [49] "2296"      "2303"      "2335"      "2348"      "2348"      "2487"
 [55] "2494"      "2494"      "2627"      "2637"      "2649"      "2649"
 [61] "2668"      "2668"      "2823"      "2932"      "3065"      "3066"
 [67] "3084"      "3084"      "3091"      "3169"      "3192"      "3192"
 [73] "3192"      "3204"      "3214"      "3233"      "3280"      "3280"
 [79] "3313"      "3357"      "3357"      "3516"      "3516"      "3651"
 [85] "3670"      "3688"      "3720"      "3720"      "3815"      "3815"
 [91] "3861"      "3911"      "3976"      "4040"      "4054"      "4070"
 [97] "4070"      "4072"      "4086"      "4086"      "4089"      "4090"
[103] "4090"      "4093"      "4208"      "4222"      "4254"      "4300"
[109] "4303"      "4487"      "4488"      "4726"      "4780"      "4851"
[115] "4851"      "4851"      "4902"      "5008"      "5076"      "5156"
[121] "5308"      "5308"      "5469"      "5591"      "5591"      "5594"
[127] "5595"      "5610"      "5610"      "5717"      "5764"      "5764"
[133] "5788"      "5928"      "5931"      "5978"      "5978"      "5978"
[139] "5979"      "6201"      "6405"      "6405"      "6422"      "6422"
[145] "6423"      "6469"      "6469"      "6495"      "6548"      "6591"
[151] "6608"      "6660"      "6662"      "6662"      "6662"      "6663"
[157] "6663"      "6722"      "6722"      "6774"      "6862"      "6886"
[163] "6899"      "6909"      "6910"      "6926"      "6928"      "6939"
[169] "6939"      "6949"      "7022"      "7042"      "7048"      "7114"
[175] "7157"      "7291"      "7473"      "7476"      "7478"      "7520"
[181] "7547"      "7869"      "8091"      "8092"      "8320"      "8321"
[187] "8463"      "8482"      "8621"      "8626"      "8820"      "8828"
[193] "8829"      "8829"      "8829"      "8854"      "8929"      "8929"
[199] "9037"      "9112"      "9219"      "9221"      "9241"      "9464"
[205] "9464"      "9698"      "9698"      "9723"      "9739"      "9839"
[211] "9965"      "10016"     "10019"     "10322"     "10322"     "10322"
[217] "10370"     "10371"     "10371"     "10413"     "10443"     "10500"
[223] "10501"     "10505"     "10507"     "10509"     "10512"     "10512"
[229] "10512"     "10524"     "10524"     "10538"     "10538"     "11051"
[235] "11051"     "11137"     "11166"     "22823"     "22823"     "23014"
[241] "23081"     "23376"     "23376"     "23528"     "23528"     "23603"
[247] "23603"     "23788"     "23788"     "25920"     "25920"     "26147"
[253] "26585"     "29072"     "29081"     "29081"     "29947"     "29947"
[259] "30812"     "30812"     "30812"     "51441"     "51441"     "51441"
[265] "51548"     "51548"     "51548"     "51755"     "53615"     "54345"
[271] "54437"     "54517"     "54815"     "54888"     "54888"     "54910"
[277] "54940"     "54940"     "55553"     "55593"     "55698"     "55716"
[283] "55716"     "55818"     "56339"     "56920"     "57348"     "57459"
[289] "57504"     "57556"     "57560"     "57715"     "58495"     "59349"
[295] "63950"     "64218"     "64321"     "64839"     "65979"     "65979"
[301] "65979"     "79727"     "79918"     "79918"     "80031"     "80237"
[307] "80237"     "80326"     "81606"     "83737"     "83746"     "83959"
[313] "83959"     "83989"     "83989"     "84541"     "84541"     "84844"
[319] "93649"     "117854"    "121340"    "124540"    "130497"    "130497"
[325] "144165"    "145258"    "157506"    "166379"    "170506"    "170506"
[331] "202018"    "221662"    "223117"    "340665"    "340665"    "388585"
[337] "399694"    "406938"    "440193"    "442894"    "442911"    "442917"
[343] "553115"    "100170841" "100170841" "101927721" "110599583"</code></pre>
</div>
<p>In most cases, because <code>OrgDb</code> is a standard Bioconductor
data structure, most tools can automatically construct GO gene sets
internally. There is no need for users to touch such low-level
processings.</p>
<div id="further-reading-2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading-2" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>Mapping between various databases can also be done with the general
<code>select()</code> interface. If the <code>OrgDb</code> object is
provided by a package such as <strong>org.Hs.eg.db</strong>, there is
also a separated object that specifically contains mapping between GO
terms and genes. Readers can check the documentation of
<code>org.Hs.egGO2ALLEGS</code>. Additional information on GO terms such
as GO names and long descriptions are available in the package
<strong>GO.db</strong>.</p>
<p>Bioconductor has already provided a large number of organism
packages. However, if the organism you are working on is not supported
there, you may consider to look for it with the
<strong>AnnotationHub</strong> package, which additionally provide
<code>OrgDb</code> objects for approximately 2000 organisms. The
<code>OrgDb</code> object can be directly used in the ORA analysis
introduced in the next section.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kegg-gene-sets">KEGG gene sets<a class="anchor" aria-label="anchor" href="#kegg-gene-sets"></a>
</h3>
<p>A biological pathway is a series of interactions among molecules in a
cell that leads to a certain product or a change in a cell<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. A pathway involves a
list of genes playing different roles which constructs the “pathway gene
set”. <a href="https://www.genome.jp/kegg/pathway.html" class="external-link">KEGG pathway</a>
is the mostly used database for pathways. It provides its data via a
REST API (<a href="https://rest.kegg.jp/" class="external-link uri">https://rest.kegg.jp/</a>). There are several commands to
retrieve specific types of data. To retrieve the pathway gene sets, we
can use the “link” command as shown in the following URL (“link” means
to link genes to pathways). When you enter the URL in the web
browser:</p>
<pre><code>https://rest.kegg.jp/link/pathway/hsa</code></pre>
<p>there will be a text table which contains a column of genes and a
column of pathway IDs.</p>
<pre><code>hsa:10327   path:hsa00010
hsa:124 path:hsa00010
hsa:125 path:hsa00010
hsa:126 path:hsa00010</code></pre>
<p>We can directly read the text output with <code>read.table()</code>.
Wrapping the URL with the function <code>url()</code>, you can pretend
to directly read data from the remote web server.</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggGeneSets</span> <span class="op">=</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://rest.kegg.jp/link/pathway/hsa"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         V1            V2
1 hsa:10327 path:hsa00010
2   hsa:124 path:hsa00010
3   hsa:125 path:hsa00010
4   hsa:126 path:hsa00010
5   hsa:127 path:hsa00010
6   hsa:128 path:hsa00010</code></pre>
</div>
<p>In this two-column table, the first column contains genes in the
Entrez ID type. Let’s remove the <code>"hsa:"</code> prefix, also we
remove the <code>"path:"</code> prefix for pathway IDs in the second
column.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"hsa:"</span>, <span class="st">""</span>, <span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"path:"</span>, <span class="st">""</span>, <span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     V1       V2
1 10327 hsa00010
2   124 hsa00010
3   125 hsa00010
4   126 hsa00010
5   127 hsa00010
6   128 hsa00010</code></pre>
</div>
<p>The full pathway names can be obtained via the “list” command.</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggNames</span> <span class="op">=</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://rest.kegg.jp/list/pathway/hsa"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggNames</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        V1                                                     V2
1 hsa01100              Metabolic pathways - Homo sapiens (human)
2 hsa01200               Carbon metabolism - Homo sapiens (human)
3 hsa01210 2-Oxocarboxylic acid metabolism - Homo sapiens (human)
4 hsa01212           Fatty acid metabolism - Homo sapiens (human)
5 hsa01230     Biosynthesis of amino acids - Homo sapiens (human)
6 hsa01232           Nucleotide metabolism - Homo sapiens (human)</code></pre>
</div>
<p>In both commands, we obtained data for human where the corresponding
KEGG code is <code>"hsa"</code>. The code for other organisms can be
found from the <a href="https://www.genome.jp/kegg/" class="external-link">KEGG website</a>
(e.g. <code>"mmu"</code> for mouse), or via <a href="https://rest.kegg.jp/list/organism" class="external-link uri">https://rest.kegg.jp/list/organism</a>.</p>
<p>Keep in mind, KEGG pathways are only free for academic users. If you
use it for commercial purposes, <a href="https://www.kegg.jp/kegg/legal.html" class="external-link">please contact the KEGG team
to get a licence</a>.</p>
<div id="further-reading-3" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading-3" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>Instead directly reading from the URLs, there are also R packages
which help to obtain data from the KEGG database, such as the
<strong>KEGGREST</strong> package or the <code>download_KEGG()</code>
function from the <strong>clusterProfiler</strong> package. But
essentially, they all obtain KEGG data with the REST API.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="msigdb-gene-sets">MSigDB gene sets<a class="anchor" aria-label="anchor" href="#msigdb-gene-sets"></a>
</h3>
<p><a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link">Molecular
signature database</a> (MSigDB) is a manually curated gene set database.
Initially, it was proposed as a supplementary dataset for <a href="https://www.nature.com/articles/ng1180" class="external-link">the original GSEA
paper</a>. Later it has been separated out and developed independently.
In the first version in 2005, there were only two gene sets collections
and in total 843 gene sets. Now in the newest version of MSigDB
(v2023.1.Hs), it has grown into nine gene sets collections, covering
&gt; 30K gene sets. It provides gene sets on a variety of topics.</p>
<p>MSigDB categorizes gene sets into nine collections where each
collection focuses on a specific topic. For some collections, they are
additionally split into sub-collections. There are several ways to
obtain gene sets from MSigDB. One convenient way is to use the
<strong>msigdbr</strong> package. The advantages is
<strong>msigdbr</strong> supports many other organisms by mapping to
orthologs.</p>
<p>Let’s check which organisms are supported and which gene sets
collections it provides.</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="fu">msigdbr_species</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 20 × 2
   species_name                    species_common_name
   &lt;chr&gt;                           &lt;chr&gt;
 1 Anolis carolinensis             Carolina anole, green anole
 2 Bos taurus                      bovine, cattle, cow, dairy cow, domestic cat…
 3 Caenorhabditis elegans          &lt;NA&gt;
 4 Canis lupus familiaris          dog, dogs
 5 Danio rerio                     leopard danio, zebra danio, zebra fish, zebr…
 6 Drosophila melanogaster         fruit fly
 7 Equus caballus                  domestic horse, equine, horse
 8 Felis catus                     cat, cats, domestic cat
 9 Gallus gallus                   bantam, chicken, chickens, Gallus domesticus
10 Homo sapiens                    human
11 Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monk…
12 Monodelphis domestica           gray short-tailed opossum
13 Mus musculus                    house mouse, mouse
14 Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, pla…
15 Pan troglodytes                 chimpanzee
16 Rattus norvegicus               brown rat, Norway rat, rat, rats
17 Saccharomyces cerevisiae        baker's yeast, brewer's yeast, S. cerevisiae
18 Schizosaccharomyces pombe 972h- &lt;NA&gt;
19 Sus scrofa                      pig, pigs, swine, wild boar
20 Xenopus tropicalis              tropical clawed frog, western clawed frog    </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr_collections</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 4
   gs_collection gs_subcollection  gs_collection_name               num_genesets
   &lt;chr&gt;         &lt;chr&gt;             &lt;chr&gt;                                   &lt;int&gt;
 1 C1            ""                "Positional"                              302
 2 C2            "CGP"             "Chemical and Genetic Perturbat…         3538
 3 C2            "CP"              "Canonical Pathways"                       19
 4 C2            "CP:BIOCARTA"     "BioCarta Pathways"                       292
 5 C2            "CP:KEGG_LEGACY"  "KEGG Legacy Pathways"                    186
 6 C2            "CP:KEGG_MEDICUS" "KEGG Medicus Pathways"                   658
 7 C2            "CP:PID"          "PID Pathways"                            196
 8 C2            "CP:REACTOME"     "Reactome Pathways"                      1787
 9 C2            "CP:WIKIPATHWAYS" "WikiPathways"                            885
10 C3            "MIR:MIRDB"       "miRDB"                                  2377
11 C3            "MIR:MIR_LEGACY"  "MIR_Legacy"                              221
12 C3            "TFT:GTRD"        "GTRD"                                    505
13 C3            "TFT:TFT_LEGACY"  "TFT_Legacy"                              610
14 C4            "3CA"             "Curated Cancer Cell Atlas gene…          148
15 C4            "CGN"             "Cancer Gene Neighborhoods"               427
16 C4            "CM"              "Cancer Modules"                          431
17 C5            "GO:BP"           "GO Biological Process"                  7583
18 C5            "GO:CC"           "GO Cellular Component"                  1042
19 C5            "GO:MF"           "GO Molecular Function"                  1855
20 C5            "HPO"             "Human Phenotype Ontology"               5748
21 C6            ""                "Oncogenic Signature"                     189
22 C7            "IMMUNESIGDB"     "ImmuneSigDB"                            4872
23 C7            "VAX"             "HIPC Vaccine Response"                   347
24 C8            ""                "Cell Type Signature"                     866
25 H             ""                "Hallmark"                                 50</code></pre>
</div>
<p>The first column in the above output is the primary category of gene
sets. Some gene sets collections may have sub-collections, and they are
shown in the second column. The description of gene sets collections
from MSigDB is in the following figure.</p>
<figure><img src="../fig/msigdb.png" alt="MSigDB gene sets collections" class="figure mx-auto d-block"></figure><p>The core function <code>msigdbr()</code> retrieves gene sets in a
specific category (or a subcategory if it exists).</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr</span><span class="op">(</span><span class="va">species</span>, <span class="va">category</span>, <span class="va">subcategory</span><span class="op">)</span></span></code></pre>
</div>
<p>For example, we want to obtain the hallmark gene sets for mouse.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MSigDBGeneSets</span> <span class="op">=</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"mouse"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">MSigDBGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 26
  gene_symbol ncbi_gene ensembl_gene db_gene_symbol db_ncbi_gene db_ensembl_gene
  &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;        &lt;chr&gt;
1 Abca1       11303     ENSMUSG0000… ABCA1          19           ENSG00000165029
2 Abcb8       74610     ENSMUSG0000… ABCB8          11194        ENSG00000197150
3 Acaa2       52538     ENSMUSG0000… ACAA2          10449        ENSG00000167315
4 Acadl       11363     ENSMUSG0000… ACADL          33           ENSG00000115361
5 Acadm       11364     ENSMUSG0000… ACADM          34           ENSG00000117054
6 Acads       11409     ENSMUSG0000… ACADS          35           ENSG00000122971
# ℹ 20 more variables: source_gene &lt;chr&gt;, gs_id &lt;chr&gt;, gs_name &lt;chr&gt;,
#   gs_collection &lt;chr&gt;, gs_subcollection &lt;chr&gt;, gs_collection_name &lt;chr&gt;,
#   gs_description &lt;chr&gt;, gs_source_species &lt;chr&gt;, gs_pmid &lt;chr&gt;,
#   gs_geoid &lt;chr&gt;, gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, db_version &lt;chr&gt;,
#   db_target_species &lt;chr&gt;, ortholog_taxon_id &lt;int&gt;, ortholog_sources &lt;chr&gt;,
#   num_ortholog_sources &lt;dbl&gt;, entrez_gene &lt;chr&gt;, gs_cat &lt;chr&gt;,
#   gs_subcat &lt;chr&gt;</code></pre>
</div>
<p>The output is in the <code>tibble</code> class. If you have no
experience with it, don’t worry, just take it as a table. As you can
see, the three major gene IDs type <code>"gene_symbol"</code>,
<code>"entrez_gene"</code> and <code>"ensembl_gene"</code> are all
included in the table. So users can easily pick one with the same gene
ID type as in the DE gene list. For example:</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MSigDBGeneSets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"ensembl_gene"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7,379 × 2
   gs_name               ensembl_gene
   &lt;chr&gt;                 &lt;chr&gt;
 1 HALLMARK_ADIPOGENESIS ENSMUSG00000015243
 2 HALLMARK_ADIPOGENESIS ENSMUSG00000028973
 3 HALLMARK_ADIPOGENESIS ENSMUSG00000036880
 4 HALLMARK_ADIPOGENESIS ENSMUSG00000026003
 5 HALLMARK_ADIPOGENESIS ENSMUSG00000062908
 6 HALLMARK_ADIPOGENESIS ENSMUSG00000029545
 7 HALLMARK_ADIPOGENESIS ENSMUSG00000020917
 8 HALLMARK_ADIPOGENESIS ENSMUSG00000022477
 9 HALLMARK_ADIPOGENESIS ENSMUSG00000020777
10 HALLMARK_ADIPOGENESIS ENSMUSG00000022994
# ℹ 7,369 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># or put genes in the first column</span></span>
<span><span class="va">MSigDBGeneSets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"ensembl_gene"</span>, <span class="st">"gs_name"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7,379 × 2
   ensembl_gene       gs_name
   &lt;chr&gt;              &lt;chr&gt;
 1 ENSMUSG00000015243 HALLMARK_ADIPOGENESIS
 2 ENSMUSG00000028973 HALLMARK_ADIPOGENESIS
 3 ENSMUSG00000036880 HALLMARK_ADIPOGENESIS
 4 ENSMUSG00000026003 HALLMARK_ADIPOGENESIS
 5 ENSMUSG00000062908 HALLMARK_ADIPOGENESIS
 6 ENSMUSG00000029545 HALLMARK_ADIPOGENESIS
 7 ENSMUSG00000020917 HALLMARK_ADIPOGENESIS
 8 ENSMUSG00000022477 HALLMARK_ADIPOGENESIS
 9 ENSMUSG00000020777 HALLMARK_ADIPOGENESIS
10 ENSMUSG00000022994 HALLMARK_ADIPOGENESIS
# ℹ 7,369 more rows</code></pre>
</div>
<p>If you only want to use a sub-category gene sets,
e.g. <code>"CP:KEGG"</code> from <code>"C2"</code>, you can simply
specify both <code>category</code> and <code>subcategory</code>
arguments:</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"mouse"</span>, category <span class="op">=</span> <span class="st">"C2"</span>, subcategory <span class="op">=</span> <span class="st">"CP:KEGG"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="ora-with-clusterprofiler">ORA with clusterProfiler<a class="anchor" aria-label="anchor" href="#ora-with-clusterprofiler"></a>
</h2>
<hr class="half-width">
<p>The ORA method itself is quite simple and it has been implemented in
a large number of R packages. Among them, the
<strong>clusterProfiler</strong> package especially does a good job in
that it has a seamless integration to the Bioconductor annotation
resources which allows extending GSEA analysis to other organisms
easily; it has pre-defined functions for common analysis tasks, e.g. GO
enrichment, KEGG enrichment; and it implements a variety of different
visualization methods on the GSEA results. In this section, we will
learn how to perform ORA analysis with
<strong>clusterProfiler</strong>.</p>
<p>Here we will use a list of DE genes from a different comparison. As
you may still remember, there are only 54 DE genes between genders,
which may not be a good case for GSEA analysis, since after overlapping
to a collection of gene sets, the majority of the gene sets will have
few or even no gene overlapped. In this example we use the list of DE
genes from the comparison between different time points.</p>
<p>Another thing worth to mention is, in the following code where we
filter DE genes, we additionally add a filtering on the log2 fold
change. This is recommended when the number of DE genes is too large.
The filtering on log2 fold change can be thought as a filtering from the
biology aspect.</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, </span>
<span>                               <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span>                       <span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "3110035E14Rik" "Sgk3"          "Kcnb2"         "Sbspon"
[5] "Gsta3"         "Lman2l"       </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1134</code></pre>
</div>
<p>Let’s confirm that there are around one thousand DE genes, and the DE
genes are in gene symbols.</p>
<div class="section level3">
<h3 id="go-enrichment">GO enrichment<a class="anchor" aria-label="anchor" href="#go-enrichment"></a>
</h3>
<p>In <strong>clusterProfiler</strong>, there is an
<code>enrichGO()</code> function which performs ORA on GO gene sets. To
use it, we need to provide the DE genes, the organism <code>OrgDb</code>
object which is from the organism package <strong>org.Mm.eg.db</strong>
(because our data is from mouse), also the GO namespace (one of
<code>"BP"</code>, <code>"CC"</code> and <code>"MF"</code>). The GO gene
sets are automatically retrieved and processed from
<code>org.Mm.eg.db</code> in <code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb74">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; No gene can be mapped....</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; Expected input gene ID: 16571,67025,57321,54326,12053,66408</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; return NULL...</code></pre>
</div>
<p>Oops, something seems wrong. Well, this is a common mistake where the
gene ID types do not match between DE genes and the gene sets.
Thankfully, the message clearly explains the reason. The ID type for
gene sets is Entrez ID and it cannot match any DE gene.</p>
<p>There are two ways to solve this problem. 1. Convert gene IDs in
<code>timeDEgenes</code> to Entrez IDs in advance; or 2. Simply specify
the ID type of DE genes and let <code>enrichGO()</code> do the
conversion job (recall various gene ID types are also stored in the
<code>OrgDb</code> object). Let’s choose the second way.</p>
<p>In the next code, we additionally specify
<code>keyType = "SYMBOL"</code> to explicitly tell the function that DE
genes are in gene symbols. Recall that all valid values for
<code>keyType</code> are in <code>keytypes(org.Mm.eg.db)</code>.</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                       Description GeneRatio   BgRatio
GO:0050900 GO:0050900               leukocyte migration    52/954 414/25730
GO:0006935 GO:0006935                        chemotaxis    56/954 479/25730
GO:0042330 GO:0042330                             taxis    56/954 481/25730
GO:0030595 GO:0030595              leukocyte chemotaxis    37/954 238/25730
GO:0060326 GO:0060326                   cell chemotaxis    43/954 333/25730
GO:0002685 GO:0002685 regulation of leukocyte migration    35/954 249/25730
           RichFactor FoldEnrichment   zScore       pvalue     p.adjust
GO:0050900  0.1256039       3.387618 9.610312 1.544498e-14 6.622355e-11
GO:0006935  0.1169102       3.153145 9.334085 3.281626e-14 6.622355e-11
GO:0042330  0.1164241       3.140034 9.296965 3.917781e-14 6.622355e-11
GO:0030595  0.1554622       4.192916 9.710556 1.522653e-13 1.930343e-10
GO:0060326  0.1291291       3.482697 8.947971 1.172757e-12 1.189410e-09
GO:0002685  0.1405622       3.791055 8.684202 1.339390e-11 1.026246e-08
                 qvalue
GO:0050900 5.208586e-11
GO:0006935 5.208586e-11
GO:0042330 5.208586e-11
GO:0030595 1.518245e-10
GO:0060326 9.354900e-10
GO:0002685 8.071583e-09
                                                                                                                                                                                                                                                                                                                                         geneID
GO:0050900                                 Tnfsf18/Sell/Slamf9/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Ascl2/Gdf15/Calr/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/C3/Il33/Ch25h
GO:0006935 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0042330 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0030595                                                                                                                 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl2/Ccl7/Ccl5/Ccr7/Ccl28/Lgals3/Ptk2b/Retnlg/Fpr2/Dusp1/C3/Ch25h
GO:0060326                                                                              Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Fpr2/Dusp1/C3/Ch25h/Plxnb3/Nox1
GO:0002685                                                                                                                                    Tnfsf18/Sell/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Edn3/P2ry12/Il12a/Nbl1/Padi2/Bst1/Ppbp/Ptn/Adam8/Ascl2/Gdf15/Calr/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Apod/Fpr2/Dusp1/Ager/Il33
           Count
GO:0050900    52
GO:0006935    56
GO:0042330    56
GO:0030595    37
GO:0060326    43
GO:0002685    35</code></pre>
</div>
<p>Now <code>enrichGO()</code> went through! The returned object
<code>resTimeGO</code> is in a special format which looks like a table
but actually is not! To get rid of the confusion, in the code it is
converted to a real data frame <code>resTimeGOTable</code>.</p>
<p>In the output data frame, there are the following columns:</p>
<ul>
<li>
<code>ID</code>: ID of the gene set. In this example analysis, it is
the GO ID.</li>
<li>
<code>Description</code>: Readable description. Here it is the name
of the GO term.</li>
<li>
<code>GeneRatio</code>: Number of DE genes in the gene set / total
number of DE genes.</li>
<li>
<code>BgRatio</code>: Size of the gene set / total number of
genes.</li>
<li>
<code>pvalue</code>: <em>p</em>-value calculated from the
hypergeometric distribution.</li>
<li>
<code>p.adjust</code>: Adjusted <em>p</em>-value by the BH
method.</li>
<li>
<code>qvalue</code>: <em>q</em>-value which is another way for
controlling false positives in multiple testings.</li>
<li>
<code>geneID</code>: A list of DE genes in the gene set.</li>
<li>
<code>Count</code>: Number of DE genes in the gene set.</li>
</ul>
<p>You may have found the total number of DE genes changes. There are
1134 in <code>timeDEgenes</code>, but only 983 DE genes are included in
the enrichment result table (in the <code>GeneRatio</code> column). The
main reason is by default DE genes not annotated to any GO gene set are
filtered out. This relates to the “universe” of all genes in ORA, which
we will touch in the end of this section.</p>
<p>There are several additional arguments in
<code>enrichGO()</code>:</p>
<ul>
<li>
<code>universe</code>: the universe set of genes, i.e. total genes
to use. By default it uses the union of the genes in all gene sets. If
it is set, DE genes and all gene sets will take intersections with it.
We will discuss it in the end of this section.</li>
<li>
<code>minGSSize</code>: Minimal size of gene sets. Normally gene
sets with very small size have very specific biological meanings which
are not helpful too much for the interpretation. Gene sets with size
smaller than it will be removed from the analysis. By default it is
10.</li>
<li>
<code>maxGSSize</code>: Maximal size of gene sets. Normally gene
sets with huge size provide too general biological meanings and are not
helpful either. By default is 500.</li>
<li>
<code>pvalueCutoff</code>: Cutoff for both <em>p</em>-values and
adjusted <em>p</em>-values. By default is 0.05.</li>
<li>
<code>qvalueCutoff</code>: Cutoff for <em>q</em>-values. by default
is 0.2.</li>
</ul>
<p>Note, <code>enrichGO()</code> only returns significant gene sets that
pass the cutoffs<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>. This function design might not be proper
because a function should return all the results no matter they are
significant or not. Later users may need to use the complete enrichment
table for downstream anlaysis. Second, the meaning of
<code>pvalueCutoff</code> is not precise and there is redundancy between
<code>pvalueCutoff</code> and <code>qvalueCutoff</code> (adjusted
<em>p</em>-values and <em>q</em>-values are always non-smaller than raw
<em>p</em>-values). Thus it is suggested to set both
<code>pvalueCutoff</code> and <code>qvalueCutoff</code> to 1 in
<code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb80">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                     pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                       Description GeneRatio   BgRatio
GO:0050900 GO:0050900               leukocyte migration    52/954 414/25730
GO:0006935 GO:0006935                        chemotaxis    56/954 479/25730
GO:0042330 GO:0042330                             taxis    56/954 481/25730
GO:0030595 GO:0030595              leukocyte chemotaxis    37/954 238/25730
GO:0060326 GO:0060326                   cell chemotaxis    43/954 333/25730
GO:0002685 GO:0002685 regulation of leukocyte migration    35/954 249/25730
           RichFactor FoldEnrichment   zScore       pvalue     p.adjust
GO:0050900  0.1256039       3.387618 9.610312 1.544498e-14 6.622355e-11
GO:0006935  0.1169102       3.153145 9.334085 3.281626e-14 6.622355e-11
GO:0042330  0.1164241       3.140034 9.296965 3.917781e-14 6.622355e-11
GO:0030595  0.1554622       4.192916 9.710556 1.522653e-13 1.930343e-10
GO:0060326  0.1291291       3.482697 8.947971 1.172757e-12 1.189410e-09
GO:0002685  0.1405622       3.791055 8.684202 1.339390e-11 1.026246e-08
                 qvalue
GO:0050900 5.208586e-11
GO:0006935 5.208586e-11
GO:0042330 5.208586e-11
GO:0030595 1.518245e-10
GO:0060326 9.354900e-10
GO:0002685 8.071583e-09
                                                                                                                                                                                                                                                                                                                                         geneID
GO:0050900                                 Tnfsf18/Sell/Slamf9/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Ascl2/Gdf15/Calr/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/C3/Il33/Ch25h
GO:0006935 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0042330 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0030595                                                                                                                 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl2/Ccl7/Ccl5/Ccr7/Ccl28/Lgals3/Ptk2b/Retnlg/Fpr2/Dusp1/C3/Ch25h
GO:0060326                                                                              Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Fpr2/Dusp1/C3/Ch25h/Plxnb3/Nox1
GO:0002685                                                                                                                                    Tnfsf18/Sell/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Edn3/P2ry12/Il12a/Nbl1/Padi2/Bst1/Ppbp/Ptn/Adam8/Ascl2/Gdf15/Calr/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Apod/Fpr2/Dusp1/Ager/Il33
           Count
GO:0050900    52
GO:0006935    56
GO:0042330    56
GO:0030595    37
GO:0060326    43
GO:0002685    35</code></pre>
</div>
<div id="perform-go-enrichment-on-other-organisms" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="perform-go-enrichment-on-other-organisms" class="callout-inner">
<h3 class="callout-title">Perform GO enrichment on other organisms</h3>
<div class="callout-content">
<p>Gene sets are provided as an <code>OrgDb</code> object in
<code>enrichGO()</code>, thus you can perform ORA analysis on any
organism as long as there is a corresponding <code>OrgDb</code>
object.</p>
<ul>
<li>For model organisms, the <code>OrgDb</code> object can be obtained
from the corresponding <strong>org.*.db</strong> package.</li>
<li>For other organisms, the <code>OrgDb</code> object can be found with
the <strong>AnnotationHub</strong> package.</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kegg-pathway-enrichment">KEGG pathway enrichment<a class="anchor" aria-label="anchor" href="#kegg-pathway-enrichment"></a>
</h3>
<p>To perform KEGG pathway enrichment analysis, there is also a function
<code>enrichKEGG()</code> for that. Unfortunately, it cannot perform
gene ID conversion automatically. Thus, if the ID type is not Entrez ID,
we have to convert it by hand.</p>
<p>Note if you have also set <code>universe</code>, it should be
converted to Entrez IDs as well.</p>
<p>We use the <code>mapIds()</code> function to convert genes from
symbols to Entrez IDs. Since the ID mapping is not always one-to-one. We
only take the first one if there are multiple hits by setting
<code>multiVals = "first"</code>(but of course you can choose other
options for <code>multiVals</code>, check the documentation). We also
remove genes with no mapping available (with <code>NA</code> after the
mapping)<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<div class="codewrapper sourceCode" id="cb82">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EntrezIDs</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keys <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                   keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, column <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EntrezIDs</span> <span class="op">=</span> <span class="va">EntrezIDs</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">EntrezIDs</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">EntrezIDs</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    Sgk3    Kcnb2   Sbspon    Gsta3   Lman2l  Ankrd39
"170755"  "98741" "226866"  "14859" "214895" "109346" </code></pre>
</div>
<p>We have to set the KEGG organism code if it is not human. Similarly
it is suggested to set <code>pvalueCutoff</code> and
<code>qvalueCutoff</code> both to 1 and convert the result to a data
frame.</p>
<div class="codewrapper sourceCode" id="cb86">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeKEGG</span> <span class="op">=</span> <span class="fu">enrichKEGG</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">EntrezIDs</span>, </span>
<span>                         organism <span class="op">=</span> <span class="st">"mmu"</span>,</span>
<span>                         pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                         qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeKEGGTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeKEGG</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeKEGGTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                     category
mmu00590                           Metabolism
mmu00591                           Metabolism
mmu00565                           Metabolism
mmu00592                           Metabolism
mmu04913                   Organismal Systems
mmu04061 Environmental Information Processing
                                 subcategory       ID
mmu00590                    Lipid metabolism mmu00590
mmu00591                    Lipid metabolism mmu00591
mmu00565                    Lipid metabolism mmu00565
mmu00592                    Lipid metabolism mmu00592
mmu04913                    Endocrine system mmu04913
mmu04061 Signaling molecules and interaction mmu04061
                                                           Description
mmu00590                                   Arachidonic acid metabolism
mmu00591                                      Linoleic acid metabolism
mmu00565                                        Ether lipid metabolism
mmu00592                               alpha-Linolenic acid metabolism
mmu04913                                       Ovarian steroidogenesis
mmu04061 Viral protein interaction with cytokine and cytokine receptor
         GeneRatio  BgRatio RichFactor FoldEnrichment   zScore       pvalue
mmu00590    16/457 89/10623  0.1797753       4.178890 6.384988 1.029723e-06
mmu00591    12/457 55/10623  0.2181818       5.071653 6.418624 2.825416e-06
mmu00565    11/457 49/10623  0.2244898       5.218282 6.274805 5.477871e-06
mmu00592     8/457 25/10623  0.3200000       7.438425 6.833196 6.249183e-06
mmu04913    12/457 65/10623  0.1846154       4.291399 5.643292 1.754012e-05
mmu04061    14/457 95/10623  0.1473684       3.425590 5.034919 5.088522e-05
             p.adjust       qvalue
mmu00590 0.0003253926 0.0002655603
mmu00591 0.0004464157 0.0003643300
mmu00565 0.0004936855 0.0004029078
mmu00592 0.0004936855 0.0004029078
mmu04913 0.0011085357 0.0009047010
mmu04061 0.0026799549 0.0021871717
                                                                                                       geneID
mmu00590 18783/19215/211429/329502/78390/19223/67103/242546/13118/18781/18784/11689/232889/15446/237625/11687
mmu00591                        18783/211429/329502/78390/242546/18781/18784/13113/622127/232889/237625/11687
mmu00565                               18783/211429/329502/78390/22239/18781/18784/232889/320981/237625/53897
mmu00592                                                  18783/211429/329502/78390/18781/18784/232889/237625
mmu04913                          18783/211429/329502/78390/242546/11689/232889/13076/13070/15485/13078/16867
mmu04061                  16174/20311/57349/56744/14825/20295/20296/20306/20304/20305/12775/56838/16185/16186
         Count
mmu00590    16
mmu00591    12
mmu00565    11
mmu00592     8
mmu04913    12
mmu04061    14</code></pre>
</div>
<div id="perform-kegg-pathway-enrichment-on-other-organisms" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="perform-kegg-pathway-enrichment-on-other-organisms" class="callout-inner">
<h3 class="callout-title">Perform KEGG pathway enrichment on other
organisms</h3>
<div class="callout-content">
<p>Extending ORA to other organisms is rather simple.</p>
<ol style="list-style-type: decimal">
<li>Make sure the DE genes are Entrez IDs.</li>
<li>Choose the corresponding KEGG organism code.</li>
</ol>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="msigdb-enrichment">MSigDB enrichment<a class="anchor" aria-label="anchor" href="#msigdb-enrichment"></a>
</h3>
<p>For MSigDB gene sets, there is no pre-defined enrichment function. We
need to directly use the low-level enrichment function
<code>enricher()</code> which accepts self-defined gene sets. The gene
sets should be in a format of a two-column data frame of genes and gene
sets (or a class that can be converted to a data frame).</p>
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="va">gene_sets</span> <span class="op">=</span> <span class="fu">msigdbr</span><span class="op">(</span>category <span class="op">=</span> <span class="st">"H"</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 26
  gene_symbol ncbi_gene ensembl_gene db_gene_symbol db_ncbi_gene db_ensembl_gene
  &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;        &lt;chr&gt;
1 Abca1       11303     ENSMUSG0000… ABCA1          19           ENSG00000165029
2 Abcb8       74610     ENSMUSG0000… ABCB8          11194        ENSG00000197150
3 Acaa2       52538     ENSMUSG0000… ACAA2          10449        ENSG00000167315
4 Acadl       11363     ENSMUSG0000… ACADL          33           ENSG00000115361
5 Acadm       11364     ENSMUSG0000… ACADM          34           ENSG00000117054
6 Acads       11409     ENSMUSG0000… ACADS          35           ENSG00000122971
# ℹ 20 more variables: source_gene &lt;chr&gt;, gs_id &lt;chr&gt;, gs_name &lt;chr&gt;,
#   gs_collection &lt;chr&gt;, gs_subcollection &lt;chr&gt;, gs_collection_name &lt;chr&gt;,
#   gs_description &lt;chr&gt;, gs_source_species &lt;chr&gt;, gs_pmid &lt;chr&gt;,
#   gs_geoid &lt;chr&gt;, gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, db_version &lt;chr&gt;,
#   db_target_species &lt;chr&gt;, ortholog_taxon_id &lt;int&gt;, ortholog_sources &lt;chr&gt;,
#   num_ortholog_sources &lt;dbl&gt;, entrez_gene &lt;chr&gt;, gs_cat &lt;chr&gt;,
#   gs_subcat &lt;chr&gt;</code></pre>
</div>
<p>As mentioned before, it is important the gene ID type in the gene
sets should be the same as in the DE genes, so here we choose the
<code>"gene_symbol"</code> column.</p>
<div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeHallmark</span> <span class="op">=</span> <span class="fu">enricher</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                           TERM2GENE <span class="op">=</span> <span class="va">gene_sets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                           pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeHallmarkTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeHallmark</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeHallmarkTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                                             ID
HALLMARK_MYOGENESIS                         HALLMARK_MYOGENESIS
HALLMARK_COMPLEMENT                         HALLMARK_COMPLEMENT
HALLMARK_COAGULATION                       HALLMARK_COAGULATION
HALLMARK_ALLOGRAFT_REJECTION       HALLMARK_ALLOGRAFT_REJECTION
HALLMARK_ESTROGEN_RESPONSE_LATE HALLMARK_ESTROGEN_RESPONSE_LATE
HALLMARK_INFLAMMATORY_RESPONSE   HALLMARK_INFLAMMATORY_RESPONSE
                                                    Description GeneRatio
HALLMARK_MYOGENESIS                         HALLMARK_MYOGENESIS    31/291
HALLMARK_COMPLEMENT                         HALLMARK_COMPLEMENT    26/291
HALLMARK_COAGULATION                       HALLMARK_COAGULATION    20/291
HALLMARK_ALLOGRAFT_REJECTION       HALLMARK_ALLOGRAFT_REJECTION    23/291
HALLMARK_ESTROGEN_RESPONSE_LATE HALLMARK_ESTROGEN_RESPONSE_LATE    23/291
HALLMARK_INFLAMMATORY_RESPONSE   HALLMARK_INFLAMMATORY_RESPONSE    22/291
                                 BgRatio RichFactor FoldEnrichment   zScore
HALLMARK_MYOGENESIS             201/4393  0.1542289       2.328273 5.133984
HALLMARK_COMPLEMENT             196/4393  0.1326531       2.002560 3.824271
HALLMARK_COAGULATION            139/4393  0.1438849       2.172118 3.739899
HALLMARK_ALLOGRAFT_REJECTION    201/4393  0.1144279       1.727428 2.811625
HALLMARK_ESTROGEN_RESPONSE_LATE 206/4393  0.1116505       1.685500 2.683920
HALLMARK_INFLAMMATORY_RESPONSE  201/4393  0.1094527       1.652323 2.521330
                                      pvalue     p.adjust       qvalue
HALLMARK_MYOGENESIS             5.736712e-06 0.0002753622 0.0002294685
HALLMARK_COMPLEMENT             4.315862e-04 0.0103580695 0.0086317246
HALLMARK_COAGULATION            7.108706e-04 0.0113739297 0.0094782747
HALLMARK_ALLOGRAFT_REJECTION    6.414928e-03 0.0769791391 0.0641492826
HALLMARK_ESTROGEN_RESPONSE_LATE 8.606752e-03 0.0826248216 0.0688540180
HALLMARK_INFLAMMATORY_RESPONSE  1.259890e-02 0.0963081013 0.0802567511
                                                                                                                                                                                                                geneID
HALLMARK_MYOGENESIS             Myl1/Casq1/Aplnr/Tnnc2/Ptgis/Gja5/Col15a1/Cav3/Tnnt1/Ryr1/Cox6a2/Tnni2/Lsp1/Nqo1/Hspb2/Cryab/Erbb3/Stc2/Gpx3/Sparc/Myh3/Myh1/Col1a1/Cacng1/Itgb4/Bdkrb2/Mapk12/Apod/Spdef/Cdkn1a/Actn3
HALLMARK_COMPLEMENT                                                Serpinb2/Pla2g4a/Cd46/Hspa5/Cp/Gnb4/S100a9/Cda/Cxcl1/Apoc1/Itgam/Irf7/Klkb1/Mmp15/Mmp8/Ccl5/Lgals3/Gzmb/Tmprss6/Maff/Plg/Psmb9/Hspa1a/C3/Dusp5/Phex
HALLMARK_COAGULATION                                                                                         Serpinb2/Ctse/C8g/Ctsk/Masp2/Pf4/Sh2b2/Vwf/Apoc1/Klkb1/Mmp15/Mmp8/Trf/Sparc/Itgb3/Dct/Tmprss6/Maff/Plg/C3
HALLMARK_ALLOGRAFT_REJECTION                                                           Il18rap/Cd247/Il12a/Pf4/Capg/Ccnd2/Igsf6/Irf7/Il12rb1/Icosl/Itk/Nos2/Ccl2/Ccl7/Ccl5/Gpr65/Gzmb/Il2rb/Tap1/Tap2/H2-T23/Cfp/Il2rg
HALLMARK_ESTROGEN_RESPONSE_LATE                                                    Clic3/Ass1/Mdk/S100a9/Cdc20/Nbl1/Cav1/Idh2/Th/Pdlim3/Ascl1/Xbp1/Top2a/Dcxr/Fos/Serpina3n/Dhrs2/Tst/Emp2/Fkbp5/Mapk13/Cyp4f15/Kif20a
HALLMARK_INFLAMMATORY_RESPONSE                                                              Il18rap/Sell/Aplnr/Abca1/Lpar1/Cxcl5/Adm/Irf7/Msr1/Bst2/Ccl17/Icosl/Ccl2/Ccl7/Ccl5/Ccr7/Itgb3/Has2/Il2rb/Cdkn1a/Cd70/Best1
                                Count
HALLMARK_MYOGENESIS                31
HALLMARK_COMPLEMENT                26
HALLMARK_COAGULATION               20
HALLMARK_ALLOGRAFT_REJECTION       23
HALLMARK_ESTROGEN_RESPONSE_LATE    23
HALLMARK_INFLAMMATORY_RESPONSE     22</code></pre>
</div>
<div id="further-reading-4" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading-4" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>Implementing ORA is rather simple. The following function
<code>ora()</code> performs ORA on a list of gene sets. Try to read and
understand the code.</p>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">genes</span>, <span class="va">gene_sets</span>, <span class="va">universe</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="fu">is.null</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co"># make sure genes are unique</span></span>
<span>    <span class="va">genes</span> <span class="op">=</span> <span class="fu">intersect</span><span class="op">(</span><span class="va">genes</span>, <span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="va">gene_sets</span> <span class="op">=</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">intersect</span>, <span class="va">universe</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># calculate different numbers</span></span>
<span>    <span class="va">n_11</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">genes</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span>
<span>    <span class="va">n_01</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">length</span><span class="op">)</span></span>
<span>    <span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># calculate p-values</span></span>
<span>    <span class="va">p</span> <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n</span> <span class="op">-</span> <span class="va">n_10</span>, <span class="va">n_01</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>        gene_set <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span>,</span>
<span>        hits <span class="op">=</span> <span class="va">n_11</span>,</span>
<span>        n_genes <span class="op">=</span> <span class="va">n_10</span>,</span>
<span>        gene_set_size <span class="op">=</span> <span class="va">n_01</span>,</span>
<span>        n_total <span class="op">=</span> <span class="va">n</span>,</span>
<span>        p_value <span class="op">=</span> <span class="va">p</span>,</span>
<span>        p_adjust <span class="op">=</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">p</span>, <span class="st">"BH"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Test on the MSigDB hallmark gene sets:</p>
<div class="codewrapper sourceCode" id="cb93">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HallmarkGeneSets</span> <span class="op">=</span> <span class="fu">split</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">$</span><span class="va">gene_symbol</span>, <span class="va">gene_sets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmarkGeneSets</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                                 gene_set hits n_genes
HALLMARK_ADIPOGENESIS               HALLMARK_ADIPOGENESIS    9    1134
HALLMARK_ALLOGRAFT_REJECTION HALLMARK_ALLOGRAFT_REJECTION   23    1134
HALLMARK_ANDROGEN_RESPONSE     HALLMARK_ANDROGEN_RESPONSE    6    1134
HALLMARK_ANGIOGENESIS               HALLMARK_ANGIOGENESIS    4    1134
HALLMARK_APICAL_JUNCTION         HALLMARK_APICAL_JUNCTION   14    1134
HALLMARK_APICAL_SURFACE           HALLMARK_APICAL_SURFACE    3    1134
                             gene_set_size n_total     p_value    p_adjust
HALLMARK_ADIPOGENESIS                  185   21198 0.662392856 0.871569547
HALLMARK_ALLOGRAFT_REJECTION           191   21198 0.000234807 0.002935087
HALLMARK_ANDROGEN_RESPONSE             107   21198 0.512665196 0.753919406
HALLMARK_ANGIOGENESIS                   36   21198 0.124351058 0.260206520
HALLMARK_APICAL_JUNCTION               186   21198 0.124899130 0.260206520
HALLMARK_APICAL_SURFACE                 41   21198 0.376915129 0.588929889</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="choose-a-proper-universe">Choose a proper universe<a class="anchor" aria-label="anchor" href="#choose-a-proper-universe"></a>
</h3>
<p>Finally, it is time to talk about the “universe” of ORA analysis
which is normally ignored in many analyses. In current tools, there are
mainly following different universe settings:</p>
<ol style="list-style-type: decimal">
<li>Using all genes in the genome, this also includes non-protein coding
genes. For human, the size of universe is 60k ~ 70k.</li>
<li>Using all protein-coding genes. For human, the size of universe is ~
20k.</li>
<li>In the era of microarray, total genes that are measured on the chip
is taken as the universe. For RNASeq, since reads are aligned to all
genes, we can set a cutoff and only use those “expressed” genes as the
universe.</li>
<li>Using all genes in a gene sets collection. Then the size of the
universe depends on the size of the gene sets collection. For example GO
gene sets collection is much larger than the KEGG pathway gene sets
collection.</li>
</ol>
<p>If the universe is set, DE genes as well as genes in the gene sets
are first intersected to the universe. However, in general the universe
affects three values of <span class="math inline">\(n_{22}\)</span>,
<span class="math inline">\(n_{02}\)</span> and <span class="math inline">\(n_{20}\)</span> more, which correspond to the
non-DE genes or non-gene-set genes.</p>
<center>
<table class="table">
<colgroup>
<col width="19%">
<col width="28%">
<col width="38%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td><span class="math inline">\(n_{11}\)</span></td>
<td><span class="math inline">\(n_{12}\)</span></td>
<td><span class="math inline">\(n_{1+}\)</span></td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td><span class="math inline">\(n_{21}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{22}}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{2+}}\)</span></td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td><span class="math inline">\(n_{+1}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{+2}}\)</span></td>
<td><span class="math inline">\(\color{red}{n}\)</span></td>
</tr>
</tbody>
</table>
</center>
<p>In the contingency table, we are testing the dependency of whether
genes being DE and whether genes being in the gene set. In the model,
each gene has a definite attribute of being either DE or non-DE and each
gene has a second definite attribute of either belonging to the gene set
or not. If a larger universe is used, such as total genes where there
are genes not measured nor will never annotated to gene sets (let’s call
them <em>non-informative genes</em>, e.g. non-protein coding genes or
not-expressed genes), all the non-informative genes are implicitly
assigned with an attribute of being non-DE or not in the gene set. This
implicit assignment is not proper because these genes provide no
information and they should not be included in the analysis. Adding them
to the analysis increases <span class="math inline">\(n_{22}\)</span>,
<span class="math inline">\(n_{02}\)</span> or <span class="math inline">\(n_{20}\)</span>, makes the observation <span class="math inline">\(n_{11}\)</span> getting further away from the null
distribution, eventually generates a smaller <em>p</em>-value. For the
similar reason, small universes tend to generate large
<em>p</em>-values.</p>
<p>In
<code>enrichGO()</code>/<code>enrichKEGG()</code>/<code>enricher()</code>,
universe genes can be set via the <code>universe</code> argument. By
default the universe is the total genes in the gene sets collection.
When a self-defined universe is provided, this might be different from
what you may think, the universe is <a href="https://github.com/YuLab-SMU/DOSE/blob/93a4b981c0251e5c6eb1f2413f4a02b8e6d06ff5/R/enricher_internal.R#L65C43-L65C51" class="external-link"><em>the
intersection of user-provided universe and total genes in the gene set
collection</em></a>. Thus the universe setting in
<strong>clusterProfiler</strong> is very conservative.</p>
<p>Check the more discusstions at <a href="https://twitter.com/mdziemann/status/1626407797939384320" class="external-link uri">https://twitter.com/mdziemann/status/1626407797939384320</a>.</p>
<p>We can do a simple experiment on the small MSigDB hallmark gene sets.
We use the <code>ora()</code> function which we have implemented in
previous “Further reading” section and we compare three different
universe settings.</p>
<div class="codewrapper sourceCode" id="cb95">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all genes in the gene sets collection ~ 4k genes</span></span>
<span><span class="va">df1</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmarkGeneSets</span><span class="op">)</span></span>
<span><span class="co"># all protein-coding genes, ~ 20k genes</span></span>
<span><span class="va">df2</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmarkGeneSets</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># all genes in org.Mm.eg.db ~ 70k genes</span></span>
<span><span class="va">df3</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmarkGeneSets</span>, </span>
<span>    <span class="fu">keys</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># df1, df2, and df3 are in the same row order, </span></span>
<span><span class="co"># so we can directly compare them</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">df1</span><span class="op">$</span><span class="va">p_value</span>, <span class="va">df2</span><span class="op">$</span><span class="va">p_value</span>, col <span class="op">=</span> <span class="fl">2</span>, pch <span class="op">=</span> <span class="fl">16</span>, </span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>    xlab <span class="op">=</span> <span class="st">"all hallmark genes as universe (p-values)"</span>, ylab <span class="op">=</span> <span class="st">"p-values"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"compare universes"</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">df1</span><span class="op">$</span><span class="va">p_value</span>, <span class="va">df3</span><span class="op">$</span><span class="va">p_value</span>, col <span class="op">=</span> <span class="fl">4</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topleft"</span>, legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"all protein-coding genes as universe"</span>, <span class="st">"all genes as universe"</span><span class="op">)</span>, </span>
<span>    pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-compare-universe-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It is very straightforward to see, with a larger universe, there are
more significant gene sets, which may produce potentially more false
positives. This is definitely worse when using all genes in the genome
as universe.</p>
<p>Based on the discussion in this section, the recommendation of using
universe is:</p>
<ol style="list-style-type: decimal">
<li>using protein-coding genes,</li>
<li>using measured genes,</li>
<li>or using a conservative way with
<strong>clusterProfiler</strong>.</li>
</ol>
</div>
</section><section><h2 class="section-heading" id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<hr class="half-width">
<p><strong>clusterProfiler</strong> provides a rich set of visualization
methods on the GSEA results, from simple visualization to complex ones.
Complex visualizations are normally visually fancy but do not transfer
too much useful information, and they should only be applied in very
specific scenarios under very specific settings; while simple graphs
normally do better jobs. Recently the visualization code in
<strong>clusterProfiler</strong> has been moved to a new package
<strong>enrichplot</strong>. Let’s first load the
<strong>enrichplot</strong> package. The full sets of visualizations
that <strong>enrichplot</strong> supports can be found from <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html" class="external-link uri">https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html</a>.</p>
<p>We first re-generate the enrichment table.</p>
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span></span>
<span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                     pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                       Description GeneRatio   BgRatio
GO:0050900 GO:0050900               leukocyte migration    52/954 414/25730
GO:0006935 GO:0006935                        chemotaxis    56/954 479/25730
GO:0042330 GO:0042330                             taxis    56/954 481/25730
GO:0030595 GO:0030595              leukocyte chemotaxis    37/954 238/25730
GO:0060326 GO:0060326                   cell chemotaxis    43/954 333/25730
GO:0002685 GO:0002685 regulation of leukocyte migration    35/954 249/25730
           RichFactor FoldEnrichment   zScore       pvalue     p.adjust
GO:0050900  0.1256039       3.387618 9.610312 1.544498e-14 6.622355e-11
GO:0006935  0.1169102       3.153145 9.334085 3.281626e-14 6.622355e-11
GO:0042330  0.1164241       3.140034 9.296965 3.917781e-14 6.622355e-11
GO:0030595  0.1554622       4.192916 9.710556 1.522653e-13 1.930343e-10
GO:0060326  0.1291291       3.482697 8.947971 1.172757e-12 1.189410e-09
GO:0002685  0.1405622       3.791055 8.684202 1.339390e-11 1.026246e-08
                 qvalue
GO:0050900 5.208586e-11
GO:0006935 5.208586e-11
GO:0042330 5.208586e-11
GO:0030595 1.518245e-10
GO:0060326 9.354900e-10
GO:0002685 8.071583e-09
                                                                                                                                                                                                                                                                                                                                         geneID
GO:0050900                                 Tnfsf18/Sell/Slamf9/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Ascl2/Gdf15/Calr/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/C3/Il33/Ch25h
GO:0006935 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0042330 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a7a/S100a8/S100a9/Lpar1/Ptgr1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/C3/Stx3/Ch25h/Plxnb3/Nox1
GO:0030595                                                                                                                 Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl2/Ccl7/Ccl5/Ccr7/Ccl28/Lgals3/Ptk2b/Retnlg/Fpr2/Dusp1/C3/Ch25h
GO:0060326                                                                              Tnfsf18/Sell/Slamf9/AI182371/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Fpr2/Dusp1/C3/Ch25h/Plxnb3/Nox1
GO:0002685                                                                                                                                    Tnfsf18/Sell/Fut7/AI182371/Itga4/Mdk/Grem1/Ada/Edn3/P2ry12/Il12a/Nbl1/Padi2/Bst1/Ppbp/Ptn/Adam8/Ascl2/Gdf15/Calr/Aire/Ccl2/Ccl7/Ccl5/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Apod/Fpr2/Dusp1/Ager/Il33
           Count
GO:0050900    52
GO:0006935    56
GO:0042330    56
GO:0030595    37
GO:0060326    43
GO:0002685    35</code></pre>
</div>
<p><code>barplot()</code> and <code>dotplot()</code> generate plots for
a small number of significant gene sets. Note the two functions are
directly applied on <code>resTimeGO</code> returned by
<code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">resTimeGO</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-more-enrichplots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb99">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">resTimeGO</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-more-enrichplots-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Barplots can map two variables to the plot, one to the height of bars
and the other to the colors of bars; while for dotplot, sizes of dots
can be mapped to a third variable. The variable names are in the colum
names of the result table. Both plots include the top 20 most
significant terms. On dotplot, terms are ordered by the values on x-axis
(the <code>GeneRatio</code>).</p>
<p>Now we need to talk about “what is a good visualization?”. The
essential two questions are “<em>what is the key message a plot
transfers to readers?</em>” and “<em>what is the major graphical element
in the plot?</em>”. In the barplot or dotplot, the major graphical
element which readers may notice the easiest is the height of bars or
the offset of dots to the origin. The most important message of the ORA
analysis is of course “the enrichment”. The two examples from
<code>barplot()</code> and <code>dotplot()</code> actually fail to
transfer such information to readers. In the first barplot where
<code>"Count"</code> is used as values on x-axis, the numer of DE genes
in gene sets is not a good measure of the enrichment because it has a
positive relation to the size of gene sets. A high value of
<code>"Count"</code> does not mean the gene set is more enriched.</p>
<p>It is the same reason for dotplot where <code>"GeneRatio"</code> is
used as values on x-axis. Gene ratio is calculated as the fraction of DE
genes from a certain gene set (GeneRatio = Count/Total_DE_Genes). The
dotplot puts multiple gene sets in the same plot and the aim is to
compare between gene sets, thus gene sets should be “scaled” to make
them comparable. <code>"GeneRatio"</code> is not scaled for different
gene sets and it still has a positive relation to the gene set size,
which can be observed in the dotplot where higher the gene ratio, larger
the dot size. Actually “GeneRatio” has the same effect as “Count”
(GeneRatio = Count/Total_DE_Genes), so as has been explained in the
previous paragraph, <code>"GeneRatio"</code> is not a good measure for
enrichment either.</p>
<p>Now let’s try to make a more reasonable barplot and dotplot to show
the enrichment of ORA.</p>
<p>First, let’s define some metrics which measure the “enrichment” of DE
genes on gene sets. Recall the denotations in the 2x2 contingency table
(we are too far from that!). Let’s take these numbers from the
enrichment table.</p>
<div class="codewrapper sourceCode" id="cb100">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fl">983</span>  <span class="co"># length(intersect(resTimeGO@gene, resTimeGO@universe))</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">28943</span>  <span class="co"># length(resTimeGO@universe)</span></span></code></pre>
</div>
<p>Instead of using <code>GeneRatio</code>, we use the fraction of DE
genes in the gene sets which are kind of like a “scaled” value for all
gene sets. Let’s calculate it:</p>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">DE_Ratio</span> <span class="op">=</span> <span class="va">n_11</span><span class="op">/</span><span class="va">n_01</span></span>
<span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">GS_size</span> <span class="op">=</span> <span class="va">n_01</span>  <span class="co"># size of gene sets</span></span></code></pre>
</div>
<p>Then intuitively, if a gene set has a higher <code>DE_Ratio</code>
value, we could say DE genes have a higher enrichment<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a> in it.</p>
<p>We can measure the enrichment in two other ways. First, the log2 fold
enrichment, defined as:</p>
<p><span class="math display">\[ \log_2(\mathrm{Fold\_enrichment}) =
\frac{n_{11}/n_{10}}{n_{01}/n} = \frac{n_{11}/n_{01}}{n_{10}/n} =
\frac{n_{11}n}{n_{10}n_{01}} \]</span></p>
<p>which is the log2 of the ratio of <em>DE% in the gene set</em> and
<em>DE% in the universe</em> or the log2 of the ratio of <em>gene_set%
in the DE genes</em> and <em>gene_set% in the universe</em>. The two are
identical.</p>
<div class="codewrapper sourceCode" id="cb102">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<p>Second, it is also common to use <em>z</em>-score which is</p>
<p><span class="math display">\[ z = \frac{n_{11} - \mu}{\sigma}
\]</span></p>
<p>where <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span> are <a href="https://en.wikipedia.org/wiki/Hypergeometric_distribution" class="external-link">the
mean and standard deviation of the hypergeometric distribution</a>. They
can be calculated as:</p>
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hyper_mean</span> <span class="op">=</span> <span class="va">n_01</span><span class="op">*</span><span class="va">n_10</span><span class="op">/</span><span class="va">n</span></span>
<span></span>
<span><span class="va">n_02</span> <span class="op">=</span> <span class="va">n</span> <span class="op">-</span> <span class="va">n_01</span></span>
<span><span class="va">n_20</span> <span class="op">=</span> <span class="va">n</span> <span class="op">-</span> <span class="va">n_10</span></span>
<span><span class="va">hyper_var</span> <span class="op">=</span> <span class="va">n_01</span><span class="op">*</span><span class="va">n_10</span><span class="op">/</span><span class="va">n</span> <span class="op">*</span> <span class="va">n_20</span><span class="op">*</span><span class="va">n_02</span><span class="op">/</span><span class="va">n</span><span class="op">/</span><span class="op">(</span><span class="va">n</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">zScore</span> <span class="op">=</span> <span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="va">hyper_mean</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">hyper_var</span><span class="op">)</span></span></code></pre>
</div>
<p>We will use log2 fold change as the primary variable to map to bar
heights and <code>DE_Ratio</code> as the secondary variable to map to
colors. This can be done directly with the <strong>ggplot2</strong>
package. We also add the adjusted <em>p</em>-values as labels on the
bars.</p>
<p>In <code>resTimeGOTable</code>, gene sets are already ordered by the
significance, so we take the first 10 gene sets which are the 10 most
significant gene sets.</p>
<div class="codewrapper sourceCode" id="cb104">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>, </span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            fill <span class="op">=</span> <span class="va">DE_Ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, </span>
<span>        label <span class="op">=</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"%.2e"</span>, <span class="va">p.adjust</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-plot-enrichment-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the next example, we use <em>z</em>-score as the primary variable
to map to the offset to origin, <code>DE_Ratio</code> and
<code>Count</code> to map to dot colors and sizes.</p>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>, </span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">zScore</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            col <span class="op">=</span> <span class="va">DE_Ratio</span>, size <span class="op">=</span> <span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-plot-z-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Both plots can highlight the gene set “leukocyte migration involved
in inflammatory response” is relatively small but highly enriched.</p>
<p>Another useful visualization is the volcano plot. You may be aware of
in differential expression analysis, in the volcano plot, x-axis
corresponds to log2 fold changes of the differential expression, and
y-axis corresponds to the adjusted <em>p</em>-values. It is actually
similar here where we use log2 fold enrichment on x-axis.</p>
<p>Since we only look at the over-representation, the volcano plot is
one-sided. We can set two cutoffs on the log2 fold enrichment and
adjusted <em>p</em>-values, then the gene sets on the top right region
can be thought as being both statistically significant and also
biologically sensible.</p>
<div class="codewrapper sourceCode" id="cb106">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span>, </span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">p.adjust</span><span class="op">)</span>, </span>
<span>        color <span class="op">=</span> <span class="va">DE_Ratio</span>, size <span class="op">=</span> <span class="va">GS_size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"#444444"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1.5</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"#444444"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-plot-enrichment-padj-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the “volcano plot”, we can observe the plot is composed by a list
of curves. The trends are especially clear in the right bottom of the
plot. Actually each “curve” corresponds to a same <code>"Count"</code>
value (number of DE genes in a gene set). The volcano plot shows very
clearly that the enrichment has a positive relation to the gene set size
where a large gene set can easily reach a small <em>p</em>-value with a
small DE_ratio and a small log2 fold enrichment, while a small gene set
needs to have a large DE ratio to be significant.</p>
<p>It is also common that we perform ORA analysis on up-regulated genes
and down-regulated separately. And we want to combine the significant
gene sets from the two ORA analysis in one plot. In the following code,
we first generate two enrichment tables for up-regulated genes and
down-regulated separately.</p>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># up-regulated genes</span></span>
<span><span class="va">timeDEup</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEupGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDEup</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGOup</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEupGenes</span>, </span>
<span>                       keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                       ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                       OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                       universe <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                       pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOupTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">@</span><span class="va">gene</span>, <span class="va">resTimeGOup</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span></span>
<span><span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># down-regulated genes</span></span>
<span><span class="va">timeDEdown</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEdownGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDEdown</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGOdown</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEdownGenes</span>, </span>
<span>                       keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                       ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                       OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                       universe <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                       pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOdownTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">gene</span>, <span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span></span>
<span><span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<p>As an example, let’s simply take the first 5 most significant terms
for up-regulated genes and the first 5 most significant terms for
down-regulated genes. The following <strong>ggplot2</strong> code should
be easy to read.</p>
<div class="codewrapper sourceCode" id="cb108">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The name of the 3rd term is too long, we wrap it into two lines.</span></span>
<span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">3</span>, <span class="st">"Description"</span><span class="op">]</span> <span class="op">=</span> <span class="fu">paste</span><span class="op">(</span><span class="fu">strwrap</span><span class="op">(</span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">3</span>, <span class="st">"Description"</span><span class="op">]</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">direction</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"up"</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="st">"down"</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>,</span>
<span>             <span class="va">resTimeGOdownTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            fill <span class="op">=</span> <span class="va">direction</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"up"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"down"</span> <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, </span>
<span>        label <span class="op">=</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"%.2e"</span>, <span class="va">p.adjust</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/07-gene-set-analysis-rendered-plot-up-down-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Specifically for GO enrichment, it is often that GO enrichment
returns a long list of significant GO terms (e.g. several hundreds).
This makes it difficult to summarize the common functions from the long
list. The last package we will introduce is the
<strong>simplifyEnrichment</strong> package which partitions GO terms
into clusters based on their semantic similarity<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a> and summarizes their
common functions via word clouds.</p>
<p>The input of the <code>simplifyGO()</code> function is a vector of GO
IDs. It is recommended to have at least 100 GO IDs for summarization and
visualization.</p>
<div class="codewrapper sourceCode" id="cb109">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GO_ID</span> <span class="op">=</span> <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">ID</span><span class="op">[</span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">simplifyEnrichment</span><span class="op">)</span></span>
<span><span class="fu">simplifyGO</span><span class="op">(</span><span class="va">GO_ID</span><span class="op">)</span></span></code></pre>
</div>
<!-- Carpentry CI cannot install RcppParallel which is indirectly depended by simplifyenrichment,
    so the plot is not generated on the fly.
 -->
<figure><img src="../fig/simplifyEnrichment.png" alt="simplifyEnrichment" width="600" class="figure mx-auto d-block"></figure><div id="further-reading-5" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="further-reading-5" class="callout-inner">
<h3 class="callout-title">Further reading</h3>
<div class="callout-content">
<p>ORA analysis actually applies a binary conversion on genes where
genes pass the cutoff are set as 1 (DE gene) and others are set as 0
(non-DE gene). This binary transformation over-simplifies the problem
and a lot of information are lost. There is second class of gene set
enrichment analysis methods which takes the continuous gene-level score
as input and weights the importance of a gene in the gene set. Please
refer to <a href="https://www.pnas.org/doi/10.1073/pnas.0506580102" class="external-link">Subramanian et.
al. Gene set enrichment analysis: A knowledge-based approach for
interpreting genome-wide expression profiles, PNAS 2005</a> for more
information.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>ORA analysis is based on the gene counts and it is based on Fisher’s
exact test or the hypergeometric distribution.</li>
<li>In R, it is easy to obtain gene sets from a large number of
sources.</li>
</ul>
</div>
</div>
</div>
<script>
$(function() {

    // if it is a general "callout", we need to filter the callout title
    // value for callout_title can be partial from the complete title
    var callout_title = "reading";  // short for "further reading"

    // then we select those callouts with the specified title
    var callout = $(".callout").filter(function(index) {
        var title = $(this).find(".callout-title").text();
        return title.indexOf(callout_title) >= 0;
    });
    callout.css("border-color", "#2b8cbe"); // select a color, this is for the left border of the box
    callout.find(".callout-square").css("background", "#2b8cbe"); // select a color, this is for the left border of the box

    // if you want to replace the default icon, go to https://feathericons.com/, and get the source of the svg icon
    var svg = '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-book-open callout-icon"><path d="M2 3h6a4 4 0 0 1 4 4v14a3 3 0 0 0-3-3H2z"></path><path d="M22 3h-6a4 4 0 0 0-4 4v14a3 3 0 0 1 3-3h7z"></path></svg>';
    // then update it with the new icon
    callout.find(".callout-square").html(svg);

});
</script><!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 --></section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Genes must be unique in each vector.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Also note <code>phyper()</code> can be vectorized.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>The definition is from Wikipedia: <a href="https://en.wikipedia.org/wiki/Biological_pathway" class="external-link uri">https://en.wikipedia.org/wiki/Biological_pathway</a>.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>This is actually not true. Indeed
<code>as.data.frame(resTimeGO)</code> only returns the significant GO
terms, but the complete enrichment table is still stored in
<code>resTimeGO@result</code>. However, directly retrieving the slot of
an S4 object is highly unrecommended.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>You can also use <code>select()</code> function:
<code>select(org.Mm.eg.db, keys = timeDEgenes, keytype = "SYMBOL", column = "ENTREZID")</code><a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>If here the term “enrichment” does mean statistically.<a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p>The semantic similarity between GO terms considers the
topological relations in the GO hierarchy.<a href="#fnref7" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-08-next-steps"><p>Content from <a href="08-next-steps.html">Next steps</a></p>
<hr>
<p>Last updated on 2025-11-13 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/08-next-steps.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 20 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to go further from here?</li>
<li>What other types of analyses can be done with RNA-seq data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Get an overview of usages of RNA-seq data that are not covered in
this workshop.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="transcript-level-analyses">Transcript-level analyses<a class="anchor" aria-label="anchor" href="#transcript-level-analyses"></a>
</h2>
<hr class="half-width">
<p>The analyses covered in this workshop all assumed that we have
generated a matrix with read counts for each <em>gene</em>. Some
questions, however, require expression estimates on a more fine-grained
level, typically individual transcript isoforms. This is the case, for
example, if we would like to look for differences in expression of
individual isoforms, or changes in splicing patterns induced by a
specific treatment. As already mentioned in episode 1, some
quantification tools (such as Salmon, kallisto and RSEM) do in fact
estimate abundances on the transcript level. To perform transcript-level
differential expression analysis, these estimates would be used
directly, without aggregating them on the gene level. Some caution is
warranted, however, as expression estimates for individual transcripts
are often more noisy than after aggregation to the gene level. This is
due to the high similarity often observed between different isoforms of
a gene, which leads to higher uncertainty when mapping reads to
transcripts. For this reason, several approaches have been developed
specifically for performing differential expression analysis on the
transcript level <span class="citation">[@Zhu2019-swish,
@Baldoni2023-catchsalmon]</span>.</p>
<p>Analysis of differential splicing, or differential transcript usage,
differs from the differential expression analyses mentioned so far as it
is concerned with changes in the <em>relative</em> abundances of the
transcripts of a gene. Hence, considering the transcripts in isolation
is no longer sufficient. A good resource for learning more about
differential transcript usage with Bioconductor is provided by <span class="citation">@Love2018-swimming</span>.</p>
</section><section><h2 class="section-heading" id="de-novo-transcript-assembly">De novo transcript assembly<a class="anchor" aria-label="anchor" href="#de-novo-transcript-assembly"></a>
</h2>
<hr class="half-width">
<p>In this lesson, we have assumed that we are working with a
well-annotated transcriptome, that is, that we know the sequence of all
expressed isoforms and how they are grouped together into genes.
Depending on the organism or disease type we are interested in, this may
or may not be a reasonable assumption. In situations where we do not
believe that the annotated transcriptome is complete enough, we can use
the RNA-seq data to <em>assemble</em> transcripts and create a custom
annotation. This assembly can be performed either guided by a genome
sequence (if one is available), or completely de novo. It should be
noted that transcript assembly is a challenging task, which requires a
deeply sequenced library to get the best results. In addition, data from
more recent long-read sequencing technologies can be very helpful, as
they are in principle able to sequence entire transcript molecules, thus
circumventing the need for assembly. Methods for transcript assembly
represent an active area of research. Recent review are provided e.g. by
<span class="citation">@Raghavan2022-denovo</span> and <span class="citation">@Amarasinghe2020-longread</span>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>RNA-seq data is very versatile and can be used for a number of
different purposes. It is important, however, to carefully plan one’s
analyses, to make sure that enough data is available and that abundances
for appropriate features (e.g., genes, transcripts, or exons) are
quantified.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2025-11-13",
  "datePublished": "2025-11-13"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

