<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>RNA-seq in practice: Hands-on workshop on RCAC systems: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/rcac/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/rcac/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/rcac/favicon-16x16.png">
<link rel="manifest" href="../favicons/rcac/site.webmanifest">
<link rel="mask-icon" href="../favicons/rcac/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav rcac"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="rcac-logo" alt="rcac" src="../assets/images/rcac-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav rcac" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="rcac" src="../assets/images/rcac-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      RNA-seq in practice: Hands-on workshop on RCAC systems
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            RNA-seq in practice: Hands-on workshop on RCAC systems
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq in practice: Hands-on workshop on RCAC systems
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress rcac">
    <div class="progress-bar rcac" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-downloading-and-organizing-files.html">2. Downloading and organizing files</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-qc-and-trimming.html">3. Quality control of RNA-seq reads</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04a-genome-based-quantification.html">4. A. Genome-based quantification (STAR + featureCounts)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04b-transcript-based-quantification.html">5. B. Transcript-based quantification (Salmon)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-deseq2-expression-analyses.html">6. Gene-level QC and differential expression (DESeq2)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05b-deseq2-expression-analyses-kallisto.html">7. Differential expression using DESeq2</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-gene-set-enrichment-analyses.html">8. Gene set enrichment analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-intro-to-rnaseq"><p>Content from <a href="01-intro-to-rnaseq.html">Introduction to RNA-seq</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/01-intro-to-rnaseq.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 100 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the different choices to consider when planning an RNA-seq
experiment?</li>
<li>How does one process the raw fastq files to generate a table with
read counts per gene and sample?</li>
<li>Where does one find information about annotated genes for a given
organism?</li>
<li>What are the typical steps in an RNA-seq analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what RNA-seq is.</li>
<li>Describe some of the most common design choices that have to be made
before running an RNA-seq experiment.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
<li>Show some common types of results and visualizations generated in
RNA-seq analyses.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-are-we-measuring-in-an-rna-seq-experiment">What are we measuring in an RNA-seq experiment?<a class="anchor" aria-label="anchor" href="#what-are-we-measuring-in-an-rna-seq-experiment"></a>
</h2>
<hr class="half-width">
<figure><img src="../fig/alternative_splicing_sub.png" alt="Illustration of part of the central dogma of molecular biology, where DNA is transcribed to RNA, and intronic sequences are spliced out" class="figure mx-auto d-block"></figure><p>In order to produce an RNA molecule, a stretch of DNA is first
transcribed into mRNA. Subsequently, intronic regions are spliced out,
and exonic regions are combined into different <em>isoforms</em> of a
gene.</p>
<figure><img src="../fig/rnaseq_martin_wang.png" alt="Illustration of the major experimental steps of an RNA-seq experiment" class="figure mx-auto d-block"></figure><p>(figure adapted from <a href="https://www.nature.com/articles/nrg3068" class="external-link">Martin &amp; Wang
(2011)</a>).</p>
<p>In a typical RNA-seq experiment, RNA molecules are first collected
from a sample of interest. After a potential enrichment for molecules
with polyA tails (predominantly mRNA), or depletion of otherwise highly
abundant ribosomal RNA, the remaining molecules are fragmented into
smaller pieces (there are also long-read protocols where entire
molecules are considered, but those are not the focus of this lesson).
It is important to keep in mind that because of the splicing excluding
intronic sequences, an RNA molecule (and hence a generated fragment) may
not correspond to an uninterrupted region of the genome. The RNA
fragments are then reverse transcribed into cDNA, whereafter sequencing
adapters are added to each end. These adapters allow the fragments to
attach to the flow cell. Once attached, each fragment will be heavily
amplified to generate a cluster of identical sequences on the flow cell.
The sequencer then determines the sequence of the first 50-200
nucleotides of the cDNA fragments in each such cluster, starting from
one end, which corresponds to a <em>read</em>. Many data sets are
generated with so called paired-end protocols, in which the fragments
are read from both ends. Millions of such reads (or pairs of reads) will
be generated in an experiment, and these will be represented in a (pair
of) FASTQ files. Each read is represented by four consecutive lines in
such a file: first a line with a unique read identifier, next the
inferred sequence of the read, then another identifier line, and finally
a line containing the base quality for each inferred nucleotide,
representing the probability that the nucleotide in the corresponding
position has been correctly identified.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>What are potential advantages and disadvantages of paired-end
protocols compared to only sequencing one end of a fragment?</li>
<li>What quality assessment can you think of that would be useful to
perform on the FASTQ files with read sequences?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="experimental-design-considerations">Experimental design considerations<a class="anchor" aria-label="anchor" href="#experimental-design-considerations"></a>
</h2>
<hr class="half-width">
<p>Before starting to collect data, it is essential to take some time to
think about the design of the experiment. Experimental design concerns
the organization of experiments with the purpose of making sure that the
right type of data, and enough of it, is available to answer the
questions of interest as efficiently as possible. Aspects such as which
conditions or sample groups to consider, how many replicates to collect,
and how to plan the data collection in practice are important questions
to consider. Many high-throughput biological experiments (including
RNA-seq) are sensitive to ambient conditions, and it is often difficult
to directly compare measurements that have been done on different days,
by different analysts, in different centers, or using different batches
of reagents. For this reason, it is very important to design experiments
properly, to make it possible to disentangle different types of (primary
and secondary) effects.</p>
<figure><img src="../fig/factors-affecting-measurements.png" alt="A classification of many different factors affecting measurements obtained from an experiment into treatment, biological, technical and error effects" class="figure mx-auto d-block"></figure><p>(figure from <a href="https://www.amazon.com/Experimental-Design-Laboratory-Biologists-Reproducibility/dp/1107424887" class="external-link">Lazic
(2017)</a>).</p>
<div id="challenge-discuss-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss with your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Why is it essential to have replicates?</li>
</ol>
</div>
</div>
</div>
<p>Importantly, not all replicates are equally useful, from a
statistical point of view. One common way to classify the different
types of replicates is as ‘biological’ and ‘technical’, where the latter
are typically used to test the reproducibility of the measurement
device, while biological replicates inform about the variability between
different samples from a population of interest. Another scheme
classifies replicates (or units) into ‘biological’, ‘experimental’ and
‘observational’. Here, biological units are entities we want to make
inferences about (e.g., animals, persons). Replication of biological
units is required to make a general statement of the effect of a
treatment - we can not draw a conclusion about the effect of drug on a
population of mice by studying a single mouse only. Experimental units
are the smallest entities that can be <em>independently assigned to a
treatment</em> (e.g., animal, litter, cage, well). Only replication of
experimental units constitute true replication. Observational units,
finally, are entities at which measurements are made.</p>
<p>To explore the impact of experimental design on the ability to answer
questions of interest, we are going to use an interactive application,
provided in the <a href="https://csoneson.github.io/ConfoundingExplorer/" class="external-link">ConfoundingExplorer</a>
package.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Launch the ConfoundingExplorer application and familiarize yourself
with the interface.</p>
</div>
</div>
</div>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>For a balanced design (equal distribution of replicates between the
two groups in each batch), what is the effect of increasing the strength
of the batch effect? Does it matter whether one adjusts for the batch
effect or not?</li>
<li>For an increasingly unbalanced design (most or all replicates of one
group coming from one batch), what is the effect of increasing the
strength of the batch effect? Does it matter whether one adjusts for the
batch effect or not?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="rna-seq-quantification-from-reads-to-count-matrix">RNA-seq quantification: from reads to count matrix<a class="anchor" aria-label="anchor" href="#rna-seq-quantification-from-reads-to-count-matrix"></a>
</h2>
<hr class="half-width">
<figure><img src="../fig/rna-quant.jpeg" alt="Illustration of a set of reads generated by a sequencer, and genomic and transcriptomic reference sequences" class="figure mx-auto d-block"></figure><p>The read sequences contained in the FASTQ files from the sequencer
are typically not directly useful as they are, since we do not have the
information about which gene or transcript they originate from. Thus,
the first processing step is to attempt to identify the location of
origin for each read, and use this to obtain an estimate of the number
of reads originating from a gene (or another features, such as an
individual transcript). This can then be used as a proxy for the
abundance, or expression level, of the gene. There is a plethora of RNA
quantification pipelines, and the most common approaches can be
categorized into three main types:</p>
<ol style="list-style-type: decimal">
<li><p>Align reads to the genome, and count the number of reads that map
within the exons of each gene. This is the one of simplest methods. For
species for which the transcriptome is poorly annotated, this would be
the preferred approach. Example: <code>STAR</code> alignment to GRCm39 +
<code>Rsubread</code> <a href="https://doi.org/10.1093%2Fnar%2Fgkz114" class="external-link">featureCounts</a></p></li>
<li><p>Align reads to the transcriptome, quantify transcript expression,
and summarize transcript expression into gene expression. This approach
can produce accurate quantification results (<a href="https://doi.org/10.1186/s13059-016-0940-1" class="external-link">independent
benchmarking</a>), particularly for high-quality samples without DNA
contamination. Example: RSEM quantification using
<code>rsem-calculate-expression --star</code> on the GENCODE GRCh38
transcriptome + <code>tximport</code></p></li>
<li><p>Pseudoalign reads against the transcriptome, using the
corresponding genome as a decoy, quantifying transcript expression in
the process, and summarize the transcript-level expression into
gene-level expression. The advantages of this approach include:
computational efficiency, mitigation of the effect of DNA contamination,
and GC bias correction. Example: <code>salmon quant --gcBias</code> +
<code>tximport</code></p></li>
</ol>
<p>At typical sequencing read depth, gene expression quantification is
often more accurate than transcript expression quantification. However,
differential gene expression analyses can be <a href="https://doi.org/10.12688/f1000research.7563.1" class="external-link">improved</a> by
having access also to transcript-level quantifications.</p>
<p>Other tools used in RNA-seq quantification include: TopHat2, bowtie2,
kallisto, HTseq, among many others.</p>
<p>The choice of an appropriate RNA-seq quantification depends on the
quality of the transcriptome annotation, the quality of the RNA-seq
library preparation, the presence of contaminating sequences, among many
factors. Often, it can be informative to compare the quantification
results of multiple approaches.</p>
<p>The quantification method is species- and experiment-dependent, and
often requires large amounts of computing resources, so we will give you
a overview of how to perform quantification, and scripts to run the
quantification. We will use pre-computed quantifications for the
exercises in this workshop.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with
your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Which of the mentioned RNA-Seq quantification tools have you heard
about? Do you know other pros and cons of the methods?</li>
<li>Have you done your own RNA-Seq experiment? If so, what
quantification tool did you use and why did you choose it?</li>
<li>Do you have access to specific tools / local bioinformatics expert /
computational resources for quantification? If you don’t, how might you
gain access?</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="finding-the-reference-sequences">Finding the reference sequences<a class="anchor" aria-label="anchor" href="#finding-the-reference-sequences"></a>
</h2>
<hr class="half-width">
<p>In order to quantify abundances of known genes or transcripts from
RNA-seq data, we need a reference database informing us of the sequence
of these features, to which we can then compare our reads. This
information can be obtained from different online repositories. It is
highly recommended to choose one of these for a particular project, and
not mix information from different sources. Depending on the
quantification tool you have chosen, you will need different types of
reference information. If you are aligning your reads to the genome and
investigating the overlap with known annotated features, you will need
the full genome sequence (provided in a fasta file) and a file that
tells you the genomic location of each annotated feature (typically
provided in a gtf file). If you are mapping your reads to the
transcriptome, you will instead need a file with the sequence of each
transcript (again, a fasta file).</p>
<ul>
<li>If you are working with mouse or human samples, the <a href="https://www.gencodegenes.org/" class="external-link">GENCODE project</a> provides
well-curated reference files.</li>
<li>
<a href="https://www.ensembl.org/info/data/ftp/index.html" class="external-link">Ensembl</a>
provides reference files for a large set of organisms, including <a href="https://plants.ensembl.org/info/data/ftp/index.html" class="external-link">plants</a>
and <a href="https://fungi.ensembl.org/info/data/ftp/index.html" class="external-link">fungi</a>.</li>
<li>
<a href="https://hgdownload.soe.ucsc.edu/downloads.html" class="external-link">UCSC</a>
also provides reference files for many organisms.</li>
</ul>
<div id="challenge-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Download the latest mouse transcriptome fasta file from GENCODE. What
do the entries look like? Tip: to read the file into R, consider the
<code>readDNAStringSet()</code> function from the
<code>Biostrings</code> package.</p>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="where-are-we-heading-towards-in-this-workshop">Where are we heading towards in this workshop?<a class="anchor" aria-label="anchor" href="#where-are-we-heading-towards-in-this-workshop"></a>
</h1>
<p>In this workshop, we will discuss and practice: downloading publicly
available RNA-seq data, performing quality assessment of the raw data,
quantifying gene expression, and performing differential expression
analysis to identify genes that are expressed at different levels
between two or more conditions. We will also cover some aspects of data
visualization and interpretation of the results.</p>
<p>The outcome of a differential expression analysis is often
represented using graphical representations, such as MA plots and
heatmaps (see below for examples).</p>
<p><img src="../fig/maPlot-example.png" alt="An example MA plot" class="figure"><img src="../fig/heatmap-example.png" alt="An example heatmap" class="figure"></p>
<p>In the following episodes we will learn, among other things, how to
generate and interpret these plots. It is also common to perform
follow-up analyses to investigate whether there is a functional
relationship among the top-ranked genes, so called gene set (enrichment)
analysis, which will also be covered in a later episode.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>RNA-seq is a technique of measuring the amount of RNA expressed
within a cell/tissue and state at a given time.</li>
<li>Many choices have to be made when planning an RNA-seq experiment,
such as whether to perform poly-A selection or ribosomal depletion,
whether to apply a stranded or an unstranded protocol, and whether to
sequence the reads in a single-end or paired-end fashion. Each of the
choices have consequences for the processing and interpretation of the
data.</li>
<li>Many approaches exist for quantification of RNA-seq data. Some
methods align reads to the genome and count the number of reads
overlapping gene loci. Other methods map reads to the transcriptome and
use a probabilistic approach to estimate the abundance of each gene or
transcript.</li>
<li>Information about annotated genes can be accessed via several
sources, including Ensembl, UCSC and GENCODE.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div></section><section id="aio-02-downloading-and-organizing-files"><p>Content from <a href="02-downloading-and-organizing-files.html">Downloading and organizing files</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/02-downloading-and-organizing-files.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What files are required to process raw RNA seq reads?</li>
<li>Where do we obtain raw reads, reference genomes, annotations, and
transcript sequences?</li>
<li>How should project directories be organized to support a smooth
workflow?</li>
<li>What practical considerations matter when downloading and preparing
RNA seq data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Identify the essential inputs for RNA seq analysis: FASTQ files,
reference genome, annotation, and transcript sequences.</li>
<li>Learn where and how to download RNA seq datasets and reference
materials.</li>
<li>Create a clean project directory structure suitable for quality
control, alignment, and quantification.</li>
<li>Understand practical issues such as compressed FASTQ files and
adapter trimming.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Before we can perform quality control, mapping, or quantification, we
must first gather the files required for RNA seq analysis and organize
them in a reproducible way. This episode introduces the essential inputs
of an RNA seq workflow and demonstrates how to obtain and structure them
on a computing system.</p>
<p>We begin by discussing raw FASTQ reads, reference genomes, annotation
files, and transcript sequences. We then download the dataset used
throughout the workshop and set up the directory structure that supports
downstream steps.</p>
<p>Most FASTQ files arrive compressed as <code>.fastq.gz</code> files.
Modern RNA seq tools accept compressed files directly, so uncompressing
them is usually unnecessary.</p>
<div id="should-i-uncompress-fastq-files" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-uncompress-fastq-files" class="callout-inner">
<h3 class="callout-title">Should I uncompress FASTQ files?</h3>
<div class="callout-content">
<p>In almost all RNA seq workflows, you should keep FASTQ files
compressed. Tools such as FastQC, HISAT2, STAR, salmon, and fastp can
read <code>.gz</code> files directly. Keeping files compressed saves
space and reduces input and output overhead.</p>
</div>
</div>
</div>
<p>Many library preparation protocols introduce adapter sequences at the
ends of reads. For most alignment based analyses, explicit trimming is
not required because aligners soft clip adapter sequences
automatically.</p>
<div id="should-i-trim-adapters-for-rna-seq-analysis" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="should-i-trim-adapters-for-rna-seq-analysis" class="callout-inner">
<h3 class="callout-title">Should I trim adapters for RNA seq
analysis?</h3>
<div class="callout-content">
<p>In most cases, no. Aligners such as HISAT2 and STAR soft clip adapter
sequences. Overly aggressive trimming can reduce mapping rates or
distort read length distributions. Trimming is needed only for specific
applications such as transcript assembly where uniform read lengths are
important.<br>
More information: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/" class="external-link uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/</a></p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="downloading-the-data-and-project-organization">Downloading the data and project organization<a class="anchor" aria-label="anchor" href="#downloading-the-data-and-project-organization"></a>
</h2>
<hr class="half-width">
<p>For this workshop, we use a publicly available dataset that
investigates the effects of cardiomyocyte specific Nat10 knockout on
heart development in mice (BioProject PRJNA1254208, GEO GSE295332). The
study includes three biological replicates per group, sequenced on an
Illumina NovaSeq X Plus in paired end mode.</p>
<p>Before downloading files, we create a reproducible directory layout
for raw data, scripts, mapping outputs, and count files.</p>
<div id="create-the-directories-needed-for-this-episode" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="create-the-directories-needed-for-this-episode" class="callout-inner">
<h3 class="callout-title">Create the directories needed for this
episode</h3>
<div class="callout-content">
<p>Create a working directory for the workshop (e.g.,
<code>rnaseq-workshop</code>). Inside it, create four
subdirectories:</p>
<ul>
<li>
<code>data</code> : raw FASTQ files, reference genome,
annotation<br>
</li>
<li>
<code>scripts</code> : custom scripts, SLURM job files<br>
</li>
<li>
<code>results</code> : alignment, counts and various other
outputs</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RCAC_SCRATCH</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> rnaseq-workshop/<span class="dt">{data</span><span class="op">,</span><span class="dt">scripts</span><span class="op">,</span><span class="dt">results}</span></span></code></pre>
</div>
<p>The resulting directory structure should look like this:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">aseethar@scholar-fe03:[aseethar]</span> $ pwd</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">/scratch/scholar/aseethar</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">aseethar@scholar-fe03:[aseethar]</span> $ tree rnaseq-workshop/</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">rnaseq-workshop/</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">├──</span> results</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">├──</span> data</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="ex">└──</span> scripts</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="what-files-do-i-need-for-rna-seq-analysis">What files do I need for RNA-seq analysis?<a class="anchor" aria-label="anchor" href="#what-files-do-i-need-for-rna-seq-analysis"></a>
</h2>
<hr class="half-width">
<div id="prereq3" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div class="callout-inner">
<div class="callout-content">
<p>You will need:</p>
<ol style="list-style-type: decimal">
<li>Raw reads (FASTQ)
<ul>
<li>Usually <code>.fastq.gz</code> files containing nucleotide sequences
and per-base quality scores.</li>
</ul>
</li>
</ol>
<p><strong><em>For alignment based workflows</em></strong></p>
<ol start="2" style="list-style-type: decimal">
<li>Reference genome (FASTA)
<ul>
<li>Contains chromosome or contig sequences.</li>
</ul>
</li>
<li>Annotation file (GTF/GFF)
<ul>
<li>Describes gene models: exons, transcripts, coordinates.</li>
</ul>
</li>
</ol>
<p><strong><em>For transcript-level quantification
workflows</em></strong></p>
<ol start="4" style="list-style-type: decimal">
<li>Transcript sequences (FASTA)
<ul>
<li>Required for transcript based quantification with tools such as
salmon, kallisto, and RSEM.</li>
</ul>
</li>
</ol>
<p>Reference files must be consistent with each other (same genome
version). To ensure a smooth analysis workflow, keep these files
logically organized. Good directory structure minimizes mistakes,
simplifies scripting, and supports reproducibility.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="downloading-the-data">Downloading the data<a class="anchor" aria-label="anchor" href="#downloading-the-data"></a>
</h2>
<hr class="half-width">
<p>We now download:</p>
<ul>
<li>the GRCm39 reference genome<br>
</li>
<li>the corresponding GENCODE annotation</li>
<li>transcript sequences (if using transcript based quantification)</li>
<li>raw FASTQ files from SRA</li>
</ul>
<div class="section level3">
<h3 id="reference-genome-and-annotation">Reference genome and annotation<a class="anchor" aria-label="anchor" href="#reference-genome-and-annotation"></a>
</h3>
<p>We will use <strong>GENCODE GRCm39 (M38)</strong> as the
reference.</p>
<div id="annotation" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="annotation" class="callout-inner">
<h3 class="callout-title">Annotation</h3>
<div class="callout-content">
<p>Visit the GENCODE mouse page: <a href="https://www.gencodegenes.org/mouse/" class="external-link uri">https://www.gencodegenes.org/mouse/</a>.<br>
Which annotation file (GTF/GFF3 files section) should you download, and
why?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Use the <strong>basic gene annotation in GTF format</strong> (e.g.,
<code>gencode.vM38.primary_assembly.basic.annotation.gtf.gz</code>).<br>
It contains essential gene and transcript models without pseudogenes or
other biotypes not typically used for RNA-seq quantification.</p>
<figure style="text-align: center"><img src="../fig/01_fq2counts/gencode-gtf.png" alt="Basic gene annotation in GTF format" class="figure mx-auto d-block"><figcaption>
Basic gene annotation in GTF format
</figcaption></figure>
</div>
</div>
</div>
</div>
<div id="reference-genome" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="reference-genome" class="callout-inner">
<h3 class="callout-title">Reference genome</h3>
<div class="callout-content">
<p>Select the corresponding <strong>Genome sequence (GRCm39)
FASTA</strong> (e.g.,
<code>GRCm39.primary_assembly.genome.fa.gz</code>).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<figure style="text-align: center"><img src="../fig/01_fq2counts/gencode-genome.png" alt="Reference genome file in FASTA format" class="figure mx-auto d-block"><figcaption>
Reference genome file in FASTA format
</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="transcript-sequences">Transcript sequences<a class="anchor" aria-label="anchor" href="#transcript-sequences"></a>
</h3>
<p>Transcript level quantification requires a transcript FASTA file
consistent with the annotation. GENCODE provides transcript sequences
for all transcripts in GRCm39
(<code>gencode.vM38.transcripts.fa.gz</code>).</p>
<div id="transcript-sequence-file" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="transcript-sequence-file" class="callout-inner">
<h3 class="callout-title">Transcript sequence file</h3>
<div class="callout-content">
<p>Where on the GENCODE page can you find the transcript FASTA file?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>The transcript FASTA file is available in the same release directory
as the genome and annotation files (under FASTA files). It is named
<code>gencode.vM38.transcripts.fa.gz</code>.</p>
<figure style="text-align: center"><img src="../fig/01_fq2counts/gencode-transcripts.png" alt="Transcript sequences file in FASTA format" class="figure mx-auto d-block"><figcaption>
Transcript sequences file in FASTA format
</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="downloading-all-reference-files">Downloading all reference files<a class="anchor" aria-label="anchor" href="#downloading-all-reference-files"></a>
</h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="va">GTFlink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/gencode.vM38.primary_assembly.basic.annotation.gtf.gz"</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="va">GENOMElink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/GRCm39.primary_assembly.genome.fa.gz"</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="va">CDSLink</span><span class="op">=</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M38/gencode.vM38.transcripts.fa.gz"</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Download to 'data'</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GENOMElink}</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${GTFlink}</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> data <span class="va">${CDSLink}</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co"># Uncompress</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="bu">cd</span> data</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="fu">gunzip</span> GRCm39.primary_assembly.genome.fa.gz</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="fu">gunzip</span> gencode.vM38.primary_assembly.basic.annotation.gtf.gz</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="fu">gunzip</span> gencode.vM38.transcripts.fa.gz</span></code></pre>
</div>
<p>The transcript file header has multiple pieces of information which
often interfere with downstream analysis. We create a simplified version
containing only the transcript IDs.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> gencode.vM38.transcripts.fa <span class="kw">|</span><span class="fu">head</span> <span class="at">-1</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 <span class="at">-d</span> <span class="st">"|"</span> gencode.vM38.transcripts.fa <span class="op">&gt;</span> gencode.vM38.transcripts-clean.fa</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> gencode.vM38.transcripts-clean.fa <span class="kw">|</span><span class="fu">head</span> <span class="at">-1</span></span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="raw-reads">Raw reads<a class="anchor" aria-label="anchor" href="#raw-reads"></a>
</h2>
<hr class="half-width">
<p>The experiment we use investigates how <strong>NAT10</strong>
influences <strong>mouse heart development</strong>.</p>
<blockquote>
<p><strong>Study:</strong> The effects of NAT10 on heart development in
mice<br><strong>Organism:</strong> Mus musculus<br><strong>Design:</strong> Nat10^flox/flox vs Nat10-CKO (n=3 per group),
P10 hearts<br><strong>Sequencing:</strong> Illumina NovaSeq X Plus, paired-end<br><strong>BioProject:</strong> PRJNA1254208<br><strong>GEO:</strong> GSE295332</p>
</blockquote>
<figure style="text-align: center"><img src="../fig/01_fq2counts/geo-db.png" alt="FASTQ files from GEO" class="figure mx-auto d-block"><figcaption>
FASTQ files from GEO
</figcaption></figure><div id="where-do-you-find-fastq-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="where-do-you-find-fastq-files" class="callout-inner">
<h3 class="callout-title">Where do you find FASTQ files?</h3>
<div class="callout-content">
<p>GEO provides metadata and experimental details, but FASTQ files are
stored in the <strong>SRA</strong>. How do you obtain the complete list
of sequencing runs?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>Navigate to the SRA page for the BioProject and use <strong>Run
Selector</strong> to list runs.<br>
On the Run Selector page, click <strong>Accession List</strong> to
download all SRR accession IDs.</p>
<figure style="text-align: center"><img src="../fig/01_fq2counts/srr-runs.png" alt="FASTQ files from SRA Run Selector" class="figure mx-auto d-block"><figcaption>
FASTQ files from SRA Run Selector
</figcaption></figure>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="download-fastq-files-with-sra-toolkit">Download FASTQ files with SRA Toolkit<a class="anchor" aria-label="anchor" href="#download-fastq-files-with-sra-toolkit"></a>
</h3>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop/data</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">module</span> load sra-tools</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">SRR</span><span class="kw">;</span> <span class="cf">do</span> </span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>   <span class="ex">fasterq-dump</span> <span class="at">--threads</span> 4 <span class="at">--progress</span> <span class="va">$SRR</span><span class="kw">;</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>   <span class="fu">gzip</span> <span class="va">${SRR}</span>_1.fastq</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>   <span class="fu">gzip</span> <span class="va">${SRR}</span>_2.fastq</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="cf">done</span><span class="op">&lt;</span>SRR_Acc_List.txt</span></code></pre>
</div>
<p>Since downloading may take a long time, we provide a pre-downloaded
copy of all FASTQ files. The directory, after downloading, should look
like this:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">data</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">├──</span> annot.tsv</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">├──</span> gencode.vM38.primary_assembly.basic.annotation.gtf</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="ex">├──</span> gencode.vM38.transcripts.fa</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="ex">├──</span> GRCm39.primary_assembly.genome.fa</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="ex">├──</span> mart.tsv</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_1.fastq.gz</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_2.fastq.gz</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_1.fastq.gz</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_2.fastq.gz</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_1.fastq.gz</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_2.fastq.gz</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_1.fastq.gz</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_2.fastq.gz</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_1.fastq.gz</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_2.fastq.gz</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_1.fastq.gz</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_2.fastq.gz</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="ex">└──</span> SRR_Acc_List.txt</span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>RNA-seq requires three main inputs: FASTQ reads, a reference genome
(FASTA), and gene annotation (GTF/GFF).</li>
<li>
<em>In lieu</em> of a reference genome and annotation, transcript
sequences (FASTA) can be used for transcript-level quantification.</li>
<li>Keeping files compressed and well-organized supports reproducible
analysis.</li>
<li>Reference files must match in genome build and version.</li>
<li>Public data repositories such as GEO and SRA provide raw reads and
metadata.</li>
<li>Tools like <code>wget</code> and <code>fasterq-dump</code> enable
programmatic, reproducible data retrieval.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-03-qc-and-trimming"><p>Content from <a href="03-qc-and-trimming.html">Quality control of RNA-seq reads</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/03-qc-and-trimming.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we check the quality of raw RNA-seq reads?</li>
<li>What information do FastQC and MultiQC provide?</li>
<li>How do we decide if trimming is required?</li>
<li>What QC issues are common in Illumina RNA-seq data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Run FastQC and MultiQC on RNA-seq FASTQ files.</li>
<li>Understand how to interpret the important FastQC modules.</li>
<li>Learn which FastQC warnings are expected for RNA-seq.</li>
<li>Understand why trimming is usually unnecessary for standard
RNA-seq.</li>
<li>Identify samples that may require closer inspection.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Before mapping reads to a genome, it is important to evaluate the
quality of the raw RNA-seq data.<br>
Quality control helps reveal problems such as adapter contamination, low
quality cycles, and unexpected technical issues.</p>
<p>In this episode, we will run FastQC and MultiQC on the mouse heart
dataset and interpret the results.<br>
We will also explain when trimming is useful and when it can be harmful
for alignment based RNA-seq workflows.</p>
<p>FastQC evaluates individual FASTQ files. MultiQC summarizes all
FastQC reports into one view, which is helpful for comparing
replicates.</p>
<div id="why-do-we-perform-qc-on-fastq-files" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-do-we-perform-qc-on-fastq-files" class="callout-inner">
<h3 class="callout-title">Why do we perform QC on FASTQ files?</h3>
<div class="callout-content">
<p>Quality control answers several questions:</p>
<ul>
<li>Are sequencing qualities stable across cycles?<br>
</li>
<li>Are adapters present at high levels?<br>
</li>
<li>Do different replicates show similar patterns?<br>
</li>
<li>Are there outliers that indicate failed libraries?</li>
</ul>
<p>QC provides confidence that downstream steps will be reliable.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h2>
<hr class="half-width">
<p>FastQC generates an HTML report for each FASTQ file. For paired-end
datasets, you can run it on both R1 and R2 together or separately.</p>
<p>We will create a directory for QC output and then run FastQC on all
FASTQ files.</p>
<div class="section level3">
<h3 id="run-fastqc-on-your-fastq-files">Run FastQC on your FASTQ files<a class="anchor" aria-label="anchor" href="#run-fastqc-on-your-fastq-files"></a>
</h3>
<p>Use the commands below to run FastQC on all reads in the
<code>data</code> directory.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/qc_fastq/</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">module</span> load fastqc</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="ex">fastqc</span> data/<span class="pp">*</span>.fastq.gz <span class="at">--outdir</span> results/qc_fastq/ <span class="at">--threads</span> 4</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">qc</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_1_fastqc.html</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_1_fastqc.zip</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_2_fastqc.html</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">├──</span> SRR33253285_2_fastqc.zip</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_1_fastqc.html</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_1_fastqc.zip</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_2_fastqc.html</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="ex">├──</span> SRR33253286_2_fastqc.zip</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_1_fastqc.html</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_1_fastqc.zip</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_2_fastqc.html</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="ex">├──</span> SRR33253287_2_fastqc.zip</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_1_fastqc.html</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_1_fastqc.zip</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_2_fastqc.html</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="ex">├──</span> SRR33253288_2_fastqc.zip</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_1_fastqc.html</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_1_fastqc.zip</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_2_fastqc.html</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="ex">├──</span> SRR33253289_2_fastqc.zip</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_1_fastqc.html</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_1_fastqc.zip</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="ex">├──</span> SRR33253290_2_fastqc.html</span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="ex">└──</span> SRR33253290_2_fastqc.zip</span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="understanding-the-fastqc-report">Understanding the FastQC report<a class="anchor" aria-label="anchor" href="#understanding-the-fastqc-report"></a>
</h2>
<hr class="half-width">
<p>FastQC produces several diagnostic modules. RNA-seq users often
misinterpret certain modules, so we focus on the ones that matter
most.</p>
<div id="which-fastqc-modules-are-important-for-rna-seq" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="which-fastqc-modules-are-important-for-rna-seq" class="callout-inner">
<h3 class="callout-title">Which FastQC modules are important for
RNA-seq?</h3>
<div class="callout-content">
<p>Most informative modules:</p>
<ul>
<li>Per base sequence quality</li>
<li>Per sequence quality scores</li>
<li>Per base sequence content</li>
<li>Per sequence GC content</li>
<li>Adapter content</li>
<li>Overrepresented sequences</li>
</ul>
<p>Modules that commonly show warnings for RNA-seq, even when libraries
are fine:</p>
<ul>
<li>Per sequence GC content (RNA-seq reflects transcript GC bias, not
whole genome)</li>
<li>K-mer content (poly A tails and priming artifacts often trigger
warnings)</li>
</ul>
<p>Warnings do not always indicate problems. Context and comparison
across samples matter more than individual icons.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="key-modules-explained">Key modules explained<a class="anchor" aria-label="anchor" href="#key-modules-explained"></a>
</h2>
<hr class="half-width">
<p>Below are guidelines for interpreting the most useful modules.</p>
<div class="section level3">
<h3 id="per-base-sequence-quality">Per base sequence quality<a class="anchor" aria-label="anchor" href="#per-base-sequence-quality"></a>
</h3>
<p>Shows the distribution of quality scores at each position across all
reads. You want high and stable quality across the read. A small quality
drop at the end of R2 is common and usually not a problem.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_sequence_quality_histogram.png" alt="Per base sequence quality for a typical RNA-seq dataset" class="figure mx-auto d-block"><figcaption>
Per base sequence quality for a typical RNA-seq dataset
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="per-tile-sequence-quality">Per tile sequence quality<a class="anchor" aria-label="anchor" href="#per-tile-sequence-quality"></a>
</h3>
<p>Reports whether specific regions of the flowcell produced lower
quality reads. Uneven tiles can indicate instrument issues or bubble
formation, but this module rarely shows problems in modern NovaSeq
data.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_sequence_content.png" alt="Per tile sequence quality plot" class="figure mx-auto d-block"><figcaption>
Per tile sequence quality plot
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="per-sequence-quality-scores">Per sequence quality scores<a class="anchor" aria-label="anchor" href="#per-sequence-quality-scores"></a>
</h3>
<p>Shows how many reads have high or low overall quality. A good dataset
has most reads with high scores and very few reads with low quality.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_sequence_quality_score.png" alt="Distribution of per sequence quality scores" class="figure mx-auto d-block"><figcaption>
Distribution of per sequence quality scores
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="per-base-sequence-content">Per base sequence content<a class="anchor" aria-label="anchor" href="#per-base-sequence-content"></a>
</h3>
<p>Shows the proportion of A, C, G, and T at each cycle. For genomic
resequencing, all four lines should be close to 25 percent. RNA-seq
usually shows imbalanced composition in early cycles due to biased
priming and transcript composition. Mild imbalance at the start of the
read is normal and does not require trimming.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_atgc_content.png" alt="Per base sequence content for RNA-seq reads" class="figure mx-auto d-block"><figcaption>
Per base sequence content for RNA-seq reads
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="per-sequence-gc-content">Per sequence GC content<a class="anchor" aria-label="anchor" href="#per-sequence-gc-content"></a>
</h3>
<p>Shows the GC distribution across all reads. RNA-seq often fails this
test because expressed transcripts are not GC neutral. Only pronounced
multi-modal or extremely shifted distributions are concerning, for
example when contamination from another organism is suspected.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_sequence_GC_content.png" alt="Per sequence GC content distribution" class="figure mx-auto d-block"><figcaption>
Per sequence GC content distribution
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="per-base-n-content">Per base N content<a class="anchor" aria-label="anchor" href="#per-base-n-content"></a>
</h3>
<p>Reports the percentage of unidentified bases (N) at each position.
This should be near zero for modern Illumina runs. High N content
indicates failed cycles or poor sequencing chemistry.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_per_base_n_content.png" alt="Per base N content plot" class="figure mx-auto d-block"><figcaption>
Per base N content plot
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="sequence-length-distribution">Sequence length distribution<a class="anchor" aria-label="anchor" href="#sequence-length-distribution"></a>
</h3>
<p>Shows the distribution of read lengths. Standard RNA-seq libraries
should have a narrow peak at the expected read length. Variable lengths
usually appear after trimming or for specialized protocols such as small
RNA.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_length_distribution.png" alt="Sequence length distribution for RNA-seq data" class="figure mx-auto d-block"><figcaption>
Sequence length distribution for RNA-seq data
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="sequence-duplication-levels">Sequence duplication levels<a class="anchor" aria-label="anchor" href="#sequence-duplication-levels"></a>
</h3>
<p>Reports how many reads are exact duplicates. Some duplication is
normal in RNA-seq because highly expressed genes produce many identical
fragments. Very high duplication, especially combined with low library
size, can indicate low library complexity or overamplification.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_sequence_duplication_level.png" alt="Sequence duplication levels" class="figure mx-auto d-block"><figcaption>
Sequence duplication levels
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="overrepresented-sequences">Overrepresented sequences<a class="anchor" aria-label="anchor" href="#overrepresented-sequences"></a>
</h3>
<p>Lists specific sequences that occur more often than expected. Common
sources include ribosomal RNA fragments, poly A tails, adapter sequence,
or reads from highly expressed transcripts. Some overrepresented
sequences are expected, but very high levels may indicate contamination
or poor rRNA depletion.</p>
</div>
<div class="section level3">
<h3 id="adapter-content">Adapter content<a class="anchor" aria-label="anchor" href="#adapter-content"></a>
</h3>
<p>Shows whether any read cycles contain adapter sequences. Small
amounts at the extreme ends are common and usually harmless.
Consistently high adapter content across many cycles suggests short
inserts or poorly prepared libraries.</p>
<figure style="text-align: center"><img src="../fig/02_qc/fastqc_adapter_content.png" alt="Adapter content across the read" class="figure mx-auto d-block"><figcaption>
Adapter content across the read
</figcaption></figure>
</div>
</section><section><h2 class="section-heading" id="aggregating-reports-with-multiqc">Aggregating reports with MultiQC<a class="anchor" aria-label="anchor" href="#aggregating-reports-with-multiqc"></a>
</h2>
<hr class="half-width">
<p>FastQC generates a separate HTML file per sample. MultiQC
consolidates all reports and allows easy comparison of replicates.</p>
<div class="section level3">
<h3 id="run-multiqc-on-the-fastqc-outputs">Run MultiQC on the FastQC outputs<a class="anchor" aria-label="anchor" href="#run-multiqc-on-the-fastqc-outputs"></a>
</h3>
<p>Generate a single combined report using MultiQC.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> rnaseq-workshop</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="ex">multiqc</span> results/qc_fastq/ <span class="at">-o</span> results/qc_fastq/</span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/02_qc/multiqc-summary.png" alt="Example MultiQC summary across all samples" class="figure mx-auto d-block"><figcaption>
Example MultiQC summary across all samples
</figcaption></figure>
</div>
</section><section><h2 class="section-heading" id="trimming-adapter-sequences">Trimming adapter sequences<a class="anchor" aria-label="anchor" href="#trimming-adapter-sequences"></a>
</h2>
<hr class="half-width">
<p>Before deciding whether to trim reads, it is essential to understand
what adapters are and why they appear in sequencing data.</p>
<p>Adapters are short synthetic DNA sequences ligated to both ends of
each fragment during library preparation. They provide primer binding
sites needed for cluster amplification and sequencing. Ideally, the
sequenced fragment is long enough that the machine reads only biological
DNA. However, when the insert is shorter than the read length, the
sequencer runs into the adapter sequence, which produces adapter
contamination at the ends of reads.</p>
<p>Many users assume that adapters must always be trimmed, but this is
not true for typical RNA-seq workflows.</p>
<div id="do-we-need-to-trim-reads-for-standard-rna-seq-workflows" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="do-we-need-to-trim-reads-for-standard-rna-seq-workflows" class="callout-inner">
<h3 class="callout-title">Do we need to trim reads for standard RNA-seq
workflows?</h3>
<div class="callout-content">
<p>For alignment based RNA-seq, adapter trimming is usually not
required. Modern aligners such as HISAT2 and STAR detect and soft clip
adapter sequence automatically. Soft-clipped bases do not contribute to
mismatches or mapping penalties, so adapters generally do not harm
alignment quality.</p>
<p>In contrast, unnecessary trimming can introduce new problems:</p>
<ul>
<li>shorter reads and lower mapping power</li>
<li>uneven trimming across samples and artificial differences</li>
<li>reduction in read complexity</li>
<li>broken read pairing for paired-end sequencing</li>
<li>more parameters to track for reproducibility</li>
</ul>
<p>Trimming is recommended only in specific situations:</p>
<ul>
<li>very short inserts, for example degraded RNA or ancient DNA</li>
<li>adapter sequence dominating read tails, for example more than 10-15
percent adapter content</li>
<li>transcript assembly workflows such as StringTie or Scallop, which
benefit from uniform read lengths</li>
<li>small RNA or miRNA libraries, where inserts are intentionally very
short and adapter presence is guaranteed</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-evaluate-the-qc-results">Exercise: evaluate the QC results<a class="anchor" aria-label="anchor" href="#exercise-evaluate-the-qc-results"></a>
</h2>
<hr class="half-width">
<div id="inspect-your-multiqc-report" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="inspect-your-multiqc-report" class="callout-inner">
<h3 class="callout-title">Inspect your MultiQC report</h3>
<div class="callout-content">
<p>Using the MultiQC HTML report, answer:</p>
<ol style="list-style-type: decimal">
<li>Do all replicates show similar per base sequence quality
profiles?</li>
<li>Is adapter content high enough to impact analysis?</li>
<li>Is any sample behaving differently from the rest, for example in
number of reads, quality, duplication, or GC percentage?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Typical interpretation for this dataset (students should compare with
their own):</p>
<ul>
<li>All replicates show high, stable sequence quality.</li>
<li>Adapter signal appears only at the extreme ends of reads for
selected samples and is less than 4 percent.</li>
<li>No samples show concerning patterns or biases.</li>
<li>Trimming is therefore unnecessary for this dataset.</li>
</ul>
</div>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What if trimming is required? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<p><em>Optional reference: how to trim adapters with <code>fastp</code>
(not part of the hands-on)</em></p>
<p>Sometimes you may encounter datasets where trimming is appropriate,
for example short inserts, small RNA, or visibly high adapter content in
QC. Although we will not trim reads in this workshop, here is a minimal
example you can use in your own projects if trimming becomes
necessary.</p>
<p><code>fastp</code> is a fast, widely used tool that can automatically
detect adapters, trim them, filter low quality reads, and generate QC
reports.</p>
<div class="section level3">
<h3 id="example-trimming-paired-end-reads-with-fastp">Example: trimming paired-end reads with <code>fastp</code>
<a class="anchor" aria-label="anchor" href="#example-trimming-paired-end-reads-with-fastp"></a>
</h3>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">module</span> load fastp</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">-i</span> SRRXXXXXXX_1.fastq.gz <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">-I</span> SRRXXXXXXX_2.fastq.gz <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">-o</span> SRRXXXXXXX_1.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">-O</span> SRRXXXXXXX_2.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>  <span class="at">--thread</span> 8 <span class="dt">\</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>  <span class="at">--html</span> fastp_report.html <span class="dt">\</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  <span class="at">--json</span> fastp_report.json</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="what-this-command-does">What this command does<a class="anchor" aria-label="anchor" href="#what-this-command-does"></a>
</h3>
<ul>
<li>automatically detects adapters
(<code>--detect_adapter_for_pe</code>)</li>
<li>trims low quality bases (default behaviour)</li>
<li>keeps paired-end reads synchronized</li>
<li>writes trimmed output (<code>*.trimmed.fastq.gz</code>)</li>
<li>generates QC reports (HTML and JSON)</li>
</ul>
</div>
<div class="section level3">
<h3 id="when-to-use-this">When to use this<a class="anchor" aria-label="anchor" href="#when-to-use-this"></a>
</h3>
<p>Use trimming only if:</p>
<ul>
<li>QC shows strong adapter presence</li>
<li>your library prep is known to produce short inserts</li>
<li>you are working with small RNA or degraded RNA</li>
<li>you are doing transcript assembly where uniform read length
matters</li>
</ul>
<p>For standard RNA-seq, trimming is usually unnecessary.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>FastQC and MultiQC provide essential diagnostics for RNA-seq
data.</li>
<li>Some FastQC warnings are normal for RNA-seq and do not indicate
problems.</li>
<li>Aligners soft clip adapters, so trimming is usually unnecessary for
alignment based RNA-seq.</li>
<li>QC helps detect outliers before alignment and quantification.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-04a-genome-based-quantification"><p>Content from <a href="04a-genome-based-quantification.html">A. Genome-based quantification (STAR + featureCounts)</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/04a-genome-based-quantification.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 70 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we map RNA-seq reads to a reference genome?</li>
<li>How do we determine library strandness before alignment?</li>
<li>How do we quantify reads per gene using featureCounts?</li>
<li>How do we submit mapping jobs on RCAC clusters with SLURM?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Check RNA-seq library strandness using Salmon.</li>
<li>Index a reference genome for STAR.</li>
<li>Map reads to the genome using STAR.</li>
<li>Quantify gene level counts with featureCounts.</li>
<li>Interpret basic alignment and counting statistics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this episode we move from raw FASTQ files to genome based gene
counts.<br>
This workflow has three major steps:</p>
<ol style="list-style-type: decimal">
<li>Determine library strandness: STAR does not infer strandness
automatically. <code>featureCounts</code> requires the correct
setting.</li>
<li>Map reads to the genome using STAR: STAR is a splice aware aligner
designed for RNA-seq data.</li>
<li>Count mapped reads per gene using <code>featureCounts</code>
</li>
</ol>
<p>This represents one of the two standard quantification strategies in
RNA-seq.<br>
The alternative method (B. transcript based quantification with Salmon
or Kallisto) will be covered in Episode 4B.</p>
</section><section><h2 class="section-heading" id="step-1-checking-rna-seq-library-strandness">Step 1: Checking RNA-seq library strandness<a class="anchor" aria-label="anchor" href="#step-1-checking-rna-seq-library-strandness"></a>
</h2>
<hr class="half-width">
<div id="why-strandness-matters" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-strandness-matters" class="callout-inner">
<h3 class="callout-title">Why strandness matters</h3>
<div class="callout-content">
<p>RNA-seq libraries can be:</p>
<ul>
<li>unstranded<br>
</li>
<li>stranded forward<br>
</li>
<li>stranded reverse</li>
</ul>
<p>Stranded protocols are common in modern kits, but old GEO and SRA
datasets are often unstranded.<br>
featureCounts must be given the correct strand setting. If the wrong
setting is used, most reads will be counted against incorrect genes.</p>
<p>STAR does not infer strandness automatically, so we must determine it
<strong>before</strong> alignment.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="detecting-strandness-with-salmon">Detecting strandness with Salmon<a class="anchor" aria-label="anchor" href="#detecting-strandness-with-salmon"></a>
</h3>
<p>Salmon can infer strandness directly from raw FASTQ files without any
alignment.<br>
We run Salmon with <code>-l A</code>, which tells it to explore all
library types.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-q</span> standby <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/strand_check</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">module</span> load salmon</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="ex">salmon</span> index <span class="at">--transcripts</span> data/gencode.vM38.transcripts-clean.fa <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>             <span class="at">--index</span> data/salmon_index <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>             <span class="at">--threads</span> 4</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="ex">salmon</span> quant <span class="at">--index</span> data/salmon_index <span class="dt">\</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>             <span class="at">--libType</span> A <span class="dt">\</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>             <span class="at">--mates1</span> data/SRR33253286_1.fastq.gz <span class="dt">\</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>             <span class="at">--mates2</span> data/SRR33253286_2.fastq.gz <span class="dt">\</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>             <span class="at">--output</span> results/strand_check <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>             <span class="at">--threads</span> 4</span></code></pre>
</div>
<p>The important output is in
<code>results/strand_check/lib_format_counts.json</code>:</p>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What does this file look like? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">JSON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode json" tabindex="0"><code class="sourceCode json"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>    <span class="dt">"read_files"</span><span class="fu">:</span> <span class="st">"[ data/SRR33253286_1.fastq.gz, data/SRR33253286_2.fastq.gz]"</span><span class="fu">,</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>    <span class="dt">"expected_format"</span><span class="fu">:</span> <span class="st">"IU"</span><span class="fu">,</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="dt">"compatible_fragment_ratio"</span><span class="fu">:</span> <span class="fl">1.0</span><span class="fu">,</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>    <span class="dt">"num_compatible_fragments"</span><span class="fu">:</span> <span class="dv">21088696</span><span class="fu">,</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>    <span class="dt">"num_assigned_fragments"</span><span class="fu">:</span> <span class="dv">21088696</span><span class="fu">,</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>    <span class="dt">"num_frags_with_concordant_consistent_mappings"</span><span class="fu">:</span> <span class="dv">19116702</span><span class="fu">,</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>    <span class="dt">"num_frags_with_inconsistent_or_orphan_mappings"</span><span class="fu">:</span> <span class="dv">2510542</span><span class="fu">,</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>    <span class="dt">"strand_mapping_bias"</span><span class="fu">:</span> <span class="dv">0</span><span class="er">.</span><span class="dv">5117980078362889</span><span class="fu">,</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>    <span class="dt">"MSF"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>    <span class="dt">"OSF"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>    <span class="dt">"ISF"</span><span class="fu">:</span> <span class="dv">9783890</span><span class="fu">,</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>    <span class="dt">"MSR"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>    <span class="dt">"OSR"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>    <span class="dt">"ISR"</span><span class="fu">:</span> <span class="dv">9332812</span><span class="fu">,</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>    <span class="dt">"SF"</span><span class="fu">:</span> <span class="dv">1234245</span><span class="fu">,</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>    <span class="dt">"SR"</span><span class="fu">:</span> <span class="dv">1276297</span><span class="fu">,</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>    <span class="dt">"MU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>    <span class="dt">"OU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>    <span class="dt">"IU"</span><span class="fu">:</span> <span class="dv">0</span><span class="fu">,</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>    <span class="dt">"U"</span><span class="fu">:</span> <span class="dv">0</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="fu">}</span></span></code></pre>
</div>
<blockquote>
<h3 id="interpreting-lib_format_counts.json">Interpreting
<code>lib_format_counts.json</code>
</h3>
<p>This JSON file reports Salmon’s automatic library type detection.</p>
<ul>
<li>
<strong><code>"expected_format": "IU"</code></strong>: This is
Salmon’s conclusion. “I” means <strong>Inward</strong> (correct for
paired-end reads) and “U” means <strong>Unstranded</strong>.</li>
<li>
<strong><code>"strand_mapping_bias": 0.512</code></strong>: This is
the key evidence. A value near 0.5 (50%) indicates that reads mapped
equally to both the sense and antisense strands, the definitive sign of
an <strong>unstranded</strong> library.</li>
<li>
<strong><code>"ISF"</code></strong> and
<strong><code>"ISR"</code></strong>: These are the counts for
Inward-Sense-Forward (9.78M) and Inward-Sense-Reverse (9.33M) fragments.
Because these values are almost equal, they confirm the ~50/50 split
seen in the strand bias.</li>
</ul>
</blockquote>
<p>Common results:</p>
<table class="table">
<thead><tr class="header">
<th>Code</th>
<th>Meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>IU</td>
<td>unstranded</td>
</tr>
<tr class="even">
<td>ISR</td>
<td>stranded reverse (paired end)</td>
</tr>
<tr class="odd">
<td>ISF</td>
<td>stranded forward (paired end)</td>
</tr>
<tr class="even">
<td>SR</td>
<td>stranded reverse (single end)</td>
</tr>
<tr class="odd">
<td>SF</td>
<td>stranded forward (single end)</td>
</tr>
</tbody>
</table>
<p><strong>Conclusion:</strong> The data is <strong>unstranded</strong>.
We will record this and supply it to STAR/featureCounts.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-2-preparing-the-reference-genome-for-alignment">Step 2: Preparing the reference genome for alignment<a class="anchor" aria-label="anchor" href="#step-2-preparing-the-reference-genome-for-alignment"></a>
</h2>
<hr class="half-width">
<p>STAR requires a genome index. Indexing builds a searchable data
structure that speeds up alignment.</p>
<p>Required input:</p>
<ul>
<li>genome FASTA file</li>
<li>annotation file (GTF or GFF3)</li>
<li>output directory for the index</li>
</ul>
<p>Optionally, annotation improves splice junction detection and yields
better alignments.</p>
<div id="why-star" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-star" class="callout-inner">
<h3 class="callout-title">Why STAR?</h3>
<div class="callout-content">
<ul>
<li>Splice-aware<br>
</li>
<li>Extremely fast<br>
</li>
<li>Produces high-quality alignments<br>
</li>
<li>Well supported and widely used<br>
</li>
<li>Compatible with downstream tools such as featureCounts, StringTie,
and RSEM</li>
</ul>
<p>Optional alternatives include HISAT2 and Bowtie2, but STAR remains
the standard for high-depth Illumina RNA-seq.</p>
</div>
</div>
</div>
<div id="exercise-read-the-star-manual" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-read-the-star-manual" class="callout-inner">
<h3 class="callout-title">Exercise: Read the STAR manual</h3>
<div class="callout-content">
<p>Check the <strong>“Generating genome indexes”</strong> section of
the<br><a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" class="external-link">STAR
manual</a></p>
<p>Which options are required for indexing?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Key options:</p>
<ul>
<li>
<code>--runMode genomeGenerate</code><br>
</li>
<li>
<code>--genomeDir</code> directory to store the index<br>
</li>
<li>
<code>--genomeFastaFiles</code> the FASTA file<br>
</li>
<li>
<code>--sjdbGTFfile</code> annotation (recommended)<br>
</li>
<li>
<code>--runThreadN</code> number of threads<br>
</li>
<li>
<code>--sjdbOverhang</code> read length minus one (100 is a safe
default)</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="indexing-the-genome-with-star">Indexing the Genome with STAR<a class="anchor" aria-label="anchor" href="#indexing-the-genome-with-star"></a>
</h3>
<p>Below is a minimal SLURM job script for building the STAR genome
index. Create a file named <code>index_genome.sh</code> in
<code>$SCRATCH/rcac_rnaseq/scripts</code> with this content:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=20</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=star_index</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="ex">module</span> load star</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="va">data_dir</span><span class="op">=</span><span class="st">"</span><span class="va">${SCRATCH}</span><span class="st">/rnaseq-workshop/data"</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/GRCm39.primary_assembly.genome.fa"</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/gencode.vM38.primary_assembly.basic.annotation.gtf"</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="va">indexdir</span><span class="op">=</span><span class="st">"</span><span class="va">${data_dir}</span><span class="st">/star_index"</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>    <span class="at">--runMode</span> genomeGenerate <span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>    <span class="at">--runThreadN</span> <span class="va">${SLURM_CPUS_ON_NODE}</span> <span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>    <span class="at">--genomeDir</span> <span class="va">${indexdir}</span> <span class="dt">\</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>    <span class="at">--genomeFastaFiles</span> <span class="va">${genome}</span> <span class="dt">\</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>    <span class="at">--sjdbGTFfile</span> <span class="va">${gtf}</span></span></code></pre>
</div>
<p>Submit the job:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">sbatch</span> index_genome.sh</span></code></pre>
</div>
<p>Indexing requires about 30 to 45 minutes.</p>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What will the output look like? </h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" aria-labelledby="headingSpoiler2" data-bs-parent="#accordionSpoiler2">
<div class="accordion-body">
<p>The index directory is now ready for use in mapping and should have
contents like this: Location: <code>$SCRATCH/rcac_rnaseq/data</code>
Contents of <code>star_index/</code>:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">star_index/</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">├──</span> chrLength.txt</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">├──</span> chrNameLength.txt</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">├──</span> chrName.txt</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="ex">├──</span> chrStart.txt</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="ex">├──</span> exonGeTrInfo.tab</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="ex">├──</span> exonInfo.tab</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="ex">├──</span> geneInfo.tab</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="ex">├──</span> Genome</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="ex">├──</span> genomeParameters.txt</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="ex">├──</span> Log.out</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="ex">├──</span> SA</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="ex">├──</span> SAindex</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="ex">├──</span> sjdbInfo.txt</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="ex">├──</span> sjdbList.fromGTF.out.tab</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="ex">├──</span> sjdbList.out.tab</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="ex">└──</span> transcriptInfo.tab</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="do-i-need-the-gtf-for-indexing" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="do-i-need-the-gtf-for-indexing" class="callout-inner">
<h3 class="callout-title">Do I need the GTF for indexing?</h3>
<div class="callout-content">
<p>No, but it is <em>strongly recommended</em>.<br>
Including it improves splice junction detection and increases mapping
accuracy.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-3-mapping-reads-with-star">Step 3: Mapping reads with STAR<a class="anchor" aria-label="anchor" href="#step-3-mapping-reads-with-star"></a>
</h2>
<hr class="half-width">
<p>Once the genome index is ready, we can align reads.</p>
<p>For aligning reads, important STAR options are:</p>
<ul>
<li><code>--runThreadN</code></li>
<li><code>--genomeDir</code></li>
<li><code>--readFilesIn</code></li>
<li>
<code>--readFilesCommand zcat</code> for gzipped FASTQ files</li>
<li><code>--outSAMtype BAM SortedByCoordinate</code></li>
<li><code>--outSAMunmapped Within</code></li>
</ul>
<p>STAR’s defaults are tuned for mammalian genomes. If working with
plants, fungi, or repetitive genomes, additional tuning may be
required.</p>
<div id="exercise-write-a-star-alignment-command" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-write-a-star-alignment-command" class="callout-inner">
<h3 class="callout-title">Exercise: Write a STAR alignment command</h3>
<div class="callout-content">
<p>Write a basic STAR alignment command using your genome index and
paired FASTQ files.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>    <span class="at">--runThreadN</span> 12 <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>    <span class="at">--genomeDir</span> data/star_index <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>    <span class="at">--readFilesIn</span> data/SRR1234567_1.fastq.gz data/SRR1234567_2.fastq.gz <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> results/mapping/SRR1234567 <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    <span class="at">--outSAMunmapped</span> Within</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="mapping-many-fastq-files-with-a-slurm-array">Mapping many FASTQ files with a SLURM array<a class="anchor" aria-label="anchor" href="#mapping-many-fastq-files-with-a-slurm-array"></a>
</h3>
<p>When having multiple samples, array jobs simplify submission.</p>
<p>First, we will create <code>samples.txt</code> file with just the
sample names</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/data</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>_1.fastq.gz <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_1.fastq.gz//'</span> <span class="op">&gt;</span> <span class="va">${SCRATCH}</span>/rnaseq-workshop/scripts/samples.txt</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=20</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#SBATCH --time=8:00:00</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=read_mapping</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#SBATCH --array=1-6</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="ex">module</span> load star</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="va">FASTQ_DIR</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data"</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="va">GENOME_INDEX</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data/star_index"</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="va">OUTDIR</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/mapping"</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUTDIR</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span><span class="va">$(</span><span class="fu">sed</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${SLURM_ARRAY_TASK_ID}</span><span class="st">p"</span> samples.txt<span class="va">)</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="va">R1</span><span class="op">=</span><span class="va">${FASTQ_DIR}</span>/<span class="va">${SAMPLE}</span>_1.fastq.gz</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="va">R2</span><span class="op">=</span><span class="va">${FASTQ_DIR}</span>/<span class="va">${SAMPLE}</span>_2.fastq.gz</span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a>    <span class="at">--runThreadN</span> <span class="va">${SLURM_CPUS_ON_NODE}</span> <span class="dt">\</span></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a>    <span class="at">--genomeDir</span> <span class="va">$GENOME_INDEX</span> <span class="dt">\</span></span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a>    <span class="at">--readFilesIn</span> <span class="va">$R1</span> <span class="va">$R2</span> <span class="dt">\</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> <span class="va">$OUTDIR</span>/<span class="va">$SAMPLE</span> <span class="dt">\</span></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a>    <span class="at">--outSAMunmapped</span> Within <span class="dt">\</span></span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span></code></pre>
</div>
<p>Submit:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">sbatch</span> map_reads.sh</span></code></pre>
</div>
<div id="why-use-an-array-job" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="why-use-an-array-job" class="callout-inner">
<h3 class="callout-title">Why use an array job?</h3>
<div class="callout-content">
<p>Think about this:</p>
<ul>
<li>We have <em>six</em> samples, each with two FASTQ files<br>
</li>
<li>Having six separate slurm jobs means:
<ul>
<li>Copy-pasting the same command six times<br>
</li>
<li>Higher chance of typos or mistakes<br>
</li>
</ul>
</li>
<li>if mapping with single slurm file, sequential execution =
slower</li>
<li>Samples can run independently</li>
</ul>
<p>Why is a job array a good choice?</p>
<p>Arrays allow us to run identical commands across many inputs:</p>
<ul>
<li>All samples are processed the same way<br>
</li>
<li>Runs in parallel<br>
</li>
<li>Simplifies submission<br>
</li>
<li>Reduces copy-paste errors<br>
</li>
<li>Minimizes runtime cost on the cluster</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-assessing-alignment-quality">Step 4: Assessing alignment quality<a class="anchor" aria-label="anchor" href="#step-4-assessing-alignment-quality"></a>
</h2>
<hr class="half-width">
<p>STAR produces several useful files, particularly:</p>
<ul>
<li><p><code>Log.final.out</code><br>
Contains mapping statistics, including % uniquely mapped.</p></li>
<li><p><code>Aligned.sortedByCoord.out.bam</code><br>
Ready for downstream quantification.</p></li>
</ul>
<p>To summarize alignment statistics across samples, we use MultiQC:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="ex">multiqc</span> results/mapping <span class="at">-o</span> results/qc_alignment</span></code></pre>
</div>
<div class="section level3">
<h3 id="key-metrics-to-review">Key metrics to review<a class="anchor" aria-label="anchor" href="#key-metrics-to-review"></a>
</h3>
<ul>
<li>
<strong>Uniquely mapped reads</strong> — typically 60–90% is
good<br>
</li>
<li>
<strong>Multi-mapped reads</strong> — depends on genome, but &gt;20%
is concerning<br>
</li>
<li>
<strong>Unmapped reads</strong> — check reasons (mismatches? too
short?)<br>
</li>
<li>
<strong>Mismatch rate</strong> — high values indicate quality or
contamination issues</li>
</ul>
<div id="exercise-examine-your-alignment-metrics" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-examine-your-alignment-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Examine your alignment metrics</h3>
<div class="callout-content">
<p>Using your MultiQC alignment report:</p>
<ol style="list-style-type: decimal">
<li>Which sample has the highest uniquely mapped percentage?<br>
</li>
<li>Are any samples clear outliers?<br>
</li>
<li>Does mapped/unmapped reads appear consistent across samples?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Example interpretation for this dataset:</p>
<ul>
<li>All samples have ~80% uniquely mapped reads.</li>
<li>No sample is an outlier.</li>
<li>mapped/unmapped reads consistent across samples</li>
</ul>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-quantifying-gene-counts-with-featurecounts">Step 5: Quantifying gene counts with featureCounts<a class="anchor" aria-label="anchor" href="#step-5-quantifying-gene-counts-with-featurecounts"></a>
</h2>
<hr class="half-width">
<p>Once the reads have been aligned and sorted, we can quantify how many
read pairs map to each gene. This gives us gene-level counts that we
will later use for differential expression. For alignment-based RNA-seq
workflows, <code>featureCounts</code> is the recommended tool because it
is fast, robust, and compatible with most gene annotation formats.</p>
<p><code>featureCounts</code> uses two inputs:</p>
<ol style="list-style-type: decimal">
<li>A set of aligned BAM files (sorted by coordinate)</li>
<li>A GTF annotation file describing gene models</li>
</ol>
<p>It assigns each read (or read pair) to a gene based on the exon
boundaries.</p>
<div id="important-stranded-or-unstranded" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="important-stranded-or-unstranded" class="callout-inner">
<h3 class="callout-title">Important: stranded or unstranded?</h3>
<div class="callout-content">
<p><code>featureCounts</code> needs to know whether your dataset is
stranded.</p>
<ul>
<li>
<code>-s 0</code> unstranded</li>
<li>
<code>-s 1</code> stranded forward</li>
<li>
<code>-s 2</code> stranded reverse</li>
</ul>
<p>From our Salmon strandness check, this dataset behaves as
<strong>unstranded</strong>, so we will use <code>-s 0</code>.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="running-featurecounts">Running featureCounts<a class="anchor" aria-label="anchor" href="#running-featurecounts"></a>
</h3>
<p>We will create a new directory for count output:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/scripts</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/results/counts</span></code></pre>
</div>
<p>Below is an example SLURM script to count all BAM files at once.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=16</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=featurecounts</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="ex">module</span> load subread</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="va">GTF</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data/gencode.vM38.annotation.gtf"</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="va">BAMS</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/mapping/*.bam"</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="va">output_dir</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/counts"</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="dt">\</span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>  <span class="at">-T</span> <span class="va">${SLURM_CPUS_PER_TASK}</span> <span class="dt">\</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>  <span class="at">-p</span> <span class="dt">\</span></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a>  <span class="at">-s</span> 0 <span class="dt">\</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>  <span class="at">-t</span> exon <span class="dt">\</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a>  <span class="at">-g</span> gene_id <span class="dt">\</span></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>  <span class="at">-a</span> <span class="va">${GTF}</span> <span class="dt">\</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>  <span class="at">-o</span> <span class="va">${output_dir}</span>/gene_counts.txt <span class="dt">\</span></span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>  <span class="va">${BAMS}</span></span></code></pre>
</div>
<p>Submit the job:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">sbatch</span> count_features.sh</span></code></pre>
</div>
<p><code>gene_counts.txt</code> will contain:</p>
<ul>
<li>one row per gene</li>
<li>one column per sample</li>
</ul>
<div id="generate-a-clean-count-matrix" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="generate-a-clean-count-matrix" class="callout-inner">
<h3 class="callout-title">Generate a clean count matrix</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/results/counts</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1,7- gene_counts.txt <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  <span class="fu">sed</span> <span class="st">'s+/scratch/negishi/'</span><span class="va">$USER</span><span class="st">'/rnaseq-workshop/results/mapping/++g'</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="fu">grep</span> <span class="at">-v</span> <span class="st">"^#"</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>  <span class="fu">sed</span> <span class="st">'s/Aligned.sortedByCoord.out.bam//g'</span> <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>  <span class="op">&gt;</span> gene_counts_clean.txt</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-6-qc-on-counts">Step 6: QC on counts<a class="anchor" aria-label="anchor" href="#step-6-qc-on-counts"></a>
</h2>
<hr class="half-width">
<p>Before proceeding to differential expression, it is good practice to
perform some QC on the count data. We can again use MultiQC to summarize
featureCounts statistics:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/qc_counts</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="ex">multiqc</span> results/counts <span class="at">-o</span> results/qc_counts</span></code></pre>
</div>
<div id="exercise-examine-your-counting-metrics" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-examine-your-counting-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Examine your counting metrics</h3>
<div class="callout-content">
<p>Using your MultiQC featureCounts report:</p>
<ol style="list-style-type: decimal">
<li>Which sample has the lowest percent assigned, and what are common
reasons for lower assignment?</li>
<li>Which unassigned category is the largest, and what does it
indicate?</li>
<li>Does the sample with the most assigned reads also have the highest
percent assigned?</li>
<li>What does “Unassigned: No Features” mean, and why might it
occur?</li>
<li>Are multi-mapped reads a concern for RNA-seq differential
expression?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>Example interpretation for this dataset:</p>
<ol style="list-style-type: decimal">
<li>SRR33253289 has the lowest assignment rate. Causes include low
complexity, incomplete annotation, or more intronic reads.</li>
<li>Unmapped or No Features dominate. These reflect reads that do not
align or do not overlap annotated exons.</li>
<li>No. Total assigned reads depend on depth, while percent assigned
reflects library quality.</li>
<li>Reads mapped outside annotated exons, often from intronic regions,
unannotated transcripts, or incomplete annotation.</li>
<li>Multi-mapping is expected for paralogs and repeats. It is usually
fine as long as the rate is consistent across samples.</li>
</ol>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>STAR is a fast splice-aware aligner widely used for RNA-seq.</li>
<li>Genome indexing requires FASTA and optionally GTF for improved
splice detection.</li>
<li>Mapping is efficiently performed using SLURM array jobs.</li>
<li>Alignment statistics (from <code>Log.final.out</code> + MultiQC)
must be reviewed.</li>
<li>Strandness should be inferred using aligned BAM files.</li>
<li>Final BAM files are ready for downstream counting.</li>
<li>
<code>featureCounts</code> is used to obtain gene-level counts for
differential expression.</li>
<li>Exons are counted and summed per gene.</li>
<li>The output is a simple matrix for downstream statistical
analysis.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-04b-transcript-based-quantification"><p>Content from <a href="04b-transcript-based-quantification.html">B. Transcript-based quantification (Salmon)</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/04b-transcript-based-quantification.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 70 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we quantify expression without genome alignment?</li>
<li>What inputs does Salmon require?</li>
<li>How do we run Salmon for paired-end RNA-seq data?</li>
<li>How do we interpret transcript-level outputs (TPM, NumReads)?</li>
<li>How do we summarize transcripts to gene-level counts using
tximport?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Build a transcriptome index for Salmon.</li>
<li>Quantify transcript abundance directly from FASTQ files.</li>
<li>Understand key Salmon output files.</li>
<li>Summarize transcript-level estimates to gene-level counts.</li>
<li>Prepare counts for downstream differential expression.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In the previous episode, we mapped reads to the
<strong>genome</strong> using STAR and generated <strong>gene-level
counts</strong> with featureCounts.</p>
<p>In this episode, we use an alternative quantification strategy:</p>
<ol style="list-style-type: decimal">
<li>Build a <em>transcriptome</em> index<br>
</li>
<li>Quantify expression directly from FASTQ files using
<strong>Salmon</strong> (<em>without alignment</em>)<br>
</li>
<li>Summarize transcript estimates back to gene-level counts</li>
</ol>
<p>This workflow is faster, uses less storage, and models
transcript-level uncertainty.</p>
<div id="when-should-i-use-transcript-based-quantification" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="when-should-i-use-transcript-based-quantification" class="callout-inner">
<h3 class="callout-title">When should I use transcript-based
quantification?</h3>
<div class="callout-content">
<p>Salmon is ideal when:</p>
<ul>
<li>You do not need alignment files.</li>
<li>You want transcript-level TPMs.</li>
<li>You want fast quantification with bias correction.</li>
<li>Storage is limited (no BAM files generated).</li>
</ul>
<p>It is <em>not</em> ideal if you need splice junctions, variant
calling, or visualization in IGV (all those depends on alignment
files).</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-1-preparing-the-transcriptome-reference">Step 1: Preparing the transcriptome reference<a class="anchor" aria-label="anchor" href="#step-1-preparing-the-transcriptome-reference"></a>
</h2>
<hr class="half-width">
<p>Salmon requires a <strong>transcriptome FASTA file</strong> (all
annotated transcripts). We previously downloaded transcripts file from
GENCODE for this purpose (<code>gencode.vM38.transcripts.fa</code>).</p>
<div id="why-transcriptome-choice-matters" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-transcriptome-choice-matters" class="callout-inner">
<h3 class="callout-title">Why transcriptome choice matters</h3>
<div class="callout-content">
<p>Transcript-level quantification <strong>inherits all assumptions of
the annotation</strong>.<br>
Missing or incorrect transcripts → biased TPM estimates.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="building-the-salmon-index">Building the Salmon index<a class="anchor" aria-label="anchor" href="#building-the-salmon-index"></a>
</h3>
<p>A Salmon index is lightweight and quick to build. <em>You may have
already built this in Episode 4A.</em></p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> workshop <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>1:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> data/salmon_index</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">module</span> load salmon</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="ex">salmon</span> index <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>    <span class="at">--transcripts</span> data/gencode.vM38.transcripts-clean.fa <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>    <span class="at">--index</span> data/salmon_index <span class="dt">\</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>    <span class="at">--threads</span> 4</span></code></pre>
</div>
<p>This creates a directory containing Salmon’s k-mer index of all
transcripts.</p>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> What will the index contain? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<pre><code>salmon_index/
├── complete_ref_lens.bin        # IMPORTANT: full transcript lengths used in TPM/effLength calculations
├── ctable.bin                   # internal k-mer lookup table (not user-relevant)
├── ctg_offsets.bin              # internal offsets for reference sequences
├── duplicate_clusters.tsv       # internal duplicate k-mer cluster info
├── info.json                    # IMPORTANT: index metadata (k-mer size, transcripts, build parameters)
├── mphf.bin                     # minimal perfect hash function for fast lookup
├── pos.bin                      # internal k-mer position table
├── pre_indexing.log             # IMPORTANT: log of preprocessing and FASTA parsing steps
├── rank.bin                     # internal rank/select bitvector structure
├── refAccumLengths.bin          # cumulative reference lengths (internal)
├── ref_indexing.log             # IMPORTANT: main indexing log; good for debugging build problems
├── reflengths.bin               # IMPORTANT: transcript lengths used in quantification math
├── refseq.bin                   # IMPORTANT: transcript sequences stored in binary format (core index content)
├── seq.bin                      # internal sequence encoding structures
└── versionInfo.json             # IMPORTANT: Salmon version + environment info for reproducibility</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-2-quantifying-transcript-abundances">Step 2: Quantifying transcript abundances<a class="anchor" aria-label="anchor" href="#step-2-quantifying-transcript-abundances"></a>
</h2>
<hr class="half-width">
<p>The core Salmon command is <code>salmon quant</code>. We use the
transcript index and paired-end FASTQ files.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/results/salmon_quant</span></code></pre>
</div>
<p>Example command:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">salmon</span> quant <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>    <span class="at">--index</span> salmon_index <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>    <span class="at">--libType</span> IU <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    <span class="at">--mates1</span> data/SRR33253286_1.fastq.gz <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>    <span class="at">--mates2</span> data/SRR33253286_2.fastq.gz <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    <span class="at">--output</span> salmon_quant/SRR33253286 <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>    <span class="at">--threads</span> 16 <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>    <span class="at">--validateMappings</span> <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>    <span class="at">--seqBias</span> <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>    <span class="at">--gcBias</span></span></code></pre>
</div>
<p>Important flags:</p>
<ul>
<li>
<code>--libType IU</code> → <strong>unstranded</strong>,
inward-oriented paired-end (Replace with <code>ISR</code>,
<code>ISF</code>, etc. based on your strandness.)</li>
<li>
<code>--validateMappings</code> → improves mapping accuracy</li>
<li>
<code>--seqBias</code>, <code>--gcBias</code> → corrects sequence-
and GC-related biases</li>
</ul>
<div id="correct-library-type-is-essential" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="correct-library-type-is-essential" class="callout-inner">
<h3 class="callout-title">Correct library type is essential</h3>
<div class="callout-content">
<p>A wrong <code>--libType</code> silently distorts transcript
abundances. Use Salmon’s <em>strandness detection</em> from Episode 4A
if unsure.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-3-running-salmon-for-many-samples">Step 3: Running Salmon for many samples<a class="anchor" aria-label="anchor" href="#step-3-running-salmon-for-many-samples"></a>
</h2>
<hr class="half-width">
<p>Again, we use a SLURM array.</p>
<p>First, create a sample list (or reuse the one from Episode 4A):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/data</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>_1.fastq.gz <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_1.fastq.gz//'</span> <span class="op">&gt;</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/scripts/samples.txt</span></code></pre>
</div>
<p>Array job:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#SBATCH --ntasks=1</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=16</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#SBATCH --account=workshop</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#SBATCH --partition=cpu</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#SBATCH --time=4:00:00</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#SBATCH --job-name=salmon_quant</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#SBATCH --array=1-6</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#SBATCH --output=cluster-%x.%j.out</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#SBATCH --error=cluster-%x.%j.err</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="ex">module</span> load salmon</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="va">DATA</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data"</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="va">INDEX</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/data/salmon_index"</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span><span class="st">"</span><span class="va">$SCRATCH</span><span class="st">/rnaseq-workshop/results/salmon_quant"</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span><span class="va">$(</span><span class="fu">sed</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${SLURM_ARRAY_TASK_ID}</span><span class="st">p"</span> samples.txt<span class="va">)</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="va">R1</span><span class="op">=</span><span class="va">${DATA}</span>/<span class="va">${SAMPLE}</span>_1.fastq.gz</span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="va">R2</span><span class="op">=</span><span class="va">${DATA}</span>/<span class="va">${SAMPLE}</span>_2.fastq.gz</span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="ex">salmon</span> quant <span class="dt">\</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a>  <span class="at">--index</span> <span class="va">$INDEX</span> <span class="dt">\</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a>  <span class="at">--libType</span> IU <span class="dt">\</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a>  <span class="at">--mates1</span> <span class="va">$R1</span> <span class="dt">\</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a>  <span class="at">--mates2</span> <span class="va">$R2</span> <span class="dt">\</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a>  <span class="at">--output</span> <span class="va">$OUT</span>/<span class="va">$SAMPLE</span> <span class="dt">\</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a>  <span class="at">--threads</span> <span class="va">${SLURM_CPUS_ON_NODE}</span> <span class="dt">\</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a>  <span class="at">--validateMappings</span> <span class="dt">\</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a>  <span class="at">--seqBias</span> <span class="dt">\</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a>  <span class="at">--gcBias</span></span></code></pre>
</div>
<p>Submit:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">sbatch</span> quant_salmon.sh</span></code></pre>
</div>
<p>The typical run time is ~5 minutes per sample.</p>
<div id="why-use-an-array-job-here" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="why-use-an-array-job-here" class="callout-inner">
<h3 class="callout-title">Why use an array job here?</h3>
<div class="callout-content">
<p>Unlike the STAR workflow in 4A, Salmon runs quickly and produces
small output. The advantage of array jobs here is not speed, but
<em>consistency</em>, all samples are quantified with identical
parameters and metadata. This ensures that transcript-level TPM and
count estimates are directly comparable across samples.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="inspect-results">Inspect results<a class="anchor" aria-label="anchor" href="#inspect-results"></a>
</h3>
<p>After Salmon finishes, each sample directory contains a small set of
output files.<br>
Only a few of these are important for routine analysis:</p>
<pre><code>SRR33253285/
├── quant.sf                       # IMPORTANT: transcript-level TPM, counts, effective lengths
├── lib_format_counts.json         # IMPORTANT: inferred library type (IU, ISR, ISF)
├── aux_info
│   ├── meta_info.json             # IMPORTANT: mapping rate, fragment counts, bias flags
│   ├── fld.gz                     # fragment length distribution (binary format)
│   ├── expected_bias.gz           # bias model used for correction
│   ├── observed_bias.gz           # observed sequence and GC bias
│   └── ambig_info.tsv             # ambiguous equivalence class summaries (advanced QC only)
├── libParams
│   └── flenDist.txt               # fragment length distribution in a human-readable format
├── cmd_info.json                  # command and parameters used to run Salmon
└── logs
└── salmon_quant.log               # main log file for troubleshooting</code></pre>
</div>
<div class="section level3">
<h3 id="what-is-inside-quant-sf">What is inside <em>quant.sf</em>?<a class="anchor" aria-label="anchor" href="#what-is-inside-quant-sf"></a>
</h3>
<p>This file contains the transcript-level quantification results:</p>
<ul>
<li>
<strong>Name</strong>: transcript ID<br>
</li>
<li>
<strong>Length</strong>: transcript length<br>
</li>
<li>
<strong>EffectiveLength</strong>: length adjusted for fragment
distribution and bias<br>
</li>
<li>
<strong>TPM</strong>: within-sample normalized abundance<br>
</li>
<li>
<strong>NumReads</strong>: estimated number of reads assigned to
that transcript</li>
</ul>
<p>Example:</p>
<pre class="text"><code>Name                       Length  EffectiveLength TPM             NumReads
ENSMUST00000193812.2       1070    775.000         0.000000        0.000
ENSMUST00000082908.3       110     110.000         0.000000        0.000
ENSMUST00000162897.2       4153    3483.630        0.013994        1.000
ENSMUST00000159265.2       2989    2694.000        0.000000        0.000
ENSMUST00000070533.5       3634    3339.000        0.000000        0.000
ENSMUST00000192857.2       480     185.000         0.000000        0.000
ENSMUST00000195335.2       2819    2524.000        0.000000        0.000
ENSMUST00000192336.2       2233    1938.000        0.000000        0.000
ENSMUST00000194099.2       2309    2014.000        0.000000        0.000</code></pre>
</div>
<div class="section level3">
<h3 id="key-qc-metrics-to-review">Key QC metrics to review<a class="anchor" aria-label="anchor" href="#key-qc-metrics-to-review"></a>
</h3>
<p>Open <code>aux_info/meta_info.json</code> and check:</p>
<ul>
<li>
<strong>percent_mapped</strong>: proportion of fragments assigned to
transcripts</li>
<li>
<strong>library_types</strong>: Salmon inferred library orientation,
which should match your <code>--libType</code>
</li>
<li>
<strong>fragment length statistics</strong>: mean and standard
deviation of inferred fragment lengths</li>
</ul>
<p>These metrics provide a first-pass sanity check before summarizing
the results to gene-level counts.</p>
<div id="what-does-salmon-count" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="what-does-salmon-count" class="callout-inner">
<h3 class="callout-title">What does Salmon count?</h3>
<div class="callout-content">
<p><code>NumReads</code> is a model-based estimate, not raw counts of
reads. TPM values are <em>within-sample</em> normalized and should not
be directly compared across samples.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-summarizing-to-gene-level-counts-tximport">Step 4: Summarizing to gene-level counts
(<code>tximport</code>)<a class="anchor" aria-label="anchor" href="#step-4-summarizing-to-gene-level-counts-tximport"></a>
</h2>
<hr class="half-width">
<p>Most differential expression tools (DESeq2, edgeR) require
<strong>gene-level</strong> counts. We convert transcript estimates →
gene-level matrix.</p>
<div class="section level3">
<h3 id="prepare-tx2gene-mapping">Prepare tx2gene mapping<a class="anchor" aria-label="anchor" href="#prepare-tx2gene-mapping"></a>
</h3>
<p>We will need a mapping file that links the transcript IDs to gene
IDs. The <code>gencode.vM38.primary_assembly.basic.annotation.gtf</code>
file contains this information.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop/data</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span><span class="st">'\t'</span> <span class="st">'$3=="transcript" {</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="st">    match($9, /transcript_id "([^"]+)"/, tx);</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="st">    match($9, /gene_id "([^"]+)"/, gene);</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="st">    print tx[1] "\t" gene[1]</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="st">}'</span> gencode.vM38.primary_assembly.basic.annotation.gtf <span class="op">&gt;</span> tx2gene.tsv</span></code></pre>
</div>
<p>Next, we need to process this mapping in R. Load the modules and
start the R session, first:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">module</span> load r-rnaseq</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="ex">R</span></span></code></pre>
</div>
<p>In the R session, read in the <code>tx2gene</code> mapping:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">setwd</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="fu">Sys.getenv</span><span class="op">(</span><span class="st">"SCRATCH"</span><span class="op">)</span>,<span class="st">"/rnaseq-workshop"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tximport</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"data/tx2gene.tsv"</span>, col_types <span class="op">=</span> <span class="st">"cc"</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"scripts/samples.txt"</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu">file.path</span><span class="op">(</span><span class="st">"results/salmon_quant"</span>, <span class="va">samples</span><span class="op">$</span><span class="va">X1</span>, <span class="st">"quant.sf"</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">$</span><span class="va">X1</span></span>
<span></span>
<span><span class="va">txi</span> <span class="op">&lt;-</span> <span class="fu">tximport</span><span class="op">(</span><span class="va">files</span>,</span>
<span>                type <span class="op">=</span> <span class="st">"salmon"</span>,</span>
<span>                tx2gene <span class="op">=</span> <span class="va">tx2gene</span>,</span>
<span>                countsFromAbundance <span class="op">=</span> <span class="st">"lengthScaledTPM"</span><span class="op">)</span></span>
<span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">txi</span>, file <span class="op">=</span> <span class="st">"results/salmon_quant/txi.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p><code>txi$counts</code> now contains gene-level counts suitable for
DESeq2.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(txi<span class="sc">$</span>counts)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>                      SRR33253285 SRR33253286 SRR33253287 SRR33253288</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>ENSMUSG00000000001<span class="fl">.5</span>     <span class="fl">473.4426</span>  <span class="fl">553.784929</span>  <span class="fl">826.996853</span>  <span class="fl">676.910126</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>ENSMUSG00000000003<span class="fl">.16</span>      <span class="fl">0.0000</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>ENSMUSG00000000028<span class="fl">.16</span>     <span class="fl">40.5783</span>   <span class="fl">33.966518</span>   <span class="fl">55.305501</span>   <span class="fl">97.845071</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>ENSMUSG00000000031<span class="fl">.20</span>      <span class="fl">0.0000</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>ENSMUSG00000000037<span class="fl">.18</span>     <span class="fl">19.2220</span>    <span class="fl">9.599673</span>   <span class="fl">16.951493</span>   <span class="fl">20.226549</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>ENSMUSG00000000049<span class="fl">.12</span>      <span class="fl">0.0000</span>    <span class="fl">5.022073</span>    <span class="fl">1.984551</span>    <span class="fl">1.016411</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>                      SRR33253289 SRR33253290</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>ENSMUSG00000000001<span class="fl">.5</span>   <span class="fl">724.117586</span>  <span class="fl">813.848375</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>ENSMUSG00000000003<span class="fl">.16</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>ENSMUSG00000000028<span class="fl">.16</span>   <span class="fl">41.691461</span>   <span class="fl">63.812279</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>ENSMUSG00000000031<span class="fl">.20</span>    <span class="fl">0.000000</span>    <span class="fl">0.000000</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>ENSMUSG00000000037<span class="fl">.18</span>   <span class="fl">19.680740</span>   <span class="fl">18.945058</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>ENSMUSG00000000049<span class="fl">.12</span>    <span class="fl">0.993724</span>    <span class="fl">9.186882</span></span></code></pre>
</div>
<div id="why-use-lengthscaledtpm" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-lengthscaledtpm" class="callout-inner">
<h3 class="callout-title">Why use lengthScaledTPM?</h3>
<div class="callout-content">
<p>This method leverages Salmon’s bias correction while generating
stable counts for statistical modeling.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-qc-of-quantification-metrics">Step 5: QC of quantification metrics<a class="anchor" aria-label="anchor" href="#step-5-qc-of-quantification-metrics"></a>
</h2>
<hr class="half-width">
<p>Use MultiQC to verify:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$SCRATCH</span>/rnaseq-workshop</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="ex">module</span> load multiqc</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="ex">multiqc</span> results/salmon_quant <span class="at">-o</span> results/qc_salmon</span></code></pre>
</div>
<p>Inspect:</p>
<ul>
<li>mapping rates</li>
<li>fragment length distributions</li>
<li>library type consistency</li>
</ul>
<div id="exercise-examine-your-salmon-outputs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-examine-your-salmon-outputs" class="callout-inner">
<h3 class="callout-title">Exercise: examine your Salmon outputs</h3>
<div class="callout-content">
<p>Using your MultiQC summary and Salmon fragment length plots:</p>
<ol style="list-style-type: decimal">
<li>What is the percent aligned for each sample, and are any samples
outliers?</li>
<li>Is the inferred library type consistent across samples?</li>
<li>How similar are the fragment length distributions across
samples?</li>
<li>Does the fragment length curve show unusual peaks or
multimodality?</li>
<li>Do CFR and M bias values suggest any strand or positional bias?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Example interpretation for this dataset:</p>
<ol style="list-style-type: decimal">
<li>All samples show about 89 to 91 percent aligned. No sample is an
outlier and the range is narrow.</li>
<li>The inferred library type is consistent across samples and matches
an unstranded protocol.</li>
<li>Fragment length distributions are nearly identical. All samples peak
around the same length and have similar spread.</li>
<li>The fragment length curves are smooth and unimodal. There are no
secondary peaks or irregular shoulders.</li>
<li>CFR is 100 percent for all samples and M bias is about 0.5,
indicating balanced, strand-neutral coverage without detectable end
bias.</li>
</ol>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Salmon performs fast, alignment-free transcript quantification.</li>
<li>A transcriptome FASTA is required for building the index.</li>
<li>
<code>salmon quant</code> uses FASTQ files directly with bias
correction.</li>
<li>Transcript-level outputs include TPM and estimated counts.</li>
<li>
<code>tximport</code> converts transcript estimates to gene-level
counts.</li>
<li>Gene-level counts from Salmon are suitable for DESeq2.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-05-deseq2-expression-analyses"><p>Content from <a href="05-deseq2-expression-analyses.html">Gene-level QC and differential expression (DESeq2)</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05-deseq2-expression-analyses.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 85 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we explore RNA seq count data before running DESeq2.</li>
<li>How do we restrict analysis to protein coding genes.</li>
<li>How do we perform differential expression with DESeq2.</li>
<li>How do we visualize sample relationships and DE genes.</li>
<li>How do we save a useful DE results table with annotation and
expression values.</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Import count and sample metadata into R.</li>
<li>Attach biotype and gene symbol annotation to Ensembl gene IDs.</li>
<li>Explore library sizes, variance stabilizing transforms, distances,
and PCA.</li>
<li>Run DESeq2 for a two group comparison.</li>
<li>Create a volcano plot with gene symbols.</li>
<li>Export joined DE results including normalized counts and
annotation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this episode we move from <strong>count matrices</strong> to
<strong>differential expression analysis</strong> and
<strong>visualization</strong>.</p>
<p>The workflow has four parts:</p>
<ol style="list-style-type: decimal">
<li>Import counts and sample metadata.</li>
<li>Exploratory analysis on protein coding genes (quality checks).</li>
<li>Differential expression analysis with DESeq2.</li>
<li>Visualization and export of annotated DE results.</li>
</ol>
<div id="what-you-need-for-this-episode" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-you-need-for-this-episode" class="callout-inner">
<h3 class="callout-title">What you need for this episode</h3>
<div class="callout-content">
<ul>
<li>
<code>gene_counts_clean.txt</code> generated from
featureCounts<br>
</li>
<li>a <code>samples.csv</code> file describing the experimental
groups<br>
</li>
<li>RStudio session on Scholar using Open OnDemand</li>
</ul>
<p>While we created the count matrix in the previous episode, we still
need to create a sample metadata file. This file should contain at least
two columns: sample names matching the count matrix column names, and
the experimental condition (e.g., control vs treatment). Simply
copy/paste the following into a text file and save it in your
<code>scripts</code> directory as <code>samples.csv</code>:</p>
<pre><code>sample,condition
SRR33253285,Nat10-CKO
SRR33253286,Nat10-CKO
SRR33253287,Nat10-CKO
SRR33253288,Nat10floxflox
SRR33253289,Nat10flox/flox
SRR33253290,Nat10flox/flox</code></pre>
<p>Also, create a direcotry for DESeq2 results:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/deseq2</span></code></pre>
</div>
</div>
</div>
</div>
<p>All analyses are performed in R through Open OnDemand</p>
</section><section><h2 class="section-heading" id="step-1-start-open-ondemand-r-session-and-prepare-data">Step 1: Start Open OnDemand R session and prepare data<a class="anchor" aria-label="anchor" href="#step-1-start-open-ondemand-r-session-and-prepare-data"></a>
</h2>
<hr class="half-width">
<p>We will be using the OOD to start an interactive session on Scholar:
<a href="https://gateway.negishi.rcac.purdue.edu/" class="external-link">https://gateway.negishi.rcac.purdue.edu</a></p>
<ol style="list-style-type: decimal">
<li>Login using your Purdue credentials after clicking the above
link</li>
<li>Click on “Interactive Apps” in the top menu, and select “RStudio
(Bioconductor)”</li>
<li>Fill the job submission form as follows:
<ul>
<li>queue: <code>workshop</code>
</li>
<li>Walltime: <code>4</code>
</li>
<li>Number of cores: <code>4</code>
</li>
</ul>
</li>
<li>Click “Launch” and wait for the RStudio session to start</li>
<li>Once the session starts, you’ll will be able to click on “Connect to
RStudio server” which will open the RStudio interface.</li>
</ol>
<figure style="text-align: center"><img src="../fig/05_deseq/open-on-demand.png" alt="Open OnDemand interface" class="figure mx-auto d-block"><figcaption>
Open OnDemand interface
</figcaption></figure><p>Once in RStudio, load the necessary libraries:</p>
<div class="section level3">
<h3 id="setup-load-packages-and-data">Setup: load packages and data<a class="anchor" aria-label="anchor" href="#setup-load-packages-and-data"></a>
</h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Mmusculus.v79</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ensembldb</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggrepel</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">vsn</span><span class="op">)</span></span>
<span><span class="co"># Modify the path as needed</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="st">"/scratch/negishi/aseethar/rnaseq-workshop"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">countsFile</span> <span class="op">&lt;-</span> <span class="st">"results/counts/gene_counts_clean.txt"</span></span>
<span><span class="co"># prepare this file first!</span></span>
<span><span class="va">groupFile</span>  <span class="op">&lt;-</span> <span class="st">"scripts/samples.csv"</span></span>
<span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="va">groupFile</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cts</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">read.delim</span><span class="op">(</span><span class="va">countsFile</span>, row.names <span class="op">=</span> <span class="fl">1</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>We also load gene level annotation tables prepared earlier.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/mart.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read.csv</span><span class="op">(</span></span>
<span>    <span class="st">"data/annot.tsv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> How were these data prepared? </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<p>Since the gene IDs in the count matrix are Ensembl IDs, it will be
hard to interpret the results without annotation. We will need to use
the <code>org.Mm.eg.db</code> package for attaching gene symbols, so we
can interpret the results better.0</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># select Mart</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>ensembl <span class="ot">=</span> <span class="fu">useMart</span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># we need to use the correct dataset for mouse</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co"># we will query the available datasets and filter for mouse (GRCm39)</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="fu">listDatasets</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(description, <span class="st">"GRCm39"</span>))</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># we can use the dataset name to set the dataset</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>ensembl <span class="ot">=</span> <span class="fu">useDataset</span>(<span class="st">"mmusculus_gene_ensembl"</span>, <span class="at">mart =</span> ensembl)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># our counts have Ensembl gene IDs with version numbers</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co"># lets find how it is referred in biomaRt</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="fu">listFilters</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(name, <span class="st">"ensembl"</span>))</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co"># so we will now set the filter type accordingly</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>filterType <span class="ot">&lt;-</span> <span class="st">"ensembl_gene_id_version"</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co"># from our counts data, get the list of gene IDs</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>filterValues <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co"># take a look at the available attributes (first 20)</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="fu">listAttributes</span>(ensembl) <span class="sc">%&gt;%</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>     <span class="fu">head</span>(<span class="dv">20</span>)</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co"># we will retrieve ensembl gene id, ensembl gene id with version and external gene name</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>attributeNames <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'ensembl_gene_id'</span>,</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>                    <span class="st">'ensembl_gene_id_version'</span>,</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>                    <span class="st">'external_gene_name'</span>,</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>                    <span class="st">'gene_biotype'</span>,</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>                    <span class="st">'description'</span>)</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="co"># get the annotation</span></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>annot <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>  <span class="at">attributes =</span> attributeNames,</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>  <span class="at">filters =</span> filterType,</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>  <span class="at">values =</span> filterValues,</span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>  <span class="at">mart =</span> ensembl</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>)</span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">saveRDS</span>(annot, <span class="at">file =</span> <span class="st">"results/counts/gene_annotation.rds"</span>)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co"># also save as tsv for future use</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a><span class="fu">write.table</span>(</span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a>  annot,</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"data/mart.tsv"</span>,</span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>  <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="cn">FALSE</span></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a>)</span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="co"># save a simplified annotation file with only essential columns</span></span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a>mart <span class="sc">%&gt;%</span> </span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a><span class="fu">select</span>(ensembl_gene_id, ensembl_gene_id_version, external_gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a>  <span class="fu">write.table</span>(</span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a>    <span class="at">file =</span> <span class="st">"data/annot.tsv"</span>,</span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a>    <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">FALSE</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The objects loaded here form the foundation for the entire workflow.
<code>cts</code> contains raw gene counts from featureCounts.
<code>coldata</code> contains the metadata that describes each sample
and defines the comparison groups. <code>mart</code> and
<code>annot</code> contain functional information for each gene so we
can attach gene symbols and biotypes later in the analysis.</p>
<div id="sample-metadata-and-design" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="sample-metadata-and-design" class="callout-inner">
<h3 class="callout-title">Sample metadata and design</h3>
<div class="callout-content">
<p><code>coldata</code> must have row names that match the count matrix
column names. The <code>condition</code> column defines the groups for
DESeq2 (<code>Nat10-CKO</code> vs <code>Nat10flox/flox</code>).</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exploratory-analysis">Exploratory analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="attach-annotation-and-filter-to-protein-coding-genes">Attach annotation and filter to protein coding genes<a class="anchor" aria-label="anchor" href="#attach-annotation-and-filter-to-protein-coding-genes"></a>
</h3>
<p>We first join counts with biotype information from <code>mart</code>
and filter to protein coding genes.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cts_tbl</span> <span class="op">&lt;-</span> <span class="va">cts</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cts_annot</span> <span class="op">&lt;-</span> <span class="va">cts_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">mart</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="fu">is.na</span><span class="op">(</span><span class="va">cts_annot</span><span class="op">$</span><span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Joining annotation allows us to move beyond raw Ensembl IDs. This
step attaches gene symbols, biotypes, and descriptions so that later
plots and tables are interpretable. At this point the dataset is still
large and includes many noncoding genes, pseudogenes, and biotypes we
won’t analyze further.</p>
<p>Summarize gene biotypes:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biotype_df</span> <span class="op">&lt;-</span> <span class="va">cts_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">count</span><span class="op">(</span><span class="va">gene_biotype</span>, name <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">biotype_df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu">reorder</span><span class="op">(</span><span class="va">gene_biotype</span>, <span class="va">n</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_col</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Gene biotype"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Number of genes"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Distribution of gene biotypes in annotation"</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/gene-biotypes.png" alt="Distribution of gene biotypes in annotation" class="figure mx-auto d-block"><figcaption>
Distribution of gene biotypes in annotation
</figcaption></figure><p>Visualizing biotype composition gives a sense of how many genes
belong to each functional class. RNA-seq quantifies all transcribed
loci, but many biotypes (snRNA, TEC, pseudogenes) are not suitable for
differential expression with DESeq2 because they tend to be lowly
expressed or unstable. This motivates filtering to protein-coding genes
for QC and visualization.</p>
<p>Filter to protein coding genes and lowly expressed genes:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cts_coding</span> <span class="op">&lt;-</span> <span class="va">cts_annot</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">ensembl_gene_id_version</span>, <span class="fu">all_of</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">cts_coding</span><span class="op">)</span></span></code></pre>
</div>
<pre><code>[1] 15839     6</code></pre>
<div id="why-filter-protein-coding-genes-and-low-expression" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-filter-protein-coding-genes-and-low-expression" class="callout-inner">
<h3 class="callout-title">Why filter protein coding genes and low
expression</h3>
<div class="callout-content">
<p>Noncoding biotypes (lncRNA, pseudogene, etc.) can be biologically
important. Filtering lowly expressed genes reduces noise in exploratory
analyses. Genes with near-zero counts across all samples inflate
variance estimates and obscure real biological structure. The threshold
of 5 total counts is a conservative filter that keeps genes with minimal
but genuine expression while removing background noise.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="library-size-differences">Library size differences<a class="anchor" aria-label="anchor" href="#library-size-differences"></a>
</h3>
<p>We summarize total counts per sample for protein coding genes.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">libSize</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="va">cts_coding</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rename</span><span class="op">(</span>total_counts <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>total_millions <span class="op">=</span> <span class="va">total_counts</span> <span class="op">/</span> <span class="fl">1e6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">libSize</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">total_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_col</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Total assigned reads"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Total reads per sample"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/libsize-difference.png" alt="Total assigned reads per sample" class="figure mx-auto d-block"><figcaption>
Total assigned reads per sample
</figcaption></figure><p>Library size differences affect normalization. Although DESeq2
corrects for this through size factors, inspecting raw library sizes
ensures no sample has abnormally low sequencing depth, which could
compromise downstream analyses.</p>
</div>
<div class="section level3">
<h3 id="create-deseq2-object-for-exploratory-plots">Create DESeq2 object for exploratory plots<a class="anchor" aria-label="anchor" href="#create-deseq2-object-for-exploratory-plots"></a>
</h3>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>  countData <span class="op">=</span> <span class="va">cts_coding</span>,</span>
<span>  colData   <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>  design    <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<p>Inspect size factors vs library size:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span></span>
<span>  <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    libSize    <span class="op">=</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    sizeFactor <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,</span>
<span>    Group      <span class="op">=</span> <span class="va">dds</span><span class="op">$</span><span class="va">condition</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">libSize</span>, y <span class="op">=</span> <span class="va">sizeFactor</span>, col <span class="op">=</span> <span class="va">Group</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size"</span>, y <span class="op">=</span> <span class="st">"Size factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/size-factors-vs-lib-size.png" alt="Size factors vs library size" class="figure mx-auto d-block"><figcaption>
Size factors vs library size
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="variance-stabilization-and-mean-sd-plots">Variance stabilization and mean SD plots<a class="anchor" aria-label="anchor" href="#variance-stabilization-and-mean-sd-plots"></a>
</h3>
<p>Raw count data have a strong mean–variance dependency: low count
genes have high variance, and high count genes show lower relative
variability. This makes Euclidean distance and PCA misleading on raw
counts. VST removes this dependency so that downstream clustering is
biologically meaningful rather than driven by count scale.</p>
<p>Inspect mean vs standard deviation plot before transformation:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/before-vsd.png" alt="average counts vs variance" class="figure mx-auto d-block"><figcaption>
average counts vs variance
</figcaption></figure><p>Apply variance stabilizing transform (VST):</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/after-vsd.png" alt="average counts vs variance after transformation" class="figure mx-auto d-block"><figcaption>
average counts vs variance after transformation
</figcaption></figure><div id="why-use-vst" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-use-vst" class="callout-inner">
<h3 class="callout-title">Why use VST</h3>
<div class="callout-content">
<p>Counts have mean dependent variance. Variance stabilizing transform
(VST) makes the variance roughly constant across the range, which
improves distance based methods and PCA.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sample-to-sample-distances">Sample to sample distances<a class="anchor" aria-label="anchor" href="#sample-to-sample-distances"></a>
</h3>
<p>The distance heatmap helps answer a key question: Do samples group by
biological condition or by an unwanted variable (batch, sequencing run,
contamination)? This is one of the fastest ways to detect outliers.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleDists</span>       <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sampleDistMatrix</span>  <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">sampleDists</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleDistMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">rev</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Blues"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span></span>
<span>  <span class="va">sampleDistMatrix</span>,</span>
<span>  clustering_distance_rows <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>  clustering_distance_cols <span class="op">=</span> <span class="va">sampleDists</span>,</span>
<span>  col                      <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/clustering.png" alt="Eucledean distance heatmap" class="figure mx-auto d-block"><figcaption>
Eucledean distance heatmap
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="pca-on-top-variable-genes">PCA on top variable genes<a class="anchor" aria-label="anchor" href="#pca-on-top-variable-genes"></a>
</h3>
<p>PCA projects samples into low-dimensional space and reveals dominant
sources of variation. Ideally, samples separate by biological condition
along one of the top PCs. If not, the signal from the condition may be
weak or confounded by other factors.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">select</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">rv</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fu">seq_len</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="fl">500</span>, <span class="fu">length</span><span class="op">(</span><span class="va">rv</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">select</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="va">pca</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">intgroup</span>    <span class="op">&lt;-</span> <span class="st">"condition"</span></span>
<span><span class="va">intgroup.df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span>, <span class="va">intgroup</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">intgroup</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">factor</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">intgroup.df</span>, <span class="fl">1</span>, <span class="va">paste</span>, collapse <span class="op">=</span> <span class="st">" : "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>  PC1 <span class="op">=</span> <span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  PC2 <span class="op">=</span> <span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  <span class="va">intgroup.df</span>,</span>
<span>  name <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">d</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_shape_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"Nat10-CKO"</span>       <span class="op">=</span> <span class="st">"#00b462"</span>,</span>
<span>    <span class="st">"Nat10flox/flox"</span>  <span class="op">=</span> <span class="st">"#980c80"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, stroke <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text_repel</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"PC1"</span>, <span class="fu">round</span><span class="op">(</span><span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"PC2"</span>, <span class="fu">round</span><span class="op">(</span><span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/pca-plot.png" alt="PCA plot showing sample clustering" class="figure mx-auto d-block"><figcaption>
PCA plot showing sample clustering
</figcaption></figure><div id="exercise-interpret-exploratory-plots" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-interpret-exploratory-plots" class="callout-inner">
<h3 class="callout-title">Exercise: interpret exploratory plots</h3>
<div class="callout-content">
<p>Using the library size, distance heatmap, and PCA:</p>
<ol style="list-style-type: decimal">
<li>Are there any obvious outlier samples.</li>
<li>Do replicates from the same condition cluster together.</li>
<li>Does any sample look inconsistent with its metadata label.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Example interpretation:</p>
<ol style="list-style-type: decimal">
<li>Library sizes are comparable across samples and fall within a narrow
range.</li>
<li>Samples cluster by condition in both distance heatmaps and PCA.</li>
<li>No sample appears as a clear outlier relative to its group.</li>
</ol>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-gene-expression-analysis">Differential gene expression analysis<a class="anchor" aria-label="anchor" href="#differential-gene-expression-analysis"></a>
</h2>
<hr class="half-width">
<p>For DE testing we use the <strong>full count matrix</strong>
(<code>cts</code>) without filtering by biotype. This preserves all
features counted by featureCounts.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>  countData <span class="op">=</span> <span class="va">cts_coding</span>,</span>
<span>  colData   <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>  design    <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>You can either run the individual DESeq2 steps:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateDispersions</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">nbinomWaldTest</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Single command to run the DESeq2 pipeline </h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" aria-labelledby="headingSpoiler2" data-bs-parent="#accordionSpoiler2">
<div class="accordion-body">
<p>Above three steps can be run together using the <code>DESeq</code>
function:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code>estimating size factors
  Note: levels of factors in the design contain characters other than
  letters, numbers, '_' and '.'. It is recommended (but not required) to use
  only letters, numbers, and delimiters '_' or '.', as these are safe characters
  for column names in R. [This is a message, not a warning or an error]
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
  Note: levels of factors in the design contain characters other than
  letters, numbers, '_' and '.'. It is recommended (but not required) to use
  only letters, numbers, and delimiters '_' or '.', as these are safe characters
  for column names in R. [This is a message, not a warning or an error]
final dispersion estimates
fitting model and testing</code></pre>
</div>
</div>
</div>
</div>
<p>Inspect dispersion estimates:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/dispersion-plot.png" alt="Dispersion estimates from DESeq2" class="figure mx-auto d-block"><figcaption>
Dispersion estimates from DESeq2
</figcaption></figure><p>Obtain results for the contrast of interest
(<code>Nat10flox/flox</code> vs <code>Nat10-CKO</code>):</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">results</span><span class="op">(</span></span>
<span>    <span class="va">dds</span>,</span>
<span>    contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>      <span class="st">"condition"</span>,</span>
<span>      <span class="st">"Nat10flox/flox"</span>,</span>
<span>      <span class="st">"Nat10-CKO"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">TXT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode txt" tabindex="0"><code class="sourceCode default"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>out of 15839 with nonzero total read count</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>adjusted p-value &lt; 0.1</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>LFC &gt; 0 (up)       : 3195, 20%</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>LFC &lt; 0 (down)     : 3398, 21%</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>outliers [1]       : 362, 2.3%</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>low counts [2]     : 308, 1.9%</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>(mean count &lt; 1)</span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>[1] see 'cooksCutoff' argument of ?results</span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>[2] see 'independentFiltering' argument of ?results</span></code></pre>
</div>
<p>Order the results by p value to see the top DE genes:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code>log2 fold change (MLE): condition Nat10flox/flox vs Nat10-CKO
Wald test p-value: condition Nat10flox.flox vs Nat10.CKO
DataFrame with 6 rows and 6 columns
       baseMean log2FoldChange     lfcSE      stat       pvalue         padj
      &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;
18512   4645.33       -5.17453 0.1335387  -38.7493  0.00000e+00  0.00000e+00
940     5041.68       -2.88015 0.1039939  -27.6953 7.94364e-169 6.02485e-165
13789  14362.27       -2.40229 0.0951377  -25.2506 1.11514e-140 5.63852e-137
16183  42766.77       -4.68585 0.1923780  -24.3575 4.82857e-131 1.83111e-127
18278   7055.06       -3.48409 0.1535305  -22.6932 5.23170e-114 1.58719e-110
270     1469.83       -2.87951 0.1298316  -22.1788 5.50606e-109 1.39202e-105</code></pre>
<p>Each row represents one gene. Key columns: <code>baseMean</code>:
average normalized count across all samples <code>log2FoldChange</code>:
direction and magnitude of change between groups <code>pvalue</code> /
<code>padj</code>: statistical significance after multiple testing
correction Genes with <code>NA</code> values often have too few reads to
estimate dispersion and should not be interpreted.</p>
<p>Create a summary table:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="va">summary_table</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span></span>
<span>    total_genes <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    up  <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span>  <span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    down<span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res_df</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">res_df</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">log2fc_cut</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">summary_table</span><span class="op">)</span></span></code></pre>
</div>
<pre class="text"><code># A tibble: 1 × 4
  total_genes   sig    up  down
        &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1       15839  5658  2180  2363</code></pre>
<p>Keep gene IDs:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">res_df</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="joining-results-with-expression-and-annotation">Joining results with expression and annotation<a class="anchor" aria-label="anchor" href="#joining-results-with-expression-and-annotation"></a>
</h2>
<hr class="half-width">
<p>Get normalized counts:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_counts</span>    <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">norm_counts_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">norm_counts</span><span class="op">)</span></span>
<span><span class="va">norm_counts_df</span><span class="op">$</span><span class="va">ensembl_gene_id_version</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">norm_counts_df</span><span class="op">)</span></span></code></pre>
</div>
<p>Join DE results, normalized counts, and annotation:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">norm_counts_df</span>,</span>
<span>                   by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">mart</span>,</span>
<span>                   by <span class="op">=</span> <span class="st">"ensembl_gene_id_version"</span><span class="op">)</span></span></code></pre>
</div>
<p>Joining normalized counts and annotation makes the output
biologically interpretable. The final table becomes a comprehensive
results object containing: - <code>gene identifiers</code> -
<code>gene symbols</code> - <code>functional descriptions</code> -
<code>differential expression statistics</code> -
<code>normalized expression values</code> This is the format most
researchers expect when examining results or importing them into
downstream tools.</p>
<p>Define significance labels for plotting:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2fc_cut</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_annot</span> <span class="op">&lt;-</span> <span class="va">res_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span></span>
<span>    label <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">coalesce</span><span class="op">(</span><span class="va">external_gene_name</span>, <span class="va">ensembl_gene_id_version</span><span class="op">)</span>,</span>
<span>    sig <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&gt;=</span>  <span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"up"</span>,</span>
<span>      <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&lt;=</span> <span class="op">-</span><span class="va">log2fc_cut</span> <span class="op">~</span> <span class="st">"down"</span>,</span>
<span>      <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"ns"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="visualization-volcano-plot">Visualization: volcano plot<a class="anchor" aria-label="anchor" href="#visualization-volcano-plot"></a>
</h2>
<hr class="half-width">
<p>The volcano plot summarizes statistical significance
(<code>padj</code>) and effect size (<code>log2FoldChange</code>)
simultaneously. Genes far left or right show strong expression
differences. Genes high on the plot pass stringent significance
thresholds. Labels help identify the most biologically meaningful
genes.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span></span>
<span>  <span class="va">res_annot</span>,</span>
<span>  <span class="fu">aes</span><span class="op">(</span></span>
<span>    x     <span class="op">=</span> <span class="va">log2FoldChange</span>,</span>
<span>    y     <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>    col   <span class="op">=</span> <span class="va">sig</span>,</span>
<span>    label <span class="op">=</span> <span class="va">label</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"up"</span>   <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>    <span class="st">"down"</span> <span class="op">=</span> <span class="st">"dodgerblue3"</span>,</span>
<span>    <span class="st">"ns"</span>   <span class="op">=</span> <span class="st">"grey70"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text_repel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span><span class="va">res_annot</span>, <span class="va">sig</span> <span class="op">!=</span> <span class="st">"ns"</span><span class="op">)</span>,</span>
<span>    max.overlaps       <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    min.segment.length <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>    box.padding        <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    seed               <span class="op">=</span> <span class="fl">42</span>,</span>
<span>    show.legend        <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10 adjusted p value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Volcano plot: Nat10-CKO vs Nat10flox/flox"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/05_deseq/volcano-plot.png" alt="Volcano plot of differential expression results" class="figure mx-auto d-block"><figcaption>
Volcano plot of differential expression results
</figcaption></figure><div id="interpreting-the-volcano" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-the-volcano" class="callout-inner">
<h3 class="callout-title">Interpreting the volcano</h3>
<div class="callout-content">
<p>Points to the right are higher in <code>Nat10flox/flox</code>, to the
left are higher in <code>Nat10-CKO</code>. Points near the top have
stronger statistical support (small adjusted p values). Colored points
mark genes that pass both fold change and padj thresholds.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="saving-results">Saving results<a class="anchor" aria-label="anchor" href="#saving-results"></a>
</h2>
<hr class="half-width">
<p>Saving both the full table and the significant subset allows
flexibility: - The full table is needed for pathway analyses and
reproducibility. - The subset table focuses on interpretable and
biologically meaningful DE genes.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">readr</span><span class="fu">::</span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>  <span class="va">res_annot</span>,</span>
<span>  <span class="st">"results/deseq2/DESeq2_results_joined.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Filter to significant genes (padj ≤ 0.05 and fold change ≥ 1.5):</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_res</span> <span class="op">&lt;-</span> <span class="va">res_annot</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span></span>
<span>    <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>    <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">log2fc_cut</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu">write_tsv</span><span class="op">(</span></span>
<span>  <span class="va">sig_res</span>,</span>
<span>  <span class="st">"results/deseq2/DESeq2_results_sig.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="exercise-explore-top-de-genes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-explore-top-de-genes" class="callout-inner">
<h3 class="callout-title">Exercise: explore top DE genes</h3>
<div class="callout-content">
<p>Using <code>sig_res</code>:</p>
<ol style="list-style-type: decimal">
<li>Which genes have the largest positive log2 fold change.</li>
<li>Which genes have the largest negative log2 fold change.</li>
<li>Are any of the top genes known to be related to Nat10 function or
the studied phenotype.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Example interpretation:</p>
<ul>
<li>Several strongly up regulated genes are involved in cell cycle and
RNA processing.</li>
<li>Strongly down regulated genes may include transcription factors and
signaling components.</li>
<li>Cross referencing top hits with prior literature or databases can
reveal mechanistic hypotheses.</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Counts and sample metadata must be aligned before building DESeq2
objects.</li>
<li>Biotype annotation allows restriction to protein coding genes for
exploratory QC.</li>
<li>Variance stabilizing transform, distance heatmaps, and PCA help
detect outliers and batch effects.</li>
<li>DESeq2 provides a complete pipeline from raw counts to differential
expression.</li>
<li>Joining DE results with normalized counts and gene annotation yields
a useful final table.</li>
<li>Volcano plots summarize differential expression using fold change
and adjusted p values.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-05b-deseq2-expression-analyses-kallisto"><p>Content from <a href="05b-deseq2-expression-analyses-kallisto.html">Differential expression using DESeq2</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/05b-deseq2-expression-analyses-kallisto.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 85 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>This script should work for many simple designs ranging from two
samples (i.e., control/treatment) to designs with large numbers of
samples, for instance a time course. The first sample group is used as
the denominator in the fold-change calculations. This version is set up
for the simple two sample case, but you can switch to the code in
comments for more complicated setups.</p>
<p>Required packages: DESeq2 (bioconductor) ggplot2 reshape2 ashr or
apelgm (for shrunken LFC calculation) scales (volcano plot)</p>
<p>For bioconductor packages such as DESeq2, you need to load using
BiocManager.</p>
</section><section><h2 class="section-heading" id="attribution">Attribution<a class="anchor" aria-label="anchor" href="#attribution"></a>
</h2>
<hr class="half-width">
<p>This section is adapted from materials developed by <strong>Michael
Gribskov</strong>, Professor of Computational Genomics &amp; Systems
Biology at Purdue University.</p>
<p>Original and related materials are available via the <strong>CGSB
Wiki</strong>: <a href="https://cgsb.miraheze.org/wiki/Main_Page" class="external-link uri">https://cgsb.miraheze.org/wiki/Main_Page</a></p>
<p>Michael Gribskov v 2.4 updated: 2025 December 12</p>
<div id="what-you-need-for-this-episode" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<span class="callout-header">Prerequisite</span>
<div id="what-you-need-for-this-episode" class="callout-inner">
<h3 class="callout-title">What you need for this episode</h3>
<div class="callout-content">

</div>
</div>
</div>
</section><section><h2 class="section-heading" id="working-directory-and-setup">Working directory and setup<a class="anchor" aria-label="anchor" href="#working-directory-and-setup"></a>
</h2>
<hr class="half-width">
<p>It’s easiest to work in the directory where your data is.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">getwd</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="st">"A:/mrg/Dropbox/projects/25plantago"</span><span class="op">)</span></span>
<span><span class="fu">dir</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="package-installation-and-loading">Package installation and loading<a class="anchor" aria-label="anchor" href="#package-installation-and-loading"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="reading-raw-read-count-data">Reading raw read count data<a class="anchor" aria-label="anchor" href="#reading-raw-read-count-data"></a>
</h2>
<hr class="half-width">
<p>Read in raw read count data from merged salmon output (NumReads
column). You may want to adjust row and column labels to be shorter or
more informative. It will be convenient if you make the first sample
group the one that will be the denominator in the log fold change
calculation.</p>
<p>Normally this will correctly read the rownames and column names from
the counts file.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">salmon</span> <span class="op">&lt;-</span> <span class="fu">read.delim</span><span class="op">(</span><span class="st">"JM13_18_2.numreads.out"</span>, row.names<span class="op">=</span><span class="fl">1</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span></code></pre>
</div>
<p>If the rownames and colnames are OK, skip ahead to raw counts.</p>
<p>Possible adjustments to rows and columns include reordering columns,
selecting subsets, manually moving row or column names, and editing
names to remove redundant strings.</p>
<p>Remove suffix <code>.salmon2</code> from column names.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names</span>  <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span>
<span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu">sub</span><span class="op">(</span><span class="st">".salmon2"</span>,<span class="st">""</span>,<span class="va">names</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="raw-counts-quality-control">Raw counts quality control<a class="anchor" aria-label="anchor" href="#raw-counts-quality-control"></a>
</h2>
<hr class="half-width">
<p>The count matrix from the previous section is named
<code>salmon</code>.</p>
<p>Total counts per sample can be checked to verify agreement with read
counts and expected dimensions.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span>
<span><span class="fu">colSums</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span></span></code></pre>
</div>
<p>Boxplot and density plots of raw counts provide a visual summary of
count distributions.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">melted</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span>, value.name<span class="op">=</span><span class="st">"count"</span>, variable<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>,<span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">sample</span>, y<span class="op">=</span><span class="va">count</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_boxplot</span><span class="op">(</span>show.legend<span class="op">=</span><span class="cn">F</span> <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_y_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Un-normalized Counts"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">"log( Count )"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Sample"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">count</span>,fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, limits<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">100000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">annotation_logticks</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Un-normalized Counts"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">count</span>,fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.3</span>, show.legend<span class="op">=</span><span class="cn">F</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, limits<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">100000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Un-normalized Counts"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="optional-low-count-filtering">Optional low-count filtering<a class="anchor" aria-label="anchor" href="#optional-low-count-filtering"></a>
</h2>
<hr class="half-width">
<p>Optional removal of very low count transcripts. A threshold of
approximately 5 reads per sample is typical, but this is experiment
specific. Be careful not to remove genes expressed in only one
group.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="fu">table</span><span class="op">(</span> <span class="fu">rowSums</span><span class="op">(</span><span class="va">salmon</span><span class="op">[</span>,<span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">cutoff</span> <span class="op">)</span></span>
<span><span class="va">cut</span> <span class="op">&lt;-</span> <span class="va">salmon</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">salmon</span><span class="op">[</span>,<span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">cutoff</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">melted</span>  <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="va">cut</span>, variable<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="va">title</span> <span class="op">&lt;-</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"Totals counts &gt; %d"</span>, <span class="va">cutoff</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">value</span>,fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, limits<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.01</span>,<span class="fl">100000000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">annotation_logticks</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="va">title</span><span class="op">)</span></span></code></pre>
</div>
<p>Rename filtered data for downstream analysis.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">salmon</span> <span class="op">&lt;-</span> <span class="va">cut</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="preliminary-normalization-with-deseq2">Preliminary normalization with DESeq2<a class="anchor" aria-label="anchor" href="#preliminary-normalization-with-deseq2"></a>
</h2>
<hr class="half-width">
<p>Preliminary normalization uses one pass of DESeq2 to estimate size
factors and enable prefiltering.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">replicates</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"Sminus"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="st">"Splus"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>sampleNames<span class="op">=</span><span class="fu">colnames</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span>, sample<span class="op">=</span><span class="va">replicates</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_init</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span> <span class="fu">round</span><span class="op">(</span><span class="va">salmon</span><span class="op">)</span>, colData<span class="op">=</span><span class="va">metadata</span>, design<span class="op">=</span><span class="op">~</span><span class="va">sample</span> <span class="op">)</span></span>
<span><span class="va">init</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">count_init</span><span class="op">)</span></span>
<span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">init</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">init</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">melted</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">init</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span>, variable<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">value</span>,fill<span class="op">=</span><span class="va">Var2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, limits<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">100000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">annotation_logticks</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Preliminary normalization"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="pre-filtering-based-on-normalized-counts">Pre-filtering based on normalized counts<a class="anchor" aria-label="anchor" href="#pre-filtering-based-on-normalized-counts"></a>
</h2>
<hr class="half-width">
<p>Prefiltering removes poorly measured genes unlikely to show
significant differential expression. Multiple cutoffs are evaluated to
guide selection.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trial</span> <span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">svector</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">15</span>, <span class="fl">20</span>, <span class="fl">25</span>, <span class="fl">30</span>, <span class="fl">40</span>, <span class="fl">50</span>, <span class="fl">60</span>, <span class="fl">70</span>, <span class="fl">80</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">min_over</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">nrep</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">nsample</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">levels</span><span class="op">(</span><span class="va">replicates</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">mincount</span> <span class="kw">in</span> <span class="va">cutoff</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">select</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">init</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">nrep</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="va">mincount</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_over</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">c</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsample</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">first</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">c</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">nrep</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>        <span class="va">last</span> <span class="op">&lt;-</span> <span class="va">first</span> <span class="op">+</span> <span class="va">nrep</span> <span class="op">-</span> <span class="fl">1</span></span>
<span>        <span class="va">select</span> <span class="op">&lt;-</span> <span class="va">select</span> <span class="op">|</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">init</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="va">first</span><span class="op">:</span><span class="va">last</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="va">mincount</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_over</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">trial</span><span class="op">[</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%d"</span>,<span class="va">mincount</span><span class="op">)</span><span class="op">]</span><span class="op">=</span><span class="fu">list</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">select</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">svector</span><span class="op">[</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%d"</span>,<span class="va">mincount</span><span class="op">)</span><span class="op">]</span><span class="op">=</span><span class="fu">list</span><span class="op">(</span><span class="va">select</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">cutoff</span> <span class="kw">in</span> <span class="fu">names</span><span class="op">(</span><span class="va">trial</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"cutoff:%4s     discard:%6d   keep:%6d"</span>, <span class="va">cutoff</span>, <span class="va">trial</span><span class="op">[[</span><span class="va">cutoff</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="st">"FALSE"</span><span class="op">]</span>, <span class="va">trial</span><span class="op">[[</span><span class="va">cutoff</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="st">"TRUE"</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="fu">print</span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cutoff_selected</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">min_selected</span> <span class="op">=</span> <span class="va">svector</span><span class="op">[[</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%d"</span>,<span class="va">cutoff_selected</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">min_selected</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">melted</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">init</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">[</span><span class="va">min_selected</span>,<span class="op">]</span><span class="op">)</span>, variable<span class="op">=</span><span class="st">"sample"</span>, varnames<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">title</span> <span class="op">&lt;-</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"Normalized counts selected for %d samples over %d counts"</span>, <span class="va">min_over</span>, <span class="va">cutoff_selected</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">melted</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">value</span>,fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, limits<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">100000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">annotation_logticks</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="va">title</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="final-differential-expression-analysis">Final differential expression analysis<a class="anchor" aria-label="anchor" href="#final-differential-expression-analysis"></a>
</h2>
<hr class="half-width">
<p>Final DESeq2 analysis is performed using the prefiltered counts.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected</span> <span class="op">&lt;-</span> <span class="va">salmon</span><span class="op">[</span><span class="va">min_selected</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">count_filtered</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span> <span class="fu">round</span><span class="op">(</span><span class="va">selected</span><span class="op">)</span>, colData<span class="op">=</span><span class="va">metadata</span>, design<span class="op">=</span><span class="op">~</span><span class="va">sample</span> <span class="op">)</span></span>
<span><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">count_filtered</span><span class="op">)</span></span>
<span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">filtered</span>,normalized<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="pca-with-variance-stabilization">PCA with variance stabilization<a class="anchor" aria-label="anchor" href="#pca-with-variance-stabilization"></a>
</h2>
<hr class="half-width">
<p>Variance stabilized PCA provides improved separation of biological
signal.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span><span class="fu">vst</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">filtered</span>, normalized<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                           colData<span class="op">=</span><span class="fu">colData</span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">sampleNames</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"C16"</span>,<span class="fl">12</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="st">"C31"</span>,<span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">times</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"T00"</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">"T48"</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">"T06"</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">"T72"</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">time</span> <span class="op">&lt;-</span>  <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="va">times</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pcaData</span> <span class="op">&lt;-</span> <span class="fu">plotPCA</span><span class="op">(</span><span class="fu">DESeqTransform</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, intgroup<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"sampleNames"</span>, <span class="st">"time"</span><span class="op">)</span>, returnData<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaData</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaData</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color<span class="op">=</span><span class="va">se</span><span class="op">$</span><span class="va">sampleNames</span>, shape<span class="op">=</span><span class="va">se</span><span class="op">$</span><span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Normalized/Filtered Counts - variance stabilizing transform"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_shape_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">15</span>,<span class="fl">16</span>,<span class="fl">17</span>,<span class="fl">18</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>,<span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>,<span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="differential-expression-contrasts">Differential expression contrasts<a class="anchor" aria-label="anchor" href="#differential-expression-contrasts"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nrep</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">foreground</span> <span class="op">&lt;-</span> <span class="fu">levels</span><span class="op">(</span><span class="va">filtered</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">background</span> <span class="op">=</span> <span class="va">foreground</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">f</span> <span class="kw">in</span> <span class="va">foreground</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">contrast_name</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="va">f</span>, <span class="va">background</span>, sep<span class="op">=</span><span class="st">"_vs_"</span><span class="op">)</span></span>
<span>    <span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">contrasts</span>, <span class="va">contrast_name</span><span class="op">)</span></span>
<span>    <span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">filtered</span>, contrast<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"sample"</span>, <span class="va">f</span>, <span class="va">background</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">assign</span><span class="op">(</span><span class="va">contrast_name</span>,<span class="va">contrast</span><span class="op">)</span></span>
<span>    <span class="fu">volcano</span><span class="op">(</span><span class="va">contrast</span>, <span class="fl">5e-1</span>, <span class="fu">log2</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, stattype<span class="op">=</span><span class="st">"padj"</span>, symmetric<span class="op">=</span><span class="cn">T</span> <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="saving-results">Saving results<a class="anchor" aria-label="anchor" href="#saving-results"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_file</span> <span class="op">=</span> <span class="st">"final.tsv"</span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">statout</span>, file<span class="op">=</span><span class="va">output_file</span>, sep<span class="op">=</span><span class="st">'\t'</span>,col.names<span class="op">=</span><span class="cn">NA</span>,row.names<span class="op">=</span><span class="cn">T</span>,quote<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">

</div>
</div>
</div>
</section><section><h2 class="section-heading" id="appendix-functions">Appendix: functions<a class="anchor" aria-label="anchor" href="#appendix-functions"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">volcano</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">contrast</span>, <span class="va">p_cut</span><span class="op">=</span><span class="fl">0.1</span>, <span class="va">lfc_cut</span><span class="op">=</span><span class="fl">1.0</span>, <span class="va">stattype</span><span class="op">=</span><span class="st">"svalue"</span>,</span>
<span>                     <span class="va">title</span><span class="op">=</span><span class="st">"Volcano Plot"</span>, <span class="va">xspace</span><span class="op">=</span><span class="fl">5</span>, <span class="va">yspace</span><span class="op">=</span><span class="fl">1</span>, <span class="va">symmetric</span><span class="op">=</span><span class="cn">FALSE</span> <span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sig</span> <span class="op">&lt;-</span> <span class="va">contrast</span><span class="op">[[</span><span class="va">stattype</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;</span><span class="va">p_cut</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">contrast</span><span class="op">[[</span><span class="va">stattype</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;</span><span class="va">p_cut</span><span class="op">)</span></span>
<span>    <span class="va">lfc</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="va">contrast</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">&gt;</span><span class="va">lfc_cut</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">contrast</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">&gt;</span><span class="va">lfc_cut</span><span class="op">)</span></span>
<span>    <span class="va">both</span> <span class="op">&lt;-</span> <span class="va">sig</span> <span class="op">&amp;</span> <span class="va">lfc</span></span>
<span>    <span class="va">neither</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="va">sig</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">lfc</span> </span>
<span>    <span class="va">other</span> <span class="op">&lt;-</span> <span class="fu">xor</span><span class="op">(</span><span class="va">sig</span>,<span class="va">lfc</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">title</span> <span class="op">==</span> <span class="st">"Volcano Plot"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"Volcano Plot -  %s"</span>, <span class="va">contrast_name</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="va">low</span> <span class="op">&lt;-</span> <span class="fu">min</span><span class="op">(</span><span class="va">contrast</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span></span>
<span>    <span class="va">high</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span><span class="va">contrast</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">symmetric</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">l</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">high</span><span class="op">)</span>, <span class="fu">abs</span><span class="op">(</span><span class="va">low</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">limits_x</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="va">l</span>, <span class="va">l</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">limits_x</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">low</span>, <span class="va">high</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">contrast</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">log2FoldChange</span>, y<span class="op">=</span><span class="fu">log10</span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">stattype</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">ggtitle</span><span class="op">(</span><span class="va">title</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">scale_x_continuous</span><span class="op">(</span>limits<span class="op">=</span><span class="va">limits_x</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_hline</span><span class="op">(</span>yintercept<span class="op">=</span><span class="fu">log10</span><span class="op">(</span><span class="va">p_cut</span><span class="op">)</span>, linewidth<span class="op">=</span><span class="fl">0.25</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_vline</span><span class="op">(</span>xintercept<span class="op">=</span><span class="va">lfc_cut</span>, linewidth<span class="op">=</span><span class="fl">0.25</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_vline</span><span class="op">(</span>xintercept<span class="op">=</span><span class="op">-</span><span class="va">lfc_cut</span>, linewidth<span class="op">=</span><span class="fl">0.25</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_point</span><span class="op">(</span>data<span class="op">=</span><span class="va">contrast</span><span class="op">[</span><span class="va">neither</span>,<span class="op">]</span>, color<span class="op">=</span><span class="st">"black"</span>, size<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_point</span><span class="op">(</span>data<span class="op">=</span><span class="va">contrast</span><span class="op">[</span><span class="va">other</span>,<span class="op">]</span>, shape<span class="op">=</span><span class="fl">21</span>, fill<span class="op">=</span><span class="st">"yellow"</span>, size<span class="op">=</span><span class="fl">2</span>, stroke<span class="op">=</span><span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_point</span><span class="op">(</span>data<span class="op">=</span><span class="va">contrast</span><span class="op">[</span><span class="va">both</span>,<span class="op">]</span>, shape<span class="op">=</span><span class="fl">21</span>, color<span class="op">=</span><span class="st">"black"</span>, fill<span class="op">=</span><span class="st">"red"</span>, size<span class="op">=</span><span class="fl">3</span>, stroke<span class="op">=</span><span class="fl">0.6</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="fu">print</span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-06-gene-set-enrichment-analyses"><p>Content from <a href="06-gene-set-enrichment-analyses.html">Gene set enrichment analysis</a></p>
<hr>
<p>Last updated on 2025-12-16 |

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/episodes/06-gene-set-enrichment-analyses.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 80 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we identify functional pathways and biological themes from DE
genes.</li>
<li>How do we run over representation analysis for GO, KEGG, and MSigDB
sets.</li>
<li>How do we create simple visualizations of enriched terms.</li>
<li>How do we interpret enrichment results in a biologically meaningful
way.</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Prepare a gene list from DESeq2 results.</li>
<li>Run GO enrichment using clusterProfiler.</li>
<li>Run KEGG pathway enrichment with organism specific data.</li>
<li>Perform MSigDB Hallmark enrichment.</li>
<li>Produce barplots and dotplots for enriched terms.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Differential expression gives a list of up regulated and down
regulated genes.<br>
To understand the underlying biology, we examine which
<strong>functional gene sets</strong> are enriched.<br>
This process is called <strong>over representation analysis</strong>
(ORA). ORA identifies pathways or gene sets that contain more DE genes
than expected by chance.</p>
<p>Here we will use <strong>clusterProfiler</strong>,
<strong>org.Mm.eg.db</strong>, and <strong>msigdbr</strong> to run GO,
KEGG, and Hallmark analyses on mouse data.</p>
<div id="what-ora-requires" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="what-ora-requires" class="callout-inner">
<h3 class="callout-title">What ORA requires</h3>
<div class="callout-content">
<ul>
<li>A background universe of all tested genes (usually all genes in the
DESeq2 dataset).</li>
<li>A subset of genes considered significant (for example padj &lt;
0.05).</li>
<li>A gene identifier type that matches the annotation database (Ensembl
or Entrez).</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-1-load-de-results-and-prepare-gene-lists">Step 1: Load DE results and prepare gene lists<a class="anchor" aria-label="anchor" href="#step-1-load-de-results-and-prepare-gene-lists"></a>
</h2>
<hr class="half-width">
<p>Start your R session:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">sinteractive</span> <span class="at">-A</span> rcac <span class="at">-p</span> cpu <span class="at">-N</span> 1 <span class="at">-n</span> 4 <span class="at">--time</span><span class="op">=</span>02:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">module</span> load biocontainers</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">module</span> load r-rnaseq</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">R</span></span></code></pre>
</div>
<p>Load data:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">readr</span><span class="op">)</span></span></code></pre>
</div>
<p>Load DESeq2 results from Episode C:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"DESeq2_results_shrunken.csv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<p>Define:</p>
<ul>
<li>universe = all genes tested<br>
</li>
<li>significant set = padj &lt; 0.05</li>
</ul>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">X1</span></span>
<span><span class="va">sig</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span></code></pre>
</div>
<p>Convert Ensembl IDs to Entrez (required for GO and KEGG):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">id_map</span> <span class="op">&lt;-</span> <span class="fu">bitr</span><span class="op">(</span><span class="va">universe</span>,</span>
<span>               fromType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>               toType <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>               OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<p>Merge IDs into results:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">res</span>, <span class="va">id_map</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"X1"</span> <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sig_entrez</span> <span class="op">&lt;-</span> <span class="va">res2</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">ENTREZID</span><span class="op">)</span></span>
<span><span class="va">universe_entrez</span> <span class="op">&lt;-</span> <span class="va">res2</span><span class="op">$</span><span class="va">ENTREZID</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="step-2-go-enrichment">Step 2: GO enrichment<a class="anchor" aria-label="anchor" href="#step-2-go-enrichment"></a>
</h2>
<hr class="half-width">
<p>Run GO Biological Process ORA:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>                universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>                OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<p>View top terms:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">ego</span><span class="op">)</span></span></code></pre>
</div>
<p>Plot:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ego</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"GO Biological Process enrichment"</span><span class="op">)</span></span></code></pre>
</div>
<div id="why-go-is-useful" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="why-go-is-useful" class="callout-inner">
<h3 class="callout-title">Why GO is useful</h3>
<div class="callout-content">
<p>GO BP terms provide high level summaries of biological
activity.<br>
Strongly biased terms (for example mitochondrial or translation) often
correspond to global shifts in cell state or stress.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-3-kegg-enrichment">Step 3: KEGG enrichment<a class="anchor" aria-label="anchor" href="#step-3-kegg-enrichment"></a>
</h2>
<hr class="half-width">
<p>KEGG requires Entrez IDs and organism code <code>"mmu"</code> for
mouse.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ekegg</span> <span class="op">&lt;-</span> <span class="fu">enrichKEGG</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>                    universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>                    organism <span class="op">=</span> <span class="st">"mmu"</span><span class="op">)</span></span></code></pre>
</div>
<p>Plot KEGG enrichment:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">ekegg</span>, showCategory <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre>
</div>
<div id="kegg-caveat" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="kegg-caveat" class="callout-inner">
<h3 class="callout-title">KEGG caveat</h3>
<div class="callout-content">
<p>KEGG Entrez annotations may not perfectly match recent Ensembl
builds.<br>
If a gene is missing, it will be silently dropped.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-4-msigdb-hallmark-enrichment">Step 4: MSigDB Hallmark enrichment<a class="anchor" aria-label="anchor" href="#step-4-msigdb-hallmark-enrichment"></a>
</h2>
<hr class="half-width">
<p>Load Hallmark gene sets for mouse:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mus musculus"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="va">hallmark_list</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">hallmark</span><span class="op">$</span><span class="va">entrez_gene</span>, <span class="va">hallmark</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span></span></code></pre>
</div>
<p>Run ORA:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ehall</span> <span class="op">&lt;-</span> <span class="fu">enricher</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">sig_entrez</span>,</span>
<span>                  universe <span class="op">=</span> <span class="va">universe_entrez</span>,</span>
<span>                  TERM2GENE <span class="op">=</span> <span class="va">hallmark_list</span><span class="op">)</span></span></code></pre>
</div>
<p>Plot:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">ehall</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"MSigDB Hallmark enrichment"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="inspect-enriched-hallmark-sets">Inspect enriched Hallmark sets<a class="anchor" aria-label="anchor" href="#inspect-enriched-hallmark-sets"></a>
</h3>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">ehall</span><span class="op">)</span></span></code></pre>
</div>
<div id="why-hallmark-sets-are-recommended" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-hallmark-sets-are-recommended" class="callout-inner">
<h3 class="callout-title">Why Hallmark sets are recommended</h3>
<div class="callout-content">
<p>Hallmark pathways are curated to reduce redundancy.<br>
They are easier to interpret compared to long lists of GO terms.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="step-5-saving-results">Step 5: Saving results<a class="anchor" aria-label="anchor" href="#step-5-saving-results"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ego</span><span class="op">)</span>, <span class="st">"GO_enrichment.csv"</span><span class="op">)</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ekegg</span><span class="op">)</span>, <span class="st">"KEGG_enrichment.csv"</span><span class="op">)</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span><span class="fu">as.data.frame</span><span class="op">(</span><span class="va">ehall</span><span class="op">)</span>, <span class="st">"Hallmark_enrichment.csv"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="step-6-interpretation-guidelines">Step 6: Interpretation guidelines<a class="anchor" aria-label="anchor" href="#step-6-interpretation-guidelines"></a>
</h2>
<hr class="half-width">
<p>Meaningful enrichment patterns should:</p>
<ul>
<li>involve pathways relevant to your experimental design<br>
</li>
<li>contain multiple DE genes in the same direction<br>
</li>
<li>show consistent themes across GO, KEGG, and Hallmark sets</li>
</ul>
<p>Pathways enriched only by one DE gene or very small sets should be
interpreted cautiously.</p>
<div id="exercise-interpret-enrichment-results" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-interpret-enrichment-results" class="callout-inner">
<h3 class="callout-title">Exercise: interpret enrichment results</h3>
<div class="callout-content">
<p>Using your GO, KEGG, and Hallmark outputs:</p>
<ol style="list-style-type: decimal">
<li>Identify the pathways with the smallest adjusted p values.<br>
</li>
<li>Are the enriched pathways consistent with the expected
phenotype.<br>
</li>
<li>Which pathway collections (GO, KEGG, Hallmark) give the clearest
signal.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Example interpretation:</p>
<ul>
<li>Several GO terms related to RNA processing and cell cycle appear
enriched.<br>
</li>
<li>Hallmark sets provide clearer summaries of biological themes
compared to GO.<br>
</li>
<li>KEGG pathways show fewer hits but corroborate the GO findings.</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>ORA requires a universe and a significant gene set.<br>
</li>
<li>Ensembl IDs must be converted to Entrez for GO and KEGG.<br>
</li>
<li>clusterProfiler provides GO BP and KEGG ORA.<br>
</li>
<li>Hallmark sets offer high quality curated pathways for
interpretation.<br>
</li>
<li>Dotplots and barplots are effective for visualizing enrichment
results.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://aseetharam.github.io/rcac_rnaseq_workshop/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:aseethar@purdue.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.3" class="external-link">sandpaper (0.17.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/aseetharam/varnish/tree/9765d26807833061ac50b41dc951cf45941290f6" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, RNA-seq, Bioconductor, R, HPC, RCAC",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "identifier": "https://rcac_rnaseq_workshop.github.io/NA/instructor/aio.html",
  "dateCreated": "2025-11-12",
  "dateModified": "2025-12-16",
  "datePublished": "2025-12-16"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

